<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>RNA-Seq Workflow Template | GEN242</title><meta property="og:title" content="RNA-Seq Workflow Template" />
<meta property="og:description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ]  Introduction This report describes the analysis of the RNA-Seq data set from Howard et al (2013)." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/sprnaseq/sprnaseqnew/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2022-04-21T17:59:16-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="RNA-Seq Workflow Template">
<meta itemprop="description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ]  Introduction This report describes the analysis of the RNA-Seq data set from Howard et al (2013).">
<meta itemprop="dateModified" content="2022-04-21T17:59:16-07:00" />
<meta itemprop="wordCount" content="4091">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="RNA-Seq Workflow Template"/>
<meta name="twitter:description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ]  Introduction This report describes the analysis of the RNA-Seq data set from Howard et al (2013)."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S22)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T19)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2022/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.828f0f0bb84f9505b932d9dc4c3393d9.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.828f0f0bb84f9505b932d9dc4c3393d9.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../tutorials/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Tutorials</a>
  </li>
  <ul>
    <li class="collapse show" id="tutorials">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsgithubgithub" href="../../../tutorials/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialslinuxlinux" href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrbasicsrbasics" href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrprogrammingrprogramming" href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrparallelrparallel" href="../../../tutorials/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrsequencesrsequences" href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempiper" href="../../../tutorials/systempiper/systempiper/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempipernew" href="../../../tutorials/systempiper/systempipernew/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseq" href="../../../tutorials/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-tutorialssprnaseqsprnaseqnew" href="../../../tutorials/sprnaseq/sprnaseqnew/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspchipseqspchipseq" href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsscrnaseqscrnaseq" href="../../../tutorials/scrnaseq/scrnaseq/">scRNA-Seq Embedding Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrmarkdownrmarkdown" href="../../../tutorials/rmarkdown/rmarkdown/">R Markdown Tutorial</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrfearfea" href="../../../tutorials/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrclusteringrclustering" href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrgraphicsrgraphics" href="../../../tutorials/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappsshinyapps" href="../../../tutorials/shinyapps/shinyapps/">Shiny Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrpackagesrpackages" href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsdplyrdplyr" href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialscmdtocwlcmdtocwl" href="../../../tutorials/cmdtocwl/cmdtocwl/">Automate Creation of CWL Instructions</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/sprnaseq/sprnaseqNew.Rmd" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/tutorials/sprnaseq/sprnaseqNew.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/tutorials/sprnaseq/sprnaseqNew.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=RNA-Seq%20Workflow%20Template" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#introduction">Introduction</a>
      <ul>
        <li><a href="#experimental-design">Experimental design</a></li>
        <li><a href="#workflow-environment">Workflow environment</a></li>
        <li><a href="#load-packages">Load packages</a></li>
        <li><a href="#import-custom-functions">Import custom functions</a></li>
        <li><a href="#experiment-definition-provided-by-targets-file">Experiment definition provided by <code>targets</code> file</a></li>
      </ul>
    </li>
    <li><a href="#workflow-steps">Workflow steps</a>
      <ul>
        <li><a href="#required-packages-and-resources">Required packages and resources</a></li>
        <li><a href="#read-preprocessing">Read preprocessing</a>
          <ul>
            <li><a href="#read-quality-filtering-and-trimming">Read quality filtering and trimming</a></li>
            <li><a href="#fastq-quality-report">FASTQ quality report</a></li>
          </ul>
        </li>
        <li><a href="#alignments">Alignments</a>
          <ul>
            <li><a href="#read-mapping-with-hisat2">Read mapping with <code>HISAT2</code></a></li>
            <li><a href="#read-and-alignment-stats">Read and alignment stats</a></li>
            <li><a href="#create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</a></li>
          </ul>
        </li>
        <li><a href="#read-quantification">Read quantification</a>
          <ul>
            <li><a href="#create-a-database-for-gene-annotation">Create a database for gene annotation</a></li>
            <li><a href="#read-counting-with-summarizeoverlaps-in-parallel-mode-using-multiple-cores">Read counting with <code>summarizeOverlaps</code> in parallel mode using multiple cores</a></li>
            <li><a href="#sample-wise-correlation-analysis">Sample-wise correlation analysis</a></li>
          </ul>
        </li>
        <li><a href="#analysis-of-degs">Analysis of DEGs</a>
          <ul>
            <li><a href="#run-edger">Run <code>edgeR</code></a></li>
            <li><a href="#add-gene-descriptions">Add gene descriptions</a></li>
          </ul>
        </li>
        <li><a href="#plot-deg-results">Plot DEG results</a></li>
        <li><a href="#venn-diagrams-of-deg-sets">Venn diagrams of DEG sets</a></li>
        <li><a href="#go-term-enrichment-analysis">GO term enrichment analysis</a>
          <ul>
            <li><a href="#obtain-gene-to-go-mappings">Obtain gene-to-GO mappings</a></li>
            <li><a href="#batch-go-term-enrichment-analysis">Batch GO term enrichment analysis</a></li>
            <li><a href="#plot-batch-go-term-results">Plot batch GO term results</a></li>
          </ul>
        </li>
        <li><a href="#clustering-and-heat-maps">Clustering and heat maps</a></li>
        <li><a href="#version-information">Version Information</a></li>
      </ul>
    </li>
    <li><a href="#running-workflow">Running workflow</a>
      <ul>
        <li><a href="#interactive-job-submissions-in-a-single-machine">Interactive job submissions in a single machine</a></li>
        <li><a href="#parallelization-on-clusters">Parallelization on clusters</a></li>
        <li><a href="#visualize-workflow">Visualize workflow</a></li>
      </ul>
    </li>
    <li><a href="#checking-workflow-status">Checking workflow status</a></li>
    <li><a href="#accessing-logs-report">Accessing logs report</a></li>
    <li><a href="#funding">Funding</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../tutorials/">Tutorials</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../tutorials/sprnaseq/sprnaseqnew/">RNA-Seq Workflow Template</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>RNA-Seq Workflow Template</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 20 minute read</p>
           
	<script src="../../../rmarkdown-libs/htmlwidgets/htmlwidgets.js"></script>
<link href="../../../rmarkdown-libs/datatables-css/datatables-crosstalk.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/datatables-binding/datatables.js"></script>
<script src="../../../rmarkdown-libs/jquery/jquery-3.6.0.min.js"></script>
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/dt-core/js/jquery.dataTables.min.js"></script>
<link href="../../../rmarkdown-libs/crosstalk/css/crosstalk.min.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/crosstalk/js/crosstalk.min.js"></script>
<script src="../../../rmarkdown-libs/htmlwidgets/htmlwidgets.js"></script>
<link href="../../../rmarkdown-libs/datatables-css/datatables-crosstalk.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/datatables-binding/datatables.js"></script>
<script src="../../../rmarkdown-libs/jquery/jquery-3.6.0.min.js"></script>
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/dt-core/js/jquery.dataTables.min.js"></script>
<link href="../../../rmarkdown-libs/crosstalk/css/crosstalk.min.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/crosstalk/js/crosstalk.min.js"></script>
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<!--
Rscript -e "rmarkdown::render('sprnaseq.Rmd', c('BiocStyle::html_document'), clean=F); knitr::knit('sprnaseq.Rmd', tangle=TRUE)"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div style="text-align: right">
<p>Source code downloads:    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/static/custom/spWFtemplates/sprnaseq.Rmd">.Rmd</a> ]    </p>
</div>
<h2 id="introduction">Introduction</h2>
<p>This report describes the analysis of the RNA-Seq data set from
Howard et al (2013). The corresponding FASTQ files were downloaded from
GEO (Accession: <a href="http://www.ncbi.nlm.nih.gov/sra/?term=SRP010938">SRP010938</a>).
This data set contains 18 paired-end (PE) read sets from <em>Arabidposis thaliana</em>.
The details about all download steps are provided <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/projects/project_data/">here</a>.</p>
<p>Users want to provide here additional background information about the design of their
RNA-Seq project.</p>
<h3 id="experimental-design">Experimental design</h3>
<p>Typically, users want to specify here all information relevant for the
analysis of their NGS study. This includes detailed descriptions of
FASTQ files, experimental design, reference genome, gene annotations,
etc.</p>
<h3 id="workflow-environment">Workflow environment</h3>
<p><font color="red">NOTE: this section</font> describes how to set up the proper
environment (directory structure) for running <code>systemPipeR</code> workflows. After
mastering this task the workflow run instructions <font color="red">can be
deleted</font> since they are not expected to be included in a final HTML/PDF
report of a workflow.</p>
<ol>
<li>
<p>If a remote system or cluster is used, then users need to log in to the
remote system first. The following applies to an HPC cluster (<em>e.g.</em> HPCC
cluster).</p>
<p>A terminal application needs to be used to log in to a user’s cluster account. Next, one
can open an interactive session on a computer node with <code>srun</code>. More details about
argument settings for <code>srun</code> are available in this <a href="http://hpcc.ucr.edu/manuals_linux-cluster_jobs.html#partitions">HPCC
manual</a> or
the HPCC section of this website
<a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/linux/linux/#job-submission-with-sbatch">here</a>.
Next, load the R version required for running the workflow with <code>module load</code>. Sometimes it may be necessary to
first unload an active software version before loading another version, <em>e.g.</em> <code>module unload R</code>.</p>
</li>
</ol>
<!-- end list -->
<pre><code class="language-sh">srun --x11 --partition=gen242 --mem=20gb --cpus-per-task 8 --ntasks 1 --time 20:00:00 --pty bash -l
module unload R; module load load R/4.1.2
</code></pre>
<ol start="2">
<li>Load a workflow template with the <code>genWorkenvir</code> function. This can be done
from the command-line or from within R. However, only one of the two options needs to be used.</li>
</ol>
<p>From command-line</p>
<pre><code class="language-sh">$ Rscript -e &quot;systemPipeRdata::genWorkenvir(workflow='rnaseq')&quot;
$ cd rnaseq
</code></pre>
<p>From R</p>
<pre><code class="language-r">library(systemPipeRdata)
genWorkenvir(workflow = &quot;rnaseq&quot;)
setwd(&quot;rnaseq&quot;)
</code></pre>
<ol start="3">
<li>
<p>Optional: if the user wishes to use another <code>Rmd</code> file than the template instance
provided by the <code>genWorkenvir</code> function, then it can be copied or downloaded
into the root directory of the workflow environment (<em>e.g.</em> with <code>cp</code> or <code>wget</code>).</p>
</li>
<li>
<p>Now one can open from the root directory of the workflow the corresponding R Markdown script (<em>e.g.</em> systemPipeChIPseq.Rmd) using an R IDE, such as <em>nvim-r</em>, <em>ESS</em> or RStudio.
Subsequently, the workflow can be run as outlined below. For learning purposes it is recommended to run workflows for the first time interactively. Once all workflow steps are understood and possibly
modified to custom needs, one can run the workflow from start to finish with a
single command using <code>runWF()</code>.</p>
</li>
</ol>
<h3 id="load-packages">Load packages</h3>
<p>The <code>systemPipeR</code> package needs to be loaded to perform the analysis
steps shown in this report (H Backman and Girke 2016). The package allows users
to run the entire analysis workflow interactively or with a single command
while also generating the corresponding analysis report. For details
see <code>systemPipeR's</code> main <a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">vignette</a>.</p>
<pre><code class="language-r">library(systemPipeR)
</code></pre>
<p>To apply workflows to custom data, the user needs to modify the <em><code>targets</code></em> file and if
necessary update the corresponding parameter (<em><code>.cwl</code></em> and <em><code>.yml</code></em>) files.
A collection of pre-generated <em><code>.cwl</code></em> and <em><code>.yml</code></em> files are provided in the <em><code>param/cwl</code></em> subdirectory
of each workflow template. They are also viewable in the GitHub repository of <em><code>systemPipeRdata</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/tree/master/inst/extdata/param/cwl">see
here</a>).
For more information of the structure of the <em>targets</em> file, please consult the documentation
<a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#25_structure_of_targets_file">here</a>. More details about the new parameter files from systemPipeR can be found <a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#26_structure_of_the_new_param_files_and_construct_sysargs2_container">here</a>.</p>
<h3 id="import-custom-functions">Import custom functions</h3>
<p>Custom functions for the challenge projects can be imported with the source
command from a local R script (here <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/content/en/tutorials/spchipseq/challengeProject_Fct.R">challengeProject_Fct.R</a>). Skip this step if such a script is not available. Alternatively, these
functions can be loaded from a custom R package.</p>
<pre><code class="language-r">source(&quot;challengeProject_Fct.R&quot;)
</code></pre>
<h3 id="experiment-definition-provided-by-targets-file">Experiment definition provided by <code>targets</code> file</h3>
<p>The <code>targets</code> file defines all FASTQ files and sample comparisons of the analysis workflow.
If needed the tab separated (TSV) version of this file can be downloaded from <a href="https://github.com/tgirke/GEN242/tree/main/content/en/assignments/Projects/targets_files">here</a>
and the corresponding Google Sheet is <a href="https://docs.google.com/spreadsheets/d/1DTgTGlZZscSPjlHOGdJC8QK4vvimN1BORjXKXzd_cfA/edit#gid=472150521">here</a>.</p>
<pre><code class="language-r">targetspath &lt;- &quot;targetsPE.txt&quot;
targets &lt;- read.delim(targetspath, comment.char = &quot;#&quot;)
DT::datatable(targets, options = list(scrollX = TRUE, autoWidth = TRUE))
</code></pre>
<div id="htmlwidget-1" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-1">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18"],["./data/SRR446027_1.fastq.gz","./data/SRR446028_1.fastq.gz","./data/SRR446029_1.fastq.gz","./data/SRR446030_1.fastq.gz","./data/SRR446031_1.fastq.gz","./data/SRR446032_1.fastq.gz","./data/SRR446033_1.fastq.gz","./data/SRR446034_1.fastq.gz","./data/SRR446035_1.fastq.gz","./data/SRR446036_1.fastq.gz","./data/SRR446037_1.fastq.gz","./data/SRR446038_1.fastq.gz","./data/SRR446039_1.fastq.gz","./data/SRR446040_1.fastq.gz","./data/SRR446041_1.fastq.gz","./data/SRR446042_1.fastq.gz","./data/SRR446043_1.fastq.gz","./data/SRR446044_1.fastq.gz"],["./data/SRR446027_2.fastq.gz","./data/SRR446028_2.fastq.gz","./data/SRR446029_2.fastq.gz","./data/SRR446030_2.fastq.gz","./data/SRR446031_2.fastq.gz","./data/SRR446032_2.fastq.gz","./data/SRR446033_2.fastq.gz","./data/SRR446034_2.fastq.gz","./data/SRR446035_2.fastq.gz","./data/SRR446036_2.fastq.gz","./data/SRR446037_2.fastq.gz","./data/SRR446038_2.fastq.gz","./data/SRR446039_2.fastq.gz","./data/SRR446040_2.fastq.gz","./data/SRR446041_2.fastq.gz","./data/SRR446042_2.fastq.gz","./data/SRR446043_2.fastq.gz","./data/SRR446044_2.fastq.gz"],["M1A","M1B","A1A","A1B","V1A","V1B","M6A","M6B","A6A","A6B","V6A","V6B","M12A","M12B","A12A","A12B","V12A","V12B"],["M1","M1","A1","A1","V1","V1","M6","M6","A6","A6","V6","V6","M12","M12","A12","A12","V12","V12"],["Mock.1h.A","Mock.1h.B","Avr.1h.A","Avr.1h.B","Vir.1h.A","Vir.1h.B","Mock.6h.A","Mock.6h.B","Avr.6h.A","Avr.6h.B","Vir.6h.A","Vir.6h.B","Mock.12h.A","Mock.12h.B","Avr.12h.A","Avr.12h.B","Vir.12h.A","Vir.12h.B"],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],["23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012","23-Mar-2012"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>FileName1<\/th>\n      <th>FileName2<\/th>\n      <th>SampleName<\/th>\n      <th>Factor<\/th>\n      <th>SampleLong<\/th>\n      <th>Experiment<\/th>\n      <th>Date<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"autoWidth":true,"columnDefs":[{"className":"dt-right","targets":6},{"orderable":false,"targets":0}],"order":[],"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<h2 id="workflow-steps">Workflow steps</h2>
<p>This tutorial will demonstrate how to build the workflow in an interactive mode,
appending each step. The workflow is constructed by connecting each step via
<code>appendStep</code> method. Each <code>SYSargsList</code> instance contains instructions for
processing a set of input files with a specific command-line or R software
and the paths to the corresponding outfiles generated by a particular command-line
software/step.</p>
<p>To create a workflow within <em><code>systemPipeR</code></em>, we can start by defining an empty
container and checking the directory structure:</p>
<pre><code class="language-r">library(systemPipeR)
sal &lt;- SPRproject()
sal
</code></pre>
<h3 id="required-packages-and-resources">Required packages and resources</h3>
<p>The <code>systemPipeR</code> package needs to be loaded (H Backman and Girke 2016).</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(systemPipeR)
}, step_name = &quot;load_SPR&quot;)
</code></pre>
<h3 id="read-preprocessing">Read preprocessing</h3>
<h4 id="read-quality-filtering-and-trimming">Read quality filtering and trimming</h4>
<p>The function <code>preprocessReads</code> allows to apply predefined or custom
read preprocessing functions to all FASTQ files referenced in a
<code>SYSargsList</code> container, such as quality filtering or adapter trimming
routines. The paths to the resulting output FASTQ files are stored in the
<code>outfiles</code> slot of the <code>SYSargsList</code> object. The following example performs adapter trimming with
the <code>trimLRPatterns</code> function from the <code>Biostrings</code> package.
After the trimming step a new targets file is generated (here
<code>targets_trim.txt</code>) containing the paths to the trimmed FASTQ files.
The new targets file can be used for the next workflow step with an updated
<code>SYSargs2</code> instance, <em>e.g.</em> running the NGS alignments using the
trimmed FASTQ files.</p>
<p>Here, we are appending this step to the <code>SYSargsList</code> object created previously.
All the parameters are defined on the <code>preprocessReads/preprocessReads-pe.yml</code> and
<code>preprocessReads/preprocessReads-pe.cwl</code> files.</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;preprocessing&quot;, targets = &quot;targetsPE.txt&quot;,
    dir = TRUE, wf_file = &quot;preprocessReads/preprocessReads-pe.cwl&quot;,
    input_file = &quot;preprocessReads/preprocessReads-pe.yml&quot;, dir_path = &quot;param/cwl&quot;,
    inputvars = c(FileName1 = &quot;_FASTQ_PATH1_&quot;, FileName2 = &quot;_FASTQ_PATH2_&quot;,
        SampleName = &quot;_SampleName_&quot;), dependency = c(&quot;load_SPR&quot;))
</code></pre>
<p>After, we can check the <code>trimLRPatterns</code> function in input parameter:</p>
<pre><code class="language-r">yamlinput(sal, &quot;preprocessing&quot;)$Fct
</code></pre>
<p>After the preprocessing step, the <code>outfiles</code> files can be used to generate the new
targets files containing the paths to the trimmed FASTQ files. The new targets
information can be used for the next workflow step instance, <em>e.g.</em> running the
NGS alignments with the trimmed FASTQ files. The <code>appendStep</code> function is
automatically handling this connectivity between steps. Please check the <code>Alignments</code>
step for more details.</p>
<h4 id="fastq-quality-report">FASTQ quality report</h4>
<p>The following <code>seeFastq</code> and <code>seeFastqPlot</code> functions generate and plot a series of useful
quality statistics for a set of FASTQ files including per cycle quality box
plots, base proportions, base-level quality trends, relative k-mer
diversity, length and occurrence distribution of reads, number of reads
above quality cutoffs and mean quality distribution. The results are
written to a PDF file named <code>fastqReport.pdf</code>.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    fastq &lt;- getColumn(sal, step = &quot;preprocessing&quot;, &quot;targetsWF&quot;,
        column = 1)
    fqlist &lt;- seeFastq(fastq = fastq, batchsize = 10000, klength = 8)
    pdf(&quot;./results/fastqReport.pdf&quot;, height = 18, width = 4 *
        length(fqlist))
    seeFastqPlot(fqlist)
    dev.off()
}, step_name = &quot;fastq_report&quot;, dependency = &quot;preprocessing&quot;)
</code></pre>
<p><img src="../results/fastqReport.png" alt=""></p>
<div data-align="center">
<p>Figure 1: FASTQ quality report for 18 samples</p>
</div>
</br>
<h3 id="alignments">Alignments</h3>
<h4 id="read-mapping-with-hisat2">Read mapping with <code>HISAT2</code></h4>
<p>The following steps will demonstrate how to use the short read aligner <code>Hisat2</code>
(Kim, Langmead, and Salzberg 2015) in both interactive job submissions and batch submissions to
queuing systems of clusters using the <em><code>systemPipeR's</code></em> new CWL command-line interface.</p>
<p>The following steps will demonstrate how to use the short read aligner <code>Hisat2</code>
(Kim, Langmead, and Salzberg 2015). First, the <code>Hisat2</code> index needs to be created.</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;hisat2_index&quot;, dir = FALSE,
    targets = NULL, wf_file = &quot;hisat2/hisat2-index.cwl&quot;, input_file = &quot;hisat2/hisat2-index.yml&quot;,
    dir_path = &quot;param/cwl&quot;, dependency = &quot;load_SPR&quot;)
</code></pre>
<p>The parameter settings of the aligner are defined in the <code>workflow_hisat2-pe.cwl</code>
and <code>workflow_hisat2-pe.yml</code> files. The following shows how to construct the
corresponding <em>SYSargsList</em> object. Please note that the targets used in this
step are the outfiles from <code>preprocessing</code> step.</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;hisat2_mapping&quot;,
    dir = TRUE, targets = &quot;preprocessing&quot;, wf_file = &quot;workflow-hisat2/workflow_hisat2-pe.cwl&quot;,
    input_file = &quot;workflow-hisat2/workflow_hisat2-pe.yml&quot;, dir_path = &quot;param/cwl&quot;,
    inputvars = c(preprocessReads_1 = &quot;_FASTQ_PATH1_&quot;, preprocessReads_2 = &quot;_FASTQ_PATH2_&quot;,
        SampleName = &quot;_SampleName_&quot;), rm_targets_col = c(&quot;FileName1&quot;,
        &quot;FileName2&quot;), dependency = c(&quot;preprocessing&quot;, &quot;hisat2_index&quot;))
</code></pre>
<p>To double-check the command line for each sample, please use the following:</p>
<pre><code class="language-r">cmdlist(sal, step = &quot;hisat2_mapping&quot;, targets = 1)
</code></pre>
<h4 id="read-and-alignment-stats">Read and alignment stats</h4>
<p>The following provides an overview of the number of reads in each sample
and how many of them aligned to the reference.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    fqpaths &lt;- getColumn(sal, step = &quot;preprocessing&quot;, &quot;targetsWF&quot;,
        column = &quot;FileName1&quot;)
    bampaths &lt;- getColumn(sal, step = &quot;hisat2_mapping&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    read_statsDF &lt;- alignStats(args = bampaths, fqpaths = fqpaths,
        pairEnd = TRUE)
    write.table(read_statsDF, &quot;results/alignStats.xls&quot;, row.names = FALSE,
        quote = FALSE, sep = &quot;\t&quot;)
}, step_name = &quot;align_stats&quot;, dependency = &quot;hisat2_mapping&quot;)
</code></pre>
<p>The following shows the alignment statistics for a sample file provided by the <code>systemPipeR</code> package.</p>
<pre><code class="language-r">read.table(&quot;results/alignStats.xls&quot;, header = TRUE)[1:4, ]
</code></pre>
<pre><code>##   FileName Nreads2x   Nalign Perc_Aligned Nalign_Primary
## 1      M1A 33609678 32136300     95.61621       32136300
## 2      M1B 53002402 43620124     82.29839       43620124
## 3      A1A 50223496 48438407     96.44571       48438407
## 4      A1B 43650000 35549889     81.44304       35549889
##   Perc_Aligned_Primary
## 1             95.61621
## 2             82.29839
## 3             96.44571
## 4             81.44304
</code></pre>
<h4 id="create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</h4>
<p>The <code>symLink2bam</code> function creates symbolic links to view the BAM alignment files in a
genome browser such as IGV without moving these large files to a local
system. The corresponding URLs are written to a file with a path
specified under <code>urlfile</code>, here <code>IGVurl.txt</code>.
Please replace the directory and the user name.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    bampaths &lt;- getColumn(sal, step = &quot;hisat2_mapping&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    symLink2bam(sysargs = bampaths, htmldir = c(&quot;~/.html/&quot;, &quot;somedir/&quot;),
        urlbase = &quot;http://cluster.hpcc.ucr.edu/~&lt;username&gt;/&quot;,
        urlfile = &quot;./results/IGVurl.txt&quot;)
}, step_name = &quot;bam_urls&quot;, dependency = &quot;hisat2_mapping&quot;, run_step = &quot;optional&quot;)
</code></pre>
<h3 id="read-quantification">Read quantification</h3>
<p>Reads overlapping with annotation ranges of interest are counted for
each sample using the <code>summarizeOverlaps</code> function (Lawrence et al. 2013). The read counting is
preformed for exonic gene regions in a non-strand-specific manner while
ignoring overlaps among different genes. Subsequently, the expression
count values are normalized by <em>reads per kp per million mapped reads</em>
(RPKM). The raw read count table (<code>countDFeByg.xls</code>) and the corresponding
RPKM table (<code>rpkmDFeByg.xls</code>) are written to separate files in the directory of
this project. Parallelization is achieved with the <code>BiocParallel</code> package, here
using 4 CPU cores.</p>
<h4 id="create-a-database-for-gene-annotation">Create a database for gene annotation</h4>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(GenomicFeatures)
    txdb &lt;- suppressWarnings(makeTxDbFromGFF(file = &quot;data/tair10.gff&quot;,
        format = &quot;gff&quot;, dataSource = &quot;TAIR&quot;, organism = &quot;Arabidopsis thaliana&quot;))
    saveDb(txdb, file = &quot;./data/tair10.sqlite&quot;)
}, step_name = &quot;create_db&quot;, dependency = &quot;hisat2_mapping&quot;)
</code></pre>
<h4 id="read-counting-with-summarizeoverlaps-in-parallel-mode-using-multiple-cores">Read counting with <code>summarizeOverlaps</code> in parallel mode using multiple cores</h4>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(GenomicFeatures)
    library(BiocParallel)
    txdb &lt;- loadDb(&quot;./data/tair10.sqlite&quot;)
    outpaths &lt;- getColumn(sal, step = &quot;hisat2_mapping&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    eByg &lt;- exonsBy(txdb, by = c(&quot;gene&quot;))
    bfl &lt;- BamFileList(outpaths, yieldSize = 50000, index = character())
    multicoreParam &lt;- MulticoreParam(workers = 4)
    register(multicoreParam)
    registered()
    counteByg &lt;- bplapply(bfl, function(x) summarizeOverlaps(eByg,
        x, mode = &quot;Union&quot;, ignore.strand = TRUE, inter.feature = FALSE,
        singleEnd = FALSE, BPPARAM = multicoreParam))
    countDFeByg &lt;- sapply(seq(along = counteByg), function(x) assays(counteByg[[x]])$counts)
    rownames(countDFeByg) &lt;- names(rowRanges(counteByg[[1]]))
    colnames(countDFeByg) &lt;- names(bfl)
    rpkmDFeByg &lt;- apply(countDFeByg, 2, function(x) returnRPKM(counts = x,
        ranges = eByg))
    write.table(countDFeByg, &quot;results/countDFeByg.xls&quot;, col.names = NA,
        quote = FALSE, sep = &quot;\t&quot;)
    write.table(rpkmDFeByg, &quot;results/rpkmDFeByg.xls&quot;, col.names = NA,
        quote = FALSE, sep = &quot;\t&quot;)
    ## Creating a SummarizedExperiment object
    colData &lt;- data.frame(row.names = SampleName(sal, &quot;hisat2_mapping&quot;),
        condition = getColumn(sal, &quot;hisat2_mapping&quot;, position = &quot;targetsWF&quot;,
            column = &quot;Factor&quot;))
    colData$condition &lt;- factor(colData$condition)
    countDF_se &lt;- SummarizedExperiment::SummarizedExperiment(assays = countDFeByg,
        colData = colData)
    ## Add results as SummarizedExperiment to the workflow
    ## object
    SE(sal, &quot;read_counting&quot;) &lt;- countDF_se
}, step_name = &quot;read_counting&quot;, dependency = &quot;create_db&quot;)
</code></pre>
<p>When providing a <code>BamFileList</code> as in the example above, <code>summarizeOverlaps</code> methods
use by default <code>bplapply</code> and use the register interface from BiocParallel package.
If the number of workers is not set, <code>MulticoreParam</code> will use the number of cores
returned by <code>parallel::detectCores()</code>. For more information,
please check <code>help(&quot;summarizeOverlaps&quot;)</code> documentation.</p>
<p>Shows count table generated in previous step (<code>countDFeByg.xls</code>).
To avoid slowdowns of the load time of this page, ony 200 rows of the source
table are imported into the below <code>datatable</code> view .</p>
<pre><code class="language-r">countDF &lt;- read.delim(&quot;results/countDFeByg.xls&quot;, row.names = 1,
    check.names = FALSE)[1:200, ]
DT::datatable(countDF, options = list(scrollX = TRUE, autoWidth = TRUE))
</code></pre>
<div id="htmlwidget-2" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-2">{"x":{"filter":"none","vertical":false,"data":[["AT1G01010","AT1G01020","AT1G01030","AT1G01040","AT1G01046","AT1G01050","AT1G01060","AT1G01070","AT1G01073","AT1G01080","AT1G01090","AT1G01100","AT1G01110","AT1G01115","AT1G01120","AT1G01130","AT1G01140","AT1G01150","AT1G01160","AT1G01170","AT1G01180","AT1G01183","AT1G01190","AT1G01200","AT1G01210","AT1G01220","AT1G01225","AT1G01230","AT1G01240","AT1G01250","AT1G01260","AT1G01270","AT1G01280","AT1G01290","AT1G01300","AT1G01305","AT1G01310","AT1G01320","AT1G01340","AT1G01350","AT1G01355","AT1G01360","AT1G01370","AT1G01380","AT1G01390","AT1G01400","AT1G01410","AT1G01420","AT1G01430","AT1G01440","AT1G01448","AT1G01450","AT1G01453","AT1G01460","AT1G01470","AT1G01471","AT1G01480","AT1G01490","AT1G01500","AT1G01510","AT1G01520","AT1G01530","AT1G01540","AT1G01550","AT1G01560","AT1G01570","AT1G01580","AT1G01590","AT1G01600","AT1G01610","AT1G01620","AT1G01630","AT1G01640","AT1G01650","AT1G01660","AT1G01670","AT1G01680","AT1G01690","AT1G01695","AT1G01700","AT1G01710","AT1G01720","AT1G01725","AT1G01730","AT1G01740","AT1G01750","AT1G01760","AT1G01770","AT1G01780","AT1G01790","AT1G01800","AT1G01810","AT1G01820","AT1G01830","AT1G01840","AT1G01860","AT1G01870","AT1G01880","AT1G01890","AT1G01900","AT1G01910","AT1G01920","AT1G01930","AT1G01940","AT1G01950","AT1G01960","AT1G01970","AT1G01980","AT1G01990","AT1G02000","AT1G02010","AT1G02020","AT1G02030","AT1G02040","AT1G02050","AT1G02060","AT1G02065","AT1G02070","AT1G02074","AT1G02080","AT1G02090","AT1G02100","AT1G02110","AT1G02120","AT1G02130","AT1G02136","AT1G02140","AT1G02145","AT1G02150","AT1G02160","AT1G02170","AT1G02180","AT1G02190","AT1G02205","AT1G02210","AT1G02220","AT1G02230","AT1G02250","AT1G02260","AT1G02270","AT1G02280","AT1G02290","AT1G02300","AT1G02305","AT1G02310","AT1G02320","AT1G02330","AT1G02335","AT1G02340","AT1G02350","AT1G02360","AT1G02370","AT1G02380","AT1G02390","AT1G02391","AT1G02400","AT1G02405","AT1G02410","AT1G02420","AT1G02430","AT1G02440","AT1G02450","AT1G02460","AT1G02470","AT1G02475","AT1G02480","AT1G02490","AT1G02500","AT1G02510","AT1G02520","AT1G02530","AT1G02540","AT1G02550","AT1G02560","AT1G02570","AT1G02575","AT1G02580","AT1G02590","AT1G02600","AT1G02610","AT1G02620","AT1G02630","AT1G02640","AT1G02650","AT1G02660","AT1G02670","AT1G02680","AT1G02681","AT1G02690","AT1G02700","AT1G02710","AT1G02720","AT1G02730","AT1G02740","AT1G02750","AT1G02760","AT1G02770","AT1G02780","AT1G02790","AT1G02800"],[405,155,185,1342,30,283,139,5,0,546,1732,1263,16,0,4059,16,2494,0,393,417,100,0,62,11,77,236,16,562,515,19,396,0,0,104,845,0,3,1209,209,305,5,165,9,1,23,0,0,153,306,212,56,30,1,0,15440,1,102,990,451,243,1,0,929,3116,105,21,28,7,22,1287,1195,411,15,1510,20,29,14,8,2,27,426,1304,189,378,42,7,19,73,34,2483,604,6,567,48,39,102,0,17,0,57,631,130,258,101,154,819,243,0,268,330,101,187,1,1,58,43,0,0,0,1227,473,368,173,1280,390,0,1199,116,1030,314,995,81,2,387,0,97,12,0,95,587,229,25,123,442,2,0,349,43,30,15,288,36,13,87,0,2419,35,225,51,30,10,5,205,1,337,0,0,3825,0,22,3,0,0,2644,0,0,1,0,0,151,6,0,131,12,1247,26,98,1,24,1,7,76,43,70,80,0,6,4107,1,0],[577,327,202,1732,29,367,698,2,0,1025,2934,1246,20,0,6526,25,5619,0,514,319,74,0,59,25,131,312,16,782,326,18,894,0,0,203,2940,0,4,3405,334,585,3,401,6,2,106,0,0,392,682,215,95,72,4,7,17055,0,240,2455,668,570,68,0,2339,3329,176,24,100,1,145,4938,1642,679,20,3121,29,20,35,25,4,34,519,3186,306,384,134,25,35,149,72,3641,697,26,657,163,26,84,0,47,1,232,1443,292,365,113,328,1638,355,2,838,608,237,478,11,9,103,103,0,0,0,2506,745,592,272,1980,418,0,1184,220,1139,255,948,36,5,441,0,166,11,2,148,782,459,55,121,461,31,0,616,68,54,16,322,86,18,123,0,4715,43,343,177,92,10,14,351,7,549,0,0,5851,0,40,11,0,0,2686,0,0,1,0,0,94,1,0,130,23,2081,34,107,0,100,1,32,160,68,107,118,0,1,5334,0,4],[625,235,281,1737,30,435,145,7,0,630,2284,1609,34,0,5461,35,4252,0,614,507,182,0,20,12,141,342,20,1095,688,34,636,0,0,164,2141,0,2,1144,667,644,6,310,12,3,36,0,0,185,424,301,106,62,1,2,28423,0,172,1588,854,302,1,1,1649,6380,319,34,64,2,48,2413,1387,747,28,2892,18,28,44,14,2,38,725,2014,293,947,105,11,29,94,41,3901,958,10,1214,117,40,100,0,15,0,99,1107,265,422,224,243,1311,247,1,421,602,188,297,2,2,71,48,0,2,0,1890,901,840,179,2567,673,0,1920,217,1218,429,2195,104,8,418,0,134,17,4,169,1152,237,35,257,861,34,0,678,55,45,20,370,54,11,195,0,5360,82,362,106,56,16,9,488,5,452,0,0,11353,2,29,6,0,0,4789,1,0,1,0,0,238,6,0,285,24,2475,41,207,0,44,4,14,102,35,120,136,0,7,5843,0,2],[329,243,100,1200,17,306,148,1,0,882,2389,904,19,0,3851,39,2804,0,333,189,68,0,40,36,95,228,24,611,394,39,707,0,0,197,2585,0,4,2326,326,460,0,239,11,3,115,0,0,226,455,182,68,38,0,0,11281,0,272,2214,550,398,26,0,1519,2903,210,6,44,1,47,2825,1241,627,18,2133,25,31,20,16,27,8,432,945,65,293,76,16,20,119,73,2949,639,7,714,226,10,39,0,11,0,287,1160,184,291,139,256,1254,210,0,713,437,108,297,3,0,54,69,0,3,0,1835,718,552,219,2130,380,0,1019,120,745,224,856,63,21,232,0,163,25,0,224,585,299,49,89,372,35,0,525,80,40,1,537,68,14,88,0,2662,28,346,118,78,7,90,204,25,429,0,0,6069,0,27,3,0,0,2535,0,0,0,0,0,80,0,0,72,13,1136,27,79,0,62,7,19,127,67,93,71,0,2,4596,1,6],[903,282,225,1605,28,356,52,16,0,759,1850,1039,12,0,4089,18,3356,0,651,625,154,1,13,2,123,343,28,1005,771,27,545,0,0,146,1238,0,1,1351,565,472,11,359,14,0,19,0,0,127,432,259,153,75,1,2,23646,2,347,1689,628,234,2,0,1478,5078,204,21,56,2,33,1890,2392,662,49,2389,24,48,28,15,2,27,618,2113,355,794,119,15,22,141,33,3396,842,7,1367,117,27,107,0,26,0,66,1245,175,373,146,183,1590,271,3,447,505,166,212,1,0,64,51,1,2,0,2247,619,757,123,2198,493,0,1650,193,1012,338,1409,185,3,557,0,110,19,1,197,891,291,31,289,770,40,0,582,50,50,14,429,56,43,171,0,4698,30,243,119,82,8,12,485,11,528,0,1,6339,1,29,3,0,0,4141,0,1,1,0,0,328,25,1,242,19,2030,36,189,0,42,2,2,89,35,102,121,0,11,4173,0,3],[517,269,329,1580,41,663,49,6,0,1128,4010,1980,54,0,6599,37,4687,0,563,709,210,0,54,18,86,370,31,1156,434,19,1125,0,3,234,3599,0,1,1849,668,653,8,205,23,1,32,0,0,198,640,337,212,135,2,4,27160,0,233,1243,881,478,3,0,2338,6861,617,43,70,0,28,1475,1904,726,52,3834,19,40,42,29,2,59,1011,1986,330,647,92,19,45,160,78,5021,1093,4,1509,156,23,83,0,16,0,183,1231,184,391,209,434,1852,367,0,526,568,208,468,7,2,97,126,0,1,0,3198,746,866,315,2033,669,0,1145,261,1661,347,2403,85,10,458,0,112,4,5,417,1534,531,38,186,1030,43,0,503,81,22,22,1383,36,19,297,0,10035,69,506,116,69,12,8,208,8,612,0,0,10358,0,37,1,0,0,6639,0,1,1,2,2,54,7,2,193,15,2832,127,132,0,67,6,4,200,90,156,137,0,3,6188,0,3],[597,349,169,2032,49,1210,12,26,0,1091,4073,4821,13,0,2965,14,3156,0,1492,1190,394,0,59,13,300,710,25,703,642,45,250,0,0,233,1014,0,2,12417,434,508,11,481,12,8,82,0,0,337,1165,462,318,131,0,0,5538,0,44,2214,1265,1000,0,1,2217,2118,119,54,8,8,71,2827,5632,1222,49,2401,27,97,175,30,5,41,844,1123,284,852,100,23,63,581,231,2885,1759,15,2107,337,156,381,0,12,0,754,1674,455,366,268,515,2506,447,0,802,463,203,765,4,2,303,135,0,0,2,2517,899,846,732,3793,976,0,2915,366,3192,1023,1951,127,4,1450,0,107,53,0,648,376,856,101,637,3444,6,0,488,34,11,15,211,188,38,48,0,483,13,307,98,49,1,47,399,17,1116,0,0,13817,1,104,6,0,0,5769,1,0,0,0,0,304,83,0,195,16,365,12,258,0,72,8,25,132,80,169,289,0,15,16024,1,0],[339,294,69,1399,26,1398,10,31,0,1716,4449,4861,28,0,1697,25,4793,0,1451,1201,418,0,148,27,250,866,42,769,542,87,321,0,0,274,1347,0,5,12836,402,467,1,479,23,4,102,0,0,259,1524,555,264,142,0,0,9000,0,123,2109,1009,1113,3,3,2099,2544,414,31,11,30,37,1979,5189,1275,65,2469,38,104,313,40,2,34,959,1000,230,608,120,27,72,727,213,4425,1853,15,2160,365,144,384,0,21,0,472,2244,394,483,349,483,3369,544,1,695,507,187,655,2,6,153,132,0,1,0,2698,1024,865,645,3255,964,0,3402,342,3014,978,1706,157,2,1065,0,178,111,1,440,470,870,135,809,2760,24,0,489,64,9,13,267,172,43,43,0,456,13,394,121,65,8,331,218,17,1428,0,1,13827,1,104,6,0,0,6038,0,0,1,0,0,865,41,1,286,19,1477,21,275,0,82,8,25,179,191,168,365,0,6,15871,2,2],[1185,294,63,2038,63,909,10,64,0,602,2469,5652,38,0,915,19,1601,1,1370,850,465,0,21,15,368,547,23,767,461,94,438,0,0,291,959,0,4,5363,2667,517,10,579,42,17,22,0,0,263,400,844,328,141,6,1,4742,0,215,1293,761,807,0,11,1411,4843,2017,42,13,14,50,1071,1628,1075,124,3004,82,146,13221,226,13,59,767,1838,338,1608,99,7,39,277,117,2245,1657,12,2147,384,109,254,0,24,0,344,1437,467,315,484,526,2546,232,0,806,343,130,686,2,6,340,168,2,2,0,2875,1190,1239,678,5816,919,0,3294,331,2339,782,3263,134,1,382,0,746,157,1,789,3303,582,143,370,2581,42,1,590,68,8,14,537,270,32,781,0,831,26,439,118,114,32,1244,474,78,1007,0,0,35482,2,684,4,1,19,5124,0,0,7,0,0,51,5,0,61,5,443,19,308,0,116,4,24,93,133,475,691,0,21,17511,0,1],[861,297,190,1764,40,1150,4,24,0,924,3411,3829,31,0,901,8,3028,0,1168,914,375,1,73,19,291,610,12,679,546,20,453,0,0,248,1002,0,12,8487,1295,544,2,335,38,5,22,0,0,173,658,464,294,159,1,0,5118,0,316,2069,697,897,1,5,1817,2780,767,35,2,5,24,1386,4496,738,83,2235,55,81,2635,91,6,41,642,1519,310,894,88,25,78,589,179,2391,1552,11,1813,280,90,247,0,14,0,539,1571,367,367,323,521,2998,309,0,803,333,173,473,2,7,233,178,0,2,0,2358,837,999,575,4360,836,0,2779,273,2272,685,2064,131,2,701,0,685,177,3,537,961,792,80,427,2259,39,0,495,79,6,12,324,118,21,246,0,612,17,337,112,64,19,530,407,38,1123,0,1,27669,1,218,7,0,4,4146,0,0,1,0,0,153,17,1,224,13,510,34,236,0,87,10,30,156,110,243,387,0,3,12786,8,0],[2141,529,290,4308,71,1674,3,71,0,971,3969,7384,47,0,2748,30,5419,1,2103,1145,511,0,47,30,516,953,26,1028,941,104,683,0,0,354,2190,0,4,16385,1892,768,13,785,29,15,30,0,0,485,1151,846,493,281,3,1,11323,0,288,2592,1011,1379,0,6,2204,5832,1049,56,17,45,85,2982,4825,1513,220,5593,80,162,4248,201,13,97,1182,2218,470,1967,223,35,107,920,369,3022,2078,28,3230,587,142,509,0,22,0,1006,1916,738,552,550,870,4648,471,0,1460,571,231,880,2,8,565,346,0,1,0,4836,1351,1924,1158,8165,1794,0,4991,560,4036,962,3048,170,12,1835,0,441,142,3,803,1368,1236,204,892,5357,29,0,731,71,20,12,636,358,45,194,0,954,27,503,156,110,16,263,980,76,1445,0,0,44059,1,719,5,0,2,6821,1,0,12,0,0,149,9,0,143,9,612,21,387,1,151,3,20,166,175,445,679,0,13,26651,0,2],[1370,575,519,3800,93,2401,14,47,0,1841,5859,6912,84,0,2380,25,6432,1,2110,1603,415,0,59,28,585,931,32,954,769,37,876,0,1,443,2415,0,15,20227,530,693,17,600,20,14,23,0,1,264,1460,470,596,341,1,3,8181,0,152,3200,1027,1773,0,1,3000,3506,236,36,12,26,85,3948,9477,1164,154,4766,52,111,549,57,5,69,1206,2792,611,1335,219,35,148,1514,477,3720,2197,18,2968,437,178,504,0,23,0,985,2425,710,594,410,679,5112,649,0,1633,611,285,851,5,12,520,392,0,4,0,4002,1203,1706,1381,4740,1366,0,4650,588,4974,1118,1735,228,1,1719,0,263,50,0,643,529,2115,174,1023,5301,33,1,553,111,23,14,129,292,51,50,0,448,26,324,114,52,2,95,716,32,1732,0,0,32119,0,260,8,0,0,6977,1,2,1,0,0,282,39,0,403,10,604,21,409,0,163,21,33,249,128,218,563,0,11,18362,0,1],[310,104,79,983,37,1013,321,14,0,639,2029,2185,21,0,950,1,1433,0,771,500,243,0,18,10,86,433,10,333,292,59,172,0,0,152,1004,0,0,9085,344,145,8,279,5,0,47,0,0,230,751,260,205,122,1,0,1426,0,63,975,264,690,0,15,1208,1743,195,25,8,32,17,1697,3967,626,71,1235,26,45,686,21,1,38,500,623,115,477,53,4,31,280,151,2431,808,3,732,440,75,200,0,17,0,243,546,210,190,82,301,1508,268,0,579,226,123,447,12,0,249,154,0,3,0,2250,337,665,588,1464,383,0,831,214,1676,229,900,61,2,470,0,116,26,0,381,250,521,70,390,2247,6,0,193,45,11,10,93,57,5,61,0,77,3,195,45,36,1,54,131,58,630,0,0,12780,0,91,0,0,0,3297,0,0,10,0,0,44,4,0,83,9,106,7,105,0,55,7,59,93,69,97,143,0,1,6524,0,6],[562,359,136,2724,43,3321,770,125,0,2603,6336,5860,109,0,2097,40,4947,3,2656,2554,607,0,194,35,417,931,29,830,680,119,538,0,0,454,959,0,14,21880,327,635,19,987,19,11,141,0,0,424,2018,732,908,543,0,0,3016,0,56,3829,664,2012,2,10,4135,3390,163,47,15,33,28,2729,15812,1285,185,2019,79,140,193,36,6,78,1306,1125,704,998,174,42,143,1115,380,6592,2814,13,2639,701,228,641,0,39,0,284,1356,538,620,307,775,3934,788,0,1440,546,475,974,4,26,874,516,0,4,1,3694,891,2043,2586,6195,1168,0,4203,586,5586,1177,1880,242,1,352,0,109,41,2,1151,398,2854,180,1208,6661,32,1,488,94,37,23,124,162,29,58,0,157,18,471,111,69,8,48,168,45,1581,0,0,21376,0,93,3,0,0,7496,1,1,4,0,0,283,25,2,397,23,243,18,335,2,96,44,98,255,172,184,380,0,11,16012,5,3],[1269,326,91,3359,86,1836,1207,161,0,887,2774,6327,32,0,1179,31,3638,0,2312,1475,512,0,20,22,314,711,18,827,933,318,513,0,0,400,956,0,12,16928,3228,648,24,1018,79,15,74,0,0,483,606,975,414,210,14,0,4146,0,459,2787,544,996,25,40,1892,5389,1238,77,22,26,80,1354,2461,1182,203,4382,132,197,7587,99,19,95,1103,2102,623,1829,149,16,61,750,323,3257,2369,23,2229,1039,146,321,0,18,0,393,1533,477,463,439,622,3349,368,0,1129,342,237,937,8,17,624,323,1,1,0,4097,1371,2093,1103,7531,1486,0,4323,545,2702,990,3080,183,1,871,0,1091,106,0,1208,1009,699,199,1040,5809,26,0,606,32,15,14,316,156,39,326,0,208,77,621,118,130,16,558,269,644,1655,0,0,44136,2,488,5,0,9,6928,0,0,19,0,0,95,2,1,61,3,462,26,485,0,73,12,22,127,71,414,800,0,23,20836,0,2],[1424,372,164,2245,40,1732,651,141,0,984,3166,2761,52,0,665,19,3642,3,1743,1350,605,0,48,5,317,521,13,584,992,31,626,0,1,291,615,0,9,13377,2505,609,6,604,45,3,46,0,0,265,619,780,440,343,2,1,2716,0,849,3391,338,863,2,29,1879,4161,1130,11,6,11,20,758,5843,667,242,2619,85,119,6642,69,5,46,860,1629,621,930,118,34,95,1039,340,2671,1879,4,2061,443,80,214,0,21,0,197,1485,409,405,274,417,2845,401,0,922,279,201,569,2,13,398,363,0,1,0,3108,788,1993,857,6542,940,0,2842,371,2495,541,1857,151,0,108,0,1125,225,2,856,879,1566,119,701,4391,49,0,460,82,23,11,278,103,22,145,0,155,38,453,156,111,18,418,324,328,1613,0,0,25994,0,316,4,0,8,4379,3,1,31,0,0,202,12,0,176,1,862,25,311,0,71,20,43,100,91,287,731,0,4,9236,0,0],[1017,143,280,1076,37,997,254,35,0,406,1864,2513,21,0,1283,5,1318,0,694,483,339,0,13,11,95,320,12,354,214,87,219,0,0,116,877,0,4,6539,910,213,4,207,11,3,38,0,0,200,423,400,195,104,2,0,3821,0,262,1140,271,560,5,5,991,2392,764,36,12,14,44,1964,3898,482,74,2012,32,53,2174,35,3,35,426,1745,165,510,60,6,44,312,144,1927,812,4,785,302,39,127,0,3,0,572,504,192,174,112,197,1446,201,0,506,270,121,392,2,0,192,172,0,0,0,1991,435,757,582,1801,419,0,1132,220,1437,164,1305,69,6,1180,0,208,36,1,361,496,675,64,265,1816,2,0,197,31,15,3,80,49,12,132,0,182,62,211,40,35,1,189,344,84,607,0,0,15287,0,120,3,0,4,3287,0,0,2,0,0,31,4,1,39,1,188,3,137,0,38,8,25,79,31,91,271,0,3,5998,0,3],[1571,366,1006,2326,54,2296,386,46,0,940,3444,4968,96,0,1683,17,2811,1,1556,1456,295,0,30,26,477,420,10,487,440,75,590,0,0,234,1549,1,6,10184,739,574,4,483,10,12,30,0,0,221,478,425,299,182,0,2,11223,1,341,3426,412,986,18,2,1519,2954,467,14,8,23,41,2446,14560,470,130,3008,75,92,1317,38,1,70,692,6477,709,1003,110,20,93,778,259,2095,1575,10,1649,377,101,220,0,16,0,434,958,418,344,227,316,1867,400,0,1030,277,212,471,2,14,420,272,0,2,0,2323,793,1495,1417,3456,867,0,3361,405,3729,619,1603,217,5,1165,0,344,229,1,393,378,2219,82,473,3445,8,0,382,78,26,7,46,88,29,33,0,271,140,319,88,45,10,160,840,57,823,0,0,21832,1,98,1,0,1,4802,0,0,2,0,0,105,5,0,90,3,258,13,390,1,66,13,34,100,76,126,568,0,5,10817,0,0]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>M1A<\/th>\n      <th>M1B<\/th>\n      <th>A1A<\/th>\n      <th>A1B<\/th>\n      <th>V1A<\/th>\n      <th>V1B<\/th>\n      <th>M6A<\/th>\n      <th>M6B<\/th>\n      <th>A6A<\/th>\n      <th>A6B<\/th>\n      <th>V6A<\/th>\n      <th>V6B<\/th>\n      <th>M12A<\/th>\n      <th>M12B<\/th>\n      <th>A12A<\/th>\n      <th>A12B<\/th>\n      <th>V12A<\/th>\n      <th>V12B<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"autoWidth":true,"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18]},{"orderable":false,"targets":0}],"order":[],"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<p>A data slice of RPKM table (<code>rpkmDFeByg.xls</code>) is shown here.</p>
<pre><code class="language-r">read.delim(&quot;results/rpkmDFeByg.xls&quot;, row.names = 1, check.names = FALSE)[1:4,
    1:4]
</code></pre>
<pre><code>##                 M1A       M1B       A1A       A1B
## AT1G01010 15.552350 15.855557 15.515099 11.482534
## AT1G01020  5.663586  8.550121  5.550872  8.069877
## AT1G01030  6.294920  4.918521  6.180994  3.092568
## AT1G01040 13.909390 12.846007 11.638283 11.304143
</code></pre>
<p>Note, for most statistical differential expression or abundance analysis
methods, such as <code>edgeR</code> or <code>DESeq2</code>, the raw count values should be used as input. The
usage of RPKM values should be restricted to specialty applications
required by some users, <em>e.g.</em> manually comparing the expression levels
among different genes or features.</p>
<h4 id="sample-wise-correlation-analysis">Sample-wise correlation analysis</h4>
<p>The following computes the sample-wise Spearman correlation coefficients from
the <code>rlog</code> transformed expression values generated with the <code>DESeq2</code> package. After
transformation to a distance matrix, hierarchical clustering is performed with
the <code>hclust</code> function and the result is plotted as a dendrogram
(also see file <code>sample_tree.pdf</code>).</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(DESeq2, quietly = TRUE)
    library(ape, warn.conflicts = FALSE)
    ## Extracting SummarizedExperiment object
    se &lt;- SE(sal, &quot;read_counting&quot;)
    dds &lt;- DESeqDataSet(se, design = ~condition)
    d &lt;- cor(assay(rlog(dds)), method = &quot;spearman&quot;)
    hc &lt;- hclust(dist(1 - d))
    pdf(&quot;results/sample_tree.pdf&quot;)
    plot.phylo(as.phylo(hc), type = &quot;p&quot;, edge.col = &quot;blue&quot;, edge.width = 2,
        show.node.label = TRUE, no.margin = TRUE)
    dev.off()
}, step_name = &quot;sample_tree&quot;, dependency = &quot;read_counting&quot;)
</code></pre>
<p><img src="../results/sample_tree.png" alt=""></p>
<div data-align="center">
<p>Figure 2: Correlation dendrogram of samples</p>
</div>
</br>
<h3 id="analysis-of-degs">Analysis of DEGs</h3>
<p>The analysis of differentially expressed genes (DEGs) is performed with
the glm method of the <code>edgeR</code> package (Robinson, McCarthy, and Smyth 2010). The sample
comparisons used by this analysis are defined in the header lines of the
<code>targets.txt</code> file starting with <code>&lt;CMP&gt;</code>.</p>
<h4 id="run-edger">Run <code>edgeR</code></h4>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(edgeR)
    countDF &lt;- read.delim(&quot;results/countDFeByg.xls&quot;, row.names = 1,
        check.names = FALSE)
    cmp &lt;- readComp(stepsWF(sal)[[&quot;hisat2_mapping&quot;]], format = &quot;matrix&quot;,
        delim = &quot;-&quot;)
    edgeDF &lt;- run_edgeR(countDF = countDF, targets = targetsWF(sal)[[&quot;hisat2_mapping&quot;]],
        cmp = cmp[[1]], independent = FALSE, mdsplot = &quot;&quot;)
}, step_name = &quot;run_edger&quot;, dependency = &quot;read_counting&quot;)
</code></pre>
<h4 id="add-gene-descriptions">Add gene descriptions</h4>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(&quot;biomaRt&quot;)
    m &lt;- useMart(&quot;plants_mart&quot;, dataset = &quot;athaliana_eg_gene&quot;,
        host = &quot;https://plants.ensembl.org&quot;)
    desc &lt;- getBM(attributes = c(&quot;tair_locus&quot;, &quot;description&quot;),
        mart = m)
    desc &lt;- desc[!duplicated(desc[, 1]), ]
    descv &lt;- as.character(desc[, 2])
    names(descv) &lt;- as.character(desc[, 1])
    edgeDF &lt;- data.frame(edgeDF, Desc = descv[rownames(edgeDF)],
        check.names = FALSE)
    write.table(edgeDF, &quot;./results/edgeRglm_allcomp.xls&quot;, quote = FALSE,
        sep = &quot;\t&quot;, col.names = NA)
}, step_name = &quot;custom_annot&quot;, dependency = &quot;run_edger&quot;)
</code></pre>
<h3 id="plot-deg-results">Plot DEG results</h3>
<p>Filter and plot DEG results for up and down regulated genes. The
definition of <em>up</em> and <em>down</em> is given in the corresponding help
file. To open it, type <code>?filterDEGs</code> in the R console.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    edgeDF &lt;- read.delim(&quot;results/edgeRglm_allcomp.xls&quot;, row.names = 1,
        check.names = FALSE)
    pdf(&quot;results/DEGcounts.pdf&quot;)
    DEG_list &lt;- filterDEGs(degDF = edgeDF, filter = c(Fold = 2,
        FDR = 20))
    dev.off()
    write.table(DEG_list$Summary, &quot;./results/DEGcounts.xls&quot;,
        quote = FALSE, sep = &quot;\t&quot;, row.names = FALSE)
}, step_name = &quot;filter_degs&quot;, dependency = &quot;custom_annot&quot;)
</code></pre>
<p><img src="../results/DEGcounts.png" alt=""></p>
<div data-align="center">
<p>Figure 3: Up and down regulated DEGs with FDR of 1%</p>
</div>
</br>
<h3 id="venn-diagrams-of-deg-sets">Venn diagrams of DEG sets</h3>
<p>The <code>overLapper</code> function can compute Venn intersects for large numbers of sample
sets (up to 20 or more) and plots 2-5 way Venn diagrams. A useful
feature is the possibility to combine the counts from several Venn
comparisons with the same number of sample sets in a single Venn diagram
(here for 4 up and down DEG sets).</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    vennsetup &lt;- overLapper(DEG_list$Up[6:9], type = &quot;vennsets&quot;)
    vennsetdown &lt;- overLapper(DEG_list$Down[6:9], type = &quot;vennsets&quot;)
    pdf(&quot;results/vennplot.pdf&quot;)
    vennPlot(list(vennsetup, vennsetdown), mymain = &quot;&quot;, mysub = &quot;&quot;,
        colmode = 2, ccol = c(&quot;blue&quot;, &quot;red&quot;))
    dev.off()
}, step_name = &quot;venn_diagram&quot;, dependency = &quot;filter_degs&quot;)
</code></pre>
<p><img src="../results/vennplot.png" alt=""></p>
<div data-align="center">
<p>Figure 4: Venn Diagram for 4 Up and Down DEG Sets</p>
</div>
</br>
<h3 id="go-term-enrichment-analysis">GO term enrichment analysis</h3>
<h4 id="obtain-gene-to-go-mappings">Obtain gene-to-GO mappings</h4>
<p>The following shows how to obtain gene-to-GO mappings from <code>biomaRt</code> (here for <em>A.
thaliana</em>) and how to organize them for the downstream GO term
enrichment analysis. Alternatively, the gene-to-GO mappings can be
obtained for many organisms from Bioconductor’s <code>*.db</code> genome annotation
packages or GO annotation files provided by various genome databases.
For each annotation this relatively slow preprocessing step needs to be
performed only once. Subsequently, the preprocessed data can be loaded
with the <code>load</code> function as shown in the next subsection.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(&quot;biomaRt&quot;)
    # listMarts() # To choose BioMart database
    # listMarts(host='plants.ensembl.org')
    m &lt;- useMart(&quot;plants_mart&quot;, host = &quot;https://plants.ensembl.org&quot;)
    m &lt;- useMart(&quot;plants_mart&quot;, dataset = &quot;athaliana_eg_gene&quot;,
        host = &quot;https://plants.ensembl.org&quot;)
    go &lt;- getBM(attributes = c(&quot;go_id&quot;, &quot;tair_locus&quot;, &quot;namespace_1003&quot;),
        mart = m)
    go &lt;- go[go[, 3] != &quot;&quot;, ]
    go[, 3] &lt;- as.character(go[, 3])
    go[go[, 3] == &quot;molecular_function&quot;, 3] &lt;- &quot;F&quot;
    go[go[, 3] == &quot;biological_process&quot;, 3] &lt;- &quot;P&quot;
    go[go[, 3] == &quot;cellular_component&quot;, 3] &lt;- &quot;C&quot;
    go[1:4, ]
    if (!dir.exists(&quot;./data/GO&quot;))
        dir.create(&quot;./data/GO&quot;)
    write.table(go, &quot;data/GO/GOannotationsBiomart_mod.txt&quot;, quote = FALSE,
        row.names = FALSE, col.names = FALSE, sep = &quot;\t&quot;)
    catdb &lt;- makeCATdb(myfile = &quot;data/GO/GOannotationsBiomart_mod.txt&quot;,
        lib = NULL, org = &quot;&quot;, colno = c(1, 2, 3), idconv = NULL)
    save(catdb, file = &quot;data/GO/catdb.RData&quot;)
}, step_name = &quot;get_go_annot&quot;, dependency = &quot;filter_degs&quot;)
</code></pre>
<h4 id="batch-go-term-enrichment-analysis">Batch GO term enrichment analysis</h4>
<p>Apply the enrichment analysis to the DEG sets obtained the above differential
expression analysis. Note, in the following example the <code>FDR</code> filter is set
here to an unreasonably high value, simply because of the small size of the toy
data set used in this vignette. Batch enrichment analysis of many gene sets is
performed with the function. When <code>method=all</code>, it returns all GO terms passing
the p-value cutoff specified under the <code>cutoff</code> arguments. When <code>method=slim</code>,
it returns only the GO terms specified under the <code>myslimv</code> argument. The given
example shows how a GO slim vector for a specific organism can be obtained from
<code>BioMart</code>.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(&quot;biomaRt&quot;)
    load(&quot;data/GO/catdb.RData&quot;)
    DEG_list &lt;- filterDEGs(degDF = edgeDF, filter = c(Fold = 2,
        FDR = 50), plot = FALSE)
    up_down &lt;- DEG_list$UporDown
    names(up_down) &lt;- paste(names(up_down), &quot;_up_down&quot;, sep = &quot;&quot;)
    up &lt;- DEG_list$Up
    names(up) &lt;- paste(names(up), &quot;_up&quot;, sep = &quot;&quot;)
    down &lt;- DEG_list$Down
    names(down) &lt;- paste(names(down), &quot;_down&quot;, sep = &quot;&quot;)
    DEGlist &lt;- c(up_down, up, down)
    DEGlist &lt;- DEGlist[sapply(DEGlist, length) &gt; 0]
    BatchResult &lt;- GOCluster_Report(catdb = catdb, setlist = DEGlist,
        method = &quot;all&quot;, id_type = &quot;gene&quot;, CLSZ = 2, cutoff = 0.9,
        gocats = c(&quot;MF&quot;, &quot;BP&quot;, &quot;CC&quot;), recordSpecGO = NULL)
    m &lt;- useMart(&quot;plants_mart&quot;, dataset = &quot;athaliana_eg_gene&quot;,
        host = &quot;https://plants.ensembl.org&quot;)
    goslimvec &lt;- as.character(getBM(attributes = c(&quot;goslim_goa_accession&quot;),
        mart = m)[, 1])
    BatchResultslim &lt;- GOCluster_Report(catdb = catdb, setlist = DEGlist,
        method = &quot;slim&quot;, id_type = &quot;gene&quot;, myslimv = goslimvec,
        CLSZ = 10, cutoff = 0.01, gocats = c(&quot;MF&quot;, &quot;BP&quot;, &quot;CC&quot;),
        recordSpecGO = NULL)
    write.table(BatchResultslim, &quot;results/GOBatchSlim.xls&quot;, row.names = FALSE,
        quote = FALSE, sep = &quot;\t&quot;)
}, step_name = &quot;go_enrich&quot;, dependency = &quot;get_go_annot&quot;)
</code></pre>
<p>Shows GO term enrichment results from previous step. The last gene identifier column (10)
of this table has been excluded in this viewing instance to minimize the complexity of the
result.
To avoid slowdowns of the load time of this page, only 10 rows of the source table are shown
below.</p>
<pre><code class="language-r">BatchResult &lt;- read.delim(&quot;results/GOBatchAll.xls&quot;)[1:10, ]
knitr::kable(BatchResult[, -10])
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left">CLID</th>
<th style="text-align:right">CLSZ</th>
<th style="text-align:left">GOID</th>
<th style="text-align:right">NodeSize</th>
<th style="text-align:right">SampleMatch</th>
<th style="text-align:right">Phyper</th>
<th style="text-align:right">Padj</th>
<th style="text-align:left">Term</th>
<th style="text-align:left">Ont</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0050291</td>
<td style="text-align:right">4</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0039621</td>
<td style="text-align:right">0.0396207</td>
<td style="text-align:left">sphingosine N-acyltransferase activity</td>
<td style="text-align:left">MF</td>
</tr>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0004345</td>
<td style="text-align:right">6</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0059375</td>
<td style="text-align:right">0.0593750</td>
<td style="text-align:left">glucose-6-phosphate dehydrogenase activity</td>
<td style="text-align:left">MF</td>
</tr>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0050664</td>
<td style="text-align:right">11</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0108597</td>
<td style="text-align:right">0.1085975</td>
<td style="text-align:left">oxidoreductase activity, acting on NAD(P)H, oxygen as acceptor</td>
<td style="text-align:left">MF</td>
</tr>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0052593</td>
<td style="text-align:right">11</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0108597</td>
<td style="text-align:right">0.1085975</td>
<td style="text-align:left">tryptamine:oxygen oxidoreductase (deaminating) activity</td>
<td style="text-align:left">MF</td>
</tr>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0052594</td>
<td style="text-align:right">11</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0108597</td>
<td style="text-align:right">0.1085975</td>
<td style="text-align:left">aminoacetone:oxygen oxidoreductase(deaminating) activity</td>
<td style="text-align:left">MF</td>
</tr>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0052595</td>
<td style="text-align:right">11</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0108597</td>
<td style="text-align:right">0.1085975</td>
<td style="text-align:left">aliphatic-amine oxidase activity</td>
<td style="text-align:left">MF</td>
</tr>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0052596</td>
<td style="text-align:right">11</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0108597</td>
<td style="text-align:right">0.1085975</td>
<td style="text-align:left">phenethylamine:oxygen oxidoreductase (deaminating) activity</td>
<td style="text-align:left">MF</td>
</tr>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0052793</td>
<td style="text-align:right">12</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0118414</td>
<td style="text-align:right">0.1184141</td>
<td style="text-align:left">pectin acetylesterase activity</td>
<td style="text-align:left">MF</td>
</tr>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0008131</td>
<td style="text-align:right">15</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0147808</td>
<td style="text-align:right">0.1478083</td>
<td style="text-align:left">primary amine oxidase activity</td>
<td style="text-align:left">MF</td>
</tr>
<tr>
<td style="text-align:left">M1-A1_up_down</td>
<td style="text-align:right">26</td>
<td style="text-align:left">GO:0016018</td>
<td style="text-align:right">16</td>
<td style="text-align:right">1</td>
<td style="text-align:right">0.0157588</td>
<td style="text-align:right">0.1575878</td>
<td style="text-align:left">cyclosporin A binding</td>
<td style="text-align:left">MF</td>
</tr>
</tbody>
</table>
<h4 id="plot-batch-go-term-results">Plot batch GO term results</h4>
<p>The <code>data.frame</code> generated by <code>GOCluster</code> can be plotted with the <code>goBarplot</code> function. Because of the
variable size of the sample sets, it may not always be desirable to show
the results from different DEG sets in the same bar plot. Plotting
single sample sets is achieved by subsetting the input data frame as
shown in the first line of the following example.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    gos &lt;- BatchResultslim[grep(&quot;M6-V6_up_down&quot;, BatchResultslim$CLID),
        ]
    gos &lt;- BatchResultslim
    png(&quot;results/GOslimbarplotMF.png&quot;)
    goBarplot(gos, gocat = &quot;MF&quot;)
    dev.off()
    png(&quot;results/GOslimbarplotBP.png&quot;)
    goBarplot(gos, gocat = &quot;BP&quot;)
    dev.off()
    png(&quot;results/GOslimbarplotCC.png&quot;)
    goBarplot(gos, gocat = &quot;CC&quot;)
    dev.off()
}, step_name = &quot;go_plot&quot;, dependency = &quot;go_enrich&quot;)
</code></pre>
<p><img src="../results/GOslimbarplotMF.png" alt=""></p>
<div data-align="center">
<p>Figure 5: GO Slim Barplot for MF Ontology</p>
</div>
</br>
<h3 id="clustering-and-heat-maps">Clustering and heat maps</h3>
<p>The following example performs hierarchical clustering on the <code>rlog</code>
transformed expression matrix subsetted by the DEGs identified in the above
differential expression analysis. It uses a Pearson correlation-based distance
measure and complete linkage for cluster joining.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(pheatmap)
    geneids &lt;- unique(as.character(unlist(DEG_list[[1]])))
    y &lt;- assay(rlog(dds))[geneids, ]
    pdf(&quot;results/heatmap1.pdf&quot;)
    pheatmap(y, scale = &quot;row&quot;, clustering_distance_rows = &quot;correlation&quot;,
        clustering_distance_cols = &quot;correlation&quot;)
    dev.off()
}, step_name = &quot;heatmap&quot;, dependency = &quot;go_enrich&quot;)
</code></pre>
<p><img src="../results/heatmap1.png" alt=""></p>
<div data-align="center">
<p>Figure 6: Heat Map with Hierarchical Clustering Dendrograms of DEGs</p>
</div>
</br>
<h3 id="version-information">Version Information</h3>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    sessionInfo()
}, step_name = &quot;sessionInfo&quot;, dependency = &quot;heatmap&quot;)
</code></pre>
<h2 id="running-workflow">Running workflow</h2>
<h3 id="interactive-job-submissions-in-a-single-machine">Interactive job submissions in a single machine</h3>
<p>For running the workflow, <code>runWF</code> function will execute all the steps store in
the <code>SYSargsList</code> workflow container. The execution will be on a single machine without
submitting to a queuing system of a computer cluster. Besides, <code>runWF</code> allows the
user to create a dedicated results folder for each workflow. This includes all the
output and log files for each step. When these options are used, the output
location will be updated by default and can be assigned to the same object.</p>
<pre><code class="language-r">sal &lt;- runWF(sal)
</code></pre>
<h3 id="parallelization-on-clusters">Parallelization on clusters</h3>
<p>Alternatively, the computation can be greatly accelerated by processing many files
in parallel using several compute nodes of a cluster, where a scheduling/queuing
system is used for load balancing.</p>
<p>The <code>resources</code> list object provides the number of independent parallel cluster
processes defined under the <code>Njobs</code> element in the list. The following example
will run 18 processes in parallel using each 4 CPU cores.
If the resources available on a cluster allow running all 18 processes at the
same time, then the shown sample submission will utilize in a total of 72 CPU cores.</p>
<p>Note, <code>runWF</code> can be used with most queueing systems as it is based on utilities
from the <code>batchtools</code> package, which supports the use of template files (<em><code>*.tmpl</code></em>)
for defining the run parameters of different schedulers. To run the following
code, one needs to have both a <code>conffile</code> (see <em><code>.batchtools.conf.R</code></em> samples <a href="https://mllg.github.io/batchtools/">here</a>)
and a <code>template</code> file (see <em><code>*.tmpl</code></em> samples <a href="https://github.com/mllg/batchtools/tree/master/inst/templates">here</a>)
for the queueing available on a system. The following example uses the sample
<code>conffile</code> and <code>template</code> files for the Slurm scheduler provided by this package.</p>
<p>The resources can be appended when the step is generated, or it is possible to
add these resources later, as the following example using the <code>addResources</code>
function:</p>
<pre><code class="language-r">resources &lt;- list(conffile=&quot;.batchtools.conf.R&quot;,
                  template=&quot;batchtools.slurm.tmpl&quot;, 
                  Njobs=18, 
                  walltime=120, ## minutes
                  ntasks=1,
                  ncpus=4, 
                  memory=1024, ## Mb
                  partition = &quot;short&quot;
                  )
sal &lt;- addResources(sal, step = c(&quot;hisat2_mapping&quot;), resources = resources)
sal &lt;- runWF(sal)
</code></pre>
<h3 id="visualize-workflow">Visualize workflow</h3>
<p><em><code>systemPipeR</code></em> workflows instances can be visualized with the <code>plotWF</code> function.</p>
<pre><code class="language-r">plotWF(sal, out_format = &quot;html&quot;, out_path = &quot;plotWF.html&quot;)
</code></pre>
<h2 id="checking-workflow-status">Checking workflow status</h2>
<p>To check the summary of the workflow, we can use:</p>
<pre><code class="language-r">sal
statusWF(sal)
</code></pre>
<h2 id="accessing-logs-report">Accessing logs report</h2>
<p><em><code>systemPipeR</code></em> compiles all the workflow execution logs in one central location,
making it easier to check any standard output (<code>stdout</code>) or standard error
(<code>stderr</code>) for any command-line tools used on the workflow or the R code stdout.</p>
<pre><code class="language-r">sal &lt;- renderLogs(sal)
</code></pre>
<h2 id="funding">Funding</h2>
<p>This project is funded by NSF award <a href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1661152">ABI-1661152</a>.</p>
<h2 id="references">References</h2>
<div id="refs" class="references">
<div id="ref-H_Backman2016-bt">
<p>H Backman, Tyler W, and Thomas Girke. 2016. “systemPipeR: NGS workflow and report generation environment.” <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">https://doi.org/10.1186/s12859-016-1241-0</a>.</p>
</div>
<div id="ref-Howard2013-fq">
<p>Howard, Brian E, Qiwen Hu, Ahmet Can Babaoglu, Manan Chandra, Monica Borghi, Xiaoping Tan, Luyan He, et al. 2013. “High-Throughput RNA Sequencing of Pseudomonas-Infected Arabidopsis Reveals Hidden Transcriptome Complexity and Novel Splice Variants.” <em>PLoS One</em> 8 (10): e74183. <a href="https://doi.org/10.1371/journal.pone.0074183">https://doi.org/10.1371/journal.pone.0074183</a>.</p>
</div>
<div id="ref-Kim2015-ve">
<p>Kim, Daehwan, Ben Langmead, and Steven L Salzberg. 2015. “HISAT: A Fast Spliced Aligner with Low Memory Requirements.” <em>Nat. Methods</em> 12 (4): 357–60.</p>
</div>
<div id="ref-Lawrence2013-kt">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8): e1003118. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.</p>
</div>
<div id="ref-Robinson2010-uk">
<p>Robinson, M D, D J McCarthy, and G K Smyth. 2010. “EdgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2022-04-21: <a  href="https://github.com/tgirke/GEN242//commit/2fb8c0daf4cbdd8f8dc6542a83c79128c1b2533e">some edits (2fb8c0da)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2022 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
