<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../../favicons/android-192x192.png" sizes="192x192">

<title>RNA-Seq Workflow Template | GEN242</title><meta property="og:title" content="RNA-Seq Workflow Template" />
<meta property="og:description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Introduction Users want to provide here background information about the design of their RNA-Seq project.
Samples and environment settings Environment settings and input data systemPipeRdata package is a helper package to generate a fully populated systemPipeR workflow environment in the current working directory with a single command." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/systempiper/rnaseq/systempipernaseq/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2023-06-20T15:15:01-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="RNA-Seq Workflow Template">
<meta itemprop="description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Introduction Users want to provide here background information about the design of their RNA-Seq project.
Samples and environment settings Environment settings and input data systemPipeRdata package is a helper package to generate a fully populated systemPipeR workflow environment in the current working directory with a single command.">
<meta itemprop="dateModified" content="2023-06-20T15:15:01-07:00" />
<meta itemprop="wordCount" content="3898">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="RNA-Seq Workflow Template"/>
<meta name="twitter:description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Introduction Users want to provide here background information about the design of their RNA-Seq project.
Samples and environment settings Environment settings and input data systemPipeRdata package is a helper package to generate a fully populated systemPipeR workflow environment in the current working directory with a single command."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../../css/prism.css"/>


<script src="../../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../../css/site.css"/>
<script src="../../../../js/site.js"></script>



  </head>
  <body class="td-page">
    <header>
      
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
        <a class="navbar-brand" href="../../../../">
		<span class="navbar-logo"><svg id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 500 500" style="enable-background:new 0 0 500 500"><g><path style="fill:#fff" d="M116.8525 421.9722c-5.7041.0-10.3442-4.3127-10.3442-9.6129V88.183c0-5.3002 4.6401-9.6117 10.3442-9.6117H320.858c3.0347.0 9.3959.5498 11.7506 2.6302l.3545.3442 58.905 63.2912c2.3101 2.491 2.9202 8.4928 2.9202 11.3184v256.2039c0 5.3002-4.6407 9.6129-10.3436 9.6129H116.8525z"/><g><g><g><path style="fill:#767676" d="M384.4445 423.2066H116.852c-6.3839.0-11.5786-4.8658-11.5786-10.8474V88.1831c0-5.9804 5.1947-10.8461 11.5786-10.8461h204.0062c.377.0 9.2786.0329 12.568 2.9389l.3947.3833 58.9508 63.337c3.2135 3.4652 3.2514 11.7924 3.2514 12.1593v256.2036C396.0231 418.3408 390.8284 423.2066 384.4445 423.2066zM116.5079 411.9189c.0848.0278.1999.0531.3441.0531h267.5925c.1442.0.2581-.0253.3441-.0531V156.1556c-.0076-.9033-.3593-3.7347-.7034-5.0037l-57.6527-61.9416c-1.4651-.3176-4.4533-.6389-5.5742-.6389H116.852c-.143.0-.2594.024-.3441.0531V411.9189zm267.4533-261.149zM327.0321 89.371v.0013V89.371z"/></g></g></g><g><g><path style="fill:#5b7fc0" d="M189.0874 210.1754l.0012-.0012c7.7751.0012 15.0295 4.1862 18.932 10.9234 1.9177 3.3159 2.9305 7.1011 2.9293 10.9378.0 5.8394-2.2733 11.3304-6.4032 15.4604-4.1288 4.1288-9.6186 6.4032-15.458 6.4032s-11.328-2.2733-15.458-6.4032-6.4032-9.6186-6.4056-15.4628c.0012-6.025 2.454-11.4897 6.4116-15.4473C177.5953 212.627 183.0601 210.1742 189.0874 210.1754zm7.993 21.8576c.0012-1.4042-.3687-2.7868-1.063-3.9887-1.4293-2.4684-4.0833-3.9995-6.9299-4.0019-4.4077.0024-7.993 3.5877-7.993 7.993.0 2.1356.832 4.1431 2.3427 5.6539 1.5083 1.5083 3.5159 2.3403 5.6503 2.3415 2.1356.0 4.1443-.8308 5.6539-2.3403S197.0816 234.1722 197.0804 232.033z"/><path style="opacity:.3;fill:#fff" d="M189.0898 210.176c7.7763.0 15.0283 4.1826 18.926 10.9151 1.9201 3.3136 2.9377 7.0988 2.9353 10.9462.0024 12.0643-9.8065 21.8636-21.8613 21.8613-12.0547.0024-21.8636-9.8066-21.8612-21.8613.0-6.0285 2.4516-11.4921 6.4116-15.452C177.5977 212.6276 183.0612 210.176 189.0898 210.176zm7.9941 21.8612c0-1.4078-.3711-2.7892-1.0702-3.9959-1.4269-2.466-4.0797-3.9983-6.924-3.9983-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0024 4.4077 3.5865 7.9918 7.9942 7.9942 2.2027.0 4.2018-.8978 5.6479-2.3439C196.1861 236.239 197.0839 234.2399 197.0839 232.0372z"/><g><defs><path id="SVGID_1_" d="M194.7376 237.6875c-1.4461 1.4461-3.4452 2.3439-5.6479 2.3439-4.4077-.0024-7.9918-3.5865-7.9942-7.9942.0024-4.4125 3.5937-7.999 7.9942-7.9942 2.8443.0 5.497 1.5323 6.924 3.9983.6991 1.2067 1.0702 2.5881 1.0702 3.9959C197.0839 234.2399 196.1861 236.239 194.7376 237.6875z"/></defs><clipPath id="SVGID_2_"><use xlink:href="#SVGID_1_" style="overflow:visible"/></clipPath><path style="clip-path:url(#SVGID_2_);fill:#fff" d="M190.0704 225.0237c-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0011 1.9546.7088 3.7452 1.8782 5.1354-1.7447-1.4674-2.8575-3.663-2.8588-6.116.0024-4.4125 3.5936-7.999 7.9942-7.9942 2.3802-1e-4 4.616 1.0833 6.1218 2.8788C193.7885 225.7247 191.9774 225.0237 190.0704 225.0237z"/><path style="opacity:.13;clip-path:url(#SVGID_2_);fill:#020202" d="M190.0704 225.0237c-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0011 1.9546.7088 3.7452 1.8782 5.1354-1.7447-1.4674-2.8575-3.663-2.8588-6.116.0024-4.4125 3.5936-7.999 7.9942-7.9942 2.3802-1e-4 4.616 1.0833 6.1218 2.8788C193.7885 225.7247 191.9774 225.0237 190.0704 225.0237z"/></g><g><defs><path id="SVGID_3_" d="M189.0898 210.176c7.7763.0 15.0283 4.1826 18.926 10.9151 1.9201 3.3136 2.9377 7.0988 2.9353 10.9462.0024 12.0643-9.8065 21.8636-21.8613 21.8613-12.0547.0024-21.8636-9.8066-21.8612-21.8613.0-6.0285 2.4516-11.4921 6.4116-15.452C177.5977 212.6276 183.0612 210.176 189.0898 210.176zm7.9941 21.8612c0-1.4078-.3711-2.7892-1.0702-3.9959-1.4269-2.466-4.0797-3.9983-6.924-3.9983-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0024 4.4077 3.5865 7.9918 7.9942 7.9942 2.2027.0 4.2018-.8978 5.6479-2.3439C196.1861 236.239 197.0839 234.2399 197.0839 232.0372z"/></defs><clipPath id="SVGID_4_"><use xlink:href="#SVGID_3_" style="overflow:visible"/></clipPath><path style="clip-path:url(#SVGID_4_);fill:#5b7fc0" d="M172.6595 215.6045c-3.96 3.96-6.4116 9.4235-6.4116 15.452-.0024 12.0547 9.8066 21.8636 21.8613 21.8612 12.0547.0024 21.8636-9.797 21.8613-21.8612.0024-3.8475-1.0151-7.6326-2.9353-10.9462-3.8977-6.7324-11.1497-10.9151-18.926-10.9151C182.0806 209.1953 176.6171 211.647 172.6595 215.6045z"/></g></g><rect x="198.8952" y="225.1043" style="fill:#5b7fc0" width="122.6266" height="13.8671"/></g><g><path style="fill:#d95140" d="M189.0874 155.7611l.0012-.0012c7.7751.0012 15.0295 4.1862 18.932 10.9234 1.9177 3.3159 2.9305 7.1011 2.9293 10.9378.0 5.8394-2.2733 11.3304-6.4032 15.4604-4.1288 4.1288-9.6186 6.4032-15.458 6.4032s-11.328-2.2733-15.458-6.4032-6.4032-9.6186-6.4056-15.4628c.0012-6.0249 2.454-11.4897 6.4116-15.4473C177.5953 158.2128 183.0601 155.7599 189.0874 155.7611zm7.993 21.8577c.0012-1.4042-.3687-2.7868-1.063-3.9887-1.4293-2.4684-4.0833-3.9995-6.9299-4.0019-4.4077.0024-7.993 3.5877-7.993 7.993.0 2.1356.832 4.1431 2.3427 5.6539 1.5083 1.5083 3.5159 2.3403 5.6503 2.3415 2.1356.0 4.1443-.8308 5.6539-2.3403C196.2508 181.7667 197.0816 179.758 197.0804 177.6188z"/><path style="opacity:.3;fill:#fff" d="M189.0898 155.7617c7.7763.0 15.0283 4.1826 18.926 10.9151 1.9201 3.3135 2.9377 7.0987 2.9353 10.9462.0024 12.0643-9.8065 21.8636-21.8613 21.8613-12.0547.0024-21.8636-9.8066-21.8612-21.8613.0-6.0285 2.4516-11.4921 6.4116-15.452C177.5977 158.2134 183.0612 155.7617 189.0898 155.7617zm7.9941 21.8613c0-1.4078-.3711-2.7892-1.0702-3.9959-1.4269-2.466-4.0797-3.9983-6.924-3.9983-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0024 4.4077 3.5865 7.9918 7.9942 7.9942 2.2027.0 4.2018-.8978 5.6479-2.3439C196.1861 181.8248 197.0839 179.8256 197.0839 177.623z"/><g><defs><path id="SVGID_5_" d="M194.7376 183.2733c-1.4461 1.4461-3.4452 2.3439-5.6479 2.3439-4.4077-.0024-7.9918-3.5865-7.9942-7.9942.0024-4.4125 3.5937-7.9989 7.9942-7.9942 2.8443.0 5.497 1.5323 6.924 3.9983.6991 1.2067 1.0702 2.5881 1.0702 3.9959C197.0839 179.8256 196.1861 181.8248 194.7376 183.2733z"/></defs><clipPath id="SVGID_6_"><use xlink:href="#SVGID_5_" style="overflow:visible"/></clipPath><path style="clip-path:url(#SVGID_6_);fill:#fff" d="M190.0704 170.6095c-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0011 1.9546.7088 3.7452 1.8782 5.1354-1.7447-1.4674-2.8575-3.663-2.8588-6.116.0024-4.4125 3.5936-7.999 7.9942-7.9942 2.3802-1e-4 4.616 1.0833 6.1218 2.8788C193.7885 171.3104 191.9774 170.6095 190.0704 170.6095z"/><path style="opacity:.13;clip-path:url(#SVGID_6_);fill:#020202" d="M190.0704 170.6095c-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0011 1.9546.7088 3.7452 1.8782 5.1354-1.7447-1.4674-2.8575-3.663-2.8588-6.116.0024-4.4125 3.5936-7.999 7.9942-7.9942 2.3802-1e-4 4.616 1.0833 6.1218 2.8788C193.7885 171.3104 191.9774 170.6095 190.0704 170.6095z"/></g><g><defs><path id="SVGID_7_" d="M189.0898 155.7617c7.7763.0 15.0283 4.1826 18.926 10.9151 1.9201 3.3135 2.9377 7.0987 2.9353 10.9462.0024 12.0643-9.8065 21.8636-21.8613 21.8613-12.0547.0024-21.8636-9.8066-21.8612-21.8613.0-6.0285 2.4516-11.4921 6.4116-15.452C177.5977 158.2134 183.0612 155.7617 189.0898 155.7617zm7.9941 21.8613c0-1.4078-.3711-2.7892-1.0702-3.9959-1.4269-2.466-4.0797-3.9983-6.924-3.9983-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0024 4.4077 3.5865 7.9918 7.9942 7.9942 2.2027.0 4.2018-.8978 5.6479-2.3439C196.1861 181.8248 197.0839 179.8256 197.0839 177.623z"/></defs><clipPath id="SVGID_8_"><use xlink:href="#SVGID_7_" style="overflow:visible"/></clipPath><path style="clip-path:url(#SVGID_8_);fill:#d95140" d="M172.6595 161.1903c-3.96 3.96-6.4116 9.4235-6.4116 15.452-.0024 12.0547 9.8066 21.8636 21.8613 21.8613 12.0547.0024 21.8636-9.797 21.8613-21.8613.0024-3.8474-1.0151-7.6326-2.9353-10.9462-3.8977-6.7324-11.1497-10.9151-18.926-10.9151C182.0806 154.7811 176.6171 157.2327 172.6595 161.1903z"/></g><rect x="198.8952" y="170.69" style="fill:#d95140" width="122.6266" height="13.8671"/></g><g><g><path style="fill:#56a55c" d="M189.5379 264.6147l.0012-.0012c7.7751.0012 15.0294 4.1862 18.932 10.9235 1.9177 3.3159 2.9305 7.1011 2.9293 10.9378.0 5.8394-2.2733 11.3304-6.4032 15.4604-4.1288 4.1288-9.6186 6.4032-15.458 6.4032-5.8394.0-11.3281-2.2733-15.458-6.4032-4.13-4.13-6.4032-9.6186-6.4056-15.4628.0012-6.0249 2.454-11.4897 6.4116-15.4472C178.0458 267.0663 183.5105 264.6135 189.5379 264.6147zm7.993 21.8576c.0012-1.4042-.3687-2.7868-1.063-3.9887-1.4293-2.4684-4.0833-3.9995-6.9299-4.0019-4.4077.0024-7.993 3.5877-7.993 7.993.0 2.1356.832 4.1431 2.3427 5.6538 1.5083 1.5083 3.5159 2.3403 5.6503 2.3415 2.1356.0 4.1443-.8308 5.6539-2.3403C196.7013 290.6202 197.5321 288.6115 197.5309 286.4723z"/><path style="opacity:.3;fill:#fff" d="M189.5403 264.6153c7.7763.0 15.0283 4.1826 18.926 10.9151 1.9201 3.3135 2.9377 7.0987 2.9353 10.9462.0024 12.0643-9.8065 21.8636-21.8613 21.8613-12.0547.0024-21.8636-9.8065-21.8612-21.8613.0-6.0285 2.4516-11.492 6.4116-15.452C178.0482 267.0669 183.5117 264.6153 189.5403 264.6153zm7.9941 21.8612c0-1.4078-.3711-2.7892-1.0702-3.9959-1.4269-2.466-4.0797-3.9983-6.924-3.9983-4.4005-.0048-7.9918 3.5817-7.9942 7.9941.0024 4.4077 3.5865 7.9918 7.9942 7.9942 2.2027.0 4.2018-.8978 5.6479-2.3439C196.6366 290.6783 197.5344 288.6792 197.5344 286.4765z"/><g><defs><path id="SVGID_9_" d="M195.1881 292.1268c-1.4461 1.4461-3.4452 2.3439-5.6479 2.3439-4.4077-.0024-7.9918-3.5865-7.9942-7.9942.0024-4.4125 3.5937-7.9989 7.9942-7.9941 2.8443.0 5.497 1.5323 6.924 3.9983.6991 1.2067 1.0702 2.5881 1.0702 3.9959C197.5344 288.6792 196.6366 290.6783 195.1881 292.1268z"/></defs><clipPath id="SVGID_10_"><use xlink:href="#SVGID_9_" style="overflow:visible"/></clipPath><path style="clip-path:url(#SVGID_10_);fill:#fff" d="M190.5209 279.463c-4.4005-.0048-7.9918 3.5817-7.9942 7.9941.0011 1.9547.7088 3.7452 1.8782 5.1354-1.7446-1.4674-2.8575-3.6631-2.8588-6.1161.0024-4.4125 3.5936-7.999 7.9942-7.9941 2.3802-1e-4 4.616 1.0833 6.1218 2.8788C194.239 280.164 192.4279 279.463 190.5209 279.463z"/><path style="opacity:.13;clip-path:url(#SVGID_10_);fill:#020202" d="M190.5209 279.463c-4.4005-.0048-7.9918 3.5817-7.9942 7.9941.0011 1.9547.7088 3.7452 1.8782 5.1354-1.7446-1.4674-2.8575-3.6631-2.8588-6.1161.0024-4.4125 3.5936-7.999 7.9942-7.9941 2.3802-1e-4 4.616 1.0833 6.1218 2.8788C194.239 280.164 192.4279 279.463 190.5209 279.463z"/></g><g><defs><path id="SVGID_11_" d="M189.5403 264.6153c7.7763.0 15.0283 4.1826 18.926 10.9151 1.9201 3.3135 2.9377 7.0987 2.9353 10.9462.0024 12.0643-9.8065 21.8636-21.8613 21.8613-12.0547.0024-21.8636-9.8065-21.8612-21.8613.0-6.0285 2.4516-11.492 6.4116-15.452C178.0482 267.0669 183.5117 264.6153 189.5403 264.6153zm7.9941 21.8612c0-1.4078-.3711-2.7892-1.0702-3.9959-1.4269-2.466-4.0797-3.9983-6.924-3.9983-4.4005-.0048-7.9918 3.5817-7.9942 7.9941.0024 4.4077 3.5865 7.9918 7.9942 7.9942 2.2027.0 4.2018-.8978 5.6479-2.3439C196.6366 290.6783 197.5344 288.6792 197.5344 286.4765z"/></defs><clipPath id="SVGID_12_"><use xlink:href="#SVGID_11_" style="overflow:visible"/></clipPath><path style="clip-path:url(#SVGID_12_);fill:#56a55c" d="M173.11 270.0439c-3.96 3.96-6.4116 9.4235-6.4116 15.452-.0024 12.0547 9.8066 21.8636 21.8613 21.8613 12.0547.0024 21.8636-9.797 21.8613-21.8613.0024-3.8474-1.0151-7.6326-2.9353-10.9462-3.8977-6.7325-11.1497-10.9151-18.926-10.9151C182.5311 263.6346 177.0676 266.0863 173.11 270.0439z"/></g></g><rect x="199.3456" y="279.5436" style="fill:#56a55c" width="122.6266" height="13.8671"/></g><g><g><path style="fill:#f1bc42" d="M189.0874 318.7208l.0012-.0012c7.7751.0012 15.0295 4.1862 18.932 10.9234 1.9177 3.3159 2.9305 7.1011 2.9293 10.9378.0 5.8394-2.2733 11.3305-6.4032 15.4604-4.1288 4.1288-9.6186 6.4032-15.458 6.4032s-11.328-2.2733-15.458-6.4032-6.4032-9.6186-6.4056-15.4628c.0012-6.025 2.454-11.4897 6.4116-15.4472C177.5953 321.1724 183.0601 318.7196 189.0874 318.7208zm7.993 21.8576c.0012-1.4042-.3687-2.7868-1.063-3.9887-1.4293-2.4684-4.0833-3.9995-6.9299-4.0019-4.4077.0024-7.993 3.5877-7.993 7.993.0 2.1356.832 4.1431 2.3427 5.6539 1.5083 1.5083 3.5159 2.3403 5.6503 2.3415 2.1356.0 4.1443-.8308 5.6539-2.3403S197.0816 342.7176 197.0804 340.5784z"/><path style="opacity:.3;fill:#fff" d="M189.0898 318.7214c7.7763.0 15.0283 4.1826 18.926 10.915 1.9201 3.3136 2.9377 7.0988 2.9353 10.9462.0024 12.0643-9.8065 21.8636-21.8613 21.8612-12.0547.0024-21.8636-9.8065-21.8612-21.8612.0-6.0285 2.4516-11.4921 6.4116-15.452C177.5977 321.173 183.0612 318.7214 189.0898 318.7214zm7.9941 21.8612c0-1.4078-.3711-2.7892-1.0702-3.9959-1.4269-2.466-4.0797-3.9983-6.924-3.9983-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0024 4.4077 3.5865 7.9918 7.9942 7.9942 2.2027.0 4.2018-.8978 5.6479-2.3439C196.1861 344.7844 197.0839 342.7853 197.0839 340.5826z"/><g><defs><path id="SVGID_13_" d="M194.7376 346.2329c-1.4461 1.4461-3.4452 2.3439-5.6479 2.3439-4.4077-.0024-7.9918-3.5865-7.9942-7.9942.0024-4.4125 3.5937-7.999 7.9942-7.9942 2.8443.0 5.497 1.5323 6.924 3.9983.6991 1.2067 1.0702 2.5881 1.0702 3.9959C197.0839 342.7853 196.1861 344.7844 194.7376 346.2329z"/></defs><clipPath id="SVGID_14_"><use xlink:href="#SVGID_13_" style="overflow:visible"/></clipPath><path style="clip-path:url(#SVGID_14_);fill:#fff" d="M190.0704 333.5691c-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0011 1.9547.7088 3.7452 1.8782 5.1354-1.7447-1.4674-2.8575-3.6631-2.8588-6.1161.0024-4.4125 3.5936-7.999 7.9942-7.9942 2.3802-1e-4 4.616 1.0834 6.1218 2.8788C193.7885 334.2701 191.9774 333.5691 190.0704 333.5691z"/><path style="opacity:.13;clip-path:url(#SVGID_14_);fill:#020202" d="M190.0704 333.5691c-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0011 1.9547.7088 3.7452 1.8782 5.1354-1.7447-1.4674-2.8575-3.6631-2.8588-6.1161.0024-4.4125 3.5936-7.999 7.9942-7.9942 2.3802-1e-4 4.616 1.0834 6.1218 2.8788C193.7885 334.2701 191.9774 333.5691 190.0704 333.5691z"/></g><g><defs><path id="SVGID_15_" d="M189.0898 318.7214c7.7763.0 15.0283 4.1826 18.926 10.915 1.9201 3.3136 2.9377 7.0988 2.9353 10.9462.0024 12.0643-9.8065 21.8636-21.8613 21.8612-12.0547.0024-21.8636-9.8065-21.8612-21.8612.0-6.0285 2.4516-11.4921 6.4116-15.452C177.5977 321.173 183.0612 318.7214 189.0898 318.7214zm7.9941 21.8612c0-1.4078-.3711-2.7892-1.0702-3.9959-1.4269-2.466-4.0797-3.9983-6.924-3.9983-4.4005-.0048-7.9918 3.5817-7.9942 7.9942.0024 4.4077 3.5865 7.9918 7.9942 7.9942 2.2027.0 4.2018-.8978 5.6479-2.3439C196.1861 344.7844 197.0839 342.7853 197.0839 340.5826z"/></defs><clipPath id="SVGID_16_"><use xlink:href="#SVGID_15_" style="overflow:visible"/></clipPath><path style="clip-path:url(#SVGID_16_);fill:#f1bc42" d="M172.6595 324.15c-3.96 3.96-6.4116 9.4235-6.4116 15.452-.0024 12.0547 9.8066 21.8636 21.8613 21.8612 12.0547.0024 21.8636-9.797 21.8613-21.8612.0024-3.8474-1.0151-7.6327-2.9353-10.9462-3.8977-6.7324-11.1497-10.9151-18.926-10.9151C182.0806 317.7407 176.6171 320.1924 172.6595 324.15z"/></g></g><rect x="198.8952" y="333.6497" style="fill:#f1bc42" width="122.6266" height="13.8671"/></g></g></svg></span><span class="font-weight-bold">GEN242</span>
	</a>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../about/introduction/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../about/syllabus/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../about/schedule/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../about/internal/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../about/internal/internal_resources/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../about/internal/install/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_01"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_02"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_03"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_04"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Git and GitHub (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_05"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux Cluster Basics (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_06"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_07"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_08"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_09"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_10"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_11"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_12"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_13"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_14"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_15"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_16/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Overview (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_17/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Enrichment Analysis (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_18/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_19/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_20/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics in R (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_21/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_22/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (S22)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_23/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflow Review (S23)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_24/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Final Discussion (S24)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../slides/slides_25/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Additional Slide Shows (S25)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/github/github/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/linux/linux/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/rbasics/rbasics/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/rprogramming/rprogramming/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/rparallel/rparallel/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/rsequences/rsequences/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/systempiper/systempiper/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/sprnaseq/sprnaseq/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/spchipseq/spchipseq/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/scrnaseq/scrnaseq/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/spvarseq/spvarseq/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 VAR-Seq Workflow (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/rmarkdown/rmarkdown/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/rfea/rfea/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/rclustering/rclustering/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/project_data/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/rgraphics/rgraphics/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/shinyapps/shinyapps/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/rpackages/rpackages/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/dplyr/dplyr/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../tutorials/cmdtocwl/cmdtocwl/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T20)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../assignments/homework/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/homework/hw01/hw01/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/homework/hw02/hw02/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/homework/hw03/hw03/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/homework/hw04/hw04/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/homework/hw05/hw05"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/homework/hw06/hw06"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/homework/hw07/hw07"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/homework/hw08/hw08"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8: R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/project_overview/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/01_rnaseq_aligners/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/02_rnaseq_deg/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/02_rnaseq_dex/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEXseq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/03_cluster_analysis/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/04_scrnaseq_embedding/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/05_chipseq_peakcaller/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/06_functional_enrichment/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/07_motif_enrichment/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/08_genome_graphics/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/09_drug_target_analysis/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Drug-target Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/10_lncrna_orf_discovery/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 lncRNA Discovery
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/projects/project_data/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../../assignments/presentations/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../../assignments/presentations/paper_presentations/"
                 style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/class/ludf7verm4t14z"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
                 style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../../offline-search-index.9df7c0197cebce03dd5a85e548d1d641.json"
  data-offline-search-base-href="../../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-default td-outer">
      <main role="main" class="td-main">
        

<!--
# Compile from command-line
Rscript -e "rmarkdown::render('systemPipeRNAseq.Rmd', c('BiocStyle::html_document'), clean=FALSE); knitr::knit('systemPipeRNAseq.Rmd', tangle=TRUE)"; Rscript -e "rmarkdown::render('systemPipeRNAseq.Rmd', c('BiocStyle::pdf_document'))"
-->
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<h1 id="introduction">Introduction</h1>
<p>Users want to provide here background information about the design of their RNA-Seq project.</p>
<h1 id="samples-and-environment-settings">Samples and environment settings</h1>
<h2 id="environment-settings-and-input-data">Environment settings and input data</h2>
<p><a href="http://bioconductor.org/packages/release/data/experiment/html/systemPipeRdata.html"><em>systemPipeRdata</em></a> package is a helper package to generate a fully populated <a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html"><em>systemPipeR</em></a>
workflow environment in the current working directory with a single command.
All the instruction for generating the workflow are provide in the <em>systemPipeRdata</em> vignette <a href="http://www.bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRdata.html#1_Introduction">here</a>.</p>
<pre><code class="language-r">systemPipeRdata::genWorkenvir(workflow = &quot;rnaseq&quot;, mydirname = &quot;rnaseq&quot;)
setwd(&quot;rnaseq&quot;)
</code></pre>
<p>Typically, the user wants to record here the sources and versions of the
reference genome sequence along with the corresponding annotations. In
the provided sample data set all data inputs are stored in a <code>data</code>
subdirectory and all results will be written to a separate <code>results</code> directory,
while the <code>systemPipeRNAseq.Rmd</code> workflow and the <code>targets</code> file are expected to be
located in the parent directory.</p>
<p>The chosen data set used by this report <a href="http://www.ncbi.nlm.nih.gov/sra/?term=SRP010938">SRP010938</a>
contains 18 paired-end (PE) read sets from <em>Arabidposis thaliana</em>
(Howard et al. 2013). To minimize processing time during testing, each FASTQ
file has been subsetted to 90,000-100,000 randomly sampled PE reads that
map to the first 100,000 nucleotides of each chromosome of the <em>A.
thaliana</em> genome. The corresponding reference genome sequence (FASTA) and
its GFF annotation files have been truncated accordingly. This way the entire
test sample data set is less than 200MB in storage space. A PE read set has been
chosen for this test data set for flexibility, because it can be used for testing both types
of analysis routines requiring either SE (single end) reads or PE reads.</p>
<p>To work with real data, users want to organize their own data similarly
and substitute all test data for their own data. To rerun an established
workflow on new data, the initial <code>targets</code> file along with the corresponding
FASTQ files are usually the only inputs the user needs to provide.</p>
<p>For more details, please consult the documentation
<a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">here</a>. More information about the <code>targets</code> files from <em>systemPipeR</em> can be found <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#42_Structure_of_initial_targets_data">here</a>.</p>
<h3 id="experiment-definition-provided-by-targets-file">Experiment definition provided by <code>targets</code> file</h3>
<p>The <code>targets</code> file defines all FASTQ files and sample comparisons of the analysis workflow.</p>
<pre><code class="language-r">targetspath &lt;- system.file(&quot;extdata&quot;, &quot;targetsPE.txt&quot;, package = &quot;systemPipeR&quot;)
targets &lt;- read.delim(targetspath, comment.char = &quot;#&quot;)
targets[1:4, -c(5, 6)]
</code></pre>
<pre><code>##                     FileName1                   FileName2
## 1 ./data/SRR446027_1.fastq.gz ./data/SRR446027_2.fastq.gz
## 2 ./data/SRR446028_1.fastq.gz ./data/SRR446028_2.fastq.gz
## 3 ./data/SRR446029_1.fastq.gz ./data/SRR446029_2.fastq.gz
## 4 ./data/SRR446030_1.fastq.gz ./data/SRR446030_2.fastq.gz
##   SampleName Factor        Date
## 1        M1A     M1 23-Mar-2012
## 2        M1B     M1 23-Mar-2012
## 3        A1A     A1 23-Mar-2012
## 4        A1B     A1 23-Mar-2012
</code></pre>
<p>To work with custom data, users need to generate a <em><code>targets</code></em> file containing
the paths to their own FASTQ files.</p>
<h1 id="workflow-environment">Workflow environment</h1>
<p><em><code>systemPipeR</code></em> workflows can be designed and built from start to finish with a
single command, importing from an R Markdown file or stepwise in interactive
mode from the R console.</p>
<p>This tutorial will demonstrate how to build the workflow in an interactive mode,
appending each step. The workflow is constructed by connecting each step via
<code>appendStep</code> method. Each <code>SYSargsList</code> instance contains instructions for
processing a set of input files with a specific command-line or R software
and the paths to the corresponding outfiles generated by a particular command-line
software/step.</p>
<p>To create a workflow within <em><code>systemPipeR</code></em>, we can start by defining an empty
container and checking the directory structure:</p>
<pre><code class="language-r">library(systemPipeR)
sal &lt;- SPRproject()
sal
</code></pre>
<h2 id="required-packages-and-resources">Required packages and resources</h2>
<p>The <code>systemPipeR</code> package needs to be loaded (H Backman and Girke 2016).</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(systemPipeR)
}, step_name = &quot;load_SPR&quot;)
</code></pre>
<h2 id="read-preprocessing">Read preprocessing</h2>
<h3 id="preprocessing-with-preprocessreads-function">Preprocessing with <code>preprocessReads</code> function</h3>
<p>The function <code>preprocessReads</code> allows to apply predefined or custom
read preprocessing functions to all FASTQ files referenced in a
<code>SYSargsList</code> container, such as quality filtering or adapter trimming
routines. Internally, <code>preprocessReads</code> uses the <code>FastqStreamer</code> function from
the <code>ShortRead</code> package to stream through large FASTQ files in a
memory-efficient manner. The following example performs adapter trimming with
the <code>trimLRPatterns</code> function from the <code>Biostrings</code> package.</p>
<p>Here, we are appending this step to the <code>SYSargsList</code> object created previously.
All the parameters are defined on the <code>preprocessReads/preprocessReads-pe.yml</code> file.</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;preprocessing&quot;, targets = &quot;targetsPE.txt&quot;,
    dir = TRUE, wf_file = &quot;preprocessReads/preprocessReads-pe.cwl&quot;,
    input_file = &quot;preprocessReads/preprocessReads-pe.yml&quot;, dir_path = system.file(&quot;extdata/cwl&quot;,
        package = &quot;systemPipeR&quot;), inputvars = c(FileName1 = &quot;_FASTQ_PATH1_&quot;,
        FileName2 = &quot;_FASTQ_PATH2_&quot;, SampleName = &quot;_SampleName_&quot;),
    dependency = c(&quot;load_SPR&quot;))
</code></pre>
<p>After the preprocessing step, the <code>outfiles</code> files can be used to generate the new
targets files containing the paths to the trimmed FASTQ files. The new targets
information can be used for the next workflow step instance, <em>e.g.</em> running the
NGS alignments with the trimmed FASTQ files. The <code>appendStep</code> function is
automatically handling this connectivity between steps. Please check the next
step for more details.</p>
<p>The following example shows how one can design a custom read <em>‘preprocessReads’</em>
function using utilities provided by the <code>ShortRead</code> package, and then run it
in batch mode with the <em>‘preprocessReads’</em> function. Here, it is possible to
replace the function used on the <code>preprocessing</code> step and modify the <code>sal</code> object.
Because it is a custom function, it is necessary to save the part in the R object,
and internally the <code>preprocessReads.doc.R</code> is loading the custom function.
If the R object is saved with a different name (here <code>&quot;param/customFCT.RData&quot;</code>),
please replace that accordingly in the <code>preprocessReads.doc.R</code>.</p>
<p>Please, note that this step is not added to the workflow, here just for
demonstration.</p>
<p>First, we defined the custom function in the workflow:</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    filterFct &lt;- function(fq, cutoff = 20, Nexceptions = 0) {
        qcount &lt;- rowSums(as(quality(fq), &quot;matrix&quot;) &lt;= cutoff,
            na.rm = TRUE)
        # Retains reads where Phred scores are &gt;= cutoff
        # with N exceptions
        fq[qcount &lt;= Nexceptions]
    }
    save(list = ls(), file = &quot;param/customFCT.RData&quot;)
}, step_name = &quot;custom_preprocessing_function&quot;, dependency = &quot;preprocessing&quot;)
</code></pre>
<p>After, we can edit the input parameter:</p>
<pre><code class="language-r">yamlinput(sal, &quot;preprocessing&quot;)$Fct
yamlinput(sal, &quot;preprocessing&quot;, &quot;Fct&quot;) &lt;- &quot;'filterFct(fq, cutoff=20, Nexceptions=0)'&quot;
yamlinput(sal, &quot;preprocessing&quot;)$Fct  ## check the new function
cmdlist(sal, &quot;preprocessing&quot;, targets = 1)  ## check if the command line was updated with success
</code></pre>
<h3 id="read-trimming-with-trimmomatic">Read trimming with Trimmomatic</h3>
<p><a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a> software (Bolger, Lohse, and Usadel 2014)
performs a variety of useful trimming tasks for Illumina paired-end and single
ended data. Here, an example of how to perform this task using parameters template
files for trimming FASTQ files.</p>
<p>This step is optional.</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;trimming&quot;, targets = &quot;targetsPE.txt&quot;,
    wf_file = &quot;trimmomatic/trimmomatic-pe.cwl&quot;, input_file = &quot;trimmomatic/trimmomatic-pe.yml&quot;,
    dir_path = system.file(&quot;extdata/cwl&quot;, package = &quot;systemPipeR&quot;),
    inputvars = c(FileName1 = &quot;_FASTQ_PATH1_&quot;, FileName2 = &quot;_FASTQ_PATH2_&quot;,
        SampleName = &quot;_SampleName_&quot;), dependency = &quot;load_SPR&quot;,
    run_step = &quot;optional&quot;)
</code></pre>
<h3 id="fastq-quality-report">FASTQ quality report</h3>
<p>The following <code>seeFastq</code> and <code>seeFastqPlot</code> functions generate and plot a series of useful
quality statistics for a set of FASTQ files, including per cycle quality box
plots, base proportions, base-level quality trends, relative k-mer
diversity, length, and occurrence distribution of reads, number of reads
above quality cutoffs and mean quality distribution. The results are
written to a png file named <code>fastqReport.png</code>.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    fastq &lt;- getColumn(sal, step = &quot;preprocessing&quot;, &quot;targetsWF&quot;,
        column = 1)
    fqlist &lt;- seeFastq(fastq = fastq, batchsize = 10000, klength = 8)
    png(&quot;./results/fastqReport.png&quot;, height = 162, width = 288 *
        length(fqlist))
    seeFastqPlot(fqlist)
    dev.off()
}, step_name = &quot;fastq_report&quot;, dependency = &quot;preprocessing&quot;)
</code></pre>
<p><img src="results/fastqReport.png" alt=""></p>
<div data-align="center">
<p>Figure 1: FASTQ quality report for 18 samples</p>
</div>
</br>
<h2 id="alignments">Alignments</h2>
<h3 id="read-mapping-with-hisat2">Read mapping with <code>HISAT2</code></h3>
<p>The following steps will demonstrate how to use the short read aligner <code>Hisat2</code>
(Kim, Langmead, and Salzberg 2015). First, the <code>Hisat2</code> index needs to be created.</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;hisat2_index&quot;, dir = FALSE,
    targets = NULL, wf_file = &quot;hisat2/hisat2-index.cwl&quot;, input_file = &quot;hisat2/hisat2-index.yml&quot;,
    dir_path = &quot;param/cwl&quot;, dependency = &quot;load_SPR&quot;)
</code></pre>
<h3 id="hisat2-mapping"><code>HISAT2</code> mapping</h3>
<p>The parameter settings of the aligner are defined in the <code>workflow_hisat2-pe.cwl</code>
and <code>workflow_hisat2-pe.yml</code> files. The following shows how to construct the
corresponding <em>SYSargsList</em> object.</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;hisat2_mapping&quot;,
    dir = TRUE, targets = &quot;preprocessing&quot;, wf_file = &quot;workflow-hisat2/workflow_hisat2-pe.cwl&quot;,
    input_file = &quot;workflow-hisat2/workflow_hisat2-pe.yml&quot;, dir_path = &quot;param/cwl&quot;,
    inputvars = c(preprocessReads_1 = &quot;_FASTQ_PATH1_&quot;, preprocessReads_2 = &quot;_FASTQ_PATH2_&quot;,
        SampleName = &quot;_SampleName_&quot;), rm_targets_col = c(&quot;FileName1&quot;,
        &quot;FileName2&quot;), dependency = c(&quot;preprocessing&quot;, &quot;hisat2_index&quot;))
</code></pre>
<p>To double-check the command line for each sample, please use the following:</p>
<pre><code class="language-r">cmdlist(sal, step = &quot;hisat2_mapping&quot;, targets = 1)
</code></pre>
<h3 id="read-and-alignment-stats">Read and alignment stats</h3>
<p>The following provides an overview of the number of reads in each sample
and how many of them aligned to the reference.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    fqpaths &lt;- getColumn(sal, step = &quot;preprocessing&quot;, &quot;targetsWF&quot;,
        column = &quot;FileName1&quot;)
    bampaths &lt;- getColumn(sal, step = &quot;hisat2_mapping&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    read_statsDF &lt;- alignStats(args = bampaths, fqpaths = fqpaths,
        pairEnd = TRUE)
    write.table(read_statsDF, &quot;results/alignStats.xls&quot;, row.names = FALSE,
        quote = FALSE, sep = &quot;\t&quot;)
}, step_name = &quot;align_stats&quot;, dependency = &quot;hisat2_mapping&quot;)
</code></pre>
<h2 id="create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</h2>
<p>The <code>symLink2bam</code> function creates symbolic links to view the BAM alignment files in a
genome browser such as IGV without moving these large files to a local
system. The corresponding URLs are written to a file with a path
specified under <code>urlfile</code>, here <code>IGVurl.txt</code>.
Please replace the directory and the user name.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    bampaths &lt;- getColumn(sal, step = &quot;hisat2_mapping&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    symLink2bam(sysargs = bampaths, htmldir = c(&quot;~/.html/&quot;, &quot;somedir/&quot;),
        urlbase = &quot;http://cluster.hpcc.ucr.edu/~tgirke/&quot;, urlfile = &quot;./results/IGVurl.txt&quot;)
}, step_name = &quot;bam_IGV&quot;, dependency = &quot;hisat2_mapping&quot;, run_step = &quot;optional&quot;)
</code></pre>
<h2 id="read-quantification">Read quantification</h2>
<p>Reads overlapping with annotation ranges of interest are counted for
each sample using the <code>summarizeOverlaps</code> function (Lawrence et al. 2013).
The read counting is preformed for exon gene regions in a non-strand-specific
manner while ignoring overlaps among different genes. Subsequently, the expression
count values are normalized by <em>reads per kp per million mapped reads</em>
(RPKM). The raw read count table (<code>countDFeByg.xls</code>) and the corresponding
RPKM table (<code>rpkmDFeByg.xls</code>) are written to separate files in the directory of
this project. Parallelization is achieved with the <code>BiocParallel</code> package,
here using 4 CPU cores.</p>
<h3 id="create-a-database-for-gene-annotation">Create a database for gene annotation</h3>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(GenomicFeatures)
    txdb &lt;- suppressWarnings(makeTxDbFromGFF(file = &quot;data/tair10.gff&quot;,
        format = &quot;gff&quot;, dataSource = &quot;TAIR&quot;, organism = &quot;Arabidopsis thaliana&quot;))
    saveDb(txdb, file = &quot;./data/tair10.sqlite&quot;)
}, step_name = &quot;create_db&quot;, dependency = &quot;hisat2_mapping&quot;)
</code></pre>
<h3 id="read-counting-with-summarizeoverlaps-in-parallel-mode-using-multiple-cores">Read counting with <code>summarizeOverlaps</code> in parallel mode using multiple cores</h3>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(GenomicFeatures)
    library(BiocParallel)
    txdb &lt;- loadDb(&quot;./data/tair10.sqlite&quot;)
    outpaths &lt;- getColumn(sal, step = &quot;hisat2_mapping&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    eByg &lt;- exonsBy(txdb, by = c(&quot;gene&quot;))
    bfl &lt;- BamFileList(outpaths, yieldSize = 50000, index = character())
    multicoreParam &lt;- MulticoreParam(workers = 4)
    register(multicoreParam)
    registered()
    counteByg &lt;- bplapply(bfl, function(x) summarizeOverlaps(eByg,
        x, mode = &quot;Union&quot;, ignore.strand = TRUE, inter.feature = FALSE,
        singleEnd = FALSE, BPPARAM = multicoreParam))
    countDFeByg &lt;- sapply(seq(along = counteByg), function(x) assays(counteByg[[x]])$counts)
    rownames(countDFeByg) &lt;- names(rowRanges(counteByg[[1]]))
    colnames(countDFeByg) &lt;- names(bfl)
    rpkmDFeByg &lt;- apply(countDFeByg, 2, function(x) returnRPKM(counts = x,
        ranges = eByg))
    write.table(countDFeByg, &quot;results/countDFeByg.xls&quot;, col.names = NA,
        quote = FALSE, sep = &quot;\t&quot;)
    write.table(rpkmDFeByg, &quot;results/rpkmDFeByg.xls&quot;, col.names = NA,
        quote = FALSE, sep = &quot;\t&quot;)
    ## Creating a SummarizedExperiment object
    colData &lt;- data.frame(row.names = SampleName(sal, &quot;hisat2_mapping&quot;),
        condition = getColumn(sal, &quot;hisat2_mapping&quot;, position = &quot;targetsWF&quot;,
            column = &quot;Factor&quot;))
    colData$condition &lt;- factor(colData$condition)
    countDF_se &lt;- SummarizedExperiment::SummarizedExperiment(assays = countDFeByg,
        colData = colData)
    ## Add results as SummarizedExperiment to the workflow
    ## object
    SE(sal, &quot;read_counting&quot;) &lt;- countDF_se
}, step_name = &quot;read_counting&quot;, dependency = &quot;create_db&quot;)
</code></pre>
<p>When providing a <code>BamFileList</code> as in the example above, <code>summarizeOverlaps</code> methods
use by default <code>bplapply</code> and use the register interface from BiocParallel package.
If the number of workers is not set, <code>MulticoreParam</code> will use the number of cores
returned by <code>parallel::detectCores()</code>. For more information,
please check <code>help(&quot;summarizeOverlaps&quot;)</code> documentation.</p>
<p>Note, for most statistical differential expression or abundance analysis
methods, such as <code>edgeR</code> or <code>DESeq2</code>, the raw count values should be used as input. The
usage of RPKM values should be restricted to specialty applications
required by some users, <em>e.g.</em> manually comparing the expression levels
among different genes or features.</p>
<h3 id="sample-wise-correlation-analysis">Sample-wise correlation analysis</h3>
<p>The following computes the sample-wise Spearman correlation coefficients from
the <code>rlog</code> transformed expression values generated with the <code>DESeq2</code> package. After
transformation to a distance matrix, hierarchical clustering is performed with
the <code>hclust</code> function and the result is plotted as a dendrogram
(also see file <code>sample_tree.png</code>).</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(DESeq2, quietly = TRUE)
    library(ape, warn.conflicts = FALSE)
    ## Extracting SummarizedExperiment object
    se &lt;- SE(sal, &quot;read_counting&quot;)
    dds &lt;- DESeqDataSet(se, design = ~condition)
    d &lt;- cor(assay(rlog(dds)), method = &quot;spearman&quot;)
    hc &lt;- hclust(dist(1 - d))
    png(&quot;results/sample_tree.png&quot;)
    plot.phylo(as.phylo(hc), type = &quot;p&quot;, edge.col = &quot;blue&quot;, edge.width = 2,
        show.node.label = TRUE, no.margin = TRUE)
    dev.off()
}, step_name = &quot;sample_tree&quot;, dependency = &quot;read_counting&quot;)
</code></pre>
<p><img src="results/sample_tree.png" alt=""></p>
<div data-align="center">
<p>Figure 2: Correlation dendrogram of samples</p>
</div>
</br>
<h2 id="analysis-of-degs">Analysis of DEGs</h2>
<p>The analysis of differentially expressed genes (DEGs) is performed with
the <code>glm</code> method of the <code>edgeR</code> package (Robinson, McCarthy, and Smyth 2010). The sample
comparisons used by this analysis are defined in the header lines of the
<code>targets.txt</code> file starting with <code>&lt;CMP&gt;</code>.</p>
<h3 id="run-edger">Run <code>edgeR</code></h3>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(edgeR)
    countDF &lt;- read.delim(&quot;results/countDFeByg.xls&quot;, row.names = 1,
        check.names = FALSE)
    cmp &lt;- readComp(stepsWF(sal)[[&quot;hisat2_mapping&quot;]], format = &quot;matrix&quot;,
        delim = &quot;-&quot;)
    edgeDF &lt;- run_edgeR(countDF = countDF, targets = targetsWF(sal)[[&quot;hisat2_mapping&quot;]],
        cmp = cmp[[1]], independent = FALSE, mdsplot = &quot;&quot;)
}, step_name = &quot;run_edger&quot;, dependency = &quot;read_counting&quot;)
</code></pre>
<h3 id="add-gene-descriptions">Add gene descriptions</h3>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(&quot;biomaRt&quot;)
    m &lt;- useMart(&quot;plants_mart&quot;, dataset = &quot;athaliana_eg_gene&quot;,
        host = &quot;https://plants.ensembl.org&quot;)
    desc &lt;- getBM(attributes = c(&quot;tair_locus&quot;, &quot;description&quot;),
        mart = m)
    desc &lt;- desc[!duplicated(desc[, 1]), ]
    descv &lt;- as.character(desc[, 2])
    names(descv) &lt;- as.character(desc[, 1])
    edgeDF &lt;- data.frame(edgeDF, Desc = descv[rownames(edgeDF)],
        check.names = FALSE)
    write.table(edgeDF, &quot;./results/edgeRglm_allcomp.xls&quot;, quote = FALSE,
        sep = &quot;\t&quot;, col.names = NA)
}, step_name = &quot;custom_annot&quot;, dependency = &quot;run_edger&quot;)
</code></pre>
<h3 id="plot-deg-results">Plot DEG results</h3>
<p>Filter and plot DEG results for up and down regulated genes. The
definition of <em>up</em> and <em>down</em> is given in the corresponding help
file. To open it, type <code>?filterDEGs</code> in the R console.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    edgeDF &lt;- read.delim(&quot;results/edgeRglm_allcomp.xls&quot;, row.names = 1,
        check.names = FALSE)
    png(&quot;results/DEGcounts.png&quot;)
    DEG_list &lt;- filterDEGs(degDF = edgeDF, filter = c(Fold = 2,
        FDR = 20))
    dev.off()
    write.table(DEG_list$Summary, &quot;./results/DEGcounts.xls&quot;,
        quote = FALSE, sep = &quot;\t&quot;, row.names = FALSE)
}, step_name = &quot;filter_degs&quot;, dependency = &quot;custom_annot&quot;)
</code></pre>
<h3 id="venn-diagrams-of-deg-sets">Venn diagrams of DEG sets</h3>
<p>The <code>overLapper</code> function can compute Venn intersects for large numbers of sample
sets (up to 20 or more) and plots 2-5 way Venn diagrams. A useful
feature is the possibility to combine the counts from several Venn
comparisons with the same number of sample sets in a single Venn diagram
(here for 4 up and down DEG sets).</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    vennsetup &lt;- overLapper(DEG_list$Up[6:9], type = &quot;vennsets&quot;)
    vennsetdown &lt;- overLapper(DEG_list$Down[6:9], type = &quot;vennsets&quot;)
    png(&quot;results/vennplot.png&quot;)
    vennPlot(list(vennsetup, vennsetdown), mymain = &quot;&quot;, mysub = &quot;&quot;,
        colmode = 2, ccol = c(&quot;blue&quot;, &quot;red&quot;))
    dev.off()
}, step_name = &quot;venn_diagram&quot;, dependency = &quot;filter_degs&quot;)
</code></pre>
<h2 id="go-term-enrichment-analysis">GO term enrichment analysis</h2>
<h3 id="obtain-gene-to-go-mappings">Obtain gene-to-GO mappings</h3>
<p>The following shows how to obtain gene-to-GO mappings from <code>biomaRt</code> (here for <em>A.
thaliana</em>) and how to organize them for the downstream GO term
enrichment analysis. Alternatively, the gene-to-GO mappings can be
obtained for many organisms from Bioconductor’s <code>*.db</code> genome annotation
packages or GO annotation files provided by various genome databases.
For each annotation this relatively slow preprocessing step needs to be
performed only once. Subsequently, the preprocessed data can be loaded
with the <code>load</code> function as shown in the next subsection.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(&quot;biomaRt&quot;)
    # listMarts() # To choose BioMart database
    # listMarts(host='plants.ensembl.org')
    m &lt;- useMart(&quot;plants_mart&quot;, host = &quot;https://plants.ensembl.org&quot;)
    # listDatasets(m)
    m &lt;- useMart(&quot;plants_mart&quot;, dataset = &quot;athaliana_eg_gene&quot;,
        host = &quot;https://plants.ensembl.org&quot;)
    # listAttributes(m) # Choose data types you want to
    # download
    go &lt;- getBM(attributes = c(&quot;go_id&quot;, &quot;tair_locus&quot;, &quot;namespace_1003&quot;),
        mart = m)
    go &lt;- go[go[, 3] != &quot;&quot;, ]
    go[, 3] &lt;- as.character(go[, 3])
    go[go[, 3] == &quot;molecular_function&quot;, 3] &lt;- &quot;F&quot;
    go[go[, 3] == &quot;biological_process&quot;, 3] &lt;- &quot;P&quot;
    go[go[, 3] == &quot;cellular_component&quot;, 3] &lt;- &quot;C&quot;
    go[1:4, ]
    if (!dir.exists(&quot;./data/GO&quot;))
        dir.create(&quot;./data/GO&quot;)
    write.table(go, &quot;data/GO/GOannotationsBiomart_mod.txt&quot;, quote = FALSE,
        row.names = FALSE, col.names = FALSE, sep = &quot;\t&quot;)
    catdb &lt;- makeCATdb(myfile = &quot;data/GO/GOannotationsBiomart_mod.txt&quot;,
        lib = NULL, org = &quot;&quot;, colno = c(1, 2, 3), idconv = NULL)
    save(catdb, file = &quot;data/GO/catdb.RData&quot;)
}, step_name = &quot;get_go_annot&quot;, dependency = &quot;filter_degs&quot;)
</code></pre>
<h3 id="batch-go-term-enrichment-analysis">Batch GO term enrichment analysis</h3>
<p>Apply the enrichment analysis to the DEG sets obtained the above differential
expression analysis. Note, in the following example the <code>FDR</code> filter is set
here to an unreasonably high value, simply because of the small size of the toy
data set used in this vignette. Batch enrichment analysis of many gene sets is
performed with the function. When <code>method=all</code>, it returns all GO terms passing
the p-value cutoff specified under the <code>cutoff</code> arguments. When <code>method=slim</code>,
it returns only the GO terms specified under the <code>myslimv</code> argument. The given
example shows how a GO slim vector for a specific organism can be obtained from
<code>BioMart</code>.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(&quot;biomaRt&quot;)
    load(&quot;data/GO/catdb.RData&quot;)
    DEG_list &lt;- filterDEGs(degDF = edgeDF, filter = c(Fold = 2,
        FDR = 50), plot = FALSE)
    up_down &lt;- DEG_list$UporDown
    names(up_down) &lt;- paste(names(up_down), &quot;_up_down&quot;, sep = &quot;&quot;)
    up &lt;- DEG_list$Up
    names(up) &lt;- paste(names(up), &quot;_up&quot;, sep = &quot;&quot;)
    down &lt;- DEG_list$Down
    names(down) &lt;- paste(names(down), &quot;_down&quot;, sep = &quot;&quot;)
    DEGlist &lt;- c(up_down, up, down)
    DEGlist &lt;- DEGlist[sapply(DEGlist, length) &gt; 0]
    BatchResult &lt;- GOCluster_Report(catdb = catdb, setlist = DEGlist,
        method = &quot;all&quot;, id_type = &quot;gene&quot;, CLSZ = 2, cutoff = 0.9,
        gocats = c(&quot;MF&quot;, &quot;BP&quot;, &quot;CC&quot;), recordSpecGO = NULL)
    m &lt;- useMart(&quot;plants_mart&quot;, dataset = &quot;athaliana_eg_gene&quot;,
        host = &quot;https://plants.ensembl.org&quot;)
    goslimvec &lt;- as.character(getBM(attributes = c(&quot;goslim_goa_accession&quot;),
        mart = m)[, 1])
    BatchResultslim &lt;- GOCluster_Report(catdb = catdb, setlist = DEGlist,
        method = &quot;slim&quot;, id_type = &quot;gene&quot;, myslimv = goslimvec,
        CLSZ = 10, cutoff = 0.01, gocats = c(&quot;MF&quot;, &quot;BP&quot;, &quot;CC&quot;),
        recordSpecGO = NULL)
    write.table(BatchResultslim, &quot;results/GOBatchSlim.xls&quot;, row.names = FALSE,
        quote = FALSE, sep = &quot;\t&quot;)
}, step_name = &quot;go_enrich&quot;, dependency = &quot;get_go_annot&quot;)
</code></pre>
<h3 id="plot-batch-go-term-results">Plot batch GO term results</h3>
<p>The <code>data.frame</code> generated by <code>GOCluster</code> can be plotted with the <code>goBarplot</code> function. Because of the
variable size of the sample sets, it may not always be desirable to show
the results from different DEG sets in the same bar plot. Plotting
single sample sets is achieved by subsetting the input data frame as
shown in the first line of the following example.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    gos &lt;- BatchResultslim[grep(&quot;M6-V6_up_down&quot;, BatchResultslim$CLID),
        ]
    gos &lt;- BatchResultslim
    png(&quot;results/GOslimbarplotMF.png&quot;, height = 8, width = 10)
    goBarplot(gos, gocat = &quot;MF&quot;)
    goBarplot(gos, gocat = &quot;BP&quot;)
    goBarplot(gos, gocat = &quot;CC&quot;)
    dev.off()
}, step_name = &quot;go_plot&quot;, dependency = &quot;go_enrich&quot;)
</code></pre>
<p><img src="results/GOslimbarplotMF.png" alt=""></p>
<div data-align="center">
<p>Figure 5: GO Slim Barplot for MF Ontology</p>
</div>
</br>
<h2 id="clustering-and-heat-maps">Clustering and heat maps</h2>
<p>The following example performs hierarchical clustering on the <code>rlog</code>
transformed expression matrix subsetted by the DEGs identified in the above
differential expression analysis. It uses a Pearson correlation-based distance
measure and complete linkage for cluster joining.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(pheatmap)
    geneids &lt;- unique(as.character(unlist(DEG_list[[1]])))
    y &lt;- assay(rlog(dds))[geneids, ]
    png(&quot;results/heatmap1.png&quot;)
    pheatmap(y, scale = &quot;row&quot;, clustering_distance_rows = &quot;correlation&quot;,
        clustering_distance_cols = &quot;correlation&quot;)
    dev.off()
}, step_name = &quot;heatmap&quot;, dependency = &quot;go_enrich&quot;)
</code></pre>
<p><img src="results/heatmap1.png" alt=""></p>
<div data-align="center">
<p>Figure 6: Heat Map with Hierarchical Clustering Dendrograms of DEGs</p>
</div>
</br>
<h2 id="version-information">Version Information</h2>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    sessionInfo()
}, step_name = &quot;sessionInfo&quot;, dependency = &quot;heatmap&quot;)
</code></pre>
<h1 id="running-workflow">Running workflow</h1>
<h2 id="interactive-job-submissions-in-a-single-machine">Interactive job submissions in a single machine</h2>
<p>For running the workflow, <code>runWF</code> function will execute all the steps store in
the workflow container. The execution will be on a single machine without
submitting to a queuing system of a computer cluster.</p>
<pre><code class="language-r">sal &lt;- runWF(sal)
</code></pre>
<h2 id="parallelization-on-clusters">Parallelization on clusters</h2>
<p>Alternatively, the computation can be greatly accelerated by processing many files
in parallel using several compute nodes of a cluster, where a scheduling/queuing
system is used for load balancing.</p>
<p>The <code>resources</code> list object provides the number of independent parallel cluster
processes defined under the <code>Njobs</code> element in the list. The following example
will run 18 processes in parallel using each 4 CPU cores.
If the resources available on a cluster allow running all 18 processes at the
same time, then the shown sample submission will utilize in a total of 72 CPU cores.</p>
<p>Note, <code>runWF</code> can be used with most queueing systems as it is based on utilities
from the <code>batchtools</code> package, which supports the use of template files (<em><code>*.tmpl</code></em>)
for defining the run parameters of different schedulers. To run the following
code, one needs to have both a <code>conffile</code> (see <em><code>.batchtools.conf.R</code></em> samples <a href="https://mllg.github.io/batchtools/">here</a>)
and a <code>template</code> file (see <em><code>*.tmpl</code></em> samples <a href="https://github.com/mllg/batchtools/tree/master/inst/templates">here</a>)
for the queueing available on a system. The following example uses the sample
<code>conffile</code> and <code>template</code> files for the Slurm scheduler provided by this package.</p>
<p>The resources can be appended when the step is generated, or it is possible to
add these resources later, as the following example using the <code>addResources</code>
function:</p>
<pre><code class="language-r">resources &lt;- list(conffile=&quot;.batchtools.conf.R&quot;,
                  template=&quot;batchtools.slurm.tmpl&quot;, 
                  Njobs=18, 
                  walltime=120, ## minutes
                  ntasks=1,
                  ncpus=4, 
                  memory=1024, ## Mb
                  partition = &quot;short&quot;
                  )
sal &lt;- addResources(sal, c(&quot;hisat2_mapping&quot;), resources = resources)
sal &lt;- runWF(sal)
</code></pre>
<h2 id="visualize-workflow">Visualize workflow</h2>
<p><em><code>systemPipeR</code></em> workflows instances can be visualized with the <code>plotWF</code> function.</p>
<pre><code class="language-r">plotWF(sal, rstudio = TRUE)
</code></pre>
<h2 id="checking-workflow-status">Checking workflow status</h2>
<p>To check the summary of the workflow, we can use:</p>
<pre><code class="language-r">sal
statusWF(sal)
</code></pre>
<h2 id="accessing-logs-report">Accessing logs report</h2>
<p><em><code>systemPipeR</code></em> compiles all the workflow execution logs in one central location,
making it easier to check any standard output (<code>stdout</code>) or standard error
(<code>stderr</code>) for any command-line tools used on the workflow or the R code stdout.</p>
<pre><code class="language-r">sal &lt;- renderLogs(sal)
</code></pre>
<h2 id="session-info">Session Info</h2>
<p>This is the session information for rendering this report. To access the session information
of workflow running, check HTML report of <code>renderLogs</code>.</p>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.3.0 (2023-04-21)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 11 (bullseye)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: America/Los_Angeles
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils    
## [6] datasets  methods   base     
## 
## other attached packages:
##  [1] systemPipeR_2.6.0           ShortRead_1.58.0           
##  [3] GenomicAlignments_1.36.0    SummarizedExperiment_1.30.0
##  [5] Biobase_2.60.0              MatrixGenerics_1.12.0      
##  [7] matrixStats_0.63.0          BiocParallel_1.34.0        
##  [9] Rsamtools_2.16.0            Biostrings_2.68.0          
## [11] XVector_0.40.0              GenomicRanges_1.52.0       
## [13] GenomeInfoDb_1.36.0         IRanges_2.34.0             
## [15] S4Vectors_0.38.0            BiocGenerics_0.46.0        
## [17] BiocStyle_2.28.0           
## 
## loaded via a namespace (and not attached):
##  [1] gtable_0.3.3            xfun_0.39              
##  [3] bslib_0.4.2             hwriter_1.3.2.1        
##  [5] ggplot2_3.4.2           htmlwidgets_1.6.2      
##  [7] latticeExtra_0.6-30     lattice_0.21-8         
##  [9] generics_0.1.3          vctrs_0.6.2            
## [11] tools_4.3.0             bitops_1.0-7           
## [13] parallel_4.3.0          tibble_3.2.1           
## [15] fansi_1.0.4             pkgconfig_2.0.3        
## [17] Matrix_1.5-4            RColorBrewer_1.1-3     
## [19] lifecycle_1.0.3         GenomeInfoDbData_1.2.10
## [21] stringr_1.5.0           compiler_4.3.0         
## [23] deldir_1.0-6            munsell_0.5.0          
## [25] codetools_0.2-19        htmltools_0.5.5        
## [27] sass_0.4.5              RCurl_1.98-1.12        
## [29] yaml_2.3.7              pillar_1.9.0           
## [31] crayon_1.5.2            jquerylib_0.1.4        
## [33] DelayedArray_0.25.0     cachem_1.0.7           
## [35] tidyselect_1.2.0        digest_0.6.31          
## [37] stringi_1.7.12          dplyr_1.1.2            
## [39] bookdown_0.33           fastmap_1.1.1          
## [41] grid_4.3.0              colorspace_2.1-0       
## [43] cli_3.6.1               magrittr_2.0.3         
## [45] utf8_1.2.3              scales_1.2.1           
## [47] rmarkdown_2.21          jpeg_0.1-10            
## [49] interp_1.1-4            blogdown_1.16          
## [51] png_0.1-8               evaluate_0.20          
## [53] knitr_1.42              rlang_1.1.1            
## [55] Rcpp_1.0.10             glue_1.6.2             
## [57] formatR_1.14            BiocManager_1.30.20    
## [59] jsonlite_1.8.4          R6_2.5.1               
## [61] zlibbioc_1.46.0
</code></pre>
<h1 id="funding">Funding</h1>
<p>This project is funded by NSF award <a href="https://www.nsf.gov/awardsearch/showAward?AWD_ID=1661152">ABI-1661152</a>.</p>
<h1 id="references">References</h1>
<div id="refs" class="references hanging-indent">
<div id="ref-Bolger2014-yr">
<p>Bolger, Anthony M, Marc Lohse, and Bjoern Usadel. 2014. “Trimmomatic: A Flexible Trimmer for Illumina Sequence Data.” <em>Bioinformatics</em> 30 (15): 2114–20.</p>
</div>
<div id="ref-H_Backman2016-bt">
<p>H Backman, Tyler W, and Thomas Girke. 2016. “systemPipeR: NGS workflow and report generation environment.” <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">https://doi.org/10.1186/s12859-016-1241-0</a>.</p>
</div>
<div id="ref-Howard2013-fq">
<p>Howard, Brian E, Qiwen Hu, Ahmet Can Babaoglu, Manan Chandra, Monica Borghi, Xiaoping Tan, Luyan He, et al. 2013. “High-Throughput RNA Sequencing of Pseudomonas-Infected Arabidopsis Reveals Hidden Transcriptome Complexity and Novel Splice Variants.” <em>PLoS One</em> 8 (10): e74183. <a href="https://doi.org/10.1371/journal.pone.0074183">https://doi.org/10.1371/journal.pone.0074183</a>.</p>
</div>
<div id="ref-Kim2015-ve">
<p>Kim, Daehwan, Ben Langmead, and Steven L Salzberg. 2015. “HISAT: A Fast Spliced Aligner with Low Memory Requirements.” <em>Nat. Methods</em> 12 (4): 357–60.</p>
</div>
<div id="ref-Lawrence2013-kt">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8): e1003118. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.</p>
</div>
<div id="ref-Robinson2010-uk">
<p>Robinson, M D, D J McCarthy, and G K Smyth. 2010. “EdgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.</p>
</div>
</div>



      </main>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2024 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../../js/prism.js'></script>



  </body>
</html>