<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>Project Data Management and Run Instructions | GEN242</title><meta property="og:title" content="Project Data Management and Run Instructions" />
<meta property="og:description" content="Big data space on HPCC All larger data sets of the course projects will be organized in a big data space under /bigdata/gen242/&lt;user_name&gt;. Within this space, each student will work in a subdirectory named after their project:
 /bigdata/gen242/&lt;user_name&gt;/&lt;github_user_name&gt;_project  Project GitHub repositories Students will work on their course projects within GitHub repositories, one for each student. These project repositories are private and have been shared with each student." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/assignments/projects/project_data/" /><meta property="article:section" content="assignments" />

<meta property="article:modified_time" content="2023-06-08T19:34:03-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="Project Data Management and Run Instructions">
<meta itemprop="description" content="Big data space on HPCC All larger data sets of the course projects will be organized in a big data space under /bigdata/gen242/&lt;user_name&gt;. Within this space, each student will work in a subdirectory named after their project:
 /bigdata/gen242/&lt;user_name&gt;/&lt;github_user_name&gt;_project  Project GitHub repositories Students will work on their course projects within GitHub repositories, one for each student. These project repositories are private and have been shared with each student.">
<meta itemprop="dateModified" content="2023-06-08T19:34:03-07:00" />
<meta itemprop="wordCount" content="1979">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Project Data Management and Run Instructions"/>
<meta name="twitter:description" content="Big data space on HPCC All larger data sets of the course projects will be organized in a big data space under /bigdata/gen242/&lt;user_name&gt;. Within this space, each student will work in a subdirectory named after their project:
 /bigdata/gen242/&lt;user_name&gt;/&lt;github_user_name&gt;_project  Project GitHub repositories Students will work on their course projects within GitHub repositories, one for each student. These project repositories are private and have been shared with each student."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Git and GitHub (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux Cluster Basics (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Enrichment Analysis (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Overview (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 VAR-Seq Overview (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics in R (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (S22)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_23/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (S23)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_24/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflow Review (S24)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_25/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Final Discussion (S25)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_26/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Additional Slide Shows (S26)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spvarseq/spvarseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 VAR-Seq Workflow (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T20)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8: R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_dex/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEXseq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/09_drug_target_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Drug-target Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/10_lncrna_orf_discovery/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 lncRNA Discovery
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/class/ludf7verm4t14z"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://docs.github.com/en/education/manage-coursework-with-github-classroom/get-started-with-github-classroom/about-github-classroom"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.9ceee5996ef809a84b6ff16e89814f94.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.9ceee5996ef809a84b6ff16e89814f94.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../assignments/" class="align-left pl-0 pr-2 td-sidebar-link td-sidebar-link__section">Assignments</a>
  </li>
  <ul>
    <li class="collapse show" id="assignments">
      
      
      
      
        
          






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../assignments/homework/" class="align-left pl-0 pr-2 collapsed td-sidebar-link td-sidebar-link__section">Homework</a>
  </li>
  <ul>
    <li class="collapse " id="assignmentshomework">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentshomeworkhw01hw01" href="../../../assignments/homework/hw01/hw01/">HW1: Intro &amp; GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentshomeworkhw02hw02" href="../../../assignments/homework/hw02/hw02/">HW2: Linux &amp; HPC</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentshomeworkhw03hw03" href="../../../assignments/homework/hw03/hw03/">HW3: R Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentshomeworkhw04hw04" href="../../../assignments/homework/hw04/hw04/">HW4: Alignments</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentshomeworkhw04hw4_solutionhw4_key" href="../../../assignments/homework/hw04/hw4_solution/hw4_key/">HW4: Pairwise Alignment Solutions</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentshomeworkhw05hw05" href="../../../assignments/homework/hw05/hw05/">HW5: R Programming</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentshomeworkhw06hw06" href="../../../assignments/homework/hw06/hw06/">HW6: NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentshomeworkhw07hw07" href="../../../assignments/homework/hw07/hw07/">HW6: RNA-Seq Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentshomeworkhw08hw08" href="../../../assignments/homework/hw08/hw08/">HW8: R Markdown</a>
        
      
      
    </li>
  </ul>
</ul>

        
      
      
      
        
          






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../assignments/projects/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Projects</a>
  </li>
  <ul>
    <li class="collapse show" id="assignmentsprojects">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojectsproject_overview" href="../../../assignments/projects/project_overview/">Overview of Projects</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects01_rnaseq_aligners" href="../../../assignments/projects/01_rnaseq_aligners/">RNA-Seq Aligners</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects02_rnaseq_deg" href="../../../assignments/projects/02_rnaseq_deg/">DEG Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects03_cluster_analysis" href="../../../assignments/projects/03_cluster_analysis/">Cluster Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects02_rnaseq_dex" href="../../../assignments/projects/02_rnaseq_dex/">DEG Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects04_scrnaseq_embedding" href="../../../assignments/projects/04_scrnaseq_embedding/">scRNA-Seq Embedding</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects05_chipseq_peakcaller" href="../../../assignments/projects/05_chipseq_peakcaller/">ChIP-Seq Peak Callers</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects06_functional_enrichment" href="../../../assignments/projects/06_functional_enrichment/">Functional Enrichment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects07_motif_enrichment" href="../../../assignments/projects/07_motif_enrichment/">Motif Enrichment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects09_drug_target_analysis" href="../../../assignments/projects/09_drug_target_analysis/">Drug-target analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects08_genome_graphics" href="../../../assignments/projects/08_genome_graphics/">Genome Graphics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojects10_lncrna_orf_discovery" href="../../../assignments/projects/10_lncrna_orf_discovery/">lncRNAs and ORFs</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-assignmentsprojectsproject_data" href="../../../assignments/projects/project_data/">Project Data</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentsprojectshelper_codealignersstar_test" href="../../../assignments/projects/helper_code/aligners/star_test/"></a>
        
      
      
    </li>
  </ul>
</ul>

        
      
      
      
        
          






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../assignments/presentations/" class="align-left pl-0 pr-2 collapsed td-sidebar-link td-sidebar-link__section">Presentations</a>
  </li>
  <ul>
    <li class="collapse " id="assignmentspresentations">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-assignmentspresentationspaper_presentations" href="../../../assignments/presentations/paper_presentations/">Paper Presentations</a>
        
      
      
    </li>
  </ul>
</ul>

        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/assignments/Projects/project_data.md" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View source code</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/assignments/Projects/project_data.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/assignments/Projects/project_data.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=Project%20Data%20Management%20and%20Run%20Instructions" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#big-data-space-on-hpcc">Big data space on HPCC</a></li>
    <li><a href="#project-github-repositories">Project GitHub repositories</a></li>
    <li><a href="#generate-workflow-environment-with-real-project-data">Generate workflow environment with real project data</a></li>
    <li><a href="#download-of-project-data">Download of project data</a>
      <ul>
        <li><a href="#fastq-files-from-sra">FASTQ files from SRA</a>
          <ul>
            <li><a href="#choose-fastq-data-for-your-project">Choose FASTQ data for your project</a></li>
            <li><a href="#load-libraries-and-modules">Load libraries and modules</a></li>
            <li><a href="#define-download-function">Define download function</a></li>
            <li><a href="#run-download">Run download</a></li>
          </ul>
        </li>
        <li><a href="#download-reference-genome-and-annotation">Download reference genome and annotation</a></li>
      </ul>
    </li>
    <li><a href="#workflow-rmd-file">Workflow Rmd file</a></li>
    <li><a href="#recommendations-for-running-workflows">Recommendations for running workflows</a>
      <ul>
        <li><a href="#run-instructions">Run instructions</a></li>
        <li><a href="#modify-a-workflow">Modify a workflow</a></li>
        <li><a href="#adding-steps-to-a-workflow">Adding steps to a workflow</a></li>
        <li><a href="#submit-workflow-from-command-line-to-cluster">Submit workflow from command-line to cluster</a></li>
      </ul>
    </li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		










<li class="breadcrumb-item" >
	<a href="../../../assignments/">Assignments</a>
</li>




<li class="breadcrumb-item" >
	<a href="../../../assignments/projects/">Projects</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../assignments/projects/project_data/">Project Data</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>Project Data Management and Run Instructions</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 10 minute read</p>
           
	<p><br></br></p>
<h2 id="big-data-space-on-hpcc">Big data space on HPCC</h2>
<p>All larger data sets of the course projects will be organized in a big data space under
<code>/bigdata/gen242/&lt;user_name&gt;</code>. Within this space, each student will work in a subdirectory named after their project:</p>
<ul>
<li><code>/bigdata/gen242/&lt;user_name&gt;/&lt;github_user_name&gt;_project</code></li>
</ul>
<h2 id="project-github-repositories">Project GitHub repositories</h2>
<p>Students will work on their course projects within GitHub repositories, one for each student.
These project repositories are private and have been shared with each student.
To populate a course project with an initial project workflow, please follow the instructions
given in the following section.</p>
<h2 id="generate-workflow-environment-with-real-project-data">Generate workflow environment with real project data</h2>
<ol>
<li>Log in to the HPCC cluster and set your working directory to <code>bigdata</code> or (<code>/bigdata/gen242/&lt;user_name&gt;</code>)</li>
<li>Clone the GitHub repository for your project with <code>git clone ...</code> (URLs listed <a href="https://bit.ly/3tJ3KuZ">here</a>) and then <code>cd</code> into this directory. As mentioned above, the project GitHub repos follow this naming convention: <code>&lt;github_user_name&gt;_project</code>.</li>
<li>Generate the workflow environment for your project on the HPCC cluster with <code>genWorkenvir</code> from <code>systemPipeRdata</code>.</li>
<li>Next, <code>cd</code> into the directory of your workflow, delete its default <code>data</code> and <code>results</code> directories, and then substitute them with empty directories outside of your project GitHub repos as follows (<code>&lt;workflow&gt;</code> needs to be replaced with actual workflow name):
<pre><code class="language-sh">mkdir ../../&lt;workflow&gt;_data
mkdir ../../&lt;workflow&gt;_results
</code></pre>
</li>
<li>Within your workflow directory create symbolic links pointing to the new directories created in the previous step. For instance, the projects using the RNA-Seq workflow should create the symbolic links for their <code>data</code> and <code>results</code> directories like this (<code>&lt;user_name&gt;</code> and <code>&lt;workflow&gt;</code> needs to be replaced with your user name and workflow name):
<pre><code class="language-sh">ln -s /bigdata/gen242/&lt;user_name&gt;/&lt;workflow&gt;_data data
ln -s /bigdata/gen242/&lt;user_name&gt;/&lt;workflow&gt;_results results
</code></pre>
</li>
<li>Add the workflow directory to the GitHub repository of your project with <code>git add -A</code> and then run <code>commit</code> and <code>push</code> as outlined in the GitHub instructions of this course <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/github/github/#github-basics-from-command-line">here</a>. After this check whether the workflow directory and its content shows up on your project&rsquo;s online repos on GitHub. Very important: make sure that the <code>data</code> and <code>results</code> are empty at this point. If not investigate why and fix the problem in the corresponding step above.</li>
<li>Download the FASTQ files of your project with <code>getSRAfastq</code> (see below) to the <code>data</code> directory of your project.</li>
<li>Generate a proper <code>targets</code> file for your project where the first column(s) point(s) to the downloaded FASTQ files. In addition, provide sample names matching the experimental design (columns: <code>SampleNames</code> and <code>Factor</code>). More details about the structure of targets files are provided <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/systempiper/systempiper/#structure-of-targets-file">here</a>. Ready to use targets files for both the RNA-Seq and ChIP-Seq project can be downloaded as tab separated (TSV) files from <a href="https://github.com/tgirke/GEN242/tree/main/content/en/assignments/Projects/targets_files">here</a>. Alternatively, one can download the corresponding Google Sheets with the <code>read_sheet</code> function from the <code>googlesheets4</code> package (<a href="https://bit.ly/2QH19Ry">RNA-Seq GSheet</a> and <a href="https://bit.ly/2QFjTAV">ChIP-Seq GSheet</a>).</li>
<li>Inspect and adjust the <code>.param</code> files you will be using. For instance, make sure the software modules you are loading and the path to the reference genome are correct.</li>
<li>Every time you start working on your project you <code>cd</code> into the directory of the repository and then run <code>git pull</code> to get the latest change. When you are done, you commit and push your changes back to GitHub with <code>git commit -am &quot;some edits&quot;; git push -u origin main</code>.</li>
</ol>
<h2 id="download-of-project-data">Download of project data</h2>
<p>After logging in to one of the computer nodes via <a href="https://bit.ly/3MD40DW"><code>srun</code></a>, open R from within the GitHub repository of your project and then run the following code section, but only those that apply to your project.</p>
<h3 id="fastq-files-from-sra">FASTQ files from SRA</h3>
<h4 id="choose-fastq-data-for-your-project">Choose FASTQ data for your project</h4>
<ul>
<li>The FASTQ files for the ChIP-Seq project are from SRA study <a href="http://www.ncbi.nlm.nih.gov/sra?term=SRP002174">SRP002174</a> (<a href="http://www.ncbi.nlm.nih.gov/pubmed/20360106">Kaufman et al. 2010</a>)</li>
</ul>
<pre><code class="language-r">sraidv &lt;- paste(&quot;SRR0388&quot;, 45:51, sep=&quot;&quot;) 
</code></pre>
<ul>
<li>The FASTQ files for the RNA-Seq project are from SRA study <a href="http://www.ncbi.nlm.nih.gov/sra?term=SRP010938">SRP010938</a> (<a href="http://www.ncbi.nlm.nih.gov/pubmed/24098335">Howard et al. 2013</a>)</li>
</ul>
<pre><code class="language-r">sraidv &lt;- paste(&quot;SRR4460&quot;, 27:44, sep=&quot;&quot;)
</code></pre>
<h4 id="load-libraries-and-modules">Load libraries and modules</h4>
<pre><code class="language-r">library(systemPipeR)                                                                                                                                                                
moduleload(&quot;sratoolkit/3.0.0&quot;)                                                                                                                                                      
system(&quot;vdb-config --prefetch-to-cwd&quot;) # sets download default to current directory                                                                                          
# system('prefetch --help') # helps to speed up fastq-dump
# system('vdb-config -i') # allows to change SRA Toolkit configuration; instructions are here: https://bit.ly/3lzfU4P
# system('fastq-dump --help') # below uses this one for backwards compatibility                                                                                                     
# system('fasterq-dump --help') # faster than fastq-dump
</code></pre>
<h4 id="define-download-function">Define download function</h4>
<p>The following function downloads and extracts the FASTQ files for each project from SRA.
Internally, it uses the <code>prefetch</code> and <code>fastq-dump</code> utilities of the SRA Toolkit from NCBI.
The faster <code>fasterq-dump</code> alternative (see comment line below) is not used here for historical reasons.</p>
<pre><code class="language-r">getSRAfastq &lt;- function(sraid, threads=1) {                                                                                                                                         
    system(paste(&quot;prefetch&quot;, sraid)) # makes download faster                                                                                                                        
    system(paste(&quot;vdb-validate&quot;, sraid)) # checks integrity of the downloaded SRA file                                                                   
    system(paste(&quot;fastq-dump --split-files --gzip&quot;, sraid)) # gzip option makes it slower but saves storage space                                                                   
    # system(paste(&quot;fasterq-dump --threads 4 --split-files --progress &quot;, sraid, &quot;--outdir .&quot;)) # Faster alternative to fastq-dump                                                   
    unlink(x=sraid, recursive = TRUE, force = TRUE) # deletes sra download directory                                                                                                
}    
</code></pre>
<p>To stop the loop after a failure is detected by <code>vdb-validate</code>, use <code>&amp;&amp;</code> operator like this: <code>prefetch sraid &amp;&amp; vdb-validate sraid &amp;&amp; fastq-dump sraid</code>.</p>
<h4 id="run-download">Run download</h4>
<p>Note the following performs the download in serialized mode for the chosen data set and saves the extracted FASTQ files to
the current working directory.</p>
<pre><code class="language-r">mydir &lt;- getwd(); setwd(&quot;data&quot;)
for(i in sraidv) getSRAfastq(sraid=i)
setwd(mydir)
</code></pre>
<p>Alternatively, the download can be performed in parallelized mode with <code>BiocParallel</code>. Please run this version only on one of the compute nodes.</p>
<pre><code class="language-r">mydir &lt;- getwd(); setwd(&quot;data&quot;)
# bplapply(sraidv, getSRAfastq, BPPARAM = MulticoreParam(workers=4))
setwd(mydir)
</code></pre>
<h3 id="download-reference-genome-and-annotation">Download reference genome and annotation</h3>
<p>The following <code>downloadRefs</code> function downloads the <em>Arabidopsis thaliana</em> genome sequence and GFF file from the <a href="ftp://ftp.arabidopsis.org/home/tair/Genes/TAIR10_genome_release/">TAIR FTP site</a>.
It also assigns consistent chromosome identifiers to make them the same among both the genome sequence and the GFF file. This is
important for many analysis routines such as the read counting in the RNA-Seq workflow.</p>
<pre><code class="language-r">downloadRefs &lt;- function(rerun=FALSE) {
    if(rerun==TRUE) {
        library(Biostrings)
        download.file(&quot;https://www.arabidopsis.org/download_files/Genes/TAIR10_genome_release/TAIR10_chromosome_files/TAIR10_chr_all.fas.gz&quot;, &quot;./data/tair10.fasta.gz&quot;)
        R.utils::gunzip(&quot;./data/tair10.fasta.gz&quot;)
        dna &lt;- readDNAStringSet(&quot;./data/tair10.fasta&quot;)
        names(dna) &lt;- paste(rep(&quot;Chr&quot;, 7), c(1:5, &quot;M&quot;, &quot;C&quot;), sep=&quot;&quot;) # Fixes chromomse ids
        writeXStringSet(dna, &quot;./data/tair10.fasta&quot;)
        download.file(&quot;https://www.arabidopsis.org/download_files/Genes/TAIR10_genome_release/TAIR10_gff3/TAIR10_GFF3_genes.gff&quot;, &quot;./data/tair10.gff&quot;)
        download.file(&quot;https://www.arabidopsis.org/download_files/Genes/TAIR10_genome_release/TAIR10_functional_descriptions&quot;, &quot;./data/tair10_functional_descriptions&quot;)
        }
}
</code></pre>
<p>After importing/sourcing the above function, execute it as follows:</p>
<pre><code class="language-r">downloadRefs(rerun=TRUE) 
</code></pre>
<h2 id="workflow-rmd-file">Workflow Rmd file</h2>
<p>To run the actual data analysis workflows, the RNA-Seq project can use the <code>systemPipeRNAseq.Rmd</code> file obtained from the <code>genWorkenvir(workflow='rnaseq')</code> call directly. However,
the ChIP-Seq group should use the Rmd linked on top right corner of the ChIP-Seq tutorial <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/spchipseq/spchipseq/">here</a>
and then name the downloaded file <code>systemPipeChIPseq.Rmd</code>. To simplify this, the ChIP-Seq group members can run from the command-line within their <code>chipseq</code> workflow directory the
following download command.</p>
<pre><code class="language-r">wget https://raw.githubusercontent.com/tgirke/GEN242/main/static/custom/spWFtemplates/spchipseq.Rmd -O systemPipeChIPseq.Rmd
</code></pre>
<p>This will assign the proper file name and overwrite the preloaded version of this file that has the same name.</p>
<h2 id="recommendations-for-running-workflows">Recommendations for running workflows</h2>
<h3 id="run-instructions">Run instructions</h3>
<p>The following provides recommendations and additional options to consider for
running and modifying workflows. This also includes parallelization settings
for the specific data used by the class projects. Note, additional details can
be found in this and other sections of the workflow introduction tutorial
<a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/systempiper/systempiper/#loading-workflows-from-an-r-markdown">here</a>.
Importantly, the following should be run from within an <a href="https://bit.ly/3MD40DW"><code>srun</code></a> session.</p>
<pre><code class="language-r">library(systemPipeR)                                                                                                                                                                
sal &lt;- SPRproject() # when running a WF for first time                                                                                                                                      
sal                                                                                                                                                                                 
sal &lt;- importWF(sal, file_path = &quot;systemPipeRNAseq.Rmd&quot;) # populates sal with WF steps defined in Rmd                                                                                                                      
sal
sal &lt;- SPRproject(resume=TRUE) # when restarting a WF, skip above steps and resume WF with this command                                                                                                                                               
getRversion() # should be 4.2.2. Note, R version can be changed with `module load ...`                                                                                                                                                     
system(&quot;hostname&quot;) # should return number of a compute node; if not close Nvim-R session, log in to a compute node with srun and then restart Nvim-R session                                                                                                                                                                     
# sal &lt;- runWF(sal) # runs WF serialized. Not recommended since this will take much longer than parallel mode introduced below by taking advantage of resource allocation
resources &lt;- list(conffile=&quot;.batchtools.conf.R&quot;,                                                                                                                                    
                  template=&quot;batchtools.slurm.tmpl&quot;,                                                                                                                                 
                  Njobs=18, # chipseq should use here number of fastq files (7 or 8)                                                                                                                                                        
                  walltime=180, ## minutes                                                                                                                                          
                  ntasks=1,                                                                                                                                                         
                  ncpus=4,                                                                                                                                                          
                  memory=4096, ## Mb                                                                                                                                                
                  partition = &quot;gen242&quot;                                                                                                                                              
                  )                                                                                                                                                                 
## For RNA-Seq project use:
sal &lt;- addResources(sal, step = c(&quot;preprocessing&quot;, &quot;trimming&quot;, &quot;hisat2_mapping&quot;), resources = resources) # parallelizes time consuming computations assigned to `step` argument                                                                           
## For ChIP-Seq project use:
sal &lt;- addResources(sal, step = c(&quot;preprocessing&quot;, &quot;bowtie2_alignment&quot;), resources = resources)
sal &lt;- runWF(sal) # runs entire workflow; specific steps can be executed by assigning their corresponding position numbers within the workflow to the `steps` argument (see ?runWF)                                                                                                                                                               
sal &lt;- renderReport(sal) # after workflow has completed render Rmd to HTML report (default name is SPR_Report.html) and view it via web browser which requires symbolic link in your ~/.html folder. 
rmarkdown::render(&quot;systemPipeRNAseq.Rmd&quot;, clean=TRUE, output_format=&quot;BiocStyle::html_document&quot;) # Alternative approach for rendering report from Rmd file instead of sal object
</code></pre>
<h3 id="modify-a-workflow">Modify a workflow</h3>
<p>If needed one can modify existing workflow steps in a pre-populated <code>SYSargsList</code> object, and potentially already executed WF, with the <code>replaceStep(sal) &lt;-</code> replacement function.
The following gives an example where step number 3 in a <code>SYSargsList</code> (sal) object will be updated with modified or new code. Note, this is a generalized example where the user
needs to insert the code lines and also adjust the values assigned to the arguments: <code>step_name</code> and <code>dependency</code>. Additional details on this topic are available in
the corresponding section of <code>systemPipeR</code>&rsquo;s introductory tutorial <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/systempiper/systempiper/#how-to-modify-workflows">here</a>.</p>
<pre><code class="language-r">replaceStep(sal, step=3) &lt;- LineWise(                                                                                                                                                        
    code = {                                                                                                                                                                        
        &lt;&lt; my modified code lines &gt;&gt;
        },                                                                                                                                                                          
    step_name = &lt;&lt; &quot;my_step_name&quot; &gt;&gt;,                                                                                                                                                        
    dependency = &lt;&lt; &quot;my_dependency&quot; &gt;&gt;)
</code></pre>
<p>Since <code>step_names</code> need to be unique, one should avoid using the same
<code>step_name</code> as before. If the previous name is used, a default name will be
assigned. Rerunning the assignment will then allow to assign the previous name. This
behavior is enforced for version tracking. Subsequently, one can view and check
the code changes with <code>codeLine()</code>, and then rerun the corresponding step (here
3) as follows:</p>
<pre><code class="language-r">codeLine(stepsWF(sal)$my_step_name)
runWF(sal, steps=3)
</code></pre>
<p>Note, any step in a workflow can only be run in isolation if its expected input exists (see <code>dependency</code>).</p>
<h3 id="adding-steps-to-a-workflow">Adding steps to a workflow</h3>
<p>New steps can be added to the Rmd file of a workflow by inserting new R Markdown code chunks starting and ending with the usual <code>appendStep&lt;-</code> syntax and then creating a new
<code>SYSargsList</code> instance with <code>importWF</code> that contain the new step(s). To add steps to a pre-populated <code>SYSargsList</code> object, one can use the <code>after</code> argument of the <code>appendStep&lt;-</code>
function. The following example will add a new step after position 3 to the corresponding <code>sal</code> object. This can be useful if a longer workflow has already been completed and
one only wants to make some refinements without re-running the entire workflow.</p>
<pre><code class="language-r">appendStep(sal, after=3) &lt;- &lt;&lt; my_step_code &gt;&gt;
</code></pre>
<h3 id="submit-workflow-from-command-line-to-cluster">Submit workflow from command-line to cluster</h3>
<p>In addition to running workflows within interactive R sessions, after logging in to a computer node with <code>srun</code>, one
can execute them entirely from the command-line by including the relevant workflow run instructions in an R script.
The R script can then be submitted via a Slurm submission script to the cluster. The following gives an example for the
RNA-Seq workflow (ChIP-Seq version requires only minor adjustments). Additional details on this topic are available in
the corresponding section of <code>systemPipeR</code>&rsquo;s introductory tutorial
<a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/systempiper/systempiper/#run-from-command-line-without-cluster">here</a>.</p>
<ul>
<li>R script: <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/static/custom/spWFtemplates/cl_sbatch_run/wf_run_script.R">wf_run_script.R</a></li>
<li>Slurm submission script: <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/static/custom/spWFtemplates/cl_sbatch_run/wf_run_script.sh">wf_run_script.sh</a></li>
</ul>
<p>To test this out, users can generate in their user account of the cluster a workflow environment populated with the toy data
as outlined <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/sprnaseq/sprnaseq/#workflow-environment">here</a>). After
this, it is best to create within the workflow directory a subdirectory, e.g. called <code>cl_sbatch_run</code>, and then save the above
two files to this subdirectory. Next, the parameters in both files need to be adjusted to match the type of workflow and
the required computing resources. This includes the name of the <code>Rmd</code> file and scheduler resource settings such as:
<code>partition</code>, <code>Njobs</code>, <code>walltime</code>, <code>memory</code>, etc. After all relevant settings have been set correctly, one can
execute the workflow with <code>sbatch</code> within the <code>cl_sbatch_run</code> directory as follows:</p>
<pre><code class="language-sh">sbatch wf_run_script.sh
</code></pre>
<p>After the submission to the cluster, one usually should check its status and progress with <code>squeue -u &lt;username&gt;</code> as well a
by monitoring the content of the <code>slurm-&lt;jobid&gt;.out</code> file generated by the scheduler in the same directory. This file
contains most of the <code>STDOUT</code> and <code>STDERROR</code> generated by a cluster job. Once everything is working on the toy data,
users can run the workflow on the real data the same way.</p>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2023-06-08: <a  href="https://github.com/tgirke/GEN242//commit/acab87dd5c763092a3ecd9c1441f446974d32850">some edits (acab87dd5)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2024 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
