<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>ChIP-Seq Workflow Template | GEN242</title><meta property="og:title" content="ChIP-Seq Workflow Template" />
<meta property="og:description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Introduction The following analyzes the ChIP-Seq data from Kaufman et al." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/spchipseq/spchipseq/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2023-05-09T18:03:28-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="ChIP-Seq Workflow Template">
<meta itemprop="description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Introduction The following analyzes the ChIP-Seq data from Kaufman et al.">
<meta itemprop="dateModified" content="2023-05-09T18:03:28-07:00" />
<meta itemprop="wordCount" content="5154">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="ChIP-Seq Workflow Template"/>
<meta name="twitter:description" content="pre code { white-space: pre !important; overflow-x: scroll !important; word-break: keep-all !important; word-wrap: initial !important; }  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Introduction The following analyzes the ChIP-Seq data from Kaufman et al."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Git and GitHub (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux Cluster Basics (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Overview (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Enrichment Analysis (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics in R (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (S22)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_23/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflow Review (S23)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_24/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Final Discussion (S24)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_25/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Additional Slide Shows (S25)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T19)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8: R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_dex/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEXseq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/09_drug_target_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Drug-target Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/10_lncrna_orf_discovery/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 lncRNA Discovery
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2023/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.5f8e4b39e1ad8d5a8572a90741c99dea.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.5f8e4b39e1ad8d5a8572a90741c99dea.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../tutorials/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Tutorials</a>
  </li>
  <ul>
    <li class="collapse show" id="tutorials">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsgithubgithub" href="../../../tutorials/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialslinuxlinux" href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrbasicsrbasics" href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrprogrammingrprogramming" href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrparallelrparallel" href="../../../tutorials/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrsequencesrsequences" href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempiper" href="../../../tutorials/systempiper/systempiper/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseq" href="../../../tutorials/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-tutorialsspchipseqspchipseq" href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsscrnaseqscrnaseq" href="../../../tutorials/scrnaseq/scrnaseq/">scRNA-Seq Embedding Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrmarkdownrmarkdown" href="../../../tutorials/rmarkdown/rmarkdown/">R Markdown Tutorial</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrfearfea" href="../../../tutorials/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrclusteringrclustering" href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrgraphicsrgraphics" href="../../../tutorials/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappslezhangshinyapps" href="../../../tutorials/shinyappslezhang/shinyapps/">Shiny Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappsshinyapps" href="../../../tutorials/shinyapps/shinyapps/">Shiny Web Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrpackagesrpackages" href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsdplyrdplyr" href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialscmdtocwlcmdtocwl" href="../../../tutorials/cmdtocwl/cmdtocwl/">Automate Creation of CWL Instructions</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/spchipseq/spchipseq.Rmd" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/tutorials/spchipseq/spchipseq.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/tutorials/spchipseq/spchipseq.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=ChIP-Seq%20Workflow%20Template" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#introduction">Introduction</a>
      <ul>
        <li><a href="#experimental-design">Experimental design</a></li>
        <li><a href="#workflow-environment">Workflow environment</a></li>
        <li><a href="#load-packages">Load packages</a></li>
        <li><a href="#import-custom-functions">Import custom functions</a></li>
        <li><a href="#experiment-definition-provided-by-targets-file">Experiment definition provided by <code>targets</code> file</a></li>
      </ul>
    </li>
    <li><a href="#workflow-steps">Workflow steps</a>
      <ul>
        <li><a href="#required-packages-and-resources">Required packages and resources</a></li>
        <li><a href="#read-preprocessing">Read preprocessing</a>
          <ul>
            <li><a href="#read-quality-filtering-and-trimming">Read quality filtering and trimming</a></li>
            <li><a href="#fastq-quality-report">FASTQ quality report</a></li>
          </ul>
        </li>
        <li><a href="#alignments">Alignments</a>
          <ul>
            <li><a href="#read-mapping-with-bowtie2">Read mapping with <code>Bowtie2</code></a></li>
            <li><a href="#read-and-alignment-stats">Read and alignment stats</a></li>
          </ul>
        </li>
        <li><a href="#create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</a></li>
        <li><a href="#peak-calling-with-macs2">Peak calling with MACS2</a></li>
        <li><a href="#merge-bam-files-of-replicates-prior-to-peak-calling">Merge BAM files of replicates prior to peak calling</a>
          <ul>
            <li><a href="#peak-calling-with-inputreference-sample">Peak calling with input/reference sample</a></li>
          </ul>
        </li>
        <li><a href="#annotate-peaks-with-genomic-context">Annotate peaks with genomic context</a>
          <ul>
            <li><a href="#annotation-with-chipseeker-package">Annotation with <code>ChIPseeker</code> package</a></li>
          </ul>
        </li>
        <li><a href="#count-reads-overlapping-peaks">Count reads overlapping peaks</a></li>
        <li><a href="#differential-binding-analysis">Differential binding analysis</a></li>
        <li><a href="#go-term-enrichment-analysis">GO term enrichment analysis</a>
          <ul>
            <li><a href="#obtain-gene-to-go-mappings">Obtain gene-to-GO mappings</a></li>
            <li><a href="#go-term-enrichment-analysis-1">GO term enrichment analysis</a></li>
            <li><a href="#go-term-enrichment-analysis-2">GO term enrichment analysis</a></li>
          </ul>
        </li>
        <li><a href="#motif-analysis">Motif analysis</a>
          <ul>
            <li><a href="#parse-dna-sequences-of-peak-regions-from-genome">Parse DNA sequences of peak regions from genome</a></li>
            <li><a href="#motif-discovery-with-bcrank">Motif discovery with <code>BCRANK</code></a></li>
          </ul>
        </li>
        <li><a href="#version-information">Version Information</a></li>
      </ul>
    </li>
    <li><a href="#running-workflow">Running workflow</a>
      <ul>
        <li><a href="#interactive-job-submissions-in-a-single-machine">Interactive job submissions in a single machine</a></li>
        <li><a href="#parallelization-on-clusters">Parallelization on clusters</a></li>
        <li><a href="#visualize-workflow">Visualize workflow</a></li>
        <li><a href="#checking-workflow-status">Checking workflow status</a></li>
      </ul>
    </li>
    <li><a href="#technical-report">Technical report</a></li>
    <li><a href="#scientific-report">Scientific report</a></li>
    <li><a href="#funding">Funding</a></li>
    <li><a href="#session-info">Session Info</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../tutorials/">Tutorials</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>ChIP-Seq Workflow Template</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 25 minute read</p>
           
	<script src="../../../rmarkdown-libs/htmlwidgets/htmlwidgets.js"></script>
<link href="../../../rmarkdown-libs/datatables-css/datatables-crosstalk.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/datatables-binding/datatables.js"></script>
<script src="../../../rmarkdown-libs/jquery/jquery-3.6.0.min.js"></script>
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/dt-core/js/jquery.dataTables.min.js"></script>
<link href="../../../rmarkdown-libs/crosstalk/css/crosstalk.min.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/crosstalk/js/crosstalk.min.js"></script>
<script src="../../../rmarkdown-libs/htmlwidgets/htmlwidgets.js"></script>
<link href="../../../rmarkdown-libs/datatables-css/datatables-crosstalk.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/datatables-binding/datatables.js"></script>
<script src="../../../rmarkdown-libs/jquery/jquery-3.6.0.min.js"></script>
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="../../../rmarkdown-libs/dt-core/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/dt-core/js/jquery.dataTables.min.js"></script>
<link href="../../../rmarkdown-libs/crosstalk/css/crosstalk.min.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/crosstalk/js/crosstalk.min.js"></script>
<script src="../../../rmarkdown-libs/htmlwidgets/htmlwidgets.js"></script>
<link href="../../../rmarkdown-libs/vizjs/plotwf.css" rel="stylesheet" />
<script src="../../../rmarkdown-libs/vizjs/viz.js"></script>
<script src="../../../rmarkdown-libs/vizjs/full.render.js"></script>
<script src="../../../rmarkdown-libs/dom_to_image/dom_to_image.js"></script>
<link id="plotwf_legend-1-attachment" rel="attachment" href="spchipseq_files/plotwf_legend/plotwf_legend.svg"/>
<script src="../../../rmarkdown-libs/plotwf-binding/plotwf.js"></script>
<!--
- Compile from command-line
Rscript -e "rmarkdown::render('spchipseq.Rmd', c('BiocStyle::html_document'), clean=F); knitr::knit('spchipseq.Rmd', tangle=TRUE)"
-->
<style type="text/css">
pre code {
white-space: pre !important;
overflow-x: scroll !important;
word-break: keep-all !important;
word-wrap: initial !important;
}
</style>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div style="text-align: right">
<p>Source code downloads:    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/static/custom/spWFtemplates/spchipseq.Rmd">.Rmd</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/content/en/tutorials/spchipseq/spchipseq.R">.R</a> ]</p>
</div>
<h2 id="introduction">Introduction</h2>
<p>The following analyzes the ChIP-Seq data from Kaufman et al. (2010) using
for peak calling MACS2 where the uninduced sample serves as input (reference). Prior to running
this analysis the corresponding FASTQ files need to be downloaded following the instructions
<a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/projects/project_data/">here</a>.</p>
<p>For learning purposes one can use the much smaller toy data set and ChIP-Seq <code>Rmd</code>
workflow instance provided by the <code>systemPipeRdata</code> package. This toy workflow instance
can be conveniently obtained by running <code>genWorkenvir(workflow='chipseq')</code> (see
below). The FASTQ data used for this toy instance are the same as for the
RNA-Seq workflow.</p>
<p>For the analysis of the Kaufman et al. (2010) data set (see download <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/projects/project_data/">here</a>)
users want to use the <code>Rmd</code> instance linked from the top right corner of this page.
Additional detail about this is provided <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/projects/project_data/#workflow-rmd-file">here</a>.</p>
<h3 id="experimental-design">Experimental design</h3>
<p>Typically, users want to specify here all information relevant for the
analysis of their NGS study. This includes detailed descriptions of
FASTQ files, experimental design, reference genome, gene annotations,
etc.</p>
<h3 id="workflow-environment">Workflow environment</h3>
<p><font color="red">NOTE: this section</font> describes how to set up the proper
environment (directory structure) for running <code>systemPipeR</code> workflows. After
mastering this task the workflow run instructions <font color="red">can be deleted</font>
since they are not expected to be included in a final HTML/PDF report of a workflow.</p>
<ol>
<li>
<p>If a remote system or cluster is used, then users need to log in to the
remote system first. The following applies to an HPC cluster (<em>e.g.</em> HPCC
cluster).</p>
<p>A terminal application needs to be used to log in to a user’s cluster account. Next, one
can open an interactive session on a computer node with <code>srun --x11</code>. More details about
argument settings for <code>srun</code> are available in this <a href="http://hpcc.ucr.edu/manuals_linux-cluster_jobs.html#partitions">HPCC
manual</a> or
the HPCC section of this website
<a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/linux/linux/#job-submission-with-sbatch">here</a>.
Next, load the R version required for running the workflow with <code>module load</code>. Sometimes it may be necessary to
first unload an active software version before loading another version, <em>e.g.</em> <code>module unload R</code>.</p>
</li>
</ol>
<!-- end list -->
<pre><code class="language-bash">srun --x11 --partition=gen242 --mem=20gb --cpus-per-task 8 --ntasks 1 --time 20:00:00 --pty bash -l
module unload R; module load R/4.2.2
</code></pre>
<ol start="2">
<li>Load a workflow template with the <code>genWorkenvir</code> function. This can be done from the command-line or from within R.
However, only one of the two options needs to be used.</li>
</ol>
<p>From command-line</p>
<pre><code class="language-bash">$ Rscript -e &quot;systemPipeRdata::genWorkenvir(workflow='chipseq')&quot;
$ cd chipseq
</code></pre>
<p>From R</p>
<pre><code class="language-r">library(systemPipeRdata)
genWorkenvir(workflow = &quot;chipseq&quot;)
setwd(&quot;chipseq&quot;)
</code></pre>
<ol start="3">
<li>
<p>Optional: if the user wishes to use another <code>Rmd</code> file than the template
instance provided by the <code>genWorkenvir</code> function, then it can be copied or downloaded
into the root directory of the workflow environment (<em>e.g.</em> with <code>cp</code> or <code>wget</code>).</p>
</li>
<li>
<p>Now one can open from the root directory of the workflow the corresponding
R Markdown script (<em>e.g.</em> systemPipeChIPseq.Rmd) using an R IDE, such as <em>nvim-r</em>,
<em>ESS</em> or RStudio. Subsequently, the workflow can be run as outlined below. For
learning purposes it is recommended to run workflows for the first time interactively.
Once all workflow steps are understood and possibly modified to custom needs,
one can run the workflow from start to finish with a single command using <code>runWF()</code>.</p>
</li>
</ol>
<h3 id="load-packages">Load packages</h3>
<p>The <code>systemPipeR</code> package needs to be loaded to perform the analysis
steps shown in this report (H Backman and Girke 2016). The package allows users
to run the entire analysis workflow interactively or with a single command
while also generating the corresponding analysis report. For details
see <code>systemPipeR's</code> main <a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">vignette</a>.</p>
<pre><code class="language-r">library(systemPipeR)
</code></pre>
<p>To apply workflows to custom data, the user needs to modify the <em><code>targets</code></em> file and if
necessary update the corresponding parameter (<em><code>.cwl</code></em> and <em><code>.yml</code></em>) files.
A collection of pre-generated <em><code>.cwl</code></em> and <em><code>.yml</code></em> files are provided in the <em><code>param/cwl</code></em> subdirectory
of each workflow template. They are also viewable in the GitHub repository of <em><code>systemPipeRdata</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/tree/master/inst/extdata/param/cwl">see
here</a>).
For more information of the structure of the <em>targets</em> file, please consult the documentation
<a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#25_structure_of_targets_file">here</a>. More details about the new parameter files from <code>systemPipeR</code> can be found <a href="http://www.bioconductor.org/packages/release/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#26_structure_of_the_new_param_files_and_construct_sysargs2_container">here</a>.</p>
<h3 id="import-custom-functions">Import custom functions</h3>
<p>Custom functions for the challenge projects can be imported with the source
command from a local R script (here <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/spchipseq/challengeProject_Fct.R">challengeProject_Fct.R</a>). Skip this step if such a script is not available. Alternatively,
these functions can be loaded from a custom R package.</p>
<pre><code class="language-r">source(&quot;challengeProject_Fct.R&quot;)
</code></pre>
<h3 id="experiment-definition-provided-by-targets-file">Experiment definition provided by <code>targets</code> file</h3>
<p>The <code>targets</code> file defines all FASTQ files and sample comparisons of the analysis workflow.
If needed the tab separated (TSV) version of this file can be downloaded from <a href="https://github.com/tgirke/GEN242/tree/main/content/en/assignments/Projects/targets_files">here</a>
and the corresponding Google Sheet is <a href="https://docs.google.com/spreadsheets/d/1w9V3JDOsXR8qW_qNqoXO0E0UR_Oev4k7-o2y6NMDTt8/edit#gid=472150521">here</a>.</p>
<pre><code class="language-r">targetspath &lt;- &quot;targets_chipseq.txt&quot;
targets &lt;- read.delim(targetspath, comment.char = &quot;#&quot;)
knitr::kable(targets)
</code></pre>
<table>
<thead>
<tr>
<th style="text-align:left">FileName</th>
<th style="text-align:left">SampleName</th>
<th style="text-align:left">Factor</th>
<th style="text-align:left">SampleLong</th>
<th style="text-align:right">Experiment</th>
<th style="text-align:left">Date</th>
<th style="text-align:left">SampleReference</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">./data/SRR038845_1.fastq.gz</td>
<td style="text-align:left">AP1_1</td>
<td style="text-align:left">AP1</td>
<td style="text-align:left">APETALA1 Induced</td>
<td style="text-align:right">1</td>
<td style="text-align:left">23-Mar-12</td>
<td style="text-align:left"></td>
</tr>
<tr>
<td style="text-align:left">./data/SRR038846_1.fastq.gz</td>
<td style="text-align:left">AP1_2A</td>
<td style="text-align:left">AP1</td>
<td style="text-align:left">APETALA1 Induced</td>
<td style="text-align:right">1</td>
<td style="text-align:left">23-Mar-12</td>
<td style="text-align:left"></td>
</tr>
<tr>
<td style="text-align:left">./data/SRR038847_1.fastq.gz</td>
<td style="text-align:left">AP1_2B</td>
<td style="text-align:left">AP1</td>
<td style="text-align:left">APETALA1 Induced</td>
<td style="text-align:right">1</td>
<td style="text-align:left">23-Mar-12</td>
<td style="text-align:left"></td>
</tr>
<tr>
<td style="text-align:left">./data/SRR038848_1.fastq.gz</td>
<td style="text-align:left">C_1A</td>
<td style="text-align:left">C</td>
<td style="text-align:left">Control Mock</td>
<td style="text-align:right">1</td>
<td style="text-align:left">23-Mar-12</td>
<td style="text-align:left">AP1_1</td>
</tr>
<tr>
<td style="text-align:left">./data/SRR038849_1.fastq.gz</td>
<td style="text-align:left">C_1B</td>
<td style="text-align:left">C</td>
<td style="text-align:left">Control Mock</td>
<td style="text-align:right">1</td>
<td style="text-align:left">23-Mar-12</td>
<td style="text-align:left">AP1_1</td>
</tr>
<tr>
<td style="text-align:left">./data/SRR038850_1.fastq.gz</td>
<td style="text-align:left">C_2A</td>
<td style="text-align:left">C</td>
<td style="text-align:left">Control Mock</td>
<td style="text-align:right">1</td>
<td style="text-align:left">23-Mar-12</td>
<td style="text-align:left">AP1_2A</td>
</tr>
<tr>
<td style="text-align:left">./data/SRR038851_1.fastq.gz</td>
<td style="text-align:left">C_2B</td>
<td style="text-align:left">C</td>
<td style="text-align:left">Control Mock</td>
<td style="text-align:right">1</td>
<td style="text-align:left">23-Mar-12</td>
<td style="text-align:left">AP1_2B</td>
</tr>
</tbody>
</table>
<h2 id="workflow-steps">Workflow steps</h2>
<p>This tutorial will demonstrate how to either build and run the workflow automatically,
or in an interactive mode by appending each step with the <code>appendStep</code> method. In both
cases the <code>SYSargsList</code> object will be populated with the instructions for running each
workflow step, while supporting both command-line steps as well as line-wise R commands
defined in the corresponding code chunks of this or any <code>Rmd</code> file that has been properly
formatted.</p>
<p>To create a workflow within <em><code>systemPipeR</code></em>, we can start by defining an empty
<code>SYSargsList</code> container. When restarting an existing workflow one
can set <code>resume=TRUE</code> under the <code>SPRproject()</code> function call.</p>
<pre><code class="language-r">library(systemPipeR)
sal &lt;- SPRproject()
</code></pre>
<pre><code>## Creating directory:  /home/tgirke/tmp/GEN242/content/en/tutorials/spchipseq/data 
## Creating directory '/home/tgirke/tmp/GEN242/content/en/tutorials/spchipseq/.SPRproject'
## Creating file '/home/tgirke/tmp/GEN242/content/en/tutorials/spchipseq/.SPRproject/SYSargsList.yml'
</code></pre>
<pre><code class="language-r">sal
</code></pre>
<pre><code>## Instance of 'SYSargsList': 
##  No workflow steps added
</code></pre>
<p>Next, the <code>importWF</code> function will load the entire workflow into the <code>SYSargsList</code> object
(here <code>sal</code>). Subsequently, the <code>runWF()</code> function will run the workflow from start to finish.
If needed, specific workflow steps can be executed by assigning their corresponding
position numbers within the workflow to the <code>steps</code> argument (see <code>?runWF</code>). After completion of the workflow
one can render a scientific analysis report in HTML format with the <code>renderReport()</code>
function that uses R Markdown internally.</p>
<pre><code class="language-r">sal &lt;- importWF(sal, file_path = &quot;systemPipeChIPseq.Rmd&quot;)  ## Import all the Workflow steps
sal
sal &lt;- runWF(sal)  # Runs workflow
sal &lt;- renderReport(sal)  # Renders report
rmarkdown::render(&quot;systemPipeChIPseq.Rmd&quot;, clean = TRUE, output_format = &quot;BiocStyle::html_document&quot;)  # Alternative report rendering
</code></pre>
<h3 id="required-packages-and-resources">Required packages and resources</h3>
<p>The <code>systemPipeR</code> package needs to be loaded (H Backman and Girke 2016).</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(systemPipeR)
}, step_name = &quot;load_SPR&quot;)
</code></pre>
<h3 id="read-preprocessing">Read preprocessing</h3>
<h4 id="read-quality-filtering-and-trimming">Read quality filtering and trimming</h4>
<p>The function <code>preprocessReads</code> allows to apply predefined or custom
read preprocessing functions to all FASTQ files referenced in a
<code>SYSargsList</code> container, such as quality filtering or adapter trimming
routines. The paths to the resulting output FASTQ files are stored in the
<code>outfiles</code> slot of the <code>SYSargsList</code> object. The following example performs adapter trimming with
the <code>trimLRPatterns</code> function from the <code>Biostrings</code> package.
After the trimming step a new targets file is generated (here
<code>targets_trim.txt</code>) containing the paths to the trimmed FASTQ files.
The new targets file can be used for the next workflow step with an updated
<code>SYSargs2</code> instance, <em>e.g.</em> running the NGS alignments using the
trimmed FASTQ files.</p>
<p>Here, we are appending this step to the <code>SYSargsList</code> object created previously.
All the parameters are defined on the <code>preprocessReads/preprocessReads-pe.yml</code> and
<code>preprocessReads/preprocessReads-pe.cwl</code> files.</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;preprocessing&quot;, targets = &quot;targets_chipseq.txt&quot;,
    dir = TRUE, wf_file = &quot;preprocessReads/preprocessReads-se.cwl&quot;,
    input_file = &quot;preprocessReads/preprocessReads-se.yml&quot;, dir_path = &quot;param/cwl&quot;,
    inputvars = c(FileName = &quot;_FASTQ_PATH1_&quot;, SampleName = &quot;_SampleName_&quot;),
    dependency = c(&quot;load_SPR&quot;))
</code></pre>
<p>After, we can check the <code>trimLRPatterns</code> function in input parameter:</p>
<pre><code class="language-r">yamlinput(sal, &quot;preprocessing&quot;)$Fct
</code></pre>
<pre><code>## [1] &quot;'trimLRPatterns(Rpattern=\&quot;GCCCGGGTAA\&quot;, subject=fq)'&quot;
</code></pre>
<p>After the preprocessing step, the <code>outfiles</code> files can be used to generate the new
targets files containing the paths to the trimmed FASTQ files. The new targets
information can be used for the next workflow step instance, <em>e.g.</em> running the
NGS alignments with the trimmed FASTQ files. The <code>appendStep</code> function is
automatically handling this connectivity between steps. Please check the <code>Alignments</code>
step for more details.</p>
<h4 id="fastq-quality-report">FASTQ quality report</h4>
<p>The following <code>seeFastq</code> and <code>seeFastqPlot</code> functions generate and plot a series of useful
quality statistics for a set of FASTQ files including per cycle quality box
plots, base proportions, base-level quality trends, relative k-mer
diversity, length and occurrence distribution of reads, number of reads
above quality cutoffs and mean quality distribution. The results are
written to a PDF file named <code>fastqReport.pdf</code>.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    fastq &lt;- getColumn(sal, step = &quot;preprocessing&quot;, &quot;targetsWF&quot;,
        column = 1)
    fqlist &lt;- seeFastq(fastq = fastq, batchsize = 10000, klength = 8)
    pdf(&quot;./results/fastqReport.pdf&quot;, height = 18, width = 4 *
        length(fqlist))
    seeFastqPlot(fqlist)
    dev.off()
}, step_name = &quot;fastq_report&quot;, dependency = &quot;preprocessing&quot;)
</code></pre>
<p><img src="../results/fastqReport.png" alt=""></p>
<div data-align="center">
<p>Figure 1: FASTQ quality report for 7 samples.</p>
</div>
</br>
<h3 id="alignments">Alignments</h3>
<h4 id="read-mapping-with-bowtie2">Read mapping with <code>Bowtie2</code></h4>
<p>The NGS reads of this project will be aligned with <code>Bowtie2</code> against the
reference genome sequence (Langmead and Salzberg 2012). The parameter settings of the
Bowtie2 index are defined in the <code>bowtie2-index.cwl</code> and <code>bowtie2-index.yml</code> files.</p>
<p>Building the index:</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;bowtie2_index&quot;, dir = FALSE,
    targets = NULL, wf_file = &quot;bowtie2/bowtie2-index.cwl&quot;, input_file = &quot;bowtie2/bowtie2-index.yml&quot;,
    dir_path = &quot;param/cwl&quot;, inputvars = NULL, dependency = c(&quot;preprocessing&quot;))
</code></pre>
<p>The parameter settings of the aligner are defined in the <code>workflow_bowtie2-se.cwl</code>
and <code>workflow_bowtie2-se.yml</code> files. The following shows how to construct the
corresponding <em>SYSargsList</em> object.</p>
<p>In ChIP-Seq experiments it is usually more appropriate to eliminate reads mapping
to multiple locations. To achieve this, users want to remove the argument setting
<code>-k 50 non-deterministic</code> in the configuration files.</p>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;bowtie2_alignment&quot;,
    dir = TRUE, targets = &quot;targets_chipseq.txt&quot;, wf_file = &quot;workflow-bowtie2/workflow_bowtie2-se.cwl&quot;,
    input_file = &quot;workflow-bowtie2/workflow_bowtie2-se.yml&quot;,
    dir_path = &quot;param/cwl&quot;, inputvars = c(FileName = &quot;_FASTQ_PATH1_&quot;,
        SampleName = &quot;_SampleName_&quot;), dependency = c(&quot;bowtie2_index&quot;))
</code></pre>
<p>To double-check the command line for each sample, please use the following:</p>
<pre><code class="language-r">cmdlist(sal, step = &quot;bowtie2_alignment&quot;, targets = 1)
</code></pre>
<pre><code>## $bowtie2_alignment
## $bowtie2_alignment$AP1_1
## $bowtie2_alignment$AP1_1$bowtie2
## [1] &quot;bowtie2 -S ./results/AP1_1.sam  -x ./data/tair10.fasta  -k 50  --non-deterministic  -U ./data/SRR038845_1.fastq.gz -p 4&quot;
## 
## $bowtie2_alignment$AP1_1$`samtools-view`
## [1] &quot;samtools view -bS -o ./results/AP1_1.bam  ./results/AP1_1.sam &quot;
## 
## $bowtie2_alignment$AP1_1$`samtools-sort`
## [1] &quot;samtools sort -o ./results/AP1_1.sorted.bam  ./results/AP1_1.bam  -@ 4&quot;
## 
## $bowtie2_alignment$AP1_1$`samtools-index`
## [1] &quot;samtools index -b results/AP1_1.sorted.bam  results/AP1_1.sorted.bam.bai  ./results/AP1_1.sorted.bam &quot;
</code></pre>
<h4 id="read-and-alignment-stats">Read and alignment stats</h4>
<p>The following provides an overview of the number of reads in each sample
and how many of them aligned to the reference.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    fqpaths &lt;- getColumn(sal, step = &quot;bowtie2_alignment&quot;, &quot;targetsWF&quot;,
        column = &quot;FileName&quot;)
    bampaths &lt;- getColumn(sal, step = &quot;bowtie2_alignment&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    read_statsDF &lt;- alignStats(args = bampaths, fqpaths = fqpaths,
        pairEnd = TRUE)
    write.table(read_statsDF, &quot;results/alignStats.xls&quot;, row.names = FALSE,
        quote = FALSE, sep = &quot;\t&quot;)
}, step_name = &quot;align_stats&quot;, dependency = &quot;bowtie2_alignment&quot;)
</code></pre>
<h3 id="create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</h3>
<p>The <code>symLink2bam</code> function creates symbolic links to view the BAM alignment
files in a genome browser such as IGV without moving these large files to a
local system. The corresponding URLs are written to a file with a path
specified under <code>urlfile</code>, here <code>IGVurl.txt</code>. Please replace the directory
and the user name. The following parameter settings will create a
subdirectory under <code>~/.html</code> called <code>somedir</code> of the user account. The user
name under <code>urlbase</code>, here <code>&lt;username&gt;</code>, needs to be changed to the corresponding
user name of the person running this function.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    bampaths &lt;- getColumn(sal, step = &quot;bowtie2_alignment&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    bampaths &lt;- setNames(normalizePath(bampaths), names(bampaths))
    symLink2bam(sysargs = bampaths, htmldir = c(&quot;~/.html/&quot;, &quot;somedir/&quot;),
        urlbase = &quot;http://cluster.hpcc.ucr.edu/~&lt;username&gt;/&quot;,
        urlfile = &quot;./results/IGVurl.txt&quot;)
}, step_name = &quot;bam_IGV&quot;, dependency = &quot;bowtie2_alignment&quot;, run_step = &quot;optional&quot;)
</code></pre>
<h3 id="peak-calling-with-macs2">Peak calling with MACS2</h3>
<h3 id="merge-bam-files-of-replicates-prior-to-peak-calling">Merge BAM files of replicates prior to peak calling</h3>
<p>Merging BAM files of technical and/or biological replicates can improve
the sensitivity of the peak calling by increasing the depth of read
coverage. The <code>mergeBamByFactor</code> function merges BAM files based on grouping information
specified by a <code>factor</code>, here the <code>Factor</code> column of the imported targets file.
It also returns an updated <code>targets</code> object containing the paths to the
merged BAM files as well as to any unmerged files without replicates.
The updated <code>targets</code> object can be used to update the <code>SYSargsList</code> object.</p>
<p>This step can be skipped if merging of BAM files is not desired.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    bampaths &lt;- getColumn(sal, step = &quot;bowtie2_alignment&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    merge_bams &lt;- mergeBamByFactor(args = bampaths, targetsDF = targetsWF(sal)[[&quot;bowtie2_alignment&quot;]],
        overwrite = TRUE)
    updateColumn(sal, step = &quot;merge_bams&quot;, position = &quot;targetsWF&quot;) &lt;- merge_bams
    writeTargets(sal, step = &quot;merge_bams&quot;, file = &quot;targets_merge_bams.txt&quot;,
        overwrite = TRUE)
}, step_name = &quot;merge_bams&quot;, dependency = &quot;bowtie2_alignment&quot;)
</code></pre>
<h4 id="peak-calling-with-inputreference-sample">Peak calling with input/reference sample</h4>
<p>MACS2 can perform peak calling on ChIP-Seq data with and without input
samples (Zhang et al. 2008).</p>
<p>The following performs peak calling with input sample. The input sample
can be most conveniently specified in the <code>SampleReference</code> column of the
initial <code>targets</code> file. The <code>writeTargetsRef</code> function uses this
information to create a <code>targets</code> file intermediate for running MACS2
with the corresponding input sample(s).</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    writeTargetsRef(infile = &quot;targets_merge_bams.txt&quot;, outfile = &quot;targets_bam_ref.txt&quot;,
        silent = FALSE, overwrite = TRUE)
}, step_name = &quot;writeTargetsRef&quot;, dependency = &quot;merge_bams&quot;)
</code></pre>
<pre><code class="language-r">appendStep(sal) &lt;- SYSargsList(step_name = &quot;call_peaks_macs_withref&quot;,
    targets = &quot;targets_bam_ref.txt&quot;, wf_file = &quot;MACS2/macs2-input.cwl&quot;,
    input_file = &quot;MACS2/macs2-input.yml&quot;, dir_path = &quot;param/cwl&quot;,
    inputvars = c(FileName1 = &quot;_FASTQ_PATH1_&quot;, FileName2 = &quot;_FASTQ_PATH2_&quot;,
        SampleReference = &quot;_SampleName_&quot;), id = &quot;SampleReference&quot;,
    dependency = c(&quot;writeTargetsRef&quot;))
</code></pre>
<p>The peak calling results from MACS2 are written for each sample to
separate files in the <code>results/call_peaks_macs_withref</code> directory. They are
named after the corresponding files with extensions used by MACS2.</p>
<h3 id="annotate-peaks-with-genomic-context">Annotate peaks with genomic context</h3>
<h4 id="annotation-with-chipseeker-package">Annotation with <code>ChIPseeker</code> package</h4>
<p>The following annotates the identified peaks with genomic context information
using the <code>ChIPseeker</code> package (Yu, Wang, and He 2015).</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(ChIPseeker)
    library(GenomicFeatures)
    peaks_files &lt;- getColumn(sal, step = &quot;call_peaks_macs_withref&quot;,
        &quot;outfiles&quot;, column = &quot;peaks_xls&quot;)
    txdb &lt;- suppressWarnings(makeTxDbFromGFF(file = &quot;data/tair10.gff&quot;,
        format = &quot;gff&quot;, dataSource = &quot;TAIR&quot;, organism = &quot;Arabidopsis thaliana&quot;))
    for (i in seq(along = peaks_files)) {
        peakAnno &lt;- annotatePeak(peaks_files[i], TxDb = txdb,
            verbose = FALSE)
        df &lt;- as.data.frame(peakAnno)
        outpaths &lt;- paste0(&quot;./results/&quot;, names(peaks_files),
            &quot;_ChIPseeker_annotated.xls&quot;)
        names(outpaths) &lt;- names(peaks_files)
        write.table(df, outpaths[i], quote = FALSE, row.names = FALSE,
            sep = &quot;\t&quot;)
    }
    updateColumn(sal, step = &quot;annotation_ChIPseeker&quot;, position = &quot;outfiles&quot;) &lt;- data.frame(outpaths)
}, step_name = &quot;annotation_ChIPseeker&quot;, dependency = &quot;call_peaks_macs_withref&quot;)
</code></pre>
<p>The peak annotation results are written to the <code>results</code> directory.</p>
<p>Summary plots provided by the <code>ChIPseeker</code> package. Here applied only to one sample
for demonstration purposes.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    peaks_files &lt;- getColumn(sal, step = &quot;call_peaks_macs_withref&quot;,
        &quot;outfiles&quot;, column = &quot;peaks_xls&quot;)
    peak &lt;- readPeakFile(peaks_files[1])
    pdf(&quot;results/peakscoverage.pdf&quot;)
    covplot(peak, weightCol = &quot;X.log10.pvalue.&quot;)
    dev.off()
    pdf(&quot;results/peaksHeatmap.pdf&quot;)
    peakHeatmap(peaks_files[1], TxDb = txdb, upstream = 1000,
        downstream = 1000, color = &quot;red&quot;)
    dev.off()
    pdf(&quot;results/peaksProfile.pdf&quot;)
    plotAvgProf2(peaks_files[1], TxDb = txdb, upstream = 1000,
        downstream = 1000, xlab = &quot;Genomic Region (5'-&gt;3')&quot;,
        ylab = &quot;Read Count Frequency&quot;, conf = 0.05)
    dev.off()
}, step_name = &quot;ChIPseeker_plots&quot;, dependency = &quot;annotation_ChIPseeker&quot;)
</code></pre>
<h3 id="count-reads-overlapping-peaks">Count reads overlapping peaks</h3>
<p>The <code>countRangeset</code> function is a convenience wrapper to perform read counting
iteratively over serveral range sets, here peak range sets. Internally,
the read counting is performed with the <code>summarizeOverlaps</code> function from the
<code>GenomicAlignments</code> package. The resulting count tables are directly saved to
files, one for each peak set.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(GenomicRanges)
    bam_files &lt;- getColumn(sal, step = &quot;bowtie2_alignment&quot;, &quot;outfiles&quot;,
        column = &quot;samtools_sort_bam&quot;)
    args &lt;- getColumn(sal, step = &quot;call_peaks_macs_withref&quot;,
        &quot;outfiles&quot;, column = &quot;peaks_xls&quot;)
    outfiles &lt;- paste0(&quot;./results/&quot;, names(args), &quot;_countDF.xls&quot;)
    bfl &lt;- BamFileList(bam_files, yieldSize = 50000, index = character())
    countDFnames &lt;- countRangeset(bfl, args, outfiles, mode = &quot;Union&quot;,
        ignore.strand = TRUE)
    updateColumn(sal, step = &quot;count_peak_ranges&quot;, position = &quot;outfiles&quot;) &lt;- data.frame(countDFnames)
}, step_name = &quot;count_peak_ranges&quot;, dependency = &quot;call_peaks_macs_withref&quot;,
    )
</code></pre>
<p>Shows count table generated in previous step (<code>results/AP1_1_countDF.xls</code>).
To avoid slowdowns of the load time of this page, ony 200 rows of the source
table are imported into the below <code>datatable</code> view .</p>
<pre><code class="language-r">countDF &lt;- read.delim(&quot;results/AP1_1_countDF.xls&quot;)[1:200, ]
colnames(countDF)[1] &lt;- &quot;PeakIDs&quot;
library(DT)
datatable(countDF)
</code></pre>
<div id="htmlwidget-1" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-1">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200"],["Chr1_10000123-10001504","Chr1_10015050-10015329","Chr1_10044084-10044292","Chr1_10053060-10053403","Chr1_10065376-10065801","Chr1_10077918-10078299","Chr1_10111557-10111830","Chr1_10137133-10137852","Chr1_10139865-10140426","Chr1_10141346-10141599","Chr1_10142995-10143536","Chr1_10143619-10144022","Chr1_10145791-10146202","Chr1_10146333-10147178","Chr1_10186868-10187102","Chr1_10187659-10187893","Chr1_10203888-10204202","Chr1_10219688-10219985","Chr1_10228577-10229094","Chr1_10229262-10229710","Chr1_10242962-10243288","Chr1_10244590-10244954","Chr1_10247304-10247528","Chr1_10251234-10251447","Chr1_10263957-10264248","Chr1_102694-103029","Chr1_10294478-10294800","Chr1_10295972-10296231","Chr1_10353709-10354318","Chr1_10364919-10365439","Chr1_10369960-10370330","Chr1_10375634-10375976","Chr1_10376847-10377070","Chr1_10489119-10489357","Chr1_10489629-10489865","Chr1_1049818-1050057","Chr1_10504608-10505267","Chr1_10505986-10506492","Chr1_10519205-10519487","Chr1_10520174-10520649","Chr1_10520971-10521455","Chr1_10530766-10530996","Chr1_10535179-10535644","Chr1_10537198-10537646","Chr1_10538872-10539144","Chr1_10541237-10541831","Chr1_10542202-10542449","Chr1_10553478-10553874","Chr1_10562570-10563068","Chr1_10563816-10565000","Chr1_10565793-10566457","Chr1_10567369-10567945","Chr1_10569384-10569620","Chr1_105715-106037","Chr1_10581383-10581685","Chr1_10622431-10622757","Chr1_10631994-10632273","Chr1_10649137-10649367","Chr1_10668660-10668958","Chr1_10669009-10669489","Chr1_10676832-10677414","Chr1_10678636-10679743","Chr1_10679884-10680612","Chr1_10683747-10684382","Chr1_10684727-10685161","Chr1_10685376-10685989","Chr1_10691341-10691562","Chr1_10691767-10692130","Chr1_10692583-10693433","Chr1_10693954-10694617","Chr1_10695463-10696565","Chr1_10701975-10702395","Chr1_10704093-10704316","Chr1_10713567-10714255","Chr1_10715301-10715933","Chr1_1072348-1072586","Chr1_10726169-10726568","Chr1_1072802-1073152","Chr1_10737905-10738222","Chr1_10793236-10793520","Chr1_10794884-10795869","Chr1_10799998-10800389","Chr1_10800942-10801268","Chr1_10801874-10802349","Chr1_10803701-10803926","Chr1_1084359-1084961","Chr1_10845312-10845874","Chr1_10864966-10865601","Chr1_10871974-10872308","Chr1_10874129-10874328","Chr1_10874476-10875185","Chr1_10891639-10891997","Chr1_10909961-10910163","Chr1_10953463-10953804","Chr1_10954240-10954628","Chr1_10956271-10956623","Chr1_10958841-10959244","Chr1_10959439-10960076","Chr1_10972812-10973048","Chr1_1103200-1103416","Chr1_11033716-11034097","Chr1_11059998-11060940","Chr1_11061272-11061800","Chr1_11062322-11062750","Chr1_11067587-11068014","Chr1_11071124-11071500","Chr1_11084496-11084695","Chr1_11128-11443","Chr1_11155158-11155410","Chr1_11196849-11197415","Chr1_11200138-11200385","Chr1_11210773-11211568","Chr1_11212068-11212348","Chr1_11212560-11212978","Chr1_1121262-1121569","Chr1_11222722-11222952","Chr1_11227689-11228070","Chr1_11231069-11231415","Chr1_1129384-1129772","Chr1_1134217-1134971","Chr1_11343316-11343530","Chr1_11350683-11350893","Chr1_11374190-11375132","Chr1_11377612-11378008","Chr1_11385968-11386241","Chr1_11386635-11386872","Chr1_11388323-11388599","Chr1_11421622-11422073","Chr1_11428546-11428802","Chr1_11429167-11429366","Chr1_11429941-11430188","Chr1_11430639-11430926","Chr1_11441925-11442147","Chr1_11446002-11446822","Chr1_11447290-11447887","Chr1_115641-116419","Chr1_1159091-1159510","Chr1_11594952-11595190","Chr1_11598295-11598860","Chr1_11600614-11600983","Chr1_11601139-11601823","Chr1_11605622-11605926","Chr1_11620054-11620600","Chr1_11621059-11621447","Chr1_11621523-11623089","Chr1_11624914-11625625","Chr1_11626261-11626946","Chr1_11628379-11628622","Chr1_11628853-11629244","Chr1_11630475-11631803","Chr1_11662337-11662575","Chr1_1172920-1173129","Chr1_1173195-1173526","Chr1_11763531-11763800","Chr1_11770202-11770807","Chr1_11771164-11771564","Chr1_11798751-11798970","Chr1_11801046-11801322","Chr1_11804660-11805026","Chr1_11807363-11807831","Chr1_11808375-11808791","Chr1_11809718-11810175","Chr1_11829516-11829982","Chr1_11830901-11831419","Chr1_11874425-11874646","Chr1_1188104-1188901","Chr1_11925063-11925373","Chr1_11927715-11928115","Chr1_11928466-11928909","Chr1_11978528-11979276","Chr1_1197874-1198387","Chr1_12004740-12004982","Chr1_12005493-12006453","Chr1_12026542-12027136","Chr1_12050647-12051422","Chr1_12051618-12051822","Chr1_12054702-12055258","Chr1_12055957-12056342","Chr1_12056485-12057547","Chr1_12059598-12059861","Chr1_12064129-12064328","Chr1_12070355-12070698","Chr1_12123473-12123764","Chr1_12132602-12132955","Chr1_12177510-12177803","Chr1_12180624-12180825","Chr1_12254744-12255260","Chr1_1233120-1233354","Chr1_1235877-1236728","Chr1_12458546-12458848","Chr1_12474908-12475186","Chr1_128465-128823","Chr1_12916443-12916688","Chr1_12986436-12986729","Chr1_12986791-12987139","Chr1_13029530-13030312","Chr1_1303036-1303264","Chr1_130507-131207","Chr1_1306355-1306663","Chr1_1307140-1307396"],[384,19,54,31,89,82,36,95,135,29,44,44,59,134,27,14,50,49,193,75,36,29,22,24,33,21,19,16,108,125,38,17,20,20,28,36,135,53,23,168,101,19,67,26,31,55,40,39,40,290,229,40,21,39,28,13,36,16,30,40,96,271,81,63,48,106,18,40,207,86,121,85,19,158,100,38,72,40,30,16,277,51,54,97,22,88,120,77,19,18,137,64,23,44,53,47,33,150,22,29,39,107,161,49,94,23,8,60,53,91,27,58,32,69,33,26,26,47,49,222,15,32,191,35,17,26,43,71,17,16,13,39,6,120,80,93,116,17,86,41,178,36,125,48,178,59,92,23,41,284,80,17,38,23,104,104,12,13,35,72,80,82,49,98,15,100,16,36,32,251,72,25,143,127,177,16,94,37,219,22,12,80,55,37,30,19,102,9,268,26,13,74,31,23,37,304,30,93,40,16],[571,50,31,71,122,115,59,201,191,38,102,72,120,296,32,44,59,56,357,129,54,62,36,35,47,64,66,43,213,149,69,48,33,45,37,19,281,95,63,301,180,39,109,83,58,86,50,80,100,625,371,100,35,69,46,78,46,44,68,106,148,447,152,152,123,205,38,71,347,159,264,96,31,196,144,37,150,42,47,37,434,93,79,135,37,132,186,137,71,38,236,85,33,54,86,63,76,243,53,29,114,319,248,135,277,73,30,92,37,165,62,157,91,138,61,34,81,82,63,330,31,32,295,78,58,48,52,131,46,26,43,54,36,167,112,154,128,31,89,49,199,60,165,90,399,162,177,29,84,485,141,32,70,41,152,139,39,46,75,127,120,169,157,141,35,199,74,82,114,343,164,44,237,228,284,28,164,61,433,37,43,99,44,75,69,33,160,34,285,59,62,153,37,52,45,481,32,281,53,38],[282,26,19,32,89,46,21,90,95,26,58,30,58,149,25,20,35,24,158,80,26,22,22,21,23,26,49,17,96,83,24,28,13,18,16,15,107,40,32,152,74,21,49,36,27,36,19,31,45,348,151,54,14,31,15,32,31,18,21,41,83,222,82,62,40,102,10,30,143,98,119,38,12,92,66,24,94,21,20,27,189,41,52,70,19,41,94,59,33,19,132,26,11,24,29,25,39,119,13,14,53,155,110,55,110,50,14,42,15,77,21,91,27,58,37,24,42,44,31,151,15,25,142,32,22,20,18,67,28,15,22,21,19,75,46,58,43,14,45,27,90,25,51,44,191,76,83,19,48,248,67,21,35,16,68,81,14,19,42,57,78,67,74,58,9,76,24,23,45,169,72,28,117,96,141,14,67,29,195,20,17,57,25,38,29,13,78,26,118,16,25,71,15,28,16,200,16,129,27,23],[56,1,4,6,1,12,3,15,7,5,10,34,1,1,2,0,4,13,1,17,1,0,0,0,4,3,1,1,15,15,10,9,2,5,3,1,6,19,0,3,13,6,11,0,8,2,18,4,5,9,7,7,2,10,0,4,5,0,10,7,0,11,3,5,1,17,0,10,12,0,25,3,0,4,19,10,4,15,5,0,12,2,1,2,2,60,16,2,5,14,2,3,2,12,12,11,3,6,4,1,2,4,9,0,24,3,14,1,7,0,6,8,2,9,4,20,1,10,6,1,1,7,10,34,7,4,3,4,6,2,0,22,1,6,13,1,23,0,33,9,1,17,61,2,21,1,19,0,9,3,5,5,10,5,0,16,3,34,6,14,0,0,5,25,0,8,11,1,1,43,9,0,10,15,10,0,22,16,14,4,0,1,14,2,19,21,3,1,34,5,11,67,29,5,15,8,1,9,1,3],[43,3,3,6,6,4,8,8,8,8,10,12,1,4,3,1,9,5,2,5,7,1,6,1,6,3,4,1,9,10,8,6,2,8,1,0,6,22,2,3,18,4,10,3,8,7,9,8,5,32,17,7,7,10,0,3,4,0,8,3,0,21,23,9,1,20,1,11,12,1,33,7,2,9,13,8,9,16,3,2,9,4,8,8,2,19,14,11,5,5,17,1,0,7,6,5,5,9,2,4,6,17,12,3,16,1,8,8,5,1,7,2,3,3,9,16,0,7,7,19,9,2,23,19,4,0,7,14,2,3,0,15,0,6,24,4,19,2,10,7,8,12,17,0,39,14,11,4,3,6,22,4,12,7,1,11,5,10,11,15,0,1,9,17,2,13,11,0,4,26,12,0,21,18,17,0,15,5,19,1,1,5,6,8,11,6,1,1,22,9,5,27,7,1,9,14,0,6,1,7],[104,12,14,25,18,10,17,20,20,5,21,11,11,15,8,10,13,16,32,16,12,10,9,7,18,11,7,4,21,27,12,8,2,6,10,4,22,16,13,30,11,17,13,15,13,16,6,14,14,33,36,11,7,2,6,11,12,12,16,19,13,41,21,20,10,25,6,22,30,15,33,15,14,17,19,9,35,10,14,9,32,22,8,8,8,15,18,17,20,14,22,15,5,7,13,5,12,11,6,7,14,41,22,20,46,8,11,14,7,16,5,9,12,10,9,5,12,19,9,31,8,12,48,15,11,4,5,36,4,4,12,5,13,20,20,18,28,6,13,15,31,10,12,7,41,23,26,5,22,36,43,5,13,2,28,29,7,5,16,18,18,18,24,12,10,30,9,9,13,55,26,3,46,37,30,6,29,5,21,11,7,21,7,13,9,3,24,5,36,23,16,25,4,6,11,31,12,43,18,4],[55,5,3,4,9,4,9,15,17,6,4,8,10,14,4,4,7,8,19,5,4,12,5,5,2,7,5,4,12,17,6,6,4,1,3,5,16,5,2,19,10,4,7,6,2,5,8,7,9,14,24,3,7,3,3,9,7,5,7,11,9,22,11,2,8,16,4,10,14,7,13,10,2,10,11,4,30,5,8,3,20,3,6,7,5,9,5,8,10,4,12,9,10,0,9,2,7,18,2,6,11,26,11,15,23,9,4,10,6,9,6,9,4,9,4,1,3,8,7,13,3,8,20,10,5,5,6,12,3,1,5,9,3,8,11,10,17,1,4,5,18,4,12,6,25,13,11,0,10,19,22,5,10,4,19,17,2,8,6,18,7,9,7,7,4,8,3,2,4,23,11,1,24,13,14,4,16,7,12,5,2,9,3,11,7,4,20,6,13,6,14,14,2,4,9,17,4,32,3,4]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>PeakIDs<\/th>\n      <th>AP1_1<\/th>\n      <th>AP1_2A<\/th>\n      <th>AP1_2B<\/th>\n      <th>C_1A<\/th>\n      <th>C_1B<\/th>\n      <th>C_2A<\/th>\n      <th>C_2B<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,3,4,5,6,7,8]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<h3 id="differential-binding-analysis">Differential binding analysis</h3>
<p>The <code>runDiff</code> function performs differential binding analysis in batch mode for
several count tables using <code>edgeR</code> or <code>DESeq2</code> (Robinson, McCarthy, and Smyth 2010; Love, Huber, and Anders 2014).
Internally, it calls the functions <code>run_edgeR</code> and <code>run_DESeq2</code>. It also returns
the filtering results and plots from the downstream <code>filterDEGs</code> function using
the fold change and FDR cutoffs provided under the <code>dbrfilter</code> argument.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    countDF_files &lt;- getColumn(sal, step = &quot;count_peak_ranges&quot;,
        &quot;outfiles&quot;)
    outfiles &lt;- paste0(&quot;./results/&quot;, names(countDF_files), &quot;_peaks_edgeR.xls&quot;)
    names(outfiles) &lt;- names(countDF_files)
    cmp &lt;- readComp(file = stepsWF(sal)[[&quot;bowtie2_alignment&quot;]],
        format = &quot;matrix&quot;)
    dbrlist &lt;- runDiff(args = countDF_files, outfiles = outfiles,
        diffFct = run_edgeR, targets = targetsWF(sal)[[&quot;bowtie2_alignment&quot;]],
        cmp = cmp[[1]], independent = TRUE, dbrfilter = c(Fold = 2,
            FDR = 1))
}, step_name = &quot;diff_bind_analysis&quot;, dependency = &quot;count_peak_ranges&quot;,
    )
</code></pre>
<h3 id="go-term-enrichment-analysis">GO term enrichment analysis</h3>
<h4 id="obtain-gene-to-go-mappings">Obtain gene-to-GO mappings</h4>
<p>The following shows how to obtain gene-to-GO mappings from <code>biomaRt</code> (here for <em>A.
thaliana</em>) and how to organize them for the downstream GO term
enrichment analysis. Alternatively, the gene-to-GO mappings can be
obtained for many organisms from Bioconductor’s <code>*.db</code> genome annotation
packages or GO annotation files provided by various genome databases.
For each annotation this relatively slow preprocessing step needs to be
performed only once. Subsequently, the preprocessed data can be loaded
with the <code>load</code> function as shown in the next subsection.</p>
<h4 id="go-term-enrichment-analysis-1">GO term enrichment analysis</h4>
<p>The following performs GO term enrichment analysis for each annotated peak
set. Note: the following assumes that the GO annotation data exists under
<code>data/GO/catdb.RData</code>.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(&quot;biomaRt&quot;)
    # listMarts() # To choose BioMart database
    # listMarts(host='plants.ensembl.org')
    m &lt;- useMart(&quot;plants_mart&quot;, host = &quot;https://plants.ensembl.org&quot;)
    m &lt;- useMart(&quot;plants_mart&quot;, dataset = &quot;athaliana_eg_gene&quot;,
        host = &quot;https://plants.ensembl.org&quot;)
    go &lt;- getBM(attributes = c(&quot;go_id&quot;, &quot;tair_locus&quot;, &quot;namespace_1003&quot;),
        mart = m)
    go &lt;- go[go[, 3] != &quot;&quot;, ]
    go[, 3] &lt;- as.character(go[, 3])
    go[go[, 3] == &quot;molecular_function&quot;, 3] &lt;- &quot;F&quot;
    go[go[, 3] == &quot;biological_process&quot;, 3] &lt;- &quot;P&quot;
    go[go[, 3] == &quot;cellular_component&quot;, 3] &lt;- &quot;C&quot;
    go[1:4, ]
    if (!dir.exists(&quot;./data/GO&quot;))
        dir.create(&quot;./data/GO&quot;)
    write.table(go, &quot;data/GO/GOannotationsBiomart_mod.txt&quot;, quote = FALSE,
        row.names = FALSE, col.names = FALSE, sep = &quot;\t&quot;)
    catdb &lt;- makeCATdb(myfile = &quot;data/GO/GOannotationsBiomart_mod.txt&quot;,
        lib = NULL, org = &quot;&quot;, colno = c(1, 2, 3), idconv = NULL)
    save(catdb, file = &quot;data/GO/catdb.RData&quot;)
}, step_name = &quot;get_go_annot&quot;, dependency = &quot;annotation_ChIPseeker&quot;,
    )
</code></pre>
<h4 id="go-term-enrichment-analysis-2">GO term enrichment analysis</h4>
<p>The following performs GO term enrichment analysis for each annotated peak
set. Note: the following assumes that the GO annotation data exists under
<code>data/GO/catdb.RData</code>.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    annofiles &lt;- getColumn(sal, step = &quot;annotation_ChIPseeker&quot;,
        &quot;outfiles&quot;)
    gene_ids &lt;- sapply(annofiles, function(x) unique(as.character(read.delim(x)[,
        &quot;geneId&quot;])), simplify = FALSE)
    load(&quot;data/GO/catdb.RData&quot;)
    BatchResult &lt;- GOCluster_Report(catdb = catdb, setlist = gene_ids,
        method = &quot;all&quot;, id_type = &quot;gene&quot;, CLSZ = 2, cutoff = 0.9,
        gocats = c(&quot;MF&quot;, &quot;BP&quot;, &quot;CC&quot;), recordSpecGO = NULL)
    write.table(BatchResult, &quot;results/GOBatchAll.xls&quot;, quote = FALSE,
        row.names = FALSE, sep = &quot;\t&quot;)
}, step_name = &quot;go_enrich&quot;, dependency = &quot;annotation_ChIPseeker&quot;,
    )
</code></pre>
<p>Shows GO term enrichment results from previous step. The last gene identifier column (10)
of this table has been excluded in this viewing instance to minimize the complexity of the
result.
To avoid slowdowns of the load time of this page, only 50 rows of the source
table are imported into the below <code>datatable</code> view .</p>
<pre><code class="language-r">BatchResult &lt;- read.delim(&quot;results/GOBatchAll.xls&quot;)[1:50, ]
library(DT)
datatable(BatchResult[, -10], options = list(scrollX = TRUE,
    autoWidth = TRUE))
</code></pre>
<div id="htmlwidget-2" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-2">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50"],["C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A","C_1A"],[4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903,4903],["GO:0003700","GO:0140110","GO:0098772","GO:0043565","GO:0000976","GO:0001067","GO:1990837","GO:0003677","GO:0003690","GO:0042802","GO:0000981","GO:0046983","GO:0000987","GO:0003676","GO:0000978","GO:1901363","GO:0033612","GO:0097159","GO:0042562","GO:0005515","GO:0000977","GO:0060089","GO:0004672","GO:0019840","GO:0016773","GO:0043178","GO:0004860","GO:0019210","GO:0016705","GO:0015291","GO:0038023","GO:0001653","GO:0016301","GO:0022857","GO:0004861","GO:0030291","GO:0022804","GO:0042803","GO:0000156","GO:0005215","GO:0004674","GO:0019900","GO:0015293","GO:0005102","GO:0004805","GO:0005516","GO:0010427","GO:0080161","GO:0008509","GO:0019207"],[1667,1837,2497,1298,1007,1007,1032,2566,1142,394,336,664,306,4598,284,7859,46,7889,32,8016,324,236,1133,23,1296,39,27,30,400,313,207,17,1498,1200,8,8,515,208,19,1273,880,138,101,126,12,235,19,19,240,111],[682,698,807,506,412,412,416,777,432,133,117,195,102,966,94,1550,26,1552,20,1566,95,73,260,14,289,19,15,16,104,85,61,11,325,266,7,7,127,60,11,276,198,42,33,39,8,62,10,10,63,34],[3.49341724975865e-120,1.67823421235582e-104,2.18308574281622e-80,2.03153636203605e-78,2.16539339765293e-70,2.16539339765293e-70,7.55925494113875e-69,8.11545366519395e-63,5.44122511323004e-62,8.45596025352572e-15,3.02911820739919e-14,5.20591490791628e-14,2.60120228948827e-11,4.80717206533446e-11,2.37684179178145e-10,3.18096489951193e-09,3.70334206206444e-09,6.71010417736292e-09,2.23753662482306e-08,4.36325237409659e-08,1.70493636893304e-07,4.830138137427e-07,2.6496064818211e-06,4.80082717610204e-06,8.57859203202179e-06,9.09042267661985e-06,1.04457669068979e-05,1.06626043165579e-05,1.95557875198908e-05,2.00638856116351e-05,2.16501006655692e-05,2.30169720521419e-05,2.58444261599642e-05,3.08098675362986e-05,3.69557186521355e-05,3.69557186521355e-05,3.83718364858986e-05,5.00968427978002e-05,9.90476514846966e-05,0.000111473574255863,0.000134651553167959,0.000179143121967622,0.000200017440650394,0.000203270191184441,0.000242118052007588,0.000569854299398944,0.000584020859104921,0.000584020859104921,0.000596646365491369,0.000620806400430897],[2.12399768785326e-117,1.02036640111234e-101,1.32731613163226e-77,1.23517410811792e-75,1.31655918577298e-67,1.31655918577298e-67,4.59602700421236e-66,4.93419582843792e-60,3.30826486884386e-59,5.14122383414364e-12,1.8417038700987e-11,3.1651962640131e-11,1.58153099200887e-08,2.92276061572335e-08,1.44511980940312e-07,1.93402665890325e-06,2.25163197373518e-06,4.07974333983666e-06,1.36042226789242e-05,2.65285744345073e-05,0.000103660131231129,0.000293672398755561,0.00161096074094723,0.00291890292307004,0.00521578395546925,0.00552697698738487,0.00635102627939394,0.00648286342446718,0.0118899188120936,0.0121988424518741,0.0131632612046661,0.0139943190077023,0.0157134111052582,0.0187323994620695,0.0224690769404984,0.0224690769404984,0.0233300765834263,0.0304588804210625,0.0602209721026955,0.0677759331475647,0.0818681443261192,0.108919018156314,0.121610603915439,0.12358827624014,0.147207775620613,0.346471414034558,0.355084682335792,0.355084682335792,0.362760990218752,0.377450291461985],["DNA-binding transcription factor activity","transcription regulator activity","molecular function regulator","sequence-specific DNA binding","transcription regulatory region sequence-specific DNA binding","regulatory region nucleic acid binding","sequence-specific double-stranded DNA binding","DNA binding","double-stranded DNA binding","identical protein binding","DNA-binding transcription factor activity, RNA polymerase II-specific","protein dimerization activity","cis-regulatory region sequence-specific DNA binding","nucleic acid binding","RNA polymerase II cis-regulatory region sequence-specific DNA binding","heterocyclic compound binding","receptor serine/threonine kinase binding","organic cyclic compound binding","hormone binding","protein binding","RNA polymerase II transcription regulatory region sequence-specific DNA binding","molecular transducer activity","protein kinase activity","isoprenoid binding","phosphotransferase activity, alcohol group as acceptor","alcohol binding","protein kinase inhibitor activity","kinase inhibitor activity","oxidoreductase activity, acting on paired donors, with incorporation or reduction of molecular oxygen","secondary active transmembrane transporter activity","signaling receptor activity","peptide receptor activity","kinase activity","transmembrane transporter activity","cyclin-dependent protein serine/threonine kinase inhibitor activity","protein serine/threonine kinase inhibitor activity","active transmembrane transporter activity","protein homodimerization activity","phosphorelay response regulator activity","transporter activity","protein serine/threonine kinase activity","kinase binding","symporter activity","signaling receptor binding","trehalose-phosphatase activity","calmodulin binding","abscisic acid binding","auxin transmembrane transporter activity","anion transmembrane transporter activity","kinase regulator activity"],["MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF","MF"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>CLID<\/th>\n      <th>CLSZ<\/th>\n      <th>GOID<\/th>\n      <th>NodeSize<\/th>\n      <th>SampleMatch<\/th>\n      <th>Phyper<\/th>\n      <th>Padj<\/th>\n      <th>Term<\/th>\n      <th>Ont<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"autoWidth":true,"columnDefs":[{"className":"dt-right","targets":[2,4,5,6,7]},{"orderable":false,"targets":0}],"order":[],"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
<h3 id="motif-analysis">Motif analysis</h3>
<h4 id="parse-dna-sequences-of-peak-regions-from-genome">Parse DNA sequences of peak regions from genome</h4>
<p>Enrichment analysis of known DNA binding motifs or <em>de novo</em> discovery
of novel motifs requires the DNA sequences of the identified peak
regions. To parse the corresponding sequences from the reference genome,
the <code>getSeq</code> function from the <code>Biostrings</code> package can be used. The
following example parses the sequences for each peak set and saves the
results to separate FASTA files, one for each peak set. In addition, the
sequences in the FASTA files are ranked (sorted) by increasing p-values
as expected by some motif discovery tools, such as <code>BCRANK</code>.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(Biostrings)
    library(seqLogo)
    library(BCRANK)
    rangefiles &lt;- getColumn(sal, step = &quot;call_peaks_macs_withref&quot;,
        &quot;outfiles&quot;)
    for (i in seq(along = rangefiles)) {
        df &lt;- read.delim(rangefiles[i], comment = &quot;#&quot;)
        peaks &lt;- as(df, &quot;GRanges&quot;)
        names(peaks) &lt;- paste0(as.character(seqnames(peaks)),
            &quot;_&quot;, start(peaks), &quot;-&quot;, end(peaks))
        peaks &lt;- peaks[order(values(peaks)$X.log10.pvalue., decreasing = TRUE)]
        pseq &lt;- getSeq(FaFile(&quot;./data/tair10.fasta&quot;), peaks)
        names(pseq) &lt;- names(peaks)
        writeXStringSet(pseq, paste0(rangefiles[i], &quot;.fasta&quot;))
    }
}, step_name = &quot;parse_peak_sequences&quot;, dependency = &quot;call_peaks_macs_withref&quot;)
</code></pre>
<h4 id="motif-discovery-with-bcrank">Motif discovery with <code>BCRANK</code></h4>
<p>The Bioconductor package <code>BCRANK</code> is one of the many tools available for
<em>de novo</em> discovery of DNA binding motifs in peak regions of ChIP-Seq
experiments. The given example applies this method on the first peak
sample set and plots the sequence logo of the highest ranking motif.</p>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    library(Biostrings)
    library(seqLogo)
    library(BCRANK)
    rangefiles &lt;- getColumn(sal, step = &quot;call_peaks_macs_withref&quot;,
        &quot;outfiles&quot;)
    set.seed(0)
    BCRANKout &lt;- bcrank(paste0(rangefiles[1], &quot;.fasta&quot;), restarts = 25,
        use.P1 = TRUE, use.P2 = TRUE)
    toptable(BCRANKout)
    topMotif &lt;- toptable(BCRANKout, 1)
    weightMatrix &lt;- pwm(topMotif, normalize = FALSE)
    weightMatrixNormalized &lt;- pwm(topMotif, normalize = TRUE)
    pdf(&quot;results/seqlogo.pdf&quot;)
    seqLogo(weightMatrixNormalized)
    dev.off()
}, step_name = &quot;bcrank_enrich&quot;, dependency = &quot;call_peaks_macs_withref&quot;)
</code></pre>
<p><img src="../results/seqlogo.png" alt=""></p>
<div data-align="center">
<p>Figure 2: One of the motifs identified by <code>BCRANK</code></p>
</div>
</br>
<h3 id="version-information">Version Information</h3>
<pre><code class="language-r">appendStep(sal) &lt;- LineWise(code = {
    sessionInfo()
}, step_name = &quot;sessionInfo&quot;, dependency = &quot;bcrank_enrich&quot;)
</code></pre>
<h2 id="running-workflow">Running workflow</h2>
<h3 id="interactive-job-submissions-in-a-single-machine">Interactive job submissions in a single machine</h3>
<p>For running the workflow, <code>runWF</code> function will execute all the steps store in
the workflow container. The execution will be on a single machine without
submitting to a queuing system of a computer cluster.</p>
<pre><code class="language-r">sal &lt;- runWF(sal)
</code></pre>
<h3 id="parallelization-on-clusters">Parallelization on clusters</h3>
<p>Alternatively, the computation can be greatly accelerated by processing many files
in parallel using several compute nodes of a cluster, where a scheduling/queuing
system is used for load balancing.</p>
<p>The <code>resources</code> list object provides the number of independent parallel cluster
processes defined under the <code>Njobs</code> element in the list. The following example
will run 18 processes in parallel using each 4 CPU cores.
If the resources available on a cluster allow running all 18 processes at the
same time, then the shown sample submission will utilize in a total of 72 CPU cores.</p>
<p>Note, <code>runWF</code> can be used with most queueing systems as it is based on utilities
from the <code>batchtools</code> package, which supports the use of template files (<em><code>*.tmpl</code></em>)
for defining the run parameters of different schedulers. To run the following
code, one needs to have both a <code>conffile</code> (see <em><code>.batchtools.conf.R</code></em> samples <a href="https://mllg.github.io/batchtools/">here</a>)
and a <code>template</code> file (see <em><code>*.tmpl</code></em> samples <a href="https://github.com/mllg/batchtools/tree/master/inst/templates">here</a>)
for the queueing available on a system. The following example uses the sample
<code>conffile</code> and <code>template</code> files for the Slurm scheduler provided by this package.</p>
<p>The resources can be appended when the step is generated, or it is possible to
add these resources later, as the following example using the <code>addResources</code>
function:</p>
<pre><code class="language-r">resources &lt;- list(conffile=&quot;.batchtools.conf.R&quot;,
                  template=&quot;batchtools.slurm.tmpl&quot;, 
                  Njobs=8, 
                  walltime=120, ## minutes
                  ntasks=1,
                  ncpus=4, 
                  memory=1024, ## Mb
                  partition = &quot;short&quot;
                  )
sal &lt;- addResources(sal, c(&quot;bowtie2_alignment&quot;), resources = resources)
sal &lt;- runWF(sal)
</code></pre>
<h3 id="visualize-workflow">Visualize workflow</h3>
<p><em><code>systemPipeR</code></em> workflows instances can be visualized with the <code>plotWF</code> function.</p>
<pre><code class="language-r">plotWF(sal)
</code></pre>
<div id="htmlwidget-3" style="width:672px;height:480px;" class="plotwf html-widget"></div>
<script type="application/json" data-for="htmlwidget-3">{"x":{"dot":"digraph {\n    node[fontsize=20];\n    subgraph {\n        load_SPR -> preprocessing -> bowtie2_index -> bowtie2_alignment -> merge_bams -> writeTargetsRef -> call_peaks_macs_withref -> bcrank_enrich -> sessionInfo\n   }\n    preprocessing -> fastq_report\n    bowtie2_alignment -> align_stats\n    bowtie2_alignment -> bam_IGV\n    call_peaks_macs_withref -> annotation_ChIPseeker\n    call_peaks_macs_withref -> count_peak_ranges\n    call_peaks_macs_withref -> parse_peak_sequences\n    annotation_ChIPseeker -> ChIPseeker_plots\n    annotation_ChIPseeker -> go_enrich\n    count_peak_ranges -> diff_bind_analysis\n    load_SPR[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">load_SPR<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step load_SPR: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    preprocessing[fillcolor=\"#d3d6eb\" style=\"filled, rounded\" label=<<b><font color=\"black\">preprocessing<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">7<\/font><\/b>; <font color=\"black\">0s<\/font>> , shape=\"box\"  tooltip=\"step preprocessing: 0 samples passed; 0 samples have warnings; 0 samples have errors; 7 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    fastq_report[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">fastq_report<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step fastq_report: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    bowtie2_index[fillcolor=\"#d3d6eb\" style=\"filled, rounded\" label=<<b><font color=\"black\">bowtie2_index<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">6<\/font><\/b>; <font color=\"black\">0s<\/font>> , shape=\"box\"  tooltip=\"step bowtie2_index: 0 samples passed; 0 samples have warnings; 0 samples have errors; 6 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    bowtie2_alignment[fillcolor=\"#d3d6eb\" style=\"filled, rounded\" label=<<b><font color=\"black\">bowtie2_alignment<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">28<\/font><\/b>; <font color=\"black\">0s<\/font>> , shape=\"box\"  tooltip=\"step bowtie2_alignment: 0 samples passed; 0 samples have warnings; 0 samples have errors; 28 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    align_stats[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">align_stats<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step align_stats: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    bam_IGV[style=\"solid, \"label=<<b><font color=\"black\">bam_IGV<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step bam_IGV: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    merge_bams[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">merge_bams<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step merge_bams: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    writeTargetsRef[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">writeTargetsRef<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step writeTargetsRef: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    call_peaks_macs_withref[fillcolor=\"#d3d6eb\" style=\"filled, rounded\" label=<<b><font color=\"black\">call_peaks_macs_withref<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">5<\/font><\/b>; <font color=\"black\">0s<\/font>> , shape=\"box\"  tooltip=\"step call_peaks_macs_withref: 0 samples passed; 0 samples have warnings; 0 samples have errors; 5 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    annotation_ChIPseeker[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">annotation_ChIPseeker<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step annotation_ChIPseeker: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    ChIPseeker_plots[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">ChIPseeker_plots<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step ChIPseeker_plots: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    count_peak_ranges[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">count_peak_ranges<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step count_peak_ranges: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    diff_bind_analysis[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">diff_bind_analysis<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step diff_bind_analysis: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    go_enrich[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">go_enrich<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step go_enrich: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    parse_peak_sequences[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">parse_peak_sequences<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step parse_peak_sequences: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    bcrank_enrich[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">bcrank_enrich<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step bcrank_enrich: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n    sessionInfo[fillcolor=\"#d3d6eb\" style=\"filled, \"label=<<b><font color=\"black\">sessionInfo<\/font><br><\/br><font color=\"#5cb85c\">0<\/font>/<font color=\"#f0ad4e\">0<\/font>/<font color=\"#d9534f\">0<\/font>/<font color=\"blue\">1<\/font><\/b>; <font color=\"black\">0s<\/font>>  tooltip=\"step sessionInfo: 0 samples passed; 0 samples have warnings; 0 samples have errors; 1 samples in total; Start time: 2022-04-30 18:12:57; End time: 2022-04-30 18:12:57; Duration: 00:00:00\"]\n        subgraph cluster_legend {\n        rankdir=TB;\n        color=\"#eeeeee\";\n        style=filled;\n        ranksep =1;\n        label=\"Legends\";\n        fontsize = 30;\n        node [style=filled, fontsize=10];\n        legend_img-> step_state[color=\"#eeeeee\"];\n\n        legend_img[shape=none, image=\"plotwf_legend-src.png\", label = \" \", height=1, width=3, style=\"\"];\n\n        step_state[style=\"filled\", shape=\"box\" color=white, label =<\n            <table>\n            <tr><td><b>Step Colors<\/b><\/td><\/tr>\n            <tr><td><font color=\"black\">Pending steps<\/font>; <font color=\"#5cb85c\">Successful steps<\/font>; <font color=\"#d9534f\">Failed steps<\/font><\/td><\/tr>\n            <tr><td><b>Targets Files / Code Chunk <\/b><\/td><\/tr><tr><td><font color=\"#5cb85c\">0 (pass) <\/font> | <font color=\"#f0ad4e\">0 (warning) <\/font> | <font color=\"#d9534f\">0 (error) <\/font> | <font color=\"blue\">0 (total)<\/font>; Duration<\/td><\/tr><\/table>\n            >];\n\n    }\n\n}\n","plotid":"sprwf-75681234","responsive":true,"width":null,"height":null,"plot_method":"renderSVGElement","rmd":true,"msg":"","plot_ctr":true,"pan_zoom":false,"legend_uri":"data:image/svg+xml;base64,<?xml version="1.0" encoding="UTF-8"?>
<!-- Do not edit this file with editors other than diagrams.net -->
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="496px" height="278px" viewBox="-0.5 -0.5 496 278" content="&lt;mxfile host=&quot;app.diagrams.net&quot; modified=&quot;2021-11-24T20:39:44.923Z&quot; agent=&quot;5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.45 Safari/537.36&quot; version=&quot;15.8.3&quot; etag=&quot;T_dExnL7SLX2buX8PaX3&quot; type=&quot;google&quot;&gt;&lt;diagram id=&quot;Vw9VViGsx-sL_NjKe7JP&quot;&gt;7VrbcuI4EP0aHpOyJd94DARmXmZ3K2zVPitYGFVkyWPLgczXr2TL+CID3sGQTRUkldgtqS2d02p1t5nAebz/lqJk+4OHmE6AFe4n8HkCgA0AmKhfK/zQEsvxS0mUklDLasGK/MJVRy3NSYizVkfBORUkaQvXnDG8Fi0ZSlO+a3fbcNp+aoIibAhWa0RN6T8kFNtSOrWsWv4dk2hbPdmrWmJUddaCbItCvmuI4GIC5ynnoryK93NMFXoVLuW45ZHWw8RSzMSQAZqJd0RzvTY9L/FRLTZKeZ5M4Ez+YyFW4yx5d5i4utFKcCrwvo8A9Fops8wJ2odlS4PBPMYi/ZBdKkWBHqJtBUB9v6uBB45byrZN0KGnCddkRwfdNR7yQkPSDw88D08blt2WCLxK0Fq17qT5S9lWxFL/sy0vN4TSOac8LcbC0MVB6Cg0RcrfcNXCOJPDZxFFWdYHd/aGxXp7DvsmxqAfY42pA4dB6oyAqHNdRJfLRfC8OIboVVEMpgNR9C9H0b0uiovlEiznt0XxLGzwcti8sWHjTOjTKTBR3LjqR8oRJRGTMoo34hioSlVjbPmRci4fToQysWBEhH3rSg7TNxDOOJWNXZjl9EUbywqktVwGlijM1CKJPHOfdENMwlANn6U4I7/0kaIwSThhopizO5u4z0pXLnhW8mIfBbxBVoODJp3ydoliQhX6f5NYRhzA+gPv5N8XHiM2lI7TBxzw/Ue3fcQB05PAHsKqw/QSvgKDrxBlW3wn7Dhh0HUNwgKDMOdKhE0Nwn4gJuPNWM3+q5Fm27fbZm7w6Hdoc81Qss8xjkFblbk0eJvzOMkFvpN2Yqs5xlbr4cy7Fmd2T7jg0eIIT+QSm6x5P3OVshW4PJT4PskOQbKv2+RVpP7Lp8hEFVe65DRKdWWrYQ9V9/8UxIV4g3Iqbre9vKDLlO8ZTLn2Y09kB0YIiG0zkX2576wTfAV2hy9o9WSB19pZZl4tvaFcXPhACbu7xFMuEXZdIrRsg7jgWsSZ6Xuq6jssoiZrdSplX+K8PiHAA0N8Gej1Ze4IIA/I7jELn1QhszbHBpadSl1hz1WdEh4wwqFR5DyLUHP9PSZWyVJMkSDvbfV9kOgn/KX2ZPMwmT6CDgPdqknG83SN9UDQKHB2dEFgG3FfV5dAaYSFoavg6bD4YdQNqDBcRJ3OzsrO/3ceZabUAd57DKxp/XF/j1PHMlLmM5pHZNiscMgELESC601+P7eOBIhdpwod06leLZQ36xx/JoJwhuidtFMnoTWAtD4PMgppZq1jlPxLwTs0+bro7Q4MPfx6quB7i+pi+/UZBGaU7/i9kcwIDAKz6jEKg4zLllvRuCw+n0ujGZLemEizFPKSM0ZYNFHqSxJe0yGE2pZktKCyQ+oKZ5n0yF/IH+s5dt4E2dbNzMIOgnYYBL3KAs74ZziGUZg1l8oUQvJ+kS2c9RKvaP0WFbv6YV0yo/SJFLGsgm+mfMTA5x2M+Yg7OYiLhekx+GdO0q9WeP98m/XbNguBGVH0VeFHsViz7rQSOBnPh82Lr2zcrWG4NUxh24MFZjGr71Xab1iDvK2/V1VmfvXX0+DiXw==&lt;/diagram&gt;&lt;/mxfile&gt;"><defs/><g><rect x="4" y="90" width="490" height="92" fill="#d5e8d4" stroke="none" pointer-events="none"/><rect x="4" y="182" width="490" height="94" fill="#ffe8de" stroke="none" pointer-events="none"/><rect x="4" y="4" width="490" height="86" fill="#eff2fc" stroke="none" pointer-events="none"/><rect x="4" y="4" width="140" height="272" fill-opacity="0.8" fill="#f5f5f5" stroke="none" pointer-events="none"/><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 11px; margin-left: 115px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 8px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: nowrap;">solid</div></div></div></foreignObject><text x="115" y="13" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="8px" text-anchor="middle">solid</text></switch></g><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 10px; margin-left: 198px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 8px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: nowrap;">dashed</div></div></div></foreignObject><text x="198" y="12" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="8px" text-anchor="middle">dashed</text></switch></g><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 32px; margin-left: 116px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 11px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: nowrap;">Management</div></div></div></foreignObject><text x="116" y="35" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="11px" text-anchor="middle">Management</text></switch></g><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 32px; margin-left: 198px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 11px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: nowrap;">Compute</div></div></div></foreignObject><text x="198" y="35" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="11px" text-anchor="middle">Compute</text></switch></g><ellipse cx="232.5" cy="123" rx="51.5" ry="27" fill="rgba(255, 255, 255, 1)" stroke="rgba(0, 0, 0, 1)" stroke-width="2" pointer-events="none"/><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 50px; height: 1px; padding-top: 62px; margin-left: 92px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 12px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: normal; overflow-wrap: normal;"><span style="font-size: 8px">ellipse</span></div></div></div></foreignObject><text x="116" y="65" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="12px" text-anchor="middle">ellipse</text></switch></g><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 85px; margin-left: 114px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 11px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: nowrap;">R</div></div></div></foreignObject><text x="114" y="88" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="11px" text-anchor="middle">R</text></switch></g><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 83px; margin-left: 198px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 11px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: nowrap;">Command-line</div></div></div></foreignObject><text x="198" y="86" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="11px" text-anchor="middle">Command-line</text></switch></g><rect x="349" y="96" width="105" height="50" rx="7.5" ry="7.5" fill="rgba(255, 255, 255, 1)" stroke="rgba(0, 0, 0, 1)" stroke-width="2" pointer-events="none"/><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 51px; height: 1px; padding-top: 61px; margin-left: 176px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 8px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: normal; overflow-wrap: normal;">rectangle</div></div></div></foreignObject><text x="201" y="63" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="8px" text-anchor="middle">rectangle</text></switch></g><path d="M 182.5 38 L 287.5 38" fill="none" stroke="rgba(0, 0, 0, 1)" stroke-width="6" stroke-miterlimit="10" pointer-events="none"/><path d="M 354 37.62 L 459 37.62" fill="none" stroke="rgba(0, 0, 0, 1)" stroke-width="6" stroke-miterlimit="10" stroke-dasharray="18 18" pointer-events="none"/><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 128px; margin-left: 115px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 11px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: nowrap;">Mandatory</div></div></div></foreignObject><text x="115" y="131" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="11px" text-anchor="middle">Mandatory</text></switch></g><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 128px; margin-left: 198px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 11px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: nowrap;">Optional</div></div></div></foreignObject><text x="198" y="131" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="11px" text-anchor="middle">Optional</text></switch></g><rect x="184" y="190" width="95" height="40" fill="#d3d6eb" stroke="none" pointer-events="none"/><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 46px; height: 1px; padding-top: 105px; margin-left: 93px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 12px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: normal; overflow-wrap: normal;"><span style="font-size: 8px">fill</span></div></div></div></foreignObject><text x="116" y="109" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="12px" text-anchor="middle">fill</text></switch></g><rect x="349" y="190" width="95" height="40" fill="#ffffff" stroke="none" pointer-events="none"/><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 46px; height: 1px; padding-top: 105px; margin-left: 176px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 12px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; white-space: normal; overflow-wrap: normal;"><span style="font-size: 8px">no fill</span></div></div></div></foreignObject><text x="198" y="109" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="12px" text-anchor="middle">no fill</text></switch></g><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 24px; margin-left: 35px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 10px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; font-weight: bold; white-space: nowrap;">Running <br style="font-size: 10px" />Session</div></div></div></foreignObject><text x="35" y="27" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="10px" text-anchor="middle" font-weight="bold">Running...</text></switch></g><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 113px; margin-left: 35px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 10px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; font-weight: bold; white-space: nowrap;"><div style="font-size: 10px"><span style="background-color: transparent ; font-size: 10px">Running</span></div>Requirement</div></div></div></foreignObject><text x="35" y="116" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="10px" text-anchor="middle" font-weight="bold">RunningRequire...</text></switch></g><g transform="translate(-0.5 -0.5)scale(2)"><switch><foreignObject pointer-events="none" width="100%" height="100%" requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility" style="overflow: visible; text-align: left;"><div xmlns="http://www.w3.org/1999/xhtml" style="display: flex; align-items: unsafe center; justify-content: unsafe center; width: 1px; height: 1px; padding-top: 68px; margin-left: 35px;"><div data-drawio-colors="color: rgba(0, 0, 0, 1); " style="box-sizing: border-box; font-size: 0px; text-align: center;"><div style="display: inline-block; font-size: 10px; font-family: &quot;Times New Roman&quot;; color: rgb(0, 0, 0); line-height: 1.2; pointer-events: none; font-weight: bold; white-space: nowrap;">Step <br style="font-size: 10px" />Class</div></div></div></foreignObject><text x="35" y="71" fill="rgba(0, 0, 0, 1)" font-family="Times New Roman" font-size="10px" text-anchor="middle" font-weight="bold">Step...</text></switch></g></g><switch><g requiredFeatures="http://www.w3.org/TR/SVG11/feature#Extensibility"/><a transform="translate(0,-5)" xlink:href="https://www.diagrams.net/doc/faq/svg-export-text-problems" target="_blank"><text text-anchor="middle" font-size="10px" x="50%" y="100%">Viewer does not support full SVG 1.1</text></a></switch></svg>"},"evals":[],"jsHooks":[]}</script>
<h3 id="checking-workflow-status">Checking workflow status</h3>
<p>To check the summary of the workflow, we can use:</p>
<pre><code class="language-r">sal
statusWF(sal)
</code></pre>
<h2 id="technical-report">Technical report</h2>
<p><em><code>systemPipeR</code></em> compiles all the workflow execution logs in one central location,
making it easier to check any standard output (<code>stdout</code>) or standard error
(<code>stderr</code>) for any command-line tools used on the workflow or the R code stdout.</p>
<pre><code class="language-r">sal &lt;- renderLogs(sal)
</code></pre>
<h2 id="scientific-report">Scientific report</h2>
<p><em><code>systemPipeR</code></em> auto-generates scientific analysis reports in HTML format.</p>
<pre><code class="language-r">sal &lt;- renderReport(sal)
</code></pre>
<p>Alternatively, scientific reports can be rendered with the <code>render</code> function from <code>rmarkdown</code>.</p>
<pre><code class="language-r">rmarkdown::render(&quot;systemPipeChIPseq.Rmd&quot;, clean = TRUE, output_format = &quot;BiocStyle::html_document&quot;)
</code></pre>
<h2 id="funding">Funding</h2>
<p>This project was supported by funds from the National Institutes of
Health (NIH) and the National Science Foundation (NSF).</p>
<h2 id="session-info">Session Info</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.3.0 (2023-04-21)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 11 (bullseye)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0 
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: America/Los_Angeles
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils    
## [6] datasets  methods   base     
## 
## other attached packages:
##  [1] DT_0.27                     systemPipeRdata_2.4.0      
##  [3] systemPipeR_2.6.0           ShortRead_1.58.0           
##  [5] GenomicAlignments_1.36.0    SummarizedExperiment_1.30.1
##  [7] Biobase_2.60.0              MatrixGenerics_1.12.0      
##  [9] matrixStats_0.63.0          BiocParallel_1.34.1        
## [11] Rsamtools_2.16.0            Biostrings_2.68.0          
## [13] XVector_0.40.0              GenomicRanges_1.52.0       
## [15] GenomeInfoDb_1.36.0         IRanges_2.34.0             
## [17] S4Vectors_0.38.1            BiocGenerics_0.46.0        
## [19] BiocStyle_2.28.0           
## 
## loaded via a namespace (and not attached):
##  [1] gtable_0.3.3            xfun_0.39              
##  [3] bslib_0.4.2             hwriter_1.3.2.1        
##  [5] ggplot2_3.4.2           remotes_2.4.2          
##  [7] htmlwidgets_1.6.2       latticeExtra_0.6-30    
##  [9] lattice_0.21-8          generics_0.1.3         
## [11] vctrs_0.6.2             tools_4.3.0            
## [13] bitops_1.0-7            parallel_4.3.0         
## [15] tibble_3.2.1            fansi_1.0.4            
## [17] pkgconfig_2.0.3         Matrix_1.5-4           
## [19] RColorBrewer_1.1-3      lifecycle_1.0.3        
## [21] GenomeInfoDbData_1.2.10 stringr_1.5.0          
## [23] compiler_4.3.0          deldir_1.0-6           
## [25] munsell_0.5.0           codetools_0.2-19       
## [27] htmltools_0.5.5         sass_0.4.6             
## [29] RCurl_1.98-1.12         yaml_2.3.7             
## [31] pillar_1.9.0            crayon_1.5.2           
## [33] jquerylib_0.1.4         DelayedArray_0.26.2    
## [35] cachem_1.0.8            tidyselect_1.2.0       
## [37] digest_0.6.31           stringi_1.7.12         
## [39] dplyr_1.1.2             bookdown_0.33          
## [41] fastmap_1.1.1           grid_4.3.0             
## [43] colorspace_2.1-0        cli_3.6.1              
## [45] magrittr_2.0.3          S4Arrays_1.0.1         
## [47] utf8_1.2.3              scales_1.2.1           
## [49] rmarkdown_2.21          jpeg_0.1-10            
## [51] interp_1.1-4            blogdown_1.16          
## [53] png_0.1-8               evaluate_0.21          
## [55] knitr_1.42              rlang_1.1.1            
## [57] Rcpp_1.0.10             glue_1.6.2             
## [59] formatR_1.14            BiocManager_1.30.20    
## [61] jsonlite_1.8.4          R6_2.5.1               
## [63] zlibbioc_1.46.0
</code></pre>
<h2 id="references">References</h2>
<div id="refs" class="references hanging-indent">
<div id="ref-H_Backman2016-bt">
<p>H Backman, Tyler W, and Thomas Girke. 2016. “systemPipeR: NGS workflow and report generation environment.” <em>BMC Bioinformatics</em> 17 (1): 388. <a href="https://doi.org/10.1186/s12859-016-1241-0">https://doi.org/10.1186/s12859-016-1241-0</a>.</p>
</div>
<div id="ref-Kaufmann2010-me">
<p>Kaufmann, Kerstin, Frank Wellmer, Jose M Muiño, Thilia Ferrier, Samuel E Wuest, Vijaya Kumar, Antonio Serrano-Mislata, et al. 2010. “Orchestration of floral initiation by APETALA1.” <em>Science</em> 328 (5974): 85–89. <a href="https://doi.org/10.1126/science.1185244">https://doi.org/10.1126/science.1185244</a>.</p>
</div>
<div id="ref-Langmead2012-bs">
<p>Langmead, Ben, and Steven L Salzberg. 2012. “Fast Gapped-Read Alignment with Bowtie 2.” <em>Nat. Methods</em> 9 (4): 357–59. <a href="https://doi.org/10.1038/nmeth.1923">https://doi.org/10.1038/nmeth.1923</a>.</p>
</div>
<div id="ref-Love2014-sh">
<p>Love, Michael, Wolfgang Huber, and Simon Anders. 2014. “Moderated Estimation of Fold Change and Dispersion for RNA-seq Data with DESeq2.” <em>Genome Biol.</em> 15 (12): 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.</p>
</div>
<div id="ref-Robinson2010-uk">
<p>Robinson, M D, D J McCarthy, and G K Smyth. 2010. “EdgeR: A Bioconductor Package for Differential Expression Analysis of Digital Gene Expression Data.” <em>Bioinformatics</em> 26 (1): 139–40. <a href="https://doi.org/10.1093/bioinformatics/btp616">https://doi.org/10.1093/bioinformatics/btp616</a>.</p>
</div>
<div id="ref-Yu2015-xu">
<p>Yu, Guangchuang, Li-Gen Wang, and Qing-Yu He. 2015. “ChIPseeker: An R/Bioconductor Package for ChIP Peak Annotation, Comparison and Visualization.” <em>Bioinformatics</em> 31 (14): 2382–3. <a href="https://doi.org/10.1093/bioinformatics/btv145">https://doi.org/10.1093/bioinformatics/btv145</a>.</p>
</div>
<div id="ref-Zhang2008-pc">
<p>Zhang, Y, T Liu, C A Meyer, J Eeckhoute, D S Johnson, B E Bernstein, C Nussbaum, et al. 2008. “Model-Based Analysis of ChIP-Seq (MACS).” <em>Genome Biol.</em> 9 (9). <a href="https://doi.org/10.1186/gb-2008-9-9-r137">https://doi.org/10.1186/gb-2008-9-9-r137</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2023-05-09: <a  href="https://github.com/tgirke/GEN242//commit/02387da79023ee2be7e7e51805de9398165e681d">some edits (02387da79)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2023 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
