<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>NGS Analysis Basics | GEN242</title><meta property="og:title" content="NGS Analysis Basics" />
<meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Overview Sequence Analysis in R and Bioconductor R Base
 Some basic string handling utilities. Wide spectrum of numeric data analysis tools.  Bioconductor
Bioconductor packages provide much more sophisticated string handling utilities for sequence analysis (Lawrence et al." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/rsequences/rsequences/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2022-04-21T12:38:29-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="NGS Analysis Basics">
<meta itemprop="description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Overview Sequence Analysis in R and Bioconductor R Base
 Some basic string handling utilities. Wide spectrum of numeric data analysis tools.  Bioconductor
Bioconductor packages provide much more sophisticated string handling utilities for sequence analysis (Lawrence et al.">
<meta itemprop="dateModified" content="2022-04-21T12:38:29-07:00" />
<meta itemprop="wordCount" content="7080">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="NGS Analysis Basics"/>
<meta name="twitter:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Overview Sequence Analysis in R and Bioconductor R Base
 Some basic string handling utilities. Wide spectrum of numeric data analysis tools.  Bioconductor
Bioconductor packages provide much more sophisticated string handling utilities for sequence analysis (Lawrence et al."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S22)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T19)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2022/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.828f0f0bb84f9505b932d9dc4c3393d9.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.828f0f0bb84f9505b932d9dc4c3393d9.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../tutorials/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Tutorials</a>
  </li>
  <ul>
    <li class="collapse show" id="tutorials">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsgithubgithub" href="../../../tutorials/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialslinuxlinux" href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrbasicsrbasics" href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrprogrammingrprogramming" href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrparallelrparallel" href="../../../tutorials/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-tutorialsrsequencesrsequences" href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempiper" href="../../../tutorials/systempiper/systempiper/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempipernew" href="../../../tutorials/systempiper/systempipernew/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseq" href="../../../tutorials/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseqnew" href="../../../tutorials/sprnaseq/sprnaseqnew/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspchipseqspchipseq" href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsscrnaseqscrnaseq" href="../../../tutorials/scrnaseq/scrnaseq/">scRNA-Seq Embedding Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrmarkdownrmarkdown" href="../../../tutorials/rmarkdown/rmarkdown/">R Markdown Tutorial</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrfearfea" href="../../../tutorials/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrclusteringrclustering" href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrgraphicsrgraphics" href="../../../tutorials/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappsshinyapps" href="../../../tutorials/shinyapps/shinyapps/">Shiny Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrpackagesrpackages" href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsdplyrdplyr" href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialscmdtocwlcmdtocwl" href="../../../tutorials/cmdtocwl/cmdtocwl/">Automate Creation of CWL Instructions</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rsequences/Rsequences.Rmd" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/tutorials/rsequences/Rsequences.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/tutorials/rsequences/Rsequences.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=NGS%20Analysis%20Basics" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#overview">Overview</a>
      <ul>
        <li><a href="#sequence-analysis-in-r-and-bioconductor">Sequence Analysis in R and Bioconductor</a></li>
      </ul>
    </li>
    <li><a href="#package-requirements">Package Requirements</a></li>
    <li><a href="#strings-in-r-base">Strings in R Base</a>
      <ul>
        <li><a href="#basic-string-matching-and-parsing">Basic String Matching and Parsing</a>
          <ul>
            <li><a href="#string-matching">String matching</a></li>
            <li><a href="#positional-parsing">Positional parsing</a></li>
          </ul>
        </li>
        <li><a href="#random-sequence-generation">Random Sequence Generation</a>
          <ul>
            <li><a href="#random-dna-sequences-of-any-length">Random DNA sequences of any length</a></li>
            <li><a href="#count-identical-sequences">Count identical sequences</a></li>
            <li><a href="#extract-reads-from-reference">Extract reads from reference</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#sequences-in-bioconductor">Sequences in Bioconductor</a>
      <ul>
        <li><a href="#important-data-objects-of-biostrings">Important Data Objects of Biostrings</a>
          <ul>
            <li><a href="#xstring-for-single-sequence"><code>XString</code> for single sequence</a></li>
            <li><a href="#xstringset-for-many-sequences"><code>XStringSet</code> for many sequences</a></li>
            <li><a href="#qualityscalexstringset-for-sequences-with-quality-data"><code>QualityScaleXStringSet</code> for sequences with quality data</a></li>
          </ul>
        </li>
        <li><a href="#sequence-import-and-export">Sequence Import and Export</a></li>
        <li><a href="#working-with-xstring-containers">Working with <code>XString</code> Containers</a></li>
        <li><a href="#working-with-xstringset-containers">Working with <code>XStringSet</code> Containers</a></li>
        <li><a href="#multiple-alignment-class">Multiple Alignment Class</a></li>
        <li><a href="#basic-sequence-manipulations">Basic Sequence Manipulations</a>
          <ul>
            <li><a href="#reverse-and-complement">Reverse and Complement</a></li>
          </ul>
        </li>
        <li><a href="#translate-dna-into-protein">Translate DNA into Protein</a></li>
        <li><a href="#pattern-matching">Pattern Matching</a>
          <ul>
            <li><a href="#pattern-matching-with-mismatches">Pattern matching with mismatches</a></li>
            <li><a href="#pattern-matching-with-regular-expression-support">Pattern matching with regular expression support</a></li>
          </ul>
        </li>
        <li><a href="#pwm-viewing-and-searching">PWM Viewing and Searching</a>
          <ul>
            <li><a href="#plot-with-seqlogo">Plot with <code>seqLogo</code></a></li>
            <li><a href="#plot-with-ggseqlogo">Plot with <code>ggseqlogo</code></a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#ngs-sequences">NGS Sequences</a>
      <ul>
        <li><a href="#sequence-and-quality-data-fastq-format">Sequence and Quality Data: FASTQ Format</a></li>
        <li><a href="#sequence-and-quality-data-qualityscalexstringset">Sequence and Quality Data: <code>QualityScaleXStringSet</code></a></li>
        <li><a href="#processing-fastq-files-with-shortread">Processing FASTQ Files with ShortRead</a></li>
        <li><a href="#fastq-quality-reports">FASTQ Quality Reports</a>
          <ul>
            <li><a href="#using-systempiper">Using <code>systemPipeR</code></a></li>
            <li><a href="#using-shortread">Using <code>ShortRead</code></a></li>
          </ul>
        </li>
        <li><a href="#filtering-and-trimming-fastq-files-with-shortread">Filtering and Trimming FASTQ Files with ShortRead</a>
          <ul>
            <li><a href="#adaptor-trimming">Adaptor trimming</a></li>
            <li><a href="#read-counting-and-duplicate-removal">Read counting and duplicate removal</a></li>
            <li><a href="#trimming-low-quality-tails">Trimming low quality tails</a></li>
            <li><a href="#removal-of-reads-with-phred-scores-below-a-threshold-value">Removal of reads with Phred scores below a threshold value</a></li>
            <li><a href="#removal-of-reads-with-x-ns-andor-low-complexity-segments">Removal of reads with x Ns and/or low complexity segments</a></li>
          </ul>
        </li>
        <li><a href="#memory-efficient-fastq-processing">Memory Efficient FASTQ Processing</a></li>
      </ul>
    </li>
    <li><a href="#range-operations">Range Operations</a>
      <ul>
        <li><a href="#important-data-objects-for-range-operations">Important Data Objects for Range Operations</a></li>
        <li><a href="#range-data-are-stored-in-iranges-and-granges-containers">Range Data Are Stored in <code>IRanges</code> and <code>GRanges</code> Containers</a>
          <ul>
            <li><a href="#construct-granges-object">Construct <code>GRanges</code> Object</a></li>
            <li><a href="#import-gff-into-granges-object">Import GFF into <code>GRanges</code> Object</a></li>
            <li><a href="#coerce-granges-object-to-dataframe">Coerce <code>GRanges</code> object to <code>data.frame</code></a></li>
          </ul>
        </li>
        <li><a href="#utilities-for-range-containers">Utilities for Range Containers</a>
          <ul>
            <li><a href="#accessor-and-subsetting-methods-for-granges-objects">Accessor and subsetting methods for GRanges objects</a></li>
            <li><a href="#useful-utilities-for-granges-objects">Useful utilities for GRanges objects</a></li>
          </ul>
        </li>
        <li><a href="#grangeslist-objects">GRangesList Objects</a></li>
      </ul>
    </li>
    <li><a href="#transcript-ranges">Transcript Ranges</a>
      <ul>
        <li><a href="#txdb-from-biomart"><code>txdb</code> from BioMart</a></li>
        <li><a href="#efficient-sequence-parsing">Efficient Sequence Parsing</a></li>
        <li><a href="#getseq"><code>getSeq</code></a>
          <ul>
            <li><a href="#extracttranscriptseqs"><code>extractTranscriptSeqs</code></a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#homework-6">Homework 6</a></li>
    <li><a href="#session-info">Session Info</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../tutorials/">Tutorials</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>NGS Analysis Basics</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 34 minute read</p>
           
	<!--
- Compile from command-line
Rscript -e "rmarkdown::render('Rsequences.Rmd', c('html_document'), clean=F); knitr::knit('Rsequences.Rmd', tangle=TRUE)"; Rscript ../md2jekyll.R Rsequences.knit.md 10; Rscript -e "rmarkdown::render('Rsequences.Rmd', c('pdf_document'))"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div style="text-align: right">
<p>Source code downloads:    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rsequences/Rsequences.Rmd">.Rmd</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rsequences/Rsequences.R">.R</a> ]</p>
</div>
<h2 id="overview">Overview</h2>
<h3 id="sequence-analysis-in-r-and-bioconductor">Sequence Analysis in R and Bioconductor</h3>
<p><strong>R Base</strong></p>
<ul>
<li>Some basic string handling utilities. Wide spectrum of numeric data analysis tools.</li>
</ul>
<p><strong>Bioconductor</strong></p>
<p>Bioconductor packages provide much more sophisticated string handling utilities for sequence analysis (Lawrence et al. 2013; Huber et al. 2015).</p>
<ul>
<li><a href="http://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a>: general sequence analysis environment</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/ShortRead.html">ShortRead</a>: pipeline for short read data</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a>: low-level infrastructure for range data</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>: high-level infrastructure for range data</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html">GenomicFeatures</a>: managing transcript centric annotations</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/GenomicAlignments.html">GenomicAlignments</a>: handling short genomic alignments</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/Rsamtools.html">Rsamtools</a>: interface to <code>samtools</code>, <code>bcftools</code> and <code>tabix</code></li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/BSgenome.html">BSgenome</a>: genome annotation data</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a>: interface to BioMart annotations</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a>: Annotation imports, interface to online genome browsers</li>
<li><a href="http://bioconductor.org/packages/release/bioc/html/HelloRanges.html">HelloRanges</a>: Bedtools semantics in Bioc’s Ranges infrastructure</li>
</ul>
<h2 id="package-requirements">Package Requirements</h2>
<p>Several Bioconductor packages are required for this tutorial. To install them, execute
the following lines in the R console. Please also make sure that you have a recent R version
installed on your system. R versions <code>4.0.x</code> or higher are recommended.</p>
<p><em>Please do not run this install on the HPCC unless you want to reinstall some of these packages in your own
user account.</em></p>
<pre><code class="language-r">source(&quot;https://bioconductor.org/biocLite.R&quot;)
if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;)
BiocManager::install(c(&quot;Biostrings&quot;, &quot;GenomicRanges&quot;, &quot;rtracklayer&quot;, &quot;systemPipeR&quot;, &quot;seqLogo&quot;, &quot;ShortRead&quot;))
</code></pre>
<h2 id="strings-in-r-base">Strings in R Base</h2>
<h3 id="basic-string-matching-and-parsing">Basic String Matching and Parsing</h3>
<h4 id="string-matching">String matching</h4>
<p>Generate sample sequence data set</p>
<pre><code class="language-r">myseq &lt;- c(&quot;ATGCAGACATAGTG&quot;, &quot;ATGAACATAGATCC&quot;, &quot;GTACAGATCAC&quot;)
</code></pre>
<p>String searching with regular expression support</p>
<pre><code class="language-r">myseq[grep(&quot;ATG&quot;, myseq)] 
</code></pre>
<pre><code>## [1] &quot;ATGCAGACATAGTG&quot; &quot;ATGAACATAGATCC&quot;
</code></pre>
<p>Searches <code>myseq</code> for first match of pattern “AT”</p>
<pre><code class="language-r">pos1 &lt;- regexpr(&quot;AT&quot;, myseq) 
as.numeric(pos1); attributes(pos1)$match.length # Returns position information of matches
</code></pre>
<pre><code>## [1] 1 1 7

## [1] 2 2 2
</code></pre>
<p>Searches <code>myseq</code> for all matches of pattern “AT”</p>
<pre><code class="language-r">pos2 &lt;- gregexpr(&quot;AT&quot;, myseq) 
as.numeric(pos2[[1]]); attributes(pos2[[1]])$match.length # Returns positions of matches in first sequence
</code></pre>
<pre><code>## [1] 1 9

## [1] 2 2
</code></pre>
<p>String substitution with regular expression support</p>
<pre><code class="language-r">gsub(&quot;^ATG&quot;, &quot;atg&quot;, myseq) 
</code></pre>
<pre><code>## [1] &quot;atgCAGACATAGTG&quot; &quot;atgAACATAGATCC&quot; &quot;GTACAGATCAC&quot;
</code></pre>
<h4 id="positional-parsing">Positional parsing</h4>
<pre><code class="language-r">nchar(myseq) # Computes length of strings
</code></pre>
<pre><code>## [1] 14 14 11
</code></pre>
<pre><code class="language-r">substring(myseq[1], c(1,3), c(2,5)) # Positional parsing of several fragments from one string
</code></pre>
<pre><code>## [1] &quot;AT&quot;  &quot;GCA&quot;
</code></pre>
<pre><code class="language-r">substring(myseq, c(1,4,7), c(2,6,10)) # Positional parsing of many strings
</code></pre>
<pre><code>## [1] &quot;AT&quot;   &quot;AAC&quot;  &quot;ATCA&quot;
</code></pre>
<h3 id="random-sequence-generation">Random Sequence Generation</h3>
<h4 id="random-dna-sequences-of-any-length">Random DNA sequences of any length</h4>
<pre><code class="language-r">rand &lt;- sapply(1:100, function(x) paste(sample(c(&quot;A&quot;,&quot;T&quot;,&quot;G&quot;,&quot;C&quot;), sample(10:20), replace=TRUE), collapse=&quot;&quot;))
rand[1:3]
</code></pre>
<pre><code>## [1] &quot;GCACACCATGTT&quot;         &quot;CCAAGGGGTCACGCAAGTAA&quot; &quot;GAATACGGAAAAG&quot;
</code></pre>
<h4 id="count-identical-sequences">Count identical sequences</h4>
<pre><code class="language-r">table(c(rand[1:4], rand[1]))
</code></pre>
<pre><code>## 
## CCAAGGGGTCACGCAAGTAA        GAATACGGAAAAG         GCACACCATGTT       TAATCATCGAAGTT 
##                    1                    1                    2                    1
</code></pre>
<h4 id="extract-reads-from-reference">Extract reads from reference</h4>
<p>Note: this requires the <code>Biostrings</code> package.</p>
<pre><code class="language-r">library(Biostrings)
ref &lt;- DNAString(paste(sample(c(&quot;A&quot;,&quot;T&quot;,&quot;G&quot;,&quot;C&quot;), 100000, replace=T), collapse=&quot;&quot;))
randstart &lt;- sample(1:(length(ref)-15), 1000)
randreads &lt;- Views(ref, randstart, width=15)
rand_set &lt;- DNAStringSet(randreads)
unlist(rand_set)
</code></pre>
<pre><code>## 15000-letter DNAString object
## seq: AATCGCGGCACTTAGCCAGCTTGTCTTATCGGGTATTTTGAATCTT...ACTGTCGCCAATGGGTCCAACGGATTCTACGCCACGACTCGTAACT
</code></pre>
<h2 id="sequences-in-bioconductor">Sequences in Bioconductor</h2>
<h3 id="important-data-objects-of-biostrings">Important Data Objects of Biostrings</h3>
<h4 id="xstring-for-single-sequence"><code>XString</code> for single sequence</h4>
<ul>
<li><code>DNAString</code>: for DNA</li>
<li><code>RNAString</code>: for RNA</li>
<li><code>AAString</code>: for amino acid</li>
<li><code>BString</code>: for any string</li>
</ul>
<h4 id="xstringset-for-many-sequences"><code>XStringSet</code> for many sequences</h4>
<ul>
<li>`DNAStringSet``: for DNA</li>
<li><code>RNAStringSet</code>: for RNA</li>
<li><code>AAStringSet</code>: for amino acid</li>
<li><code>BStringSet</code>: for any string</li>
</ul>
<h4 id="qualityscalexstringset-for-sequences-with-quality-data"><code>QualityScaleXStringSet</code> for sequences with quality data</h4>
<ul>
<li><code>QualityScaledDNAStringSet</code>: for DNA</li>
<li><code>QualityScaledRNAStringSet</code>: for RNA</li>
<li><code>QualityScaledAAStringSet</code>: for amino acid</li>
<li><code>QualityScaledBStringSet</code>: for any string</li>
</ul>
<h3 id="sequence-import-and-export">Sequence Import and Export</h3>
<p>Download the following sequences to your current working directory and then import them into R:
<a href="https://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.ffn">https://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.ffn</a></p>
<pre><code class="language-r">dir.create(&quot;data&quot;, showWarnings = FALSE)
# system(&quot;wget https://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.ffn&quot;)
download.file(&quot;https://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.ffn&quot;, &quot;data/AE004437.ffn&quot;)
</code></pre>
<p>Import FASTA file with <code>readDNAStringSet</code></p>
<pre><code class="language-r">myseq &lt;- readDNAStringSet(&quot;data/AE004437.ffn&quot;)
myseq[1:3]
</code></pre>
<pre><code>## DNAStringSet object of length 3:
##     width seq                                                                   names               
## [1]  1206 ATGACTCGGCGGTCTCGTGTCGGTGCCGGCCTC...GTCGTCGTTGTTCGACGCTGGCGGAACCCATGA gi|12057215|gb|AE...
## [2]   666 ATGAGCATCATCGAACTCGAAGGCGTGGTCAAA...GTCAACCTCGTCGATGGGGTGTTACACACGTGA gi|12057215|gb|AE...
## [3]  1110 ATGGCGTGGCGGAACCTCGGGCGGAACCGCGTG...AACGATCCGCCCGTCGAGGCGCTCGGCGAATGA gi|12057215|gb|AE...
</code></pre>
<p>Subset sequences with regular expression on sequence name field</p>
<pre><code class="language-r">sub &lt;- myseq[grep(&quot;99.*&quot;, names(myseq))]
length(sub)
</code></pre>
<pre><code>## [1] 170
</code></pre>
<p>Export subsetted sequences to FASTA file</p>
<pre><code class="language-r">writeXStringSet(sub, file=&quot;./data/AE004437sub.ffn&quot;, width=80)
</code></pre>
<p>Now inspect exported sequence file <code>AE004437sub.ffn</code> in a text editor</p>
<h3 id="working-with-xstring-containers">Working with <code>XString</code> Containers</h3>
<p>The <code>XString</code> stores the different types of biosequences in dedicated containers</p>
<pre><code class="language-r">library(Biostrings)
d &lt;- DNAString(&quot;GCATAT-TAC&quot;)
d
</code></pre>
<pre><code>## 10-letter DNAString object
## seq: GCATAT-TAC
</code></pre>
<pre><code class="language-r">d[1:4]
</code></pre>
<pre><code>## 4-letter DNAString object
## seq: GCAT
</code></pre>
<p>RNA sequences</p>
<pre><code class="language-r">r &lt;- RNAString(&quot;GCAUAU-UAC&quot;) 
r &lt;- RNAString(d) # Converts d to RNAString object
r
</code></pre>
<pre><code>## 10-letter RNAString object
## seq: GCAUAU-UAC
</code></pre>
<p>Protein sequences</p>
<pre><code class="language-r">p &lt;- AAString(&quot;HCWYHH&quot;)
p
</code></pre>
<pre><code>## 6-letter AAString object
## seq: HCWYHH
</code></pre>
<p>Any type of character strings</p>
<pre><code class="language-r">b &lt;- BString(&quot;I store any set of characters. Other XString objects store only the IUPAC characters.&quot;)
b
</code></pre>
<pre><code>## 85-letter BString object
## seq: I store any set of characters. Other XString objects store only the IUPAC characters.
</code></pre>
<h3 id="working-with-xstringset-containers">Working with <code>XStringSet</code> Containers</h3>
<p><code>XStringSet</code> containers allow to store many biosequences in one object</p>
<pre><code class="language-r">dset &lt;- DNAStringSet(c(&quot;GCATATTAC&quot;, &quot;AATCGATCC&quot;, &quot;GCATATTAC&quot;)) 
names(dset) &lt;- c(&quot;seq1&quot;, &quot;seq2&quot;, &quot;seq3&quot;) # Assigns names
dset[1:2]
</code></pre>
<pre><code>## DNAStringSet object of length 2:
##     width seq                                                                   names               
## [1]     9 GCATATTAC                                                             seq1
## [2]     9 AATCGATCC                                                             seq2
</code></pre>
<p>Important utilities for <code>XStringSet</code> containers</p>
<pre><code class="language-r">width(dset) # Returns the length of each sequences
</code></pre>
<pre><code>## [1] 9 9 9
</code></pre>
<pre><code class="language-r">d &lt;- dset[[1]] # The [[ subsetting operator returns a single entry as XString object
dset2 &lt;- c(dset, dset) # Appends/concatenates two XStringSet objects
dsetchar &lt;- as.character(dset) # Converts XStringSet to named vector 
dsetone &lt;- unlist(dset) # Collapses many sequences to a single one stored in a DNAString container
</code></pre>
<p>Sequence subsetting by positions:</p>
<pre><code class="language-r">DNAStringSet(dset, start=c(1,2,3), end=c(4,8,5)) 
</code></pre>
<pre><code>## DNAStringSet object of length 3:
##     width seq                                                                   names               
## [1]     4 GCAT                                                                  seq1
## [2]     7 ATCGATC                                                               seq2
## [3]     3 ATA                                                                   seq3
</code></pre>
<h3 id="multiple-alignment-class">Multiple Alignment Class</h3>
<p>The <code>XMultipleAlignment</code> class stores the different types of multiple sequence alignments:</p>
<pre><code class="language-r">origMAlign &lt;- readDNAMultipleAlignment(filepath = system.file(&quot;extdata&quot;,
              &quot;msx2_mRNA.aln&quot;, package = &quot;Biostrings&quot;), format = &quot;clustal&quot;)
origMAlign
</code></pre>
<pre><code>## DNAMultipleAlignment with 8 rows and 2343 columns
##      aln                                                                        names               
## [1] -----TCCCGTCTCCGCAGCAAAAAAGTTTGAGTCG...TTGTCCAAACTCACAATTAAAAAAAAAAAAAAAAA gi|84452153|ref|N...
## [2] ------------------------------------...----------------------------------- gi|208431713|ref|...
## [3] ------------------------------------...----------------------------------- gi|118601823|ref|...
## [4] ----------------------AAAAGTTGGAGTCT...----------------------------------- gi|114326503|ref|...
## [5] ------------------------------------...----------------------------------- gi|119220589|ref|...
## [6] ------------------------------------...----------------------------------- gi|148540149|ref|...
## [7] --------------CGGCTCCGCAGCGCCTCACTCG...----------------------------------- gi|45383056|ref|N...
## [8] GGGGGAGACTTCAGAAGTTGTTGTCCTCTCCGCTGA...----------------------------------- gi|213515133|ref|...
</code></pre>
<h3 id="basic-sequence-manipulations">Basic Sequence Manipulations</h3>
<h4 id="reverse-and-complement">Reverse and Complement</h4>
<pre><code class="language-r">randset &lt;- DNAStringSet(rand)
complement(randset[1:2])
</code></pre>
<pre><code>## DNAStringSet object of length 2:
##     width seq
## [1]    12 CGTGTGGTACAA
## [2]    20 GGTTCCCCAGTGCGTTCATT
</code></pre>
<pre><code class="language-r">reverse(randset[1:2])
</code></pre>
<pre><code>## DNAStringSet object of length 2:
##     width seq
## [1]    12 TTGTACCACACG
## [2]    20 AATGAACGCACTGGGGAACC
</code></pre>
<pre><code class="language-r">reverseComplement(randset[1:2])
</code></pre>
<pre><code>## DNAStringSet object of length 2:
##     width seq
## [1]    12 AACATGGTGTGC
## [2]    20 TTACTTGCGTGACCCCTTGG
</code></pre>
<h3 id="translate-dna-into-protein">Translate DNA into Protein</h3>
<pre><code class="language-r">translate(randset[1:2])
</code></pre>
<pre><code>## Warning in .Call2(&quot;DNAStringSet_translate&quot;, x, skip_code, dna_codes[codon_alphabet], : in 'x[[2]]':
## last 2 bases were ignored

## AAStringSet object of length 2:
##     width seq
## [1]     4 AHHV
## [2]     6 PRGHAS
</code></pre>
<h3 id="pattern-matching">Pattern Matching</h3>
<h4 id="pattern-matching-with-mismatches">Pattern matching with mismatches</h4>
<p>Find pattern matches in reference</p>
<pre><code class="language-r">myseq1 &lt;- readDNAStringSet(&quot;./data/AE004437.ffn&quot;) 
mypos &lt;- matchPattern(&quot;ATGGTG&quot;, myseq1[[1]], max.mismatch=1) 
</code></pre>
<p>Count only the corresponding matches</p>
<pre><code class="language-r">countPattern(&quot;ATGGCT&quot;, myseq1[[1]], max.mismatch=1) 
</code></pre>
<pre><code>## [1] 3
</code></pre>
<p>Count matches in many sequences</p>
<pre><code class="language-r">vcountPattern(&quot;ATGGCT&quot;, myseq1, max.mismatch=1)[1:20]
</code></pre>
<pre><code>##  [1] 3 0 5 4 1 2 2 1 4 3 0 0 1 2 0 1 4 0 0 1
</code></pre>
<p>Results shown in DNAStringSet object</p>
<pre><code class="language-r">tmp &lt;- c(DNAStringSet(&quot;ATGGTG&quot;), DNAStringSet(mypos)) 
</code></pre>
<p>Return a consensus matrix for query and hits</p>
<pre><code class="language-r">consensusMatrix(tmp)[1:4,] 
</code></pre>
<pre><code>##   [,1] [,2] [,3] [,4] [,5] [,6]
## A    3    0    0    0    0    0
## C    1    1    0    0    0    0
## G    0    0    4    4    1    4
## T    0    3    0    0    3    0
</code></pre>
<p>Find all pattern matches in reference</p>
<pre><code class="language-r">myvpos &lt;- vmatchPattern(&quot;ATGGCT&quot;, myseq1, max.mismatch=1) 
myvpos # The results are stored as MIndex object.
</code></pre>
<pre><code>## MIndex object of length 2058
## $`gi|12057215|gb|AE004437.1|:248-1453 Halobacterium sp. NRC-1, complete genome`
## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1         6         6
##   [2]       383       388         6
##   [3]       928       933         6
## 
## $`gi|12057215|gb|AE004437.1|:1450-2115 Halobacterium sp. NRC-1, complete genome`
## IRanges object with 0 ranges and 0 metadata columns:
##        start       end     width
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## 
## $`gi|12057215|gb|AE004437.1|:2145-3254 Halobacterium sp. NRC-1, complete genome`
## IRanges object with 5 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1         6         6
##   [2]        94        99         6
##   [3]       221       226         6
##   [4]       535       540         6
##   [5]       601       606         6
## 
## ...
## &lt;2055 more elements&gt;
</code></pre>
<pre><code class="language-r">Views(myseq1[[1]], start(myvpos[[1]]), end(myvpos[[1]])) # Retrieves the result for single entry
</code></pre>
<pre><code>## Views on a 1206-letter DNAString subject
## subject: ATGACTCGGCGGTCTCGTGTCGGTGCCGGCCTCGCAGCCATTGT...TTGCGATCGTCGTCGTCGTTGTTCGACGCTGGCGGAACCCATGA
## views:
##       start end width
##   [1]     1   6     6 [ATGACT]
##   [2]   383 388     6 [ATGGCA]
##   [3]   928 933     6 [ATGACT]
</code></pre>
<p>Return all matches</p>
<pre><code class="language-r">sapply(names(myseq1), function(x) 
       as.character(Views(myseq1[[x]], start(myvpos[[x]]), end(myvpos[[x]]))))[1:4] 
</code></pre>
<h4 id="pattern-matching-with-regular-expression-support">Pattern matching with regular expression support</h4>
<pre><code class="language-r">myseq &lt;- DNAStringSet(c(&quot;ATGCAGACATAGTG&quot;, &quot;ATGAACATAGATCC&quot;, &quot;GTACAGATCAC&quot;))
myseq[grep(&quot;^ATG&quot;, myseq, perl=TRUE)] # String searching with regular expression support
</code></pre>
<pre><code>## DNAStringSet object of length 2:
##     width seq
## [1]    14 ATGCAGACATAGTG
## [2]    14 ATGAACATAGATCC
</code></pre>
<pre><code class="language-r">pos1 &lt;- regexpr(&quot;AT&quot;, myseq) # Searches 'myseq' for first match of pattern &quot;AT&quot;
as.numeric(pos1); attributes(pos1)$match.length # Returns position information of matches
</code></pre>
<pre><code>## [1] 1 1 7

## [1] 2 2 2
</code></pre>
<pre><code class="language-r">pos2 &lt;- gregexpr(&quot;AT&quot;, myseq) # Searches 'myseq' for all matches of pattern &quot;AT&quot;
as.numeric(pos2[[1]]); attributes(pos2[[1]])$match.length # Match positions in first sequence
</code></pre>
<pre><code>## [1] 1 9

## [1] 2 2
</code></pre>
<pre><code class="language-r">DNAStringSet(gsub(&quot;^ATG&quot;, &quot;NNN&quot;, myseq)) # String substitution with regular expression support
</code></pre>
<pre><code>## DNAStringSet object of length 3:
##     width seq
## [1]    14 NNNCAGACATAGTG
## [2]    14 NNNAACATAGATCC
## [3]    11 GTACAGATCAC
</code></pre>
<h3 id="pwm-viewing-and-searching">PWM Viewing and Searching</h3>
<h4 id="plot-with-seqlogo">Plot with <code>seqLogo</code></h4>
<pre><code class="language-r">library(seqLogo) 
pwm &lt;- PWM(DNAStringSet(c(&quot;GCT&quot;, &quot;GGT&quot;, &quot;GCA&quot;))) 
pwm
</code></pre>
<pre><code>##        [,1]      [,2]      [,3]
## A 0.0000000 0.0000000 0.2312611
## C 0.0000000 0.3157205 0.0000000
## G 0.3685591 0.2312611 0.0000000
## T 0.0000000 0.0000000 0.3157205
</code></pre>
<pre><code class="language-r">seqLogo(t(t(pwm) * 1/colSums(pwm)))
</code></pre>
<img src="../../../en/tutorials/rsequences/Rsequences_files/figure-html/pwm_logo-1.png" width="672" />
<h4 id="plot-with-ggseqlogo">Plot with <code>ggseqlogo</code></h4>
<p>The <code>ggseqlogo</code> package (<a href="https://omarwagih.github.io/ggseqlogo/">manual</a>)
provides many customization options for plotting sequence logos. It also supports
various alphabets including sequence logos for amino acid sequences.</p>
<pre><code class="language-r">library(ggplot2); library(ggseqlogo)
pwm &lt;- PWM(DNAStringSet(c(&quot;GCT&quot;, &quot;GGT&quot;, &quot;GCA&quot;))) 
ggseqlogo(pwm)
</code></pre>
<pre><code>## Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = &quot;none&quot;)` instead.
</code></pre>
<img src="../../../en/tutorials/rsequences/Rsequences_files/figure-html/pwm_logo2-1.png" width="672" />
<p>Search sequence for PWM matches with score better than <code>min.score</code></p>
<pre><code class="language-r">chr &lt;- DNAString(&quot;AAAGCTAAAGGTAAAGCAAAA&quot;) 
matchPWM(pwm, chr, min.score=0.9) 
</code></pre>
<pre><code>## Views on a 21-letter DNAString subject
## subject: AAAGCTAAAGGTAAAGCAAAA
## views:
##       start end width
##   [1]     4   6     3 [GCT]
##   [2]    10  12     3 [GGT]
##   [3]    16  18     3 [GCA]
</code></pre>
<h2 id="ngs-sequences">NGS Sequences</h2>
<h3 id="sequence-and-quality-data-fastq-format">Sequence and Quality Data: FASTQ Format</h3>
<p>Four lines per sequence:</p>
<ol>
<li>ID</li>
<li>Sequence</li>
<li>ID</li>
<li>Base call qualities (Phred scores) as ASCII characters</li>
</ol>
<p>The following gives an example of 3 Illumina reads in a FASTQ file. The numbers
at the beginning of each line are not part of the FASTQ format. They have been added
solely for illustration purposes.</p>
<pre><code>1. @SRR038845.3 HWI-EAS038:6:1:0:1938 length=36
2. CAACGAGTTCACACCTTGGCCGACAGGCCCGGGTAA
3. +SRR038845.3 HWI-EAS038:6:1:0:1938 length=36
4. BA@7&gt;B=&gt;:&gt;&gt;7@7@&gt;&gt;9=BAA?;&gt;52;&gt;:9=8.=A
1. @SRR038845.41 HWI-EAS038:6:1:0:1474 length=36
2. CCAATGATTTTTTTCCGTGTTTCAGAATACGGTTAA
3. +SRR038845.41 HWI-EAS038:6:1:0:1474 length=36
4. BCCBA@BB@BBBBAB@B9B@=BABA@A:@693:@B=
1. @SRR038845.53 HWI-EAS038:6:1:1:360 length=36
2. GTTCAAAAAGAACTAAATTGTGTCAATAGAAAACTC
3. +SRR038845.53 HWI-EAS038:6:1:1:360 length=36
4. BBCBBBBBB@@BAB?BBBBCBC&gt;BBBAA8&gt;BBBAA@
</code></pre>
<h3 id="sequence-and-quality-data-qualityscalexstringset">Sequence and Quality Data: <code>QualityScaleXStringSet</code></h3>
<p>Phred quality scores are integers from 0-50 that are
stored as ASCII characters after adding 33. The basic R functions <code>rawToChar</code> and
<code>charToRaw</code> can be used to interconvert among their representations.</p>
<p>Phred score interconversion</p>
<pre><code class="language-r">phred &lt;- 1:9
phreda &lt;- paste(sapply(as.raw((phred)+33), rawToChar), collapse=&quot;&quot;)
phreda
</code></pre>
<pre><code>## [1] &quot;\&quot;#$%&amp;'()*&quot;
</code></pre>
<pre><code class="language-r">as.integer(charToRaw(phreda))-33 
</code></pre>
<pre><code>## [1] 1 2 3 4 5 6 7 8 9
</code></pre>
<p>Construct <code>QualityScaledDNAStringSet</code> from scratch</p>
<pre><code class="language-r">dset &lt;- DNAStringSet(sapply(1:100, function(x) paste(sample(c(&quot;A&quot;,&quot;T&quot;,&quot;G&quot;,&quot;C&quot;), 20, replace=T), collapse=&quot;&quot;))) # Creates random sample sequence.
myqlist &lt;- lapply(1:100, function(x) sample(1:40, 20, replace=T)) # Creates random Phred score list.
myqual &lt;- sapply(myqlist, function(x) toString(PhredQuality(x))) # Converts integer scores into ASCII characters.
myqual &lt;- PhredQuality(myqual) # Converts to a PhredQuality object.
dsetq1 &lt;- QualityScaledDNAStringSet(dset, myqual) # Combines DNAStringSet and quality data in QualityScaledDNAStringSet object.
dsetq1[1:2]
</code></pre>
<pre><code>##   A QualityScaledDNAStringSet instance containing:
## 
## DNAStringSet object of length 2:
##     width seq
## [1]    20 TAAAGGGCCCTACGCCTTGC
## [2]    20 TGCCTACACACAGTTAGTAC
## 
## PhredQuality object of length 2:
##     width seq
## [1]    20 DE4&lt;,-D;7:E94G724C1E
## [2]    20 H#)I&quot;/+2)3&gt;/I@?9'8/)
</code></pre>
<h3 id="processing-fastq-files-with-shortread">Processing FASTQ Files with ShortRead</h3>
<p>The following explains the basic usage of <code>ShortReadQ</code> objects. To make the sample code work,
download and unzip this <a href="http://cluster.hpcc.ucr.edu/~tgirke/HTML_Presentations/Manuals/Workshop_Dec_6_10_2012/Rsequences/data.zip">file</a> to your current working directory.
The following code performs the download for you.</p>
<pre><code class="language-r">library(ShortRead)
download.file(&quot;http://cluster.hpcc.ucr.edu/~tgirke/HTML_Presentations/Manuals/testdata/samplefastq/data.zip&quot;, &quot;data.zip&quot;)
unzip(&quot;data.zip&quot;)
</code></pre>
<p>Important utilities for accessing FASTQ files</p>
<pre><code class="language-r">fastq &lt;- list.files(&quot;data&quot;, &quot;*.fastq$&quot;); fastq &lt;- paste(&quot;data/&quot;, fastq, sep=&quot;&quot;)
names(fastq) &lt;- paste(&quot;flowcell6_lane&quot;, 1:length(fastq), sep=&quot;_&quot;) 
(fq &lt;- readFastq(fastq[1])) # Imports first FASTQ file
</code></pre>
<pre><code>## class: ShortReadQ
## length: 1000 reads; width: 36 cycles
</code></pre>
<pre><code class="language-r">countLines(dirPath=&quot;./data&quot;, pattern=&quot;.fastq$&quot;)/4 # Counts numbers of reads in FASTQ files
</code></pre>
<pre><code>## SRR038845.fastq SRR038846.fastq SRR038848.fastq SRR038850.fastq 
##            1000            1000            1000            1000
</code></pre>
<pre><code class="language-r">id(fq)[1] # Returns ID field
</code></pre>
<pre><code>## BStringSet object of length 1:
##     width seq
## [1]    43 SRR038845.3 HWI-EAS038:6:1:0:1938 length=36
</code></pre>
<pre><code class="language-r">sread(fq)[1] # Returns sequence
</code></pre>
<pre><code>## DNAStringSet object of length 1:
##     width seq
## [1]    36 CAACGAGTTCACACCTTGGCCGACAGGCCCGGGTAA
</code></pre>
<pre><code class="language-r">quality(fq)[1] # Returns Phred scores 
</code></pre>
<pre><code>## class: FastqQuality
## quality:
## BStringSet object of length 1:
##     width seq
## [1]    36 BA@7&gt;B=&gt;:&gt;&gt;7@7@&gt;&gt;9=BAA?;&gt;52;&gt;:9=8.=A
</code></pre>
<pre><code class="language-r">as(quality(fq), &quot;matrix&quot;)[1:4,1:12] # Coerces Phred scores to numeric matrix
</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12]
## [1,]   33   32   31   22   29   33   28   29   25    29    29    22
## [2,]   33   34   34   33   32   31   33   33   31    33    33    33
## [3,]   33   33   34   33   33   33   33   33   33    31    31    33
## [4,]   33   33   33   33   31   33   28   31   28    32    33    33
</code></pre>
<pre><code class="language-r">ShortReadQ(sread=sread(fq), quality=quality(fq), id=id(fq)) # Constructs a ShortReadQ from components
</code></pre>
<pre><code>## class: ShortReadQ
## length: 1000 reads; width: 36 cycles
</code></pre>
<h3 id="fastq-quality-reports">FASTQ Quality Reports</h3>
<h4 id="using-systempiper">Using <code>systemPipeR</code></h4>
<p>The following <code>seeFastq</code> and <code>seeFastqPlot</code> functions generate and plot a series of useful quality statistics for a set of FASTQ files.</p>
<pre><code class="language-r">library(systemPipeR)
fqlist &lt;- seeFastq(fastq=fastq, batchsize=800, klength=8) # For real data set batchsize to at least 10^5 
seeFastqPlot(fqlist)
</code></pre>
<img src="../../../en/tutorials/rsequences/Rsequences_files/figure-html/see_fastq-1.png" width="1344" />
<p>Handles many samples in one PDF file. For more details see <a href="http://bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">here</a></p>
<h4 id="using-shortread">Using <code>ShortRead</code></h4>
<p>The <code>ShortRead</code> package contains several FASTQ quality reporting functions.</p>
<pre><code class="language-r">sp &lt;- SolexaPath(system.file('extdata', package='ShortRead'))
fl &lt;- file.path(analysisPath(sp), &quot;s_1_sequence.txt&quot;) 
fls &lt;- c(fl, fl) 
coll &lt;- QACollate(QAFastqSource(fls), QAReadQuality(), QAAdapterContamination(), 
        QANucleotideUse(), QAQualityUse(), QASequenceUse(), QAFrequentSequence(n=10), 
        QANucleotideByCycle(), QAQualityByCycle())
x &lt;- qa2(coll, verbose=TRUE)
res &lt;- report(x)
if(interactive())
browseURL(res) 
</code></pre>
<h3 id="filtering-and-trimming-fastq-files-with-shortread">Filtering and Trimming FASTQ Files with ShortRead</h3>
<h4 id="adaptor-trimming">Adaptor trimming</h4>
<pre><code class="language-r">fqtrim &lt;- trimLRPatterns(Rpattern=&quot;GCCCGGGTAA&quot;, subject=fq)
sread(fq)[1:2] # Before trimming
</code></pre>
<pre><code>## DNAStringSet object of length 2:
##     width seq
## [1]    36 CAACGAGTTCACACCTTGGCCGACAGGCCCGGGTAA
## [2]    36 CCAATGATTTTTTTCCGTGTTTCAGAATACGGTTAA
</code></pre>
<pre><code class="language-r">sread(fqtrim)[1:2] # After trimming
</code></pre>
<pre><code>## DNAStringSet object of length 2:
##     width seq
## [1]    26 CAACGAGTTCACACCTTGGCCGACAG
## [2]    36 CCAATGATTTTTTTCCGTGTTTCAGAATACGGTTAA
</code></pre>
<h4 id="read-counting-and-duplicate-removal">Read counting and duplicate removal</h4>
<pre><code class="language-r">tables(fq)$distribution # Counts read occurences
</code></pre>
<pre><code>##   nOccurrences nReads
## 1            1    948
## 2            2     26
</code></pre>
<pre><code class="language-r">sum(srduplicated(fq)) # Identifies duplicated reads
</code></pre>
<pre><code>## [1] 26
</code></pre>
<pre><code class="language-r">fq[!srduplicated(fq)]
</code></pre>
<pre><code>## class: ShortReadQ
## length: 974 reads; width: 36 cycles
</code></pre>
<h4 id="trimming-low-quality-tails">Trimming low quality tails</h4>
<pre><code class="language-r">cutoff &lt;- 30
cutoff &lt;- rawToChar(as.raw(cutoff+33))
sread(trimTails(fq, k=2, a=cutoff, successive=FALSE))[1:2]
</code></pre>
<pre><code>## DNAStringSet object of length 2:
##     width seq
## [1]     4 CAAC
## [2]    20 CCAATGATTTTTTTCCGTGT
</code></pre>
<h4 id="removal-of-reads-with-phred-scores-below-a-threshold-value">Removal of reads with Phred scores below a threshold value</h4>
<pre><code class="language-r">cutoff &lt;- 30
qcount &lt;- rowSums(as(quality(fq), &quot;matrix&quot;) &lt;= 20) 
fq[qcount == 0] # Number of reads where all Phred scores &gt;= 20
</code></pre>
<pre><code>## class: ShortReadQ
## length: 349 reads; width: 36 cycles
</code></pre>
<h4 id="removal-of-reads-with-x-ns-andor-low-complexity-segments">Removal of reads with x Ns and/or low complexity segments</h4>
<pre><code class="language-r">filter1 &lt;- nFilter(threshold=1) # Keeps only reads without Ns
filter2 &lt;- polynFilter(threshold=20, nuc=c(&quot;A&quot;,&quot;T&quot;,&quot;G&quot;,&quot;C&quot;)) # Removes reads with nucleotide bias, &gt;=20 of any base
filter &lt;- compose(filter1, filter2)
fq[filter(fq)]
</code></pre>
<pre><code>## class: ShortReadQ
## length: 989 reads; width: 36 cycles
</code></pre>
<h3 id="memory-efficient-fastq-processing">Memory Efficient FASTQ Processing</h3>
<p>Streaming through FASTQ files with <code>FastqStreamer</code> and random sampling reads with <code>FastqSampler</code></p>
<pre><code class="language-r">fq &lt;- yield(FastqStreamer(fastq[1], 50)) # Imports first 50 reads 
fq &lt;- yield(FastqSampler(fastq[1], 50)) # Random samples 50 reads 
</code></pre>
<p>Streaming through a FASTQ file while applying filtering/trimming functions and writing the results to a new file
here <code>SRR038845.fastq_sub</code> in <code>data</code> directory.</p>
<pre><code class="language-r">f &lt;- FastqStreamer(fastq[1], 50) 
while(length(fq &lt;- yield(f))) {
    fqsub &lt;- fq[grepl(&quot;^TT&quot;, sread(fq))] 
    writeFastq(fqsub, paste(fastq[1], &quot;sub&quot;, sep=&quot;_&quot;), mode=&quot;a&quot;, compress=FALSE)
}
close(f)
</code></pre>
<h2 id="range-operations">Range Operations</h2>
<h3 id="important-data-objects-for-range-operations">Important Data Objects for Range Operations</h3>
<ul>
<li><code>IRanges</code>: stores range data only (IRanges library)</li>
<li><code>GRanges</code>: stores ranges and annotations (GenomicRanges library)</li>
<li><code>GRangesList</code>: list version of GRanges container (GenomicRanges library)</li>
</ul>
<h3 id="range-data-are-stored-in-iranges-and-granges-containers">Range Data Are Stored in <code>IRanges</code> and <code>GRanges</code> Containers</h3>
<h4 id="construct-granges-object">Construct <code>GRanges</code> Object</h4>
<pre><code class="language-r">library(GenomicRanges); library(rtracklayer)
gr &lt;- GRanges(seqnames = Rle(c(&quot;chr1&quot;, &quot;chr2&quot;, &quot;chr1&quot;, &quot;chr3&quot;), c(1, 3, 2, 4)), ranges = IRanges(1:10, end = 7:16, names = head(letters, 10)), strand = Rle(strand(c(&quot;-&quot;, &quot;+&quot;, &quot;*&quot;, &quot;+&quot;, &quot;-&quot;)), c(1, 2, 2, 3, 2)), score = 1:10, GC = seq(1, 0, length = 10)) # Example of creating a GRanges object with its constructor function.
</code></pre>
<h4 id="import-gff-into-granges-object">Import GFF into <code>GRanges</code> Object</h4>
<pre><code class="language-r">gff &lt;- import.gff(&quot;http://cluster.hpcc.ucr.edu/~tgirke/Documents/R_BioCond/Samples/gff3.gff&quot;) # Imports a simplified GFF3 genome annotation file.
seqlengths(gff) &lt;- end(ranges(gff[which(values(gff)[,&quot;type&quot;]==&quot;chromosome&quot;),])) 
names(gff) &lt;- 1:length(gff) # Assigns names to corresponding slot
gff[1:4,]
</code></pre>
<pre><code>## GRanges object with 4 ranges and 10 metadata columns:
##     seqnames     ranges strand |   source       type     score     phase                  ID
##        &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt;
##   1     Chr1 1-30427671      + |   TAIR10 chromosome        NA      &lt;NA&gt;                Chr1
##   2     Chr1  3631-5899      + |   TAIR10 gene              NA      &lt;NA&gt;           AT1G01010
##   3     Chr1  3631-5899      + |   TAIR10 mRNA              NA      &lt;NA&gt;         AT1G01010.1
##   4     Chr1  3760-5630      + |   TAIR10 protein           NA      &lt;NA&gt; AT1G01010.1-Protein
##            Name                Note          Parent       Index Derives_from
##     &lt;character&gt;     &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##   1        Chr1                                            &lt;NA&gt;         &lt;NA&gt;
##   2   AT1G01010 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   3 AT1G01010.1                           AT1G01010           1         &lt;NA&gt;
##   4 AT1G01010.1                                            &lt;NA&gt;  AT1G01010.1
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<pre><code class="language-r">seqinfo(gff)
</code></pre>
<pre><code>## Seqinfo object with 7 sequences from an unspecified genome:
##   seqnames seqlengths isCircular genome
##   Chr1       30427671         NA   &lt;NA&gt;
##   Chr2       19698289         NA   &lt;NA&gt;
##   Chr3       23459830         NA   &lt;NA&gt;
##   Chr4       18585056         NA   &lt;NA&gt;
##   Chr5       26975502         NA   &lt;NA&gt;
##   ChrC         154478         NA   &lt;NA&gt;
##   ChrM         366924         NA   &lt;NA&gt;
</code></pre>
<h4 id="coerce-granges-object-to-dataframe">Coerce <code>GRanges</code> object to <code>data.frame</code></h4>
<pre><code class="language-r">as.data.frame(gff)[1:4, 1:7]
</code></pre>
<pre><code>##   seqnames start      end    width strand source       type
## 1     Chr1     1 30427671 30427671      + TAIR10 chromosome
## 2     Chr1  3631     5899     2269      + TAIR10       gene
## 3     Chr1  3631     5899     2269      + TAIR10       mRNA
## 4     Chr1  3760     5630     1871      + TAIR10    protein
</code></pre>
<h3 id="utilities-for-range-containers">Utilities for Range Containers</h3>
<h4 id="accessor-and-subsetting-methods-for-granges-objects">Accessor and subsetting methods for GRanges objects</h4>
<p>Subsetting and replacement</p>
<pre><code class="language-r">gff[1:4]
</code></pre>
<pre><code>## GRanges object with 4 ranges and 10 metadata columns:
##     seqnames     ranges strand |   source       type     score     phase                  ID
##        &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt;
##   1     Chr1 1-30427671      + |   TAIR10 chromosome        NA      &lt;NA&gt;                Chr1
##   2     Chr1  3631-5899      + |   TAIR10 gene              NA      &lt;NA&gt;           AT1G01010
##   3     Chr1  3631-5899      + |   TAIR10 mRNA              NA      &lt;NA&gt;         AT1G01010.1
##   4     Chr1  3760-5630      + |   TAIR10 protein           NA      &lt;NA&gt; AT1G01010.1-Protein
##            Name                Note          Parent       Index Derives_from
##     &lt;character&gt;     &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##   1        Chr1                                            &lt;NA&gt;         &lt;NA&gt;
##   2   AT1G01010 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   3 AT1G01010.1                           AT1G01010           1         &lt;NA&gt;
##   4 AT1G01010.1                                            &lt;NA&gt;  AT1G01010.1
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<pre><code class="language-r">gff[1:4, c(&quot;type&quot;, &quot;ID&quot;)] 
</code></pre>
<pre><code>## GRanges object with 4 ranges and 2 metadata columns:
##     seqnames     ranges strand |       type                  ID
##        &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;factor&gt;         &lt;character&gt;
##   1     Chr1 1-30427671      + | chromosome                Chr1
##   2     Chr1  3631-5899      + | gene                 AT1G01010
##   3     Chr1  3631-5899      + | mRNA               AT1G01010.1
##   4     Chr1  3760-5630      + | protein    AT1G01010.1-Protein
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<pre><code class="language-r">gff[2] &lt;- gff[3] 
</code></pre>
<p>GRanges objects can be concatenated with the <code>c</code> function</p>
<pre><code class="language-r">c(gff[1:2], gff[401:402]) 
</code></pre>
<pre><code>## GRanges object with 4 ranges and 10 metadata columns:
##       seqnames     ranges strand |   source           type     score     phase                  ID
##          &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt;
##     1     Chr1 1-30427671      + |   TAIR10 chromosome            NA      &lt;NA&gt;                Chr1
##     2     Chr1  3631-5899      + |   TAIR10 mRNA                  NA      &lt;NA&gt;         AT1G01010.1
##   401     Chr5  5516-5769      - |   TAIR10 protein               NA      &lt;NA&gt; AT5G01015.2-Protein
##   402     Chr5  5770-5801      - |   TAIR10 five_prime_UTR        NA      &lt;NA&gt;                &lt;NA&gt;
##              Name            Note          Parent       Index Derives_from
##       &lt;character&gt; &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##     1        Chr1                                        &lt;NA&gt;         &lt;NA&gt;
##     2 AT1G01010.1                       AT1G01010           1         &lt;NA&gt;
##   401 AT5G01015.2                                        &lt;NA&gt;  AT5G01015.2
##   402        &lt;NA&gt;                     AT5G01015.2        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<p>Acessor functions</p>
<pre><code class="language-r">seqnames(gff)
</code></pre>
<pre><code>## factor-Rle of length 449 with 7 runs
##   Lengths:   72   22   38  118  172   13   14
##   Values : Chr1 Chr2 Chr3 Chr4 Chr5 ChrC ChrM
## Levels(7): Chr1 Chr2 Chr3 Chr4 Chr5 ChrC ChrM
</code></pre>
<pre><code class="language-r">ranges(gff)
</code></pre>
<pre><code>## IRanges object with 449 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##     1         1  30427671  30427671
##     2      3631      5899      2269
##     3      3631      5899      2269
##     4      3760      5630      1871
##     5      3631      3913       283
##   ...       ...       ...       ...
##   445     11918     12241       324
##   446     11918     12241       324
##   447     11918     12241       324
##   448     11918     12241       324
##   449     11918     12241       324
</code></pre>
<pre><code class="language-r">strand(gff)
</code></pre>
<pre><code>## factor-Rle of length 449 with 13 runs
##   Lengths:  18  54  28  21  12 117   1 171   1  12   1   8   5
##   Values :   +   -   +   -   +   -   +   -   +   -   +   -   +
## Levels(3): + - *
</code></pre>
<pre><code class="language-r">seqlengths(gff) 
</code></pre>
<pre><code>##     Chr1     Chr2     Chr3     Chr4     Chr5     ChrC     ChrM 
## 30427671 19698289 23459830 18585056 26975502   154478   366924
</code></pre>
<pre><code class="language-r">start(gff[1:4])
</code></pre>
<pre><code>## [1]    1 3631 3631 3760
</code></pre>
<pre><code class="language-r">end(gff[1:4])
</code></pre>
<pre><code>## [1] 30427671     5899     5899     5630
</code></pre>
<pre><code class="language-r">width(gff[1:4]) 
</code></pre>
<pre><code>## [1] 30427671     2269     2269     1871
</code></pre>
<p>Accessing metadata component</p>
<pre><code class="language-r">values(gff) # or elementMetadata(gff)
</code></pre>
<pre><code>## DataFrame with 449 rows and 10 columns
##       source       type     score     phase                  ID        Name                Note
##     &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt; &lt;character&gt;     &lt;CharacterList&gt;
## 1     TAIR10 chromosome        NA        NA                Chr1        Chr1                    
## 2     TAIR10 mRNA              NA        NA         AT1G01010.1 AT1G01010.1                    
## 3     TAIR10 mRNA              NA        NA         AT1G01010.1 AT1G01010.1                    
## 4     TAIR10 protein           NA        NA AT1G01010.1-Protein AT1G01010.1                    
## 5     TAIR10 exon              NA        NA                  NA          NA                    
## ...      ...        ...       ...       ...                 ...         ...                 ...
## 445   TAIR10    gene           NA        NA           ATMG00030   ATMG00030 protein_coding_gene
## 446   TAIR10    mRNA           NA        NA         ATMG00030.1 ATMG00030.1                    
## 447   TAIR10    protein        NA        NA ATMG00030.1-Protein ATMG00030.1                    
## 448   TAIR10    exon           NA        NA                  NA          NA                    
## 449   TAIR10    CDS            NA         0                  NA          NA                    
##                              Parent       Index Derives_from
##                     &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
## 1                                            NA           NA
## 2                         AT1G01010           1           NA
## 3                         AT1G01010           1           NA
## 4                                            NA  AT1G01010.1
## 5                       AT1G01010.1          NA           NA
## ...                             ...         ...          ...
## 445                                          NA           NA
## 446                       ATMG00030           1           NA
## 447                                          NA  ATMG00030.1
## 448                     ATMG00030.1          NA           NA
## 449 ATMG00030.1,ATMG00030.1-Protein          NA           NA
</code></pre>
<pre><code class="language-r">values(gff)[, &quot;type&quot;][1:20] 
</code></pre>
<pre><code>##  [1] chromosome      mRNA            mRNA            protein         exon            five_prime_UTR 
##  [7] CDS             exon            CDS             exon            CDS             exon           
## [13] CDS             exon            CDS             exon            CDS             three_prime_UTR
## [19] gene            mRNA           
## Levels: chromosome gene mRNA protein exon five_prime_UTR CDS three_prime_UTR rRNA tRNA
</code></pre>
<pre><code class="language-r">gff[values(gff)[ ,&quot;type&quot;] == &quot;gene&quot;] 
</code></pre>
<pre><code>## GRanges object with 21 ranges and 10 metadata columns:
##       seqnames      ranges strand |   source     type     score     phase          ID        Name
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt; &lt;character&gt; &lt;character&gt;
##    19     Chr1   5928-8737      - |   TAIR10     gene        NA      &lt;NA&gt;   AT1G01020   AT1G01020
##    64     Chr1 11649-13714      - |   TAIR10     gene        NA      &lt;NA&gt;   AT1G01030   AT1G01030
##    74     Chr2   1025-2810      + |   TAIR10     gene        NA      &lt;NA&gt;   AT2G01008   AT2G01008
##    84     Chr2   3706-5513      + |   TAIR10     gene        NA      &lt;NA&gt;   AT2G01010   AT2G01010
##    87     Chr2   5782-5945      + |   TAIR10     gene        NA      &lt;NA&gt;   AT2G01020   AT2G01020
##   ...      ...         ...    ... .      ...      ...       ...       ...         ...         ...
##   427     ChrC    383-1444      - |   TAIR10     gene        NA      &lt;NA&gt;   ATCG00020   ATCG00020
##   432     ChrC   1717-4347      - |   TAIR10     gene        NA      &lt;NA&gt;   ATCG00030   ATCG00030
##   437     ChrM     273-734      - |   TAIR10     gene        NA      &lt;NA&gt;   ATMG00010   ATMG00010
##   442     ChrM  8848-11415      - |   TAIR10     gene        NA      &lt;NA&gt;   ATMG00020   ATMG00020
##   445     ChrM 11918-12241      + |   TAIR10     gene        NA      &lt;NA&gt;   ATMG00030   ATMG00030
##                      Note          Parent       Index Derives_from
##           &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##    19 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##    64 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##    74 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##    84                rRNA                        &lt;NA&gt;         &lt;NA&gt;
##    87                rRNA                        &lt;NA&gt;         &lt;NA&gt;
##   ...                 ...             ...         ...          ...
##   427 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   432                tRNA                        &lt;NA&gt;         &lt;NA&gt;
##   437 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   442                rRNA                        &lt;NA&gt;         &lt;NA&gt;
##   445 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<h4 id="useful-utilities-for-granges-objects">Useful utilities for GRanges objects</h4>
<p>Remove chromosome ranges</p>
<pre><code class="language-r">gff &lt;- gff[values(gff)$type != &quot;chromosome&quot;] 
</code></pre>
<p>Erase the strand information</p>
<pre><code class="language-r">strand(gff) &lt;- &quot;*&quot; 
</code></pre>
<p>Collapses overlapping ranges to continuous ranges.</p>
<pre><code class="language-r">reduce(gff) 
</code></pre>
<pre><code>## GRanges object with 22 ranges and 0 metadata columns:
##        seqnames      ranges strand
##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     Chr1   3631-5899      *
##    [2]     Chr1   5928-8737      *
##    [3]     Chr1 11649-13714      *
##    [4]     Chr2   1025-2810      *
##    [5]     Chr2   3706-5513      *
##    ...      ...         ...    ...
##   [18]     ChrC    383-1444      *
##   [19]     ChrC   1717-4347      *
##   [20]     ChrM     273-734      *
##   [21]     ChrM  8848-11415      *
##   [22]     ChrM 11918-12241      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<p>Return uncovered regions</p>
<pre><code class="language-r">gaps(gff) 
</code></pre>
<pre><code>## GRanges object with 43 ranges and 0 metadata columns:
##        seqnames       ranges strand
##           &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     Chr1   1-30427671      +
##    [2]     Chr1   1-30427671      -
##    [3]     Chr1       1-3630      *
##    [4]     Chr1    5900-5927      *
##    [5]     Chr1   8738-11648      *
##    ...      ...          ...    ...
##   [39]     ChrM     1-366924      -
##   [40]     ChrM        1-272      *
##   [41]     ChrM     735-8847      *
##   [42]     ChrM  11416-11917      *
##   [43]     ChrM 12242-366924      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<p>More intuitive way to get uncovered regions</p>
<pre><code class="language-r">setdiff(as(seqinfo(gff), &quot;GRanges&quot;), gff) 
</code></pre>
<pre><code>## GRanges object with 29 ranges and 0 metadata columns:
##        seqnames         ranges strand
##           &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     Chr1         1-3630      *
##    [2]     Chr1      5900-5927      *
##    [3]     Chr1     8738-11648      *
##    [4]     Chr1 13715-30427671      *
##    [5]     Chr2         1-1024      *
##    ...      ...            ...    ...
##   [25]     ChrC    4348-154478      *
##   [26]     ChrM          1-272      *
##   [27]     ChrM       735-8847      *
##   [28]     ChrM    11416-11917      *
##   [29]     ChrM   12242-366924      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<p>Return disjoint ranges</p>
<pre><code class="language-r">disjoin(gff)
</code></pre>
<pre><code>## GRanges object with 211 ranges and 0 metadata columns:
##         seqnames      ranges strand
##            &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt;
##     [1]     Chr1   3631-3759      *
##     [2]     Chr1   3760-3913      *
##     [3]     Chr1   3914-3995      *
##     [4]     Chr1   3996-4276      *
##     [5]     Chr1   4277-4485      *
##     ...      ...         ...    ...
##   [207]     ChrC   1752-4310      *
##   [208]     ChrC   4311-4347      *
##   [209]     ChrM     273-734      *
##   [210]     ChrM  8848-11415      *
##   [211]     ChrM 11918-12241      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<p>Returns coverage of ranges</p>
<pre><code class="language-r">coverage(gff)
</code></pre>
<pre><code>## RleList of length 7
## $Chr1
## integer-Rle of length 30427671 with 45 runs
##   Lengths:     3630      129      154       82      281 ...      233      161      380 30413957
##   Values :        0        4        5        3        5 ...        4        2        4        0
## 
## $Chr2
## integer-Rle of length 19698289 with 14 runs
##   Lengths:     1024      248      185       53      362 ...      164      625      102 19691617
##   Values :        0        5        3        5        3 ...        3        0        5        0
## 
## $Chr3
## integer-Rle of length 23459830 with 29 runs
##   Lengths:     1652      145      139      111       95 ...      155      148      156 23453781
##   Values :        0        4        5        3        5 ...        3        5        4        0
## 
## $Chr4
## integer-Rle of length 18585056 with 72 runs
##   Lengths:     1179      357     1358      128      872 ...      212      114       74 18571697
##   Values :        0        5        0        5        3 ...        3        5        4        0
## 
## $Chr5
## integer-Rle of length 26975502 with 64 runs
##   Lengths:     1222       28       28      109       72 ...       76       55      174 26967058
##   Values :        0        4        7       13       16 ...        3        5        4        0
## 
## ...
## &lt;2 more elements&gt;
</code></pre>
<p>Return the index pairings for overlapping ranges</p>
<pre><code class="language-r">findOverlaps(gff, gff[1:4]) 
</code></pre>
<pre><code>## Hits object with 55 hits and 0 metadata columns:
##        queryHits subjectHits
##        &lt;integer&gt;   &lt;integer&gt;
##    [1]         1           1
##    [2]         1           2
##    [3]         1           4
##    [4]         1           3
##    [5]         2           1
##    ...       ...         ...
##   [51]        16           1
##   [52]        16           2
##   [53]        16           3
##   [54]        17           1
##   [55]        17           2
##   -------
##   queryLength: 442 / subjectLength: 4
</code></pre>
<p>Counts overlapping ranges</p>
<pre><code class="language-r">countOverlaps(gff, gff[1:4])[1:40]
</code></pre>
<pre><code>##  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 
##  4  4  4  4  3  4  3  3  3  3  3  3  3  3  3  3  2  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 
## 35 36 37 38 39 40 41 
##  0  0  0  0  0  0  0
</code></pre>
<p>Return only overlapping ranges</p>
<pre><code class="language-r">subsetByOverlaps(gff, gff[1:4]) 
</code></pre>
<pre><code>## GRanges object with 17 ranges and 10 metadata columns:
##      seqnames    ranges strand |   source            type     score     phase                  ID
##         &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;        &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt;
##    2     Chr1 3631-5899      * |   TAIR10  mRNA                  NA      &lt;NA&gt;         AT1G01010.1
##    3     Chr1 3631-5899      * |   TAIR10  mRNA                  NA      &lt;NA&gt;         AT1G01010.1
##    4     Chr1 3760-5630      * |   TAIR10  protein               NA      &lt;NA&gt; AT1G01010.1-Protein
##    5     Chr1 3631-3913      * |   TAIR10  exon                  NA      &lt;NA&gt;                &lt;NA&gt;
##    6     Chr1 3631-3759      * |   TAIR10  five_prime_UTR        NA      &lt;NA&gt;                &lt;NA&gt;
##   ..      ...       ...    ... .      ...             ...       ...       ...                 ...
##   14     Chr1 5174-5326      * |   TAIR10 exon                   NA      &lt;NA&gt;                &lt;NA&gt;
##   15     Chr1 5174-5326      * |   TAIR10 CDS                    NA         0                &lt;NA&gt;
##   16     Chr1 5439-5899      * |   TAIR10 exon                   NA      &lt;NA&gt;                &lt;NA&gt;
##   17     Chr1 5439-5630      * |   TAIR10 CDS                    NA         0                &lt;NA&gt;
##   18     Chr1 5631-5899      * |   TAIR10 three_prime_UTR        NA      &lt;NA&gt;                &lt;NA&gt;
##             Name            Note                          Parent       Index Derives_from
##      &lt;character&gt; &lt;CharacterList&gt;                 &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##    2 AT1G01010.1                                       AT1G01010           1         &lt;NA&gt;
##    3 AT1G01010.1                                       AT1G01010           1         &lt;NA&gt;
##    4 AT1G01010.1                                                        &lt;NA&gt;  AT1G01010.1
##    5        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;         &lt;NA&gt;
##    6        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;         &lt;NA&gt;
##   ..         ...             ...                             ...         ...          ...
##   14        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;         &lt;NA&gt;
##   15        &lt;NA&gt;                 AT1G01010.1,AT1G01010.1-Protein        &lt;NA&gt;         &lt;NA&gt;
##   16        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;         &lt;NA&gt;
##   17        &lt;NA&gt;                 AT1G01010.1,AT1G01010.1-Protein        &lt;NA&gt;         &lt;NA&gt;
##   18        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<h3 id="grangeslist-objects">GRangesList Objects</h3>
<pre><code class="language-r">sp &lt;- split(gff, seq(along=gff)) # Stores every range in separate component of a GRangesList object
split(gff, seqnames(gff)) # Stores ranges of each chromosome in separate component.
</code></pre>
<pre><code>## GRangesList object of length 7:
## $Chr1
## GRanges object with 71 ranges and 10 metadata columns:
##      seqnames      ranges strand |   source            type     score     phase                  ID
##         &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;        &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt;
##    2     Chr1   3631-5899      * |   TAIR10  mRNA                  NA      &lt;NA&gt;         AT1G01010.1
##    3     Chr1   3631-5899      * |   TAIR10  mRNA                  NA      &lt;NA&gt;         AT1G01010.1
##    4     Chr1   3760-5630      * |   TAIR10  protein               NA      &lt;NA&gt; AT1G01010.1-Protein
##    5     Chr1   3631-3913      * |   TAIR10  exon                  NA      &lt;NA&gt;                &lt;NA&gt;
##    6     Chr1   3631-3759      * |   TAIR10  five_prime_UTR        NA      &lt;NA&gt;                &lt;NA&gt;
##   ..      ...         ...    ... .      ...             ...       ...       ...                 ...
##   68     Chr1 13335-13714      * |   TAIR10 exon                   NA      &lt;NA&gt;                &lt;NA&gt;
##   69     Chr1 12941-13173      * |   TAIR10 five_prime_UTR         NA      &lt;NA&gt;                &lt;NA&gt;
##   70     Chr1 11864-12940      * |   TAIR10 CDS                    NA         0                &lt;NA&gt;
##   71     Chr1 11649-11863      * |   TAIR10 three_prime_UTR        NA      &lt;NA&gt;                &lt;NA&gt;
##   72     Chr1 11649-13173      * |   TAIR10 exon                   NA      &lt;NA&gt;                &lt;NA&gt;
##             Name            Note                          Parent       Index Derives_from
##      &lt;character&gt; &lt;CharacterList&gt;                 &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##    2 AT1G01010.1                                       AT1G01010           1         &lt;NA&gt;
##    3 AT1G01010.1                                       AT1G01010           1         &lt;NA&gt;
##    4 AT1G01010.1                                                        &lt;NA&gt;  AT1G01010.1
##    5        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;         &lt;NA&gt;
##    6        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;         &lt;NA&gt;
##   ..         ...             ...                             ...         ...          ...
##   68        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;         &lt;NA&gt;
##   69        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;         &lt;NA&gt;
##   70        &lt;NA&gt;                 AT1G01030.1,AT1G01030.1-Protein        &lt;NA&gt;         &lt;NA&gt;
##   71        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;         &lt;NA&gt;
##   72        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
## 
## ...
## &lt;6 more elements&gt;
</code></pre>
<pre><code class="language-r">unlist(sp) # Returns data as GRanges object
</code></pre>
<pre><code>## GRanges object with 442 ranges and 10 metadata columns:
##           seqnames      ranges strand |   source           type     score     phase
##              &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##       1.2     Chr1   3631-5899      * |   TAIR10 mRNA                  NA      &lt;NA&gt;
##       2.3     Chr1   3631-5899      * |   TAIR10 mRNA                  NA      &lt;NA&gt;
##       3.4     Chr1   3760-5630      * |   TAIR10 protein               NA      &lt;NA&gt;
##       4.5     Chr1   3631-3913      * |   TAIR10 exon                  NA      &lt;NA&gt;
##       5.6     Chr1   3631-3759      * |   TAIR10 five_prime_UTR        NA      &lt;NA&gt;
##       ...      ...         ...    ... .      ...            ...       ...       ...
##   438.445     ChrM 11918-12241      * |   TAIR10        gene           NA      &lt;NA&gt;
##   439.446     ChrM 11918-12241      * |   TAIR10        mRNA           NA      &lt;NA&gt;
##   440.447     ChrM 11918-12241      * |   TAIR10        protein        NA      &lt;NA&gt;
##   441.448     ChrM 11918-12241      * |   TAIR10        exon           NA      &lt;NA&gt;
##   442.449     ChrM 11918-12241      * |   TAIR10        CDS            NA         0
##                            ID        Name                Note                          Parent
##                   &lt;character&gt; &lt;character&gt;     &lt;CharacterList&gt;                 &lt;CharacterList&gt;
##       1.2         AT1G01010.1 AT1G01010.1                                           AT1G01010
##       2.3         AT1G01010.1 AT1G01010.1                                           AT1G01010
##       3.4 AT1G01010.1-Protein AT1G01010.1                                                    
##       4.5                &lt;NA&gt;        &lt;NA&gt;                                         AT1G01010.1
##       5.6                &lt;NA&gt;        &lt;NA&gt;                                         AT1G01010.1
##       ...                 ...         ...                 ...                             ...
##   438.445           ATMG00030   ATMG00030 protein_coding_gene                                
##   439.446         ATMG00030.1 ATMG00030.1                                           ATMG00030
##   440.447 ATMG00030.1-Protein ATMG00030.1                                                    
##   441.448                &lt;NA&gt;        &lt;NA&gt;                                         ATMG00030.1
##   442.449                &lt;NA&gt;        &lt;NA&gt;                     ATMG00030.1,ATMG00030.1-Protein
##                 Index Derives_from
##           &lt;character&gt;  &lt;character&gt;
##       1.2           1         &lt;NA&gt;
##       2.3           1         &lt;NA&gt;
##       3.4        &lt;NA&gt;  AT1G01010.1
##       4.5        &lt;NA&gt;         &lt;NA&gt;
##       5.6        &lt;NA&gt;         &lt;NA&gt;
##       ...         ...          ...
##   438.445        &lt;NA&gt;         &lt;NA&gt;
##   439.446           1         &lt;NA&gt;
##   440.447        &lt;NA&gt;  ATMG00030.1
##   441.448        &lt;NA&gt;         &lt;NA&gt;
##   442.449        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<pre><code class="language-r">sp[1:4, &quot;type&quot;] # Subsetting of GRangesList objects is similar to GRanges objects.
</code></pre>
<pre><code>## GRangesList object of length 4:
## $`1`
## GRanges object with 1 range and 1 metadata column:
##     seqnames    ranges strand |     type
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   2     Chr1 3631-5899      * |     mRNA
##   -------
##   seqinfo: 7 sequences from an unspecified genome
## 
## $`2`
## GRanges object with 1 range and 1 metadata column:
##     seqnames    ranges strand |     type
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   3     Chr1 3631-5899      * |     mRNA
##   -------
##   seqinfo: 7 sequences from an unspecified genome
## 
## $`3`
## GRanges object with 1 range and 1 metadata column:
##     seqnames    ranges strand |     type
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   4     Chr1 3760-5630      * |  protein
##   -------
##   seqinfo: 7 sequences from an unspecified genome
## 
## $`4`
## GRanges object with 1 range and 1 metadata column:
##     seqnames    ranges strand |     type
##        &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   5     Chr1 3631-3913      * |     exon
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
<pre><code class="language-r">lapply(sp[1:4], length) # Looping over GRangesList objects similar to lists
</code></pre>
<pre><code>## $`1`
## [1] 1
## 
## $`2`
## [1] 1
## 
## $`3`
## [1] 1
## 
## $`4`
## [1] 1
</code></pre>
<h2 id="transcript-ranges">Transcript Ranges</h2>
<p>Storing annotation ranges in <code>TranscriptDb</code> databases makes many operations more robust and convenient.</p>
<pre><code class="language-r">library(GenomicFeatures)
download.file(&quot;http://cluster.hpcc.ucr.edu/~tgirke/Documents/R_BioCond/Samples/gff3.gff&quot;, &quot;data/gff3.gff&quot;)
txdb &lt;- makeTxDbFromGFF(file=&quot;data/gff3.gff&quot;, format=&quot;gff&quot;, dataSource=&quot;TAIR&quot;, organism=&quot;Arabidopsis thaliana&quot;)
</code></pre>
<pre><code>## Warning in .extract_exons_from_GRanges(cds_IDX, gr, mcols0, tx_IDX, feature = &quot;cds&quot;, : 163 CDS couldn't be linked to a transcript so were dropped (showing only the first 6):
##   seqid start  end strand   ID Name              Parent Parent_type
## 1  Chr1  3760 3913      + &lt;NA&gt; &lt;NA&gt; AT1G01010.1-Protein        &lt;NA&gt;
## 2  Chr1  3996 4276      + &lt;NA&gt; &lt;NA&gt; AT1G01010.1-Protein        &lt;NA&gt;
## 3  Chr1  4486 4605      + &lt;NA&gt; &lt;NA&gt; AT1G01010.1-Protein        &lt;NA&gt;
## 4  Chr1  4706 5095      + &lt;NA&gt; &lt;NA&gt; AT1G01010.1-Protein        &lt;NA&gt;
## 5  Chr1  5174 5326      + &lt;NA&gt; &lt;NA&gt; AT1G01010.1-Protein        &lt;NA&gt;
## 6  Chr1  5439 5630      + &lt;NA&gt; &lt;NA&gt; AT1G01010.1-Protein        &lt;NA&gt;
</code></pre>
<pre><code class="language-r">saveDb(txdb, file=&quot;./data/TAIR10.sqlite&quot;)
</code></pre>
<pre><code>## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: TAIR
## # Organism: Arabidopsis thaliana
## # Taxonomy ID: 3702
## # miRBase build ID: NA
## # Genome: NA
## # Nb of transcripts: 28
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2022-04-20 18:50:11 -0700 (Wed, 20 Apr 2022)
## # GenomicFeatures version at creation time: 1.46.1
## # RSQLite version at creation time: 2.2.9
## # DBSCHEMAVERSION: 1.2
</code></pre>
<pre><code class="language-r">txdb &lt;- loadDb(&quot;./data/TAIR10.sqlite&quot;)
transcripts(txdb)
</code></pre>
<pre><code>## GRanges object with 28 ranges and 2 metadata columns:
##        seqnames      ranges strand |     tx_id     tx_name
##           &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##    [1]     Chr1   3631-5899      + |         1 AT1G01010.1
##    [2]     Chr1   5928-8737      - |         2 AT1G01020.1
##    [3]     Chr1   6790-8737      - |         3 AT1G01020.2
##    [4]     Chr1 11649-13714      - |         4 AT1G01030.1
##    [5]     Chr2   1025-2810      + |         5 AT2G01008.1
##    ...      ...         ...    ... .       ...         ...
##   [24]     ChrC    383-1444      - |        24 ATCG00020.1
##   [25]     ChrC   1717-4347      - |        25 ATCG00030.1
##   [26]     ChrM 11918-12241      + |        26 ATMG00030.1
##   [27]     ChrM     273-734      - |        27 ATMG00010.1
##   [28]     ChrM  8848-11415      - |        28 ATMG00020.1
##   -------
##   seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
</code></pre>
<pre><code class="language-r">transcriptsBy(txdb, by = &quot;gene&quot;)
</code></pre>
<pre><code>## GRangesList object of length 22:
## $AT1G01010
## GRanges object with 1 range and 2 metadata columns:
##       seqnames    ranges strand |     tx_id     tx_name
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     Chr1 3631-5899      + |         1 AT1G01010.1
##   -------
##   seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
## 
## $AT1G01020
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |     tx_id     tx_name
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     Chr1 5928-8737      - |         2 AT1G01020.1
##   [2]     Chr1 6790-8737      - |         3 AT1G01020.2
##   -------
##   seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
## 
## $AT1G01030
## GRanges object with 1 range and 2 metadata columns:
##       seqnames      ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     Chr1 11649-13714      - |         4 AT1G01030.1
##   -------
##   seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
## 
## ...
## &lt;19 more elements&gt;
</code></pre>
<pre><code class="language-r">exonsBy(txdb, by = &quot;gene&quot;)
</code></pre>
<pre><code>## GRangesList object of length 22:
## $AT1G01010
## GRanges object with 6 ranges and 2 metadata columns:
##       seqnames    ranges strand |   exon_id   exon_name
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     Chr1 3631-3913      + |         1        &lt;NA&gt;
##   [2]     Chr1 3996-4276      + |         2        &lt;NA&gt;
##   [3]     Chr1 4486-4605      + |         3        &lt;NA&gt;
##   [4]     Chr1 4706-5095      + |         4        &lt;NA&gt;
##   [5]     Chr1 5174-5326      + |         5        &lt;NA&gt;
##   [6]     Chr1 5439-5899      + |         6        &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
## 
## $AT1G01020
## GRanges object with 12 ranges and 2 metadata columns:
##        seqnames    ranges strand |   exon_id   exon_name
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##    [1]     Chr1 5928-6263      - |         7        &lt;NA&gt;
##    [2]     Chr1 6437-7069      - |         8        &lt;NA&gt;
##    [3]     Chr1 6790-7069      - |         9        &lt;NA&gt;
##    [4]     Chr1 7157-7232      - |        10        &lt;NA&gt;
##    [5]     Chr1 7157-7450      - |        11        &lt;NA&gt;
##    ...      ...       ...    ... .       ...         ...
##    [8]     Chr1 7762-7835      - |        14        &lt;NA&gt;
##    [9]     Chr1 7942-7987      - |        15        &lt;NA&gt;
##   [10]     Chr1 8236-8325      - |        16        &lt;NA&gt;
##   [11]     Chr1 8417-8464      - |        17        &lt;NA&gt;
##   [12]     Chr1 8571-8737      - |        18        &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
## 
## $AT1G01030
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames      ranges strand |   exon_id   exon_name
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     Chr1 11649-13173      - |        19        &lt;NA&gt;
##   [2]     Chr1 13335-13714      - |        20        &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
## 
## ...
## &lt;19 more elements&gt;
</code></pre>
<h3 id="txdb-from-biomart"><code>txdb</code> from BioMart</h3>
<p>Alternative sources for creating <code>txdb</code> databases are BioMart, Bioc annotation packages, UCSC, etc. The following shows how to create a <code>txdb</code> from BioMart.</p>
<pre><code class="language-r">library(GenomicFeatures); library(&quot;biomaRt&quot;)
txdb &lt;- makeTxDbFromBiomart(biomart = &quot;plants_mart&quot;, dataset = &quot;athaliana_eg_gene&quot;, host=&quot;plants.ensembl.org&quot;)
</code></pre>
<p>The following steps are useful to find out what is availble in BioMart.</p>
<pre><code class="language-r">listMarts() # Lists BioMart databases
listMarts(host=&quot;plants.ensembl.org&quot;)
mymart &lt;- useMart(&quot;plants_mart&quot;, host=&quot;plants.ensembl.org&quot;) # Select one, here plants_mart
listDatasets(mymart) # List datasets available in the selected BioMart database
mymart &lt;- useMart(&quot;plants_mart&quot;, dataset=&quot;athaliana_eg_gene&quot;, host=&quot;plants.ensembl.org&quot;)
listAttributes(mymart) # List available features 
getBM(attributes=c(&quot;ensembl_gene_id&quot;, &quot;description&quot;), mart=mymart)[1:4,]
</code></pre>
<h3 id="efficient-sequence-parsing">Efficient Sequence Parsing</h3>
<h3 id="getseq"><code>getSeq</code></h3>
<p>The following parses all annotation ranges provided by a <code>GRanges</code> object (e.g. <code>gff</code>) from a genome sequence stored in a local file.</p>
<pre><code class="language-r">gff &lt;- gff[values(gff)$type != &quot;chromosome&quot;] # Remove chromosome ranges
rand &lt;- DNAStringSet(sapply(unique(as.character(seqnames(gff))), function(x) paste(sample(c(&quot;A&quot;,&quot;T&quot;,&quot;G&quot;,&quot;C&quot;), 200000, replace=T), collapse=&quot;&quot;)))
writeXStringSet(DNAStringSet(rand), &quot;./data/test&quot;)
getSeq(FaFile(&quot;./data/test&quot;), gff)
</code></pre>
<pre><code>## DNAStringSet object of length 442:
##       width seq                                                                 names               
##   [1]  2269 GCCAGGTAGAATTCCAAAAATTGAAAGGCGTT...TGATCTGCCAGCGCCATTCGCTACGTCCGAAC Chr1
##   [2]  2269 GCCAGGTAGAATTCCAAAAATTGAAAGGCGTT...TGATCTGCCAGCGCCATTCGCTACGTCCGAAC Chr1
##   [3]  1871 CCTCCAAGTCTTAAAGCCTGCCTGCGACATTC...AAACCGCGAAATCTTATCGACCATTGTGTCTC Chr1
##   [4]   283 GCCAGGTAGAATTCCAAAAATTGAAAGGCGTT...GCGGTCACTTTGCTAGGGCTGGTAGCCATGCA Chr1
##   [5]   129 GCCAGGTAGAATTCCAAAAATTGAAAGGCGTT...TATCACACCTCTCTCTACTCGAATGACACGGC Chr1
##   ...   ... ...
## [438]   324 CTGGGTTGGAATGGGGCAAATACTTTCAAGGA...CAGGTGCAAATTGGCATCGCCGCACGGACGTT ChrM
## [439]   324 CTGGGTTGGAATGGGGCAAATACTTTCAAGGA...CAGGTGCAAATTGGCATCGCCGCACGGACGTT ChrM
## [440]   324 CTGGGTTGGAATGGGGCAAATACTTTCAAGGA...CAGGTGCAAATTGGCATCGCCGCACGGACGTT ChrM
## [441]   324 CTGGGTTGGAATGGGGCAAATACTTTCAAGGA...CAGGTGCAAATTGGCATCGCCGCACGGACGTT ChrM
## [442]   324 CTGGGTTGGAATGGGGCAAATACTTTCAAGGA...CAGGTGCAAATTGGCATCGCCGCACGGACGTT ChrM
</code></pre>
<h4 id="extracttranscriptseqs"><code>extractTranscriptSeqs</code></h4>
<p>Sequences composed of several ranges, such as transcripts (or CDSs) with several exons, can be parsed with <code>extractTranscriptSeqs</code>.
Note: the following expects the genome sequence in a file with path <code>data/test</code> and a valid <code>txdb</code> defining the ranges for that
genome.</p>
<pre><code class="language-r">library(GenomicFeatures); library(Biostrings); library(Rsamtools)
txdb &lt;- loadDb(&quot;./data/TAIR10.sqlite&quot;)
indexFa(&quot;data/test&quot;) # Creates index for genome fasta
</code></pre>
<pre><code>## [1] &quot;data/test.fai&quot;
</code></pre>
<pre><code class="language-r">txseq &lt;- extractTranscriptSeqs(FaFile(&quot;data/test&quot;), txdb, use.names=TRUE) 
txseq
</code></pre>
<pre><code>## DNAStringSet object of length 28:
##      width seq                                                                  names               
##  [1]  1688 GCCAGGTAGAATTCCAAAAATTGAAAGGCGTTT...TGATCTGCCAGCGCCATTCGCTACGTCCGAAC AT1G01010.1
##  [2]  1623 CGCTTGTCCGTTCAAGAAAAGAATCTATTCCCC...CCACGCGTAGATTAGTTGGGGACGTAAGGACC AT1G01020.1
##  [3]  1085 CGCTTGTCCGTTCAAGAAAAGAATCTATTCCCC...GAGCAGGTTCGTTTTTAGAGACTCTATACCGT AT1G01020.2
##  [4]  1905 GTCCAGCCTACCCTATCGATACCTGTCACCATT...CAAAATCCGGCAGCATTTATCATGATCACGTA AT1G01030.1
##  [5]  1239 CGCGGGAGAAACATTCGTCGAGGAATGATGCTG...ATGGTTTGCTCGACGAAGCTTATTCTGTCCGC AT2G01008.1
##  ...   ... ...
## [24]  1062 GCCGTCCACAATGACGGTGTGGATCTGAGAACG...TGGGTATCAGCCTACCGGCTGTAGGGAAATAG ATCG00020.1
## [25]    72 CCCGAGACTAGCCGCACCACGTGAGCTTTGGTG...AAACATCCTGTGCACGTTTTTTTGCCGGCGAT ATCG00030.1
## [26]   324 CTGGGTTGGAATGGGGCAAATACTTTCAAGGAG...CAGGTGCAAATTGGCATCGCCGCACGGACGTT ATMG00030.1
## [27]   462 GGATCTGACGCCACCAGAACCACAACTCCTGGA...GTGCTATGATGGCGTATGAGCGGGGGGTCGTA ATMG00010.1
## [28]  2568 CCGCTGATAGAGGTCAAATTACCAACAATAGTG...TGATAAATGGGATGCTTTAAGGACTCGTCGGC ATMG00020.1
</code></pre>
<h2 id="homework-6">Homework 6</h2>
<p>See <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw06/hw06/">here</a>.</p>
<h2 id="session-info">Session Info</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.1.3 (2022-03-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] grid      stats4    stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] GenomicFeatures_1.46.1      AnnotationDbi_1.56.2        rtracklayer_1.54.0         
##  [4] systemPipeR_2.0.4           ShortRead_1.52.0            GenomicAlignments_1.30.0   
##  [7] SummarizedExperiment_1.24.0 Biobase_2.54.0              MatrixGenerics_1.6.0       
## [10] matrixStats_0.61.0          Rsamtools_2.10.0            GenomicRanges_1.46.1       
## [13] BiocParallel_1.28.2         ggseqlogo_0.1               ggplot2_3.3.5              
## [16] seqLogo_1.60.0              Biostrings_2.62.0           GenomeInfoDb_1.30.0        
## [19] XVector_0.34.0              IRanges_2.28.0              S4Vectors_0.32.3           
## [22] BiocGenerics_0.40.0         BiocStyle_2.22.0           
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-7           bit64_4.0.5            filelock_1.0.2         progress_1.2.2        
##  [5] RColorBrewer_1.1-2     httr_1.4.2             tools_4.1.3            bslib_0.3.1           
##  [9] utf8_1.2.2             R6_2.5.1               DBI_1.1.1              colorspace_2.0-2      
## [13] withr_2.4.3            prettyunits_1.1.1      tidyselect_1.1.1       curl_4.3.2            
## [17] bit_4.0.4              compiler_4.1.3         cli_3.1.0              xml2_1.3.3            
## [21] DelayedArray_0.20.0    labeling_0.4.2         bookdown_0.24          sass_0.4.0            
## [25] scales_1.1.1           rappdirs_0.3.3         stringr_1.4.0          digest_0.6.29         
## [29] rmarkdown_2.13         jpeg_0.1-9             pkgconfig_2.0.3        htmltools_0.5.2       
## [33] highr_0.9              dbplyr_2.1.1           fastmap_1.1.0          htmlwidgets_1.5.4     
## [37] rlang_1.0.2            RSQLite_2.2.9          farver_2.1.0           jquerylib_0.1.4       
## [41] BiocIO_1.4.0           generics_0.1.1         hwriter_1.3.2          jsonlite_1.8.0        
## [45] dplyr_1.0.7            RCurl_1.98-1.5         magrittr_2.0.2         GenomeInfoDbData_1.2.7
## [49] Matrix_1.4-0           Rcpp_1.0.8.2           munsell_0.5.0          fansi_0.5.0           
## [53] lifecycle_1.0.1        stringi_1.7.6          yaml_2.3.5             zlibbioc_1.40.0       
## [57] BiocFileCache_2.2.0    blob_1.2.2             parallel_4.1.3         crayon_1.4.2          
## [61] lattice_0.20-45        hms_1.1.1              KEGGREST_1.34.0        knitr_1.37            
## [65] pillar_1.6.4           rjson_0.2.20           codetools_0.2-18       biomaRt_2.50.1        
## [69] XML_3.99-0.8           glue_1.6.2             evaluate_0.15          blogdown_1.8.2        
## [73] latticeExtra_0.6-29    BiocManager_1.30.16    png_0.1-7              vctrs_0.3.8           
## [77] gtable_0.3.0           purrr_0.3.4            assertthat_0.2.1       cachem_1.0.6          
## [81] xfun_0.30              restfulr_0.0.13        viridisLite_0.4.0      tibble_3.1.6          
## [85] memoise_2.0.1          ellipsis_0.3.2
</code></pre>
<h2 id="references">References</h2>
<div id="refs" class="references">
<div id="ref-Huber2015-ag">
<p>Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. “Orchestrating High-Throughput Genomic Analysis with Bioconductor.” <em>Nat. Methods</em> 12 (2): 115–21. <a href="https://doi.org/10.1038/nmeth.3252">https://doi.org/10.1038/nmeth.3252</a>.</p>
</div>
<div id="ref-Lawrence2013-kt">
<p>Lawrence, Michael, Wolfgang Huber, Hervé Pagès, Patrick Aboyoun, Marc Carlson, Robert Gentleman, Martin T Morgan, and Vincent J Carey. 2013. “Software for Computing and Annotating Genomic Ranges.” <em>PLoS Comput. Biol.</em> 9 (8): e1003118. <a href="https://doi.org/10.1371/journal.pcbi.1003118">https://doi.org/10.1371/journal.pcbi.1003118</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2022-04-21: <a  href="https://github.com/tgirke/GEN242//commit/4c2292bab75043db66271c31521a4cee3c6729de">some edits (4c2292ba)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2022 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
