<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>Introduction to HPCC Cluster and Linux | GEN242</title><meta property="og:title" content="Introduction to HPCC Cluster and Linux" />
<meta property="og:description" content="HPCC Cluster Overview The HPCC Cluster (formerly called biocluster) is a shared research computing system available at UCR. The HPCC website is available here.
What Is a Computer Cluster?   A computer cluster is an assembly of CPU units, so called computer nodes that work together to perform many computations in parallel. To achieve this, an internal network (e.g. Infiniband interconnect) connects the nodes to a larger unit, while a head node controls the load and traffic across the entire system." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/linux/linux/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2022-04-20T16:19:29-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="Introduction to HPCC Cluster and Linux">
<meta itemprop="description" content="HPCC Cluster Overview The HPCC Cluster (formerly called biocluster) is a shared research computing system available at UCR. The HPCC website is available here.
What Is a Computer Cluster?   A computer cluster is an assembly of CPU units, so called computer nodes that work together to perform many computations in parallel. To achieve this, an internal network (e.g. Infiniband interconnect) connects the nodes to a larger unit, while a head node controls the load and traffic across the entire system.">
<meta itemprop="dateModified" content="2022-04-20T16:19:29-07:00" />
<meta itemprop="wordCount" content="2953">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Introduction to HPCC Cluster and Linux"/>
<meta name="twitter:description" content="HPCC Cluster Overview The HPCC Cluster (formerly called biocluster) is a shared research computing system available at UCR. The HPCC website is available here.
What Is a Computer Cluster?   A computer cluster is an assembly of CPU units, so called computer nodes that work together to perform many computations in parallel. To achieve this, an internal network (e.g. Infiniband interconnect) connects the nodes to a larger unit, while a head node controls the load and traffic across the entire system."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S22)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T19)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2022/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.f6d5d5051795429311a98a36db40ddc5.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.f6d5d5051795429311a98a36db40ddc5.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../tutorials/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Tutorials</a>
  </li>
  <ul>
    <li class="collapse show" id="tutorials">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsgithubgithub" href="../../../tutorials/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-tutorialslinuxlinux" href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrbasicsrbasics" href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrprogrammingrprogramming" href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrparallelrparallel" href="../../../tutorials/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrsequencesrsequences" href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempiper" href="../../../tutorials/systempiper/systempiper/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseqnew" href="../../../tutorials/sprnaseq/sprnaseqnew/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspchipseqspchipseq" href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsscrnaseqscrnaseq" href="../../../tutorials/scrnaseq/scrnaseq/">scRNA-Seq Embedding Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrmarkdownrmarkdown" href="../../../tutorials/rmarkdown/rmarkdown/">R Markdown Tutorial</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrfearfea" href="../../../tutorials/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrclusteringrclustering" href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrgraphicsrgraphics" href="../../../tutorials/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappsshinyapps" href="../../../tutorials/shinyapps/shinyapps/">Shiny Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrpackagesrpackages" href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsdplyrdplyr" href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialscmdtocwlcmdtocwl" href="../../../tutorials/cmdtocwl/cmdtocwl/">Automate Creation of CWL Instructions</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempiperrnaseqsystempipernaseq" href="../../../tutorials/systempiper/rnaseq/systempipernaseq/">RNA-Seq Workflow Template</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/linux/linux.md" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View source code</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/tutorials/linux/linux.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/tutorials/linux/linux.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=Introduction%20to%20HPCC%20Cluster%20and%20Linux" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#hpcc-cluster-overview">HPCC Cluster Overview</a>
      <ul>
        <li><a href="#what-is-a-computer-cluster">What Is a Computer Cluster?</a></li>
        <li><a href="#hardware-infrastructure">Hardware Infrastructure</a></li>
        <li><a href="#computer-nodes">Computer nodes</a></li>
        <li><a href="#interconnect">Interconnect</a></li>
        <li><a href="#storage">Storage</a></li>
        <li><a href="#user-traffic">User traffic</a></li>
        <li><a href="#manuals">Manuals</a></li>
      </ul>
    </li>
    <li><a href="#linux-basics">Linux Basics</a>
      <ul>
        <li><a href="#log-into-hpcc-cluster">Log into HPCC Cluster</a></li>
        <li><a href="#important-linux-commands">Important Linux Commands</a></li>
        <li><a href="#file-exchange">File Exchange</a></li>
        <li><a href="#std-inouterr-redirect--wildcards">STD IN/OUT/ERR, Redirect &amp; Wildcards</a></li>
        <li><a href="#homework-assignment-hw2">Homework Assignment (HW2)</a></li>
        <li><a href="#permissions-and-ownership">Permissions and ownership</a></li>
        <li><a href="#symbolic-links">Symbolic Links</a></li>
      </ul>
    </li>
    <li><a href="#software-and-module-system">Software and module system</a></li>
    <li><a href="#big-data-storage">Big data storage</a></li>
    <li><a href="#queuing-system-slurm">Queuing system: <code>Slurm</code></a>
      <ul>
        <li><a href="#job-submission-with-sbatch">Job submission with <code>sbatch</code></a></li>
        <li><a href="#interactive-sessions-with-srun">Interactive sessions with <code>srun</code></a></li>
        <li><a href="#monitoring-jobs-with-squeue">Monitoring jobs with <code>squeue</code></a></li>
        <li><a href="#deleting-and-altering-jobs">Deleting and altering jobs</a></li>
        <li><a href="#resource-limits">Resource limits</a></li>
      </ul>
    </li>
    <li><a href="#textcode-editors">Text/code editors</a>
      <ul>
        <li><a href="#why-does-it-matter">Why does it matter?</a></li>
        <li><a href="#vim-overview">Vim overview</a></li>
        <li><a href="#useful-resources-for-learning-vimnvim">Useful resources for learning vim/nvim</a></li>
      </ul>
    </li>
    <li><a href="#nvim-r-tmux-essentials">Nvim-R-Tmux essentials</a>
      <ul>
        <li><a href="#basics">Basics</a></li>
        <li><a href="#quick-configuration-in-user-accounts-of-ucrs-hpcc">Quick configuration in user accounts of UCR&rsquo;s HPCC</a></li>
        <li><a href="#basic-usage-of-nvim-r-tmux">Basic usage of Nvim-R-Tmux</a></li>
        <li><a href="#important-keybindings-for-nvim">Important keybindings for nvim</a></li>
        <li><a href="#important-keybindings-for-tmux">Important keybindings for tmux</a></li>
      </ul>
    </li>
    <li><a href="#nvim-ides-for-other-languages">Nvim IDEs for other languages</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../tutorials/">Tutorials</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>Introduction to HPCC Cluster and Linux</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 14 minute read</p>
           
	<br/>
<br/>
<h2 id="hpcc-cluster-overview">HPCC Cluster Overview</h2>
<p>The HPCC Cluster (formerly called biocluster) is a shared research computing system available at UCR. The HPCC website is available <a href="http://hpcc.ucr.edu/index.html">here</a>.</p>
<h3 id="what-is-a-computer-cluster">What Is a Computer Cluster?</h3>
<ul>
<li>
<p>A computer cluster is an assembly of CPU units, so called computer nodes that work together to perform many computations in parallel. To achieve this, an internal network (e.g. Infiniband interconnect) connects the nodes to a larger unit, while a head node controls the load and traffic across the entire system.</p>
</li>
<li>
<p>Usually, users log into the head node to submit their computer requests via <code>srun</code> to a queuing system provided by resource management and scheduling software, such as SGE, Slurm or TORQUE/MAUI. The queuing system distributes the processes to the computer nodes in a controlled fashion.</p>
</li>
<li>
<p>Because the head node controls the entire system, users should never run computing jobs on the head node directly!</p>
</li>
<li>
<p>For code testing purposes, one can log into one of the nodes with <code>srun --pty bash -l</code> and run jobs interactively. Alternatively, one can log into the test node owl via ssh.</p>
</li>
</ul>
<h3 id="hardware-infrastructure">Hardware Infrastructure</h3>
<h3 id="computer-nodes">Computer nodes</h3>
<ul>
<li>Over 8,000 CPU cores</li>
<li>130 Intel, AMD and GPU nodes</li>
<li>32-128 CPU cores per node</li>
<li>256-1,024 GB of RAM per node</li>
<li>12 GPU nodes, each with total of over 80,000 cuda cores</li>
</ul>
<h3 id="interconnect">Interconnect</h3>
<ul>
<li>FDR IB @56Gbs</li>
</ul>
<h3 id="storage">Storage</h3>
<ul>
<li>Parallel GPFS storage system with 3.0 PB usable space</li>
<li>File system scales to over 50 PB</li>
<li>Backup of same architecture and similar amount</li>
</ul>
<h3 id="user-traffic">User traffic</h3>
<ul>
<li>Computing tasks need to be submitted via <code>sbatch</code> or <code>srun</code></li>
<li>HPCC Cluster headnode only for login, not for computing tasks!</li>
<li>Monitor cluster activity: <code>squeue</code> or <code>jobMonitor</code> (<code>qstatMonitor</code>)</li>
</ul>
<h3 id="manuals">Manuals</h3>
<ul>
<li><a href="https://hpcc.ucr.edu/manuals/hpc_cluster/">HPCC Cluster Manual</a></li>
<li><a href="https://hpcc.ucr.edu/manuals/linux_basics/">Linux Manual</a></li>
</ul>
<h2 id="linux-basics">Linux Basics</h2>
<h3 id="log-into-hpcc-cluster">Log into HPCC Cluster</h3>
<ul>
<li>Login command on OS X or Linux</li>
</ul>
<pre><code class="language-sh">ssh -XY user@cluster.hpcc.ucr.edu
</code></pre>
<p>Type password</p>
<ul>
<li>
<p>Windows: provide same information in a terminal application like <a href="http://www.chiark.greenend.org.uk/~sgtatham/putty/download.html">Putty</a> or <a href="http://mobaxterm.mobatek.net/">MobaXterm</a>.
<a href="https://mobaxterm.mobatek.net/demo.html">Here</a> is an annimated usage introduction for MobaXterm.</p>
<ul>
<li>Host name: <code>cluster.hpcc.ucr.edu</code></li>
<li>User name: &hellip;</li>
<li>Password: &hellip;</li>
</ul>
</li>
<li>
<p>Additional login information can be found on the corresponding HPCC manuals:</p>
<ul>
<li>Login page: <a href="https://hpcc.ucr.edu/manuals/hpc_cluster/login/">here</a></li>
<li>SSH Keys: <a href="https://hpcc.ucr.edu/manuals/hpc_cluster/sshkeys/">here</a></li>
<li>Duo Multifactor Authenication: <a href="https://hpcc.ucr.edu/manuals/hpc_cluster/login/#secure-authentication">here</a></li>
<li>UCR Duo Manual: <a href="https://its.ucr.edu/sites/g/files/rcwecm321/files/2018-06/Multi-Factor%20Authentication%20Handout.pdf">here</a></li>
</ul>
</li>
</ul>
<h3 id="important-linux-commands">Important Linux Commands</h3>
<p>The following provides a short overview of important shell commands. Much more detailed information can be found on <a href="https://hpcc.ucr.edu/manuals/linux_basics/">HPCC&rsquo;s Linux tutorials</a>.</p>
<p>Finding help</p>
<pre><code class="language-sh">man &lt;program_name&gt;
</code></pre>
<p>List content of current directory</p>
<pre><code class="language-sh">ls
</code></pre>
<p>Print current working directory</p>
<pre><code class="language-sh">pwd
</code></pre>
<p>Search in files and directories</p>
<pre><code class="language-sh">grep
</code></pre>
<p>Word count</p>
<pre><code class="language-sh">wc
</code></pre>
<p>Create directory</p>
<pre><code class="language-sh">mkdir
</code></pre>
<p>Delete files and directories</p>
<pre><code class="language-sh">rm
</code></pre>
<p>Move and rename files</p>
<pre><code class="language-sh">mv
</code></pre>
<p>Copy files from internet to <code>pwd</code></p>
<pre><code class="language-sh">wget
</code></pre>
<p>Viewing files</p>
<pre><code class="language-sh">less
</code></pre>
<h3 id="file-exchange">File Exchange</h3>
<p><strong>GUI applications</strong></p>
<ul>
<li>Windows: <a href="http://winscp.net/eng/index.php">WinSCP</a></li>
<li>Mac OS X: <a href="http://cyberduck.en.softonic.com/mac">CyberDuck</a></li>
<li>Win/OS X/Linux: <a href="https://filezilla-project.org/">FileZilla</a></li>
</ul>
<center><img title="FileZilla_ssh_key" src="../images/FileZilla_ssh_key.png" width="600"><img/></center>
<center>FileZilla settings with an SSH key. For generating SSH keys see <a href="https://hpcc.ucr.edu/manuals/hpc_cluster/sshkeys/">here</a>.</center>
<p><br></br></p>
<p><strong>SCP: via command-line</strong> (<a href="https://linux.die.net/man/1/scp">Manual</a>)</p>
<p>Advantages of this method include: batch up/downloads and ease of automation.</p>
<pre><code class="language-sh">scp file user@remotehost:/home/user/ # From local to remote 
scp user@remotehost:/home/user/file . # From remote to local 
</code></pre>
<p><strong>RSYNC: via command-line</strong> (<a href="https://linux.die.net/man/1/rsync">Manual</a>)</p>
<p>Advantages of this method include: same as SCP plus differential update options and viewing of directory content.</p>
<p>Print (view) content of remote directory</p>
<pre><code class="language-sh">rsync user@remotehost:~/somedirectory/*
</code></pre>
<p>Download directory or file(s)</p>
<pre><code class="language-sh">rsync -avzhe ssh user@remotehost:~/somedirectory .
  # -a: recursive archive mode (thus -r not required), also preserves permissions, time stamps, etc 
  # -v: verbose
  # -z: compress data during transfer
  # -h: print messages in human-readable format
  # -e: specifies transfer protocol; using ssh here provides encryption during transfer
  # --delete: files that were deleted on source will be deleted also in backup-destination
  # -n: for testing use this dry-run option, but drop '-e ssh' in this case
</code></pre>
<p>Upload directory or file(s)</p>
<pre><code class="language-sh">rsync -avzhe ssh somedirectory user@hostname:~/
</code></pre>
<h3 id="std-inouterr-redirect--wildcards">STD IN/OUT/ERR, Redirect &amp; Wildcards</h3>
<p>Wildcard <code>*</code> to specify many files</p>
<pre><code class="language-sh">file.*                        
</code></pre>
<p>Redirect <code>ls</code> output to file</p>
<pre><code class="language-sh">ls &gt; file                     
</code></pre>
<p>Specify file as input to command</p>
<pre><code class="language-sh">command &lt; myfile              
</code></pre>
<p>Append output of command to file</p>
<pre><code class="language-sh">command &gt;&gt; myfile             
</code></pre>
<p>Pipe <code>STDOUT</code> of one command to another command</p>
<pre><code class="language-sh">command1 | command2     
</code></pre>
<p>Turn off progress info</p>
<pre><code class="language-sh">command &gt; /dev/null 
</code></pre>
<p>Pipe output of <code>grep</code> to <code>wc</code></p>
<pre><code class="language-sh">grep pattern file | wc        
</code></pre>
<p>Print <code>STDERR</code> to file</p>
<pre><code class="language-sh">grep pattern nonexistingfile 2 &gt; mystderr 
</code></pre>
<h3 id="homework-assignment-hw2">Homework Assignment (HW2)</h3>
<p>See HW2 page <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw02/hw02/">here</a>.</p>
<h3 id="permissions-and-ownership">Permissions and ownership</h3>
<p>List directories and files</p>
<pre><code class="language-sh">ls -al 
</code></pre>
<p>The previous command shows something like this for each file/dir: <code>drwxrwxrwx</code>. The
meaning of this syntax is as follows:</p>
<ul>
<li><code>d</code>: directory</li>
<li><code>rwx</code>: read, write and execute permissions, respectively
<ul>
<li>first triplet: user permissions (u)</li>
<li>second triplet: group permissions (g)</li>
<li>third triplet: world permissions (o)</li>
</ul>
</li>
</ul>
<p>Example for assigning write and execute permissions to user, group and world</p>
<pre><code class="language-sh">chmod ugo+rx my_file
</code></pre>
<ul>
<li><code>+</code> causes the permissions selected to be added</li>
<li><code>-</code> causes them to be removed</li>
<li><code>=</code> causes them to be the only permissions that the file has.</li>
</ul>
<p>When performing the same operation on many files with subdirectories then one can
use <code>-R</code> for recursive behavior.</p>
<pre><code class="language-sh">chmod -R ugo+rx my_dir
</code></pre>
<p>Since directories have to be executable the capital <code>X</code> option can be useful which
applies only to directories but not to files. The following will assign <code>drwxr-xr-x</code> to directories
and <code>-rw-r--r--</code> to files and hidden files.</p>
<pre><code class="language-sh">chmod -R ugo-x,u+rwX,go+rX,go-w ./* ./.[!.]*
</code></pre>
<p>Syntax for changing user &amp; group ownership</p>
<pre><code class="language-sh">chown &lt;user&gt;:&lt;group&gt; &lt;file or dir&gt; 
</code></pre>
<h3 id="symbolic-links">Symbolic Links</h3>
<p>Symbolic links are short nicknames to files and directories that save typing of their full paths.</p>
<pre><code class="language-sh">ln -s original_filename new_nickname
</code></pre>
<h2 id="software-and-module-system">Software and module system</h2>
<ul>
<li>Over 2,000 software tools are currently installed on HPCC Cluster</li>
<li>Custom installs in user accounts via various mechanisms, e.g. environment management systems such as <a href="https://conda.io/projects/conda/en/latest/index.html">conda</a></li>
<li>Most common research databases used in bioinformatics are available</li>
<li>Support of most common programming languages used in research computing</li>
<li>A module system is used to facilitate the management of software tools. This includes any number of versions of each software.</li>
<li>New software install requests can be sent to <a href="mailto:support@hpcc.ucr.edu">support@hpcc.ucr.edu</a>.</li>
<li>To use software manged under the module system, users need to learn using some basic commands. The most common commands are listed below.</li>
</ul>
<p>Print available modules</p>
<pre><code class="language-sh">module avail
</code></pre>
<p>Print available modules starting with R</p>
<pre><code class="language-sh">module avail R
</code></pre>
<p>Load default module R</p>
<pre><code class="language-sh">module load R
</code></pre>
<p>Load specific module R version</p>
<pre><code class="language-sh">module load R/3.2.2
</code></pre>
<p>List loaded modules</p>
<pre><code class="language-sh">module list
</code></pre>
<p>Unload module R</p>
<pre><code class="language-sh">module unload R
</code></pre>
<p>Unload specific module R</p>
<pre><code class="language-sh">module unload R/3.2.3-dev
</code></pre>
<h2 id="big-data-storage">Big data storage</h2>
<p>Each user account on HPCC Cluster comes only with 20GB of disk space. Much more disk space is
available in a dedicated <code>bigdata</code> directory. How much space depends on the subscription
of each user group. The path of <code>bigdata</code> and <code>bigdata-shared</code> is as follows:</p>
<ul>
<li><code>/bigdata/labname/username</code></li>
<li><code>/bigdata/labname/shared</code></li>
</ul>
<p>All lab members share the same bigdata pool. The course number <code>gen242</code> is used as <code>labname</code>
for user accounts adminstered under GEN242 (here /bigdata/gen242/shared).</p>
<p>The disk usage of <code>home</code> and <code>bigdata</code> can be monitored on the <a href="https://dashboard.hpcc.ucr.edu/">HPCC Cluster Dashboard</a>.</p>
<h2 id="queuing-system-slurm">Queuing system: <code>Slurm</code></h2>
<p>HPCC Cluster uses <code>Slurm</code> as queuing and load balancing system. To control user traffic, any
type of compute intensive jobs need to be submitted via <code>sbatch</code> or <code>srun</code> (see below) to the computer
nodes. Much more detailed information on this topic can be found on these sites:</p>
<ul>
<li><a href="http://hpcc.ucr.edu/manuals_linux-cluster_jobs.html">UCR HPCC Manual</a></li>
<li><a href="https://slurm.schedmd.com/documentation.html">Slurm Documentation</a></li>
<li><a href="http://www.nersc.gov/users/computational-systems/cori/running-jobs/for-edison-users/torque-moab-vs-slurm-comparisons/">Torque/Slurm Comparison</a></li>
<li><a href="https://sites.google.com/a/case.edu/hpc-upgraded-cluster/slurm-cluster-commands">Switching from Torque to Slurm</a></li>
<li><a href="http://www.ceci-hpc.be/slurm_tutorial.html">Slurm Quick Start Tutorial</a></li>
</ul>
<h3 id="job-submission-with-sbatch">Job submission with <code>sbatch</code></h3>
<p>Print information about queues/partitions available on a cluster.</p>
<pre><code class="language-sh">sinfo
</code></pre>
<p>Compute jobs are submitted with <code>sbatch</code> via a submission script (here <code>script_name.sh</code>).</p>
<pre><code class="language-sh">sbatch script_name.sh
</code></pre>
<p>The following sample submission script (<code>script_name.sh</code>) executes an R script named <code>my_script.R</code>.</p>
<pre><code class="language-sh">#!/bin/bash -l

#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=1G
#SBATCH --time=1-00:15:00 # 1 day and 15 minutes
#SBATCH --mail-user=useremail@address.com
#SBATCH --mail-type=ALL
#SBATCH --job-name=&quot;some_test&quot;
#SBATCH -p batch # Choose queue/parition from: intel, batch, highmem, gpu, short

Rscript my_script.R
</code></pre>
<p><code>STDOUT</code> and <code>STDERROR</code> of jobs will be written to files named
<code>slurm-&lt;jobid&gt;.out</code> or to a custom file specified under <code>#SBATCH --output</code> in
the submission script.</p>
<h3 id="interactive-sessions-with-srun">Interactive sessions with <code>srun</code></h3>
<p>This option logs a user in to a computer node of a specified partition (queue), while Slurm monitors and controls the resource request.</p>
<pre><code class="language-sh">srun --pty bash -l
</code></pre>
<p>Interactive session with specific resource requests</p>
<pre><code class="language-sh">srun --x11 --partition=short --mem=2gb --cpus-per-task 4 --ntasks 1 --time 1:00:00 --pty bash -l
</code></pre>
<p>The argument <code>--mem</code> limits the amount of RAM, <code>--cpus</code> the number of CPU
cores, <code>--time</code> the time how long a session will be active. Under
<code>--parition</code> one can choose among different queues and node architectures.
Current options under <code>--partition</code> for most users of the HPCC cluster are: <code>intel</code>, <code>batch</code>, <code>highmem</code>, <code>gpu</code>,
and <code>short</code>. The latter has a time limit of 2 hours.</p>
<h3 id="monitoring-jobs-with-squeue">Monitoring jobs with <code>squeue</code></h3>
<p>List all jobs in queue</p>
<pre><code class="language-sh">squeue
</code></pre>
<p>List jobs of a specific user</p>
<pre><code class="language-sh">squeue -u &lt;user&gt;
</code></pre>
<p>Print more detailed information about a job</p>
<pre><code class="language-sh">scontrol show job &lt;JOBID&gt;
</code></pre>
<p>Custom command to summarize and visualize cluster activity</p>
<pre><code class="language-sh">jobMonitor
</code></pre>
<h3 id="deleting-and-altering-jobs">Deleting and altering jobs</h3>
<p>Delete a single job</p>
<pre><code class="language-sh">scancel -i &lt;JOBID&gt;
</code></pre>
<p>Delete all jobs of a user</p>
<pre><code class="language-sh">scancel -u &lt;username&gt; 
</code></pre>
<p>Delete all jobs of a certain name</p>
<pre><code class="language-sh">scancel --name &lt;myJobName&gt;
</code></pre>
<p>Altering jobs with <code>scontrol update</code>. The below example changes the walltime (<code>&lt;NEW_TIME&gt;</code>) of a specific job (<code>&lt;JOBID&gt;</code>).</p>
<pre><code class="language-sh">scontrol update jobid=&lt;JOBID&gt; TimeLimit=&lt;NEW_TIME&gt;
</code></pre>
<h3 id="resource-limits">Resource limits</h3>
<p>Resourse limits for users can be viewed as follows.</p>
<pre><code class="language-sh">sacctmgr show account $GROUP format=Account,User,Partition,GrpCPUs,GrpMem,GrpNodes --ass | grep $USER
</code></pre>
<p>Similarly, one can view the limits of the group a user belongs to.</p>
<pre><code class="language-sh">sacctmgr show account $GROUP format=Account,User,Partition,GrpCPUs,GrpMem,GrpNodes,GrpTRES%30 --ass | head -3
</code></pre>
<h2 id="textcode-editors">Text/code editors</h2>
<p>The following list includes examples of several widely used code editors.</p>
<ul>
<li><strong>Vi/Vim/Neovim</strong>: Non-graphical (terminal-based) editor. Vi is guaranteed to be available on any system. Vim and Nvim (Neovim) are the improved versions of vi.</li>
<li><strong>Emacs</strong>: Non-graphical or window-based editor. You still need to know keystroke commands to use it. Installed on all Linux distributions and on most other Unix systems.</li>
<li><strong>Pico</strong>: Simple terminal-based editor available on most versions of Unix. Uses keystroke commands, but they are listed in logical fashion at bottom of screen.</li>
<li><strong>Nano</strong>: A simple terminal-based editor which is default on modern Debian systems.</li>
<li><strong>Atom</strong>: Modern text editor developed by GitHub project.</li>
</ul>
<h3 id="why-does-it-matter">Why does it matter?</h3>
<p>To work efficiently on remote systems like a computer cluster, it is essential
to learn how to work in a pure command-line interface. GUI environments like
RStudio and similar coding environments are not suitable for this. In addition,
there is a lot of value of knowing how to work in an environment that is not
restricted to a specific programming language. Therefore, this class embraces
RStudio where it is useful, but for working on remote systems like HPCC Cluster, it
uses Nvim and Tmux. Both are useful for many programming languages.
Combinded with the <code>nvim-r</code> plugin they also provide a powerful command-line working
environment for R. The following provides a brief introduction to this environment.</p>
<h3 id="vim-overview">Vim overview</h3>
<p>The following opens a file (here <code>myfile</code>) with nvim (or vim)</p>
<pre><code class="language-sh">nvim myfile.txt # for neovim (or 'vim myfile.txt' for vim)
</code></pre>
<p>Once you are in Nvim, there are three main modes: normal, insert and command mode. The most important commands for switching between the three modes are:</p>
<ul>
<li><code>i</code>: The <code>i</code> key brings you from the normal mode to the insert mode. The latter is used for typing.</li>
<li><code>Esc</code>: The <code>Esc</code> key brings you from the insert mode back to the normal mode.</li>
<li><code>:</code>: The <code>:</code> key starts the command mode at the bottom of the screen.</li>
</ul>
<p>Use the arrow keys to move your cursor in the text. Using <code>Fn Up/Down key</code> allows to page through
the text quicker. In the following command overview, all commands starting with <code>:</code> need to be typed in the command mode.
All other commands are typed in the normal mode after pushing the <code>Esc</code> key.</p>
<p>Important modifier keys to control vim/nvim</p>
<ul>
<li><code>:w</code>: save changes to file. If you are in editing mode you have to hit <code>Esc</code> first.</li>
<li><code>:q</code>: quit file that has not been changed</li>
<li><code>:wq</code>: save and quit file</li>
<li><code>:!q</code>: quit file without saving any changes</li>
</ul>
<h3 id="useful-resources-for-learning-vimnvim">Useful resources for learning vim/nvim</h3>
<ul>
<li><a href="http://www.openvim.com">Interactive Vim Tutorial</a></li>
<li><a href="http://vimdoc.sourceforge.net/">Official Vim Documentation</a></li>
<li><a href="http://hpcc.ucr.edu/manuals_linux-basics_vim.html">HPCC Linux Manual</a></li>
</ul>
<h2 id="nvim-r-tmux-essentials">Nvim-R-Tmux essentials</h2>
<p>Terminal-based Working Environment for R: <a href="https://gist.github.com/tgirke/7a7c197b443243937f68c422e5471899">Nvim-R-Tmux</a>.</p>
<center><img title="Nvim-R-Tmux" src="https://raw.githubusercontent.com/jalvesaq/Nvim-R/master/Nvim-R.gif" ></center>
<center>Nvim-R-Tmux IDE for R</center>
<h3 id="basics">Basics</h3>
<p>Tmux is a terminal multiplexer that allows to split terminal windows and to detach/reattach to
existing terminal sessions. Combinded with the <code>nvim-r</code> plugin it provides a powerful command-line working
environment for R where users can send code from a script to the R console or command-line.
Both tmux and the <code>nvim-r</code> plugin need to be installed on a system. On HPCC Cluster both are configured
in each user account. If this is not the case then follow the quick configuration instructions given in the following subsection.</p>
<h3 id="quick-configuration-in-user-accounts-of-ucrs-hpcc">Quick configuration in user accounts of UCR&rsquo;s HPCC</h3>
<p>Skip these steps if Nvim-R-Tmux is already configured in your account. Or follow the <a href="https://github.com/tgirke/Nvim-R_Tmux">detailed
instructions</a> to install Nvim-R-Tmux from scratch on your own system.</p>
<ol>
<li>Log in to your user account on HPCC and execute <code>Install_Nvim-R_Tmux</code> (old version: <code>install_nvimRtmux</code>). Alternatively, follow these step-by-step <a href="https://github.com/tgirke/Nvim-R_Tmux">install commands</a>.</li>
<li>To enable the nvim-R-tmux environment, log out and in again.</li>
<li>Follow usage instructions of next section.</li>
</ol>
<h3 id="basic-usage-of-nvim-r-tmux">Basic usage of Nvim-R-Tmux</h3>
<p>The official and much more detailed user manual for <code>Nvim-R</code> is available <a href="https://github.com/jalvesaq/Nvim-R/blob/master/doc/Nvim-R.txt">here</a>.
The following gives a short introduction into the basic usage of Nvim-R-Tmux:</p>
<p><strong>1. Start tmux session</strong> (optional)</p>
<p>Note, running Nvim from within a tmux session is optional. Skip this step if tmux functionality is not required (<em>e.g.</em> reattaching to sessions on remote systems).</p>
<pre><code class="language-sh">tmux # starts a new tmux session 
tmux a # attaches to an existing session 
</code></pre>
<p><strong>2. Open nvim-connected R session</strong></p>
<p>Open a <code>*.R</code> or <code>*.Rmd</code> file with <code>nvim</code> and intialize a connected R session with <code>\rf</code>. This command can be remapped to other key combinations, e.g. uncommenting lines 10-12 in <code>.config/nvim/init.vim</code> will remap it to the <code>F2</code> key. Note, the resulting split window among Nvim and R behaves like a split viewport in <code>nvim</code> or <code>vim</code> meaning the usage of <code>Ctrl-w w</code> followed by <code>i</code> and <code>Esc</code> is important for navigation.</p>
<pre><code class="language-sh">nvim myscript.R # or *.Rmd file
</code></pre>
<p><strong>3. Send R code from nvim to the R pane</strong></p>
<p>Single lines of code can be sent from nvim to the R console by pressing the space bar. To send
several lines at once, one can select them in nvim&rsquo;s visual mode and then hit the space bar.
Please note, the default command for sending code lines in the nvim-r-plugin is <code>\l</code>. This key
binding has been remapped in the provided <code>.config/nvim/init.vim</code> file to the space bar. Most other key bindings (shortcuts) still start with the <code>\</code> as LocalLeader, <em>e.g.</em> <code>\rh</code> opens the help for a function/object where the curser is located in nvim. More details on this are given below.</p>
<h3 id="important-keybindings-for-nvim">Important keybindings for nvim</h3>
<p>The main advantages of Neovim compared to Vim are its better performance and its built-in terminal emulator facilitating the communication among Neovim and interactive programming environments such as R. Since the Vim and Neovim environments are managed independently, one can run them in parallel on the same system without interfering with each other. The usage of Neovim is almost identical to Vim.</p>
<p><strong>Nvim commands</strong></p>
<ul>
<li><code>\rf</code>: opens vim-connected R session. If you do this the first time in your user account, you might be asked to create an <code>R</code> directory under <code>~/</code>. If so approve this action by pressing <code>y</code>.</li>
<li><code>spacebar</code>: sends code from vim to R; here remapped in <code>init.vim</code> from default <code>\l</code></li>
<li><code>:split</code> or <code>:vsplit</code>: splits viewport (similar to pane split in tmux)</li>
<li><code>gz</code>: maximizes size of viewport in normal mode (similar to Tmux&rsquo;s <code>Ctrl-a z</code> zoom utility)</li>
<li><code>Ctrl-w w</code>: jumps cursor to R viewport and back; toggle between insert (<code>i</code>) and command (<code>Esc</code>) mode is required for navigation and controlling the environment.</li>
<li><code>Ctrl-w r</code>: swaps viewports</li>
<li><code>Ctrl-w =</code>: resizes splits to equal size</li>
<li><code>:resize &lt;+5 or -5&gt;</code>: resizes height by specified value</li>
<li><code>:vertical resize &lt;+5 or -5&gt;</code>: resizes width by specified value</li>
<li><code>Ctrl-w H</code> or <code>Ctrl-w K</code>: toggles between horizontal/vertical splits</li>
<li><code>Ctrl-spacebar</code>: omni completion for R objects/functions when nvim is in insert mode. Note, this has been remapped in <code>init.vim</code> from difficult to type default <code>Ctrl-x Ctrl-o</code>.</li>
<li><code>:h nvim-R</code>: opens nvim-R&rsquo;s user manual; navigation works the same as for any Vim/Nvim help document</li>
<li><code>:Rhelp fct_name</code>: opens help for a function from nvim&rsquo;s command mode with text completion support</li>
<li><code>Ctrl-s and Ctrl-x</code>: freezes/unfreezes vim (some systems)</li>
</ul>
<h3 id="important-keybindings-for-tmux">Important keybindings for tmux</h3>
<p><strong>Pane-level commands</strong></p>
<ul>
<li><code>Ctrl-a %</code>: splits pane vertically</li>
<li><code>Ctrl-a &quot;</code>: splits pane horizontally</li>
<li><code>Ctrl-a o</code>: jumps cursor to next pane</li>
<li><code>Ctrl-a Ctrl-o</code>: swaps panes</li>
<li><code>Ctrl-a &lt;space bar&gt;</code>: rotates pane arrangement</li>
<li><code>Ctrl-a Alt &lt;left or right&gt;</code>: resizes to left or right</li>
<li><code>Ctrl-a Esc &lt;up or down&gt;</code>: resizes to left or right</li>
</ul>
<p><strong>Window-level comands</strong></p>
<ul>
<li><code>Ctrl-a n</code>: switches to next tmux window</li>
<li><code>Ctrl-a Ctrl-a</code>: switches to previous tmux window</li>
<li><code>Ctrl-a c</code>: creates a new tmux window</li>
<li><code>Ctrl-a 1</code>: switches to specific tmux window selected by number</li>
</ul>
<p><strong>Session-level comands</strong></p>
<ul>
<li><code>Ctrl-a d</code>: detaches from current session</li>
<li><code>Ctrl-a s</code>: switch between available tmux sesssions</li>
<li><code>$ tmux new -s &lt;name&gt;</code>: starts new session with a specific name</li>
<li><code>$ tmux ls</code>: lists available tmux session(s)</li>
<li><code>$ tmux attach -t &lt;id&gt;</code>: attaches to specific tmux session</li>
<li><code>$ tmux attach</code>: reattaches to session</li>
<li><code>$ tmux kill-session -t &lt;id&gt;</code>: kills a specific tmux session</li>
<li><code>Ctrl-a : kill-session</code>: kills a session from tmux command mode that can be initiated with <code>Ctrl-a :</code></li>
</ul>
<h2 id="nvim-ides-for-other-languages">Nvim IDEs for other languages</h2>
<p>For other languages, such as Bash, Python and Ruby, one can use the
<a href="https://github.com/jalvesaq/vimcmdline">vimcmdline</a> plugin for nvim (or vim). To
install it, one needs to copy from the <code>vimcmdline</code> resository the directories
<code>ftplugin</code>, <code>plugin</code> and <code>syntax</code> and their files to <code>~/.config/nvim/</code>. For
user accounts of UCR’s HPCC, the above install script <code>install_nvimRtmux</code> includes the
install of <code>vimcmdline</code> (since 09-Jun-18).</p>
<p>The usage of <code>vimcmdline</code> is very similar to <code>nvim-R</code>. To start a connected terminal session, one
opens with nvim a code file with the extension of a given language (<em>e.g.</em> <code>*.sh</code> for Bash or <code>*.py</code> for Python),
while the corresponding interactive interpreter session is initiated
by pressing the key sequence <code>\s</code> (corresponds to <code>\rf</code> under <code>nvim-R</code>). Subsequently, code lines can be sent
with the space bar. More details are available <a href="https://github.com/jalvesaq/vimcmdline">here</a>.</p>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2022-04-20: <a  href="https://github.com/tgirke/GEN242//commit/58778e72f949a2d378c84725535f7606e7c10843">some edits (58778e72)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2022 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
