<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>Environments dplyr, tidyr and some SQLite | GEN242</title><meta property="og:title" content="Environments dplyr, tidyr and some SQLite" />
<meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ Slides ] [ .Rmd ] [ .R ]
 Overview Modern object classes and methods for handling data.frame like structures are provided by the dplyr (tidyr) and data.table packages. A related example is Bioconductor’s DataTable object class (“Learn the tidyverse,” n.d.). This tutorial provide a short introduction to the usage and functionalities of the dplyr and related packages." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/dplyr/dplyr/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2023-05-30T11:33:40-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="Environments dplyr, tidyr and some SQLite">
<meta itemprop="description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ Slides ] [ .Rmd ] [ .R ]
 Overview Modern object classes and methods for handling data.frame like structures are provided by the dplyr (tidyr) and data.table packages. A related example is Bioconductor’s DataTable object class (“Learn the tidyverse,” n.d.). This tutorial provide a short introduction to the usage and functionalities of the dplyr and related packages.">
<meta itemprop="dateModified" content="2023-05-30T11:33:40-07:00" />
<meta itemprop="wordCount" content="4852">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Environments dplyr, tidyr and some SQLite"/>
<meta name="twitter:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ Slides ] [ .Rmd ] [ .R ]
 Overview Modern object classes and methods for handling data.frame like structures are provided by the dplyr (tidyr) and data.table packages. A related example is Bioconductor’s DataTable object class (“Learn the tidyverse,” n.d.). This tutorial provide a short introduction to the usage and functionalities of the dplyr and related packages."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Git and GitHub (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux Cluster Basics (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Overview (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Enrichment Analysis (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics in R (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (S22)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_23/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflow Review (S23)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_24/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Final Discussion (S24)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_25/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Additional Slide Shows (S25)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spvarseq/spvarseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 VAR-Seq Workflow (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T20)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8: R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_dex/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEXseq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/09_drug_target_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Drug-target Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/10_lncrna_orf_discovery/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 lncRNA Discovery
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/class/ludf7verm4t14z"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.cead41dc5c3100193d09979d5e668d42.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.cead41dc5c3100193d09979d5e668d42.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../tutorials/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Tutorials</a>
  </li>
  <ul>
    <li class="collapse show" id="tutorials">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsgithubgithub" href="../../../tutorials/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialslinuxlinux" href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrbasicsrbasics" href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrprogrammingrprogramming" href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrparallelrparallel" href="../../../tutorials/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrsequencesrsequences" href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempiper" href="../../../tutorials/systempiper/systempiper/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseq" href="../../../tutorials/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspvarseqspvarseq" href="../../../tutorials/spvarseq/spvarseq/">VAR-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspchipseqspchipseq" href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsscrnaseqscrnaseq" href="../../../tutorials/scrnaseq/scrnaseq/">scRNA-Seq Embedding Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrmarkdownrmarkdown" href="../../../tutorials/rmarkdown/rmarkdown/">R Markdown Tutorial</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrfearfea" href="../../../tutorials/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrclusteringrclustering" href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrgraphicsrgraphics" href="../../../tutorials/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappslezhangshinyapps" href="../../../tutorials/shinyappslezhang/shinyapps/">Shiny Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappsshinyapps" href="../../../tutorials/shinyapps/shinyapps/">Shiny Web Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrpackagesrpackages" href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-tutorialsdplyrdplyr" href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialscmdtocwlcmdtocwl" href="../../../tutorials/cmdtocwl/cmdtocwl/">Automate Creation of CWL Instructions</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempiperrnaseqsystempipernaseq" href="../../../tutorials/systempiper/rnaseq/systempipernaseq/">RNA-Seq Workflow Template</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/dplyr/dplyr.Rmd" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/tutorials/dplyr/dplyr.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/tutorials/dplyr/dplyr.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=Environments%20dplyr,%20tidyr%20and%20some%20SQLite" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#overview">Overview</a>
      <ul>
        <li><a href="#related-documentation">Related documentation</a></li>
        <li><a href="#installation">Installation</a></li>
        <li><a href="#construct-a-tibble-tibble">Construct a <code>tibble</code> (<code>tibble</code>)</a></li>
        <li><a href="#reading-and-writing-tabular-files">Reading and writing tabular files</a>
          <ul>
            <li><a href="#import-with-readr">Import with <code>readr</code></a></li>
            <li><a href="#fast-table-import-with-fread">Fast table import with <code>fread</code></a></li>
            <li><a href="#export-with-readr">Export with <code>readr</code></a></li>
          </ul>
        </li>
        <li><a href="#column-and-row-binds">Column and row binds</a></li>
        <li><a href="#extract-column-as-vector">Extract column as vector</a></li>
        <li><a href="#important-dplyr-functions">Important <code>dplyr</code> functions</a></li>
        <li><a href="#slice-and-filter-functions">Slice and filter functions</a>
          <ul>
            <li><a href="#filter-function">Filter function</a></li>
            <li><a href="#base-r-code-equivalent">Base R code equivalent</a></li>
            <li><a href="#including-boolean-operators">Including boolean operators</a></li>
            <li><a href="#subset-rows-by-position">Subset rows by position</a></li>
            <li><a href="#subset-rows-by-names">Subset rows by names</a></li>
          </ul>
        </li>
        <li><a href="#sorting-with-arrange">Sorting with <code>arrange</code></a></li>
        <li><a href="#select-columns-with-select">Select columns with <code>select</code></a></li>
        <li><a href="#change-column-order-with-relocate">Change column order with <code>relocate</code></a></li>
        <li><a href="#renaming-columns-with-rename">Renaming columns with <code>rename</code></a></li>
        <li><a href="#obtain-unique-rows-with-distinct">Obtain unique rows with <code>distinct</code></a></li>
        <li><a href="#add-columns">Add columns</a>
          <ul>
            <li><a href="#mutate"><code>mutate</code></a></li>
            <li><a href="#transmute"><code>transmute</code></a></li>
            <li><a href="#bind_cols"><code>bind_cols</code></a></li>
          </ul>
        </li>
        <li><a href="#summarize-data">Summarize data</a></li>
        <li><a href="#merging-tibbles">Merging tibbles</a>
          <ul>
            <li><a href="#inner-join">Inner join</a></li>
            <li><a href="#left-join">Left join</a></li>
            <li><a href="#right-join">Right join</a></li>
            <li><a href="#full-join">Full join</a></li>
            <li><a href="#anti-join">Anti join</a></li>
          </ul>
        </li>
        <li><a href="#chaining">Chaining</a>
          <ul>
            <li><a href="#example-1">Example 1</a></li>
            <li><a href="#example-2">Example 2</a></li>
            <li><a href="#example-3">Example 3</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#sqlite-databases">SQLite Databases</a></li>
    <li><a href="#loading-data-into-sqlite-databases">Loading data into SQLite databases</a>
      <ul>
        <li><a href="#list-names-of-tables-in-database">List names of tables in database</a></li>
        <li><a href="#import-table-into-dataframe">Import table into <code>data.frame</code></a></li>
        <li><a href="#query-database">Query database</a></li>
        <li><a href="#join-tables">Join tables</a></li>
      </ul>
    </li>
    <li><a href="#session-info">Session Info</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../tutorials/">Tutorials</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>Environments dplyr, tidyr and some SQLite</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 23 minute read</p>
           
	<!---
- Compile from command-line
Rscript -e "rmarkdown::render('dplyr.Rmd', c('html_document'), clean=FALSE); knitr::knit('dplyr.Rmd', tangle=TRUE)"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div style="text-align: right">
<p>Source code downloads:    
[ <a href="https://girke.bioinformatics.ucr.edu/GEN242/slides/slides_22/">Slides</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/dplyr/dplyr.Rmd">.Rmd</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/dplyr/dplyr.R">.R</a> ]</p>
</div>
<h2 id="overview">Overview</h2>
<p>Modern object classes and methods for handling <code>data.frame</code> like structures
are provided by the <code>dplyr</code> (<code>tidyr</code>) and <code>data.table</code> packages. A related example is Bioconductor’s
<code>DataTable</code> object class (“Learn the tidyverse,” n.d.). This tutorial provide a short introduction to the usage and
functionalities of the <code>dplyr</code> and related packages.</p>
<h3 id="related-documentation">Related documentation</h3>
<p>More detailed tutorials on this topic can be found here:</p>
<ul>
<li><a href="https://rdrr.io/cran/dplyr/">dplyr: A Grammar of Data Manipulation</a></li>
<li><a href="https://cran.r-project.org/web/packages/dplyr/vignettes/dplyr.html">Introduction to <code>dplyr</code></a></li>
<li><a href="http://genomicsclass.github.io/book/pages/dplyr_tutorial.html">Tutorial on <code>dplyr</code></a></li>
<li><a href="http://stat545.com/bit001_dplyr-cheatsheet.html">Cheatsheet for Joins from Jenny Bryan</a></li>
<li><a href="https://cran.r-project.org/web/packages/tibble/vignettes/tibble.html">Tibbles</a></li>
<li><a href="https://www.r-bloggers.com/intro-to-the-data-table-package/">Intro to <code>data.table</code> package</a></li>
<li><a href="https://www.r-bloggers.com/working-with-large-datasets-with-dplyr-and-data-table/">Big data with <code>dplyr</code> and <code>data.table</code></a></li>
<li><a href="https://www.r-bloggers.com/fast-data-lookups-in-r-dplyr-vs-data-table/">Fast lookups with <code>dplyr</code> and <code>data.table</code></a></li>
</ul>
<h3 id="installation">Installation</h3>
<p>The <code>dplyr</code> (<code>tidyr</code>) environment has evolved into an ecosystem of packages. To simplify
package management, one can install and load the entire collection via the
<code>tidyverse</code> package. For more details on <code>tidyverse</code> see
<a href="http://tidyverse.org/">here</a>.</p>
<pre><code class="language-r">install.packages(&quot;tidyverse&quot;)
</code></pre>
<h3 id="construct-a-tibble-tibble">Construct a <code>tibble</code> (<code>tibble</code>)</h3>
<pre><code class="language-r">library(tidyverse)
as_tibble(iris) # coerce data.frame to tibble tbl
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;fct&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre>
<h3 id="reading-and-writing-tabular-files">Reading and writing tabular files</h3>
<p>While the base R read/write utilities can be used for <code>data.frames</code>, best time
performance with the least amount of typing is achieved with the export/import
functions from the <code>readr</code> package. For very large files the <code>fread</code> function from
the <code>data.table</code> package achieves the best time performance.</p>
<h4 id="import-with-readr">Import with <code>readr</code></h4>
<p>Import functions provided by <code>readr</code> include:</p>
<ul>
<li><code>read_csv()</code>: comma separated (CSV) files</li>
<li><code>read_tsv()</code>: tab separated files</li>
<li><code>read_delim()</code>: general delimited files</li>
<li><code>read_fwf()</code>: fixed width files</li>
<li><code>read_table()</code>: tabular files where colums are separated by white-space.</li>
<li><code>read_log()</code>: web log files</li>
</ul>
<p>Create a sample tab delimited file for import</p>
<pre><code class="language-r">write_tsv(iris, &quot;iris.txt&quot;) # Creates sample file
</code></pre>
<p>Import with <code>read_tsv</code></p>
<pre><code class="language-r">iris_df &lt;- read_tsv(&quot;iris.txt&quot;) # Import with read_tbv from readr package
iris_df
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre>
<p>To import Google Sheets directly into R, see <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/rbasics/rbasics/#reading-and-writing-external-data">here</a>.</p>
<h4 id="fast-table-import-with-fread">Fast table import with <code>fread</code></h4>
<p>The <code>fread</code> function from the <code>data.table</code> package provides the best time performance for reading large
tabular files into R.</p>
<pre><code class="language-r">library(data.table)
iris_df &lt;- as_tibble(fread(&quot;iris.txt&quot;)) # Import with fread and conversion to tibble
iris_df
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre>
<p>Note: to ignore lines starting with comment signs, one can pass on to <code>fread</code> a shell
command for preprocessing the file. The following example illustrates this option.</p>
<pre><code class="language-r">fread(&quot;grep -v '^#' iris.txt&quot;) 
</code></pre>
<h4 id="export-with-readr">Export with <code>readr</code></h4>
<p>Export function provided by <code>readr</code> inlcude</p>
<ul>
<li><code>write_delim()</code>: general delimited files</li>
<li><code>write_csv()</code>: comma separated (CSV) files</li>
<li><code>write_excel_csv()</code>: excel style CSV files</li>
<li><code>write_tsv()</code>: tab separated files</li>
</ul>
<p>For instance, the <code>write_tsv</code> function writes a <code>data.frame</code> or <code>tibble</code> to a tab delimited file with much nicer
default settings than the base R <code>write.table</code> function.</p>
<pre><code class="language-r">write_tsv(iris_df, &quot;iris.txt&quot;)
</code></pre>
<h3 id="column-and-row-binds">Column and row binds</h3>
<p>The equivalents to base R’s <code>rbind</code> and <code>cbind</code> are <code>bind_rows</code> and <code>bind_cols</code>, respectively.</p>
<pre><code class="language-r">bind_cols(iris_df, iris_df)
</code></pre>
<pre><code>## # A tibble: 150 × 10
##    Sepal.Length...1 Sepal.Width...2 Petal.Length...3 Petal.Width...4 Species...5 Sepal.Length...6
##               &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt;                  &lt;dbl&gt;
##  1              5.1             3.5              1.4             0.2 setosa                   5.1
##  2              4.9             3                1.4             0.2 setosa                   4.9
##  3              4.7             3.2              1.3             0.2 setosa                   4.7
##  4              4.6             3.1              1.5             0.2 setosa                   4.6
##  5              5               3.6              1.4             0.2 setosa                   5  
##  6              5.4             3.9              1.7             0.4 setosa                   5.4
##  7              4.6             3.4              1.4             0.3 setosa                   4.6
##  8              5               3.4              1.5             0.2 setosa                   5  
##  9              4.4             2.9              1.4             0.2 setosa                   4.4
## 10              4.9             3.1              1.5             0.1 setosa                   4.9
## # … with 140 more rows, and 4 more variables: Sepal.Width...7 &lt;dbl&gt;, Petal.Length...8 &lt;dbl&gt;,
## #   Petal.Width...9 &lt;dbl&gt;, Species...10 &lt;chr&gt;
</code></pre>
<pre><code class="language-r">bind_rows(iris_df, iris_df)
</code></pre>
<pre><code>## # A tibble: 300 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 290 more rows
</code></pre>
<h3 id="extract-column-as-vector">Extract column as vector</h3>
<p>The subsetting operators <code>[[</code> and <code>$</code>can be used to extract from a <code>tibble</code> single columns as vector.</p>
<pre><code class="language-r">iris_df[[5]][1:12]
</code></pre>
<pre><code>##  [1] &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot;
## [11] &quot;setosa&quot; &quot;setosa&quot;
</code></pre>
<pre><code class="language-r">iris_df$Species[1:12]
</code></pre>
<pre><code>##  [1] &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot; &quot;setosa&quot;
## [11] &quot;setosa&quot; &quot;setosa&quot;
</code></pre>
<h3 id="important-dplyr-functions">Important <code>dplyr</code> functions</h3>
<ol>
<li><code>filter()</code> and <code>slice()</code></li>
<li><code>arrange()</code></li>
<li><code>select()</code> and <code>rename()</code></li>
<li><code>distinct()</code></li>
<li><code>mutate()</code> and <code>transmute()</code></li>
<li><code>summarise()</code></li>
<li><code>sample_n()</code> and <code>sample_frac()</code></li>
</ol>
<h3 id="slice-and-filter-functions">Slice and filter functions</h3>
<h4 id="filter-function">Filter function</h4>
<pre><code class="language-r">filter(iris_df, Sepal.Length &gt; 7.5, Species==&quot;virginica&quot;)
</code></pre>
<pre><code>## # A tibble: 6 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
## 1          7.6         3            6.6         2.1 virginica
## 2          7.7         3.8          6.7         2.2 virginica
## 3          7.7         2.6          6.9         2.3 virginica
## 4          7.7         2.8          6.7         2   virginica
## 5          7.9         3.8          6.4         2   virginica
## 6          7.7         3            6.1         2.3 virginica
</code></pre>
<h4 id="base-r-code-equivalent">Base R code equivalent</h4>
<pre><code class="language-r">iris_df[iris_df[, &quot;Sepal.Length&quot;] &gt; 7.5 &amp; iris_df[, &quot;Species&quot;]==&quot;virginica&quot;, ]
</code></pre>
<pre><code>## # A tibble: 6 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
## 1          7.6         3            6.6         2.1 virginica
## 2          7.7         3.8          6.7         2.2 virginica
## 3          7.7         2.6          6.9         2.3 virginica
## 4          7.7         2.8          6.7         2   virginica
## 5          7.9         3.8          6.4         2   virginica
## 6          7.7         3            6.1         2.3 virginica
</code></pre>
<h4 id="including-boolean-operators">Including boolean operators</h4>
<pre><code class="language-r">filter(iris_df, Sepal.Length &gt; 7.5 | Sepal.Length &lt; 5.5, Species==&quot;virginica&quot;)
</code></pre>
<pre><code>## # A tibble: 7 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
## 1          7.6         3            6.6         2.1 virginica
## 2          4.9         2.5          4.5         1.7 virginica
## 3          7.7         3.8          6.7         2.2 virginica
## 4          7.7         2.6          6.9         2.3 virginica
## 5          7.7         2.8          6.7         2   virginica
## 6          7.9         3.8          6.4         2   virginica
## 7          7.7         3            6.1         2.3 virginica
</code></pre>
<h4 id="subset-rows-by-position">Subset rows by position</h4>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">slice(iris_df, 1:2)
</code></pre>
<pre><code>## # A tibble: 2 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
## 1          5.1         3.5          1.4         0.2 setosa 
## 2          4.9         3            1.4         0.2 setosa
</code></pre>
<p>Base R code equivalent</p>
<pre><code class="language-r">iris_df[1:2,]
</code></pre>
<pre><code>## # A tibble: 2 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
## 1          5.1         3.5          1.4         0.2 setosa 
## 2          4.9         3            1.4         0.2 setosa
</code></pre>
<h4 id="subset-rows-by-names">Subset rows by names</h4>
<p>Since <code>tibbles</code> do not contain row names, row wise subsetting via the <code>[,]</code> operator cannot be used.
However, the corresponding behavior can be achieved by passing to <code>select</code> a row position index
obtained by basic R intersect utilities such as <code>match</code>.</p>
<p>Create a suitable test <code>tibble</code></p>
<pre><code class="language-r">df1 &lt;- bind_cols(tibble(ids1=paste0(&quot;g&quot;, 1:10)), as_tibble(matrix(1:40, 10, 4, dimnames=list(1:10, paste0(&quot;CA&quot;, 1:4)))))
df1
</code></pre>
<pre><code>## # A tibble: 10 × 5
##    ids1    CA1   CA2   CA3   CA4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31
##  2 g2        2    12    22    32
##  3 g3        3    13    23    33
##  4 g4        4    14    24    34
##  5 g5        5    15    25    35
##  6 g6        6    16    26    36
##  7 g7        7    17    27    37
##  8 g8        8    18    28    38
##  9 g9        9    19    29    39
## 10 g10      10    20    30    40
</code></pre>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">slice(df1, match(c(&quot;g10&quot;, &quot;g4&quot;, &quot;g4&quot;), ids1))
</code></pre>
<pre><code>## # A tibble: 3 × 5
##   ids1    CA1   CA2   CA3   CA4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g10      10    20    30    40
## 2 g4        4    14    24    34
## 3 g4        4    14    24    34
</code></pre>
<p>Base R equivalent</p>
<pre><code class="language-r">df1_old &lt;- as.data.frame(df1)
rownames(df1_old) &lt;- df1_old[,1]
df1_old[c(&quot;g10&quot;, &quot;g4&quot;, &quot;g4&quot;),]
</code></pre>
<pre><code>##      ids1 CA1 CA2 CA3 CA4
## g10   g10  10  20  30  40
## g4     g4   4  14  24  34
## g4.1   g4   4  14  24  34
</code></pre>
<h3 id="sorting-with-arrange">Sorting with <code>arrange</code></h3>
<p>Row-wise ordering based on specific columns</p>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">arrange(iris_df, Species, Sepal.Length, Sepal.Width)
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          4.3         3            1.1         0.1 setosa 
##  2          4.4         2.9          1.4         0.2 setosa 
##  3          4.4         3            1.3         0.2 setosa 
##  4          4.4         3.2          1.3         0.2 setosa 
##  5          4.5         2.3          1.3         0.3 setosa 
##  6          4.6         3.1          1.5         0.2 setosa 
##  7          4.6         3.2          1.4         0.2 setosa 
##  8          4.6         3.4          1.4         0.3 setosa 
##  9          4.6         3.6          1           0.2 setosa 
## 10          4.7         3.2          1.3         0.2 setosa 
## # … with 140 more rows
</code></pre>
<p>For ordering descendingly use <code>desc()</code> function</p>
<pre><code class="language-r">arrange(iris_df, desc(Species), Sepal.Length, Sepal.Width)
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
##  1          4.9         2.5          4.5         1.7 virginica
##  2          5.6         2.8          4.9         2   virginica
##  3          5.7         2.5          5           2   virginica
##  4          5.8         2.7          5.1         1.9 virginica
##  5          5.8         2.7          5.1         1.9 virginica
##  6          5.8         2.8          5.1         2.4 virginica
##  7          5.9         3            5.1         1.8 virginica
##  8          6           2.2          5           1.5 virginica
##  9          6           3            4.8         1.8 virginica
## 10          6.1         2.6          5.6         1.4 virginica
## # … with 140 more rows
</code></pre>
<p>Base R code equivalent</p>
<pre><code class="language-r">iris_df[order(iris_df$Species, iris_df$Sepal.Length, iris_df$Sepal.Width), ]
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          4.3         3            1.1         0.1 setosa 
##  2          4.4         2.9          1.4         0.2 setosa 
##  3          4.4         3            1.3         0.2 setosa 
##  4          4.4         3.2          1.3         0.2 setosa 
##  5          4.5         2.3          1.3         0.3 setosa 
##  6          4.6         3.1          1.5         0.2 setosa 
##  7          4.6         3.2          1.4         0.2 setosa 
##  8          4.6         3.4          1.4         0.3 setosa 
##  9          4.6         3.6          1           0.2 setosa 
## 10          4.7         3.2          1.3         0.2 setosa 
## # … with 140 more rows
</code></pre>
<pre><code class="language-r">iris_df[order(iris_df$Species, decreasing=TRUE), ] 
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
##  1          6.3         3.3          6           2.5 virginica
##  2          5.8         2.7          5.1         1.9 virginica
##  3          7.1         3            5.9         2.1 virginica
##  4          6.3         2.9          5.6         1.8 virginica
##  5          6.5         3            5.8         2.2 virginica
##  6          7.6         3            6.6         2.1 virginica
##  7          4.9         2.5          4.5         1.7 virginica
##  8          7.3         2.9          6.3         1.8 virginica
##  9          6.7         2.5          5.8         1.8 virginica
## 10          7.2         3.6          6.1         2.5 virginica
## # … with 140 more rows
</code></pre>
<h3 id="select-columns-with-select">Select columns with <code>select</code></h3>
<p>Select specific columns</p>
<pre><code class="language-r">select(iris_df, Species, Petal.Length, Sepal.Length)
</code></pre>
<pre><code>## # A tibble: 150 × 3
##    Species Petal.Length Sepal.Length
##    &lt;chr&gt;          &lt;dbl&gt;        &lt;dbl&gt;
##  1 setosa           1.4          5.1
##  2 setosa           1.4          4.9
##  3 setosa           1.3          4.7
##  4 setosa           1.5          4.6
##  5 setosa           1.4          5  
##  6 setosa           1.7          5.4
##  7 setosa           1.4          4.6
##  8 setosa           1.5          5  
##  9 setosa           1.4          4.4
## 10 setosa           1.5          4.9
## # … with 140 more rows
</code></pre>
<p>Select range of columns by name</p>
<pre><code class="language-r">select(iris_df, Sepal.Length : Petal.Width)
</code></pre>
<pre><code>## # A tibble: 150 × 4
##    Sepal.Length Sepal.Width Petal.Length Petal.Width
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
##  1          5.1         3.5          1.4         0.2
##  2          4.9         3            1.4         0.2
##  3          4.7         3.2          1.3         0.2
##  4          4.6         3.1          1.5         0.2
##  5          5           3.6          1.4         0.2
##  6          5.4         3.9          1.7         0.4
##  7          4.6         3.4          1.4         0.3
##  8          5           3.4          1.5         0.2
##  9          4.4         2.9          1.4         0.2
## 10          4.9         3.1          1.5         0.1
## # … with 140 more rows
</code></pre>
<p>Drop specific columns (here range)</p>
<pre><code class="language-r">select(iris_df, -(Sepal.Length : Petal.Width))
</code></pre>
<pre><code>## # A tibble: 150 × 1
##    Species
##    &lt;chr&gt;  
##  1 setosa 
##  2 setosa 
##  3 setosa 
##  4 setosa 
##  5 setosa 
##  6 setosa 
##  7 setosa 
##  8 setosa 
##  9 setosa 
## 10 setosa 
## # … with 140 more rows
</code></pre>
<h3 id="change-column-order-with-relocate">Change column order with <code>relocate</code></h3>
<p><code>dplyr</code> approach</p>
<p>For details and examples see <code>?relocate</code></p>
<pre><code class="language-r">relocate(iris_df, Species)
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Species Sepal.Length Sepal.Width Petal.Length Petal.Width
##    &lt;chr&gt;          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
##  1 setosa           5.1         3.5          1.4         0.2
##  2 setosa           4.9         3            1.4         0.2
##  3 setosa           4.7         3.2          1.3         0.2
##  4 setosa           4.6         3.1          1.5         0.2
##  5 setosa           5           3.6          1.4         0.2
##  6 setosa           5.4         3.9          1.7         0.4
##  7 setosa           4.6         3.4          1.4         0.3
##  8 setosa           5           3.4          1.5         0.2
##  9 setosa           4.4         2.9          1.4         0.2
## 10 setosa           4.9         3.1          1.5         0.1
## # … with 140 more rows
</code></pre>
<p>Base R code approach</p>
<pre><code class="language-r">iris[,c(5, 1:4)]
</code></pre>
<h3 id="renaming-columns-with-rename">Renaming columns with <code>rename</code></h3>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">rename(iris_df, new_col_name = Species)
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width new_col_name
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;       
##  1          5.1         3.5          1.4         0.2 setosa      
##  2          4.9         3            1.4         0.2 setosa      
##  3          4.7         3.2          1.3         0.2 setosa      
##  4          4.6         3.1          1.5         0.2 setosa      
##  5          5           3.6          1.4         0.2 setosa      
##  6          5.4         3.9          1.7         0.4 setosa      
##  7          4.6         3.4          1.4         0.3 setosa      
##  8          5           3.4          1.5         0.2 setosa      
##  9          4.4         2.9          1.4         0.2 setosa      
## 10          4.9         3.1          1.5         0.1 setosa      
## # … with 140 more rows
</code></pre>
<p>Base R code approach</p>
<pre><code class="language-r">colnames(iris_df)[colnames(iris_df)==&quot;Species&quot;] &lt;- &quot;new_col_names&quot;
</code></pre>
<h3 id="obtain-unique-rows-with-distinct">Obtain unique rows with <code>distinct</code></h3>
<p><code>dplyr</code> approach</p>
<pre><code class="language-r">distinct(iris_df, Species, .keep_all=TRUE)
</code></pre>
<pre><code>## # A tibble: 3 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;     
## 1          5.1         3.5          1.4         0.2 setosa    
## 2          7           3.2          4.7         1.4 versicolor
## 3          6.3         3.3          6           2.5 virginica
</code></pre>
<p>Base R code approach</p>
<pre><code class="language-r">iris_df[!duplicated(iris_df$Species),]
</code></pre>
<pre><code>## # A tibble: 3 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;     
## 1          5.1         3.5          1.4         0.2 setosa    
## 2          7           3.2          4.7         1.4 versicolor
## 3          6.3         3.3          6           2.5 virginica
</code></pre>
<h3 id="add-columns">Add columns</h3>
<h4 id="mutate"><code>mutate</code></h4>
<p>The <code>mutate</code> function allows to append columns to existing ones.</p>
<pre><code class="language-r">mutate(iris_df, Ratio = Sepal.Length / Sepal.Width, Sum = Sepal.Length + Sepal.Width)
</code></pre>
<pre><code>## # A tibble: 150 × 7
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Ratio   Sum
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;
##  1          5.1         3.5          1.4         0.2 setosa   1.46   8.6
##  2          4.9         3            1.4         0.2 setosa   1.63   7.9
##  3          4.7         3.2          1.3         0.2 setosa   1.47   7.9
##  4          4.6         3.1          1.5         0.2 setosa   1.48   7.7
##  5          5           3.6          1.4         0.2 setosa   1.39   8.6
##  6          5.4         3.9          1.7         0.4 setosa   1.38   9.3
##  7          4.6         3.4          1.4         0.3 setosa   1.35   8  
##  8          5           3.4          1.5         0.2 setosa   1.47   8.4
##  9          4.4         2.9          1.4         0.2 setosa   1.52   7.3
## 10          4.9         3.1          1.5         0.1 setosa   1.58   8  
## # … with 140 more rows
</code></pre>
<h4 id="transmute"><code>transmute</code></h4>
<p>The <code>transmute</code> function does the same as <code>mutate</code> but drops existing columns</p>
<pre><code class="language-r">transmute(iris_df, Ratio = Sepal.Length / Sepal.Width, Sum = Sepal.Length + Sepal.Width)
</code></pre>
<pre><code>## # A tibble: 150 × 2
##    Ratio   Sum
##    &lt;dbl&gt; &lt;dbl&gt;
##  1  1.46   8.6
##  2  1.63   7.9
##  3  1.47   7.9
##  4  1.48   7.7
##  5  1.39   8.6
##  6  1.38   9.3
##  7  1.35   8  
##  8  1.47   8.4
##  9  1.52   7.3
## 10  1.58   8  
## # … with 140 more rows
</code></pre>
<h4 id="bind_cols"><code>bind_cols</code></h4>
<p>The <code>bind_cols</code> function is the equivalent of <code>cbind</code> in base R. To add rows, use the corresponding
<code>bind_rows</code> function.</p>
<pre><code class="language-r">bind_cols(iris_df, iris_df)
</code></pre>
<pre><code>## # A tibble: 150 × 10
##    Sepal.Length...1 Sepal.Width...2 Petal.Length...3 Petal.Width...4 Species...5 Sepal.Length...6
##               &lt;dbl&gt;           &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt; &lt;chr&gt;                  &lt;dbl&gt;
##  1              5.1             3.5              1.4             0.2 setosa                   5.1
##  2              4.9             3                1.4             0.2 setosa                   4.9
##  3              4.7             3.2              1.3             0.2 setosa                   4.7
##  4              4.6             3.1              1.5             0.2 setosa                   4.6
##  5              5               3.6              1.4             0.2 setosa                   5  
##  6              5.4             3.9              1.7             0.4 setosa                   5.4
##  7              4.6             3.4              1.4             0.3 setosa                   4.6
##  8              5               3.4              1.5             0.2 setosa                   5  
##  9              4.4             2.9              1.4             0.2 setosa                   4.4
## 10              4.9             3.1              1.5             0.1 setosa                   4.9
## # … with 140 more rows, and 4 more variables: Sepal.Width...7 &lt;dbl&gt;, Petal.Length...8 &lt;dbl&gt;,
## #   Petal.Width...9 &lt;dbl&gt;, Species...10 &lt;chr&gt;
</code></pre>
<h3 id="summarize-data">Summarize data</h3>
<p>Summary calculation on single column</p>
<pre><code class="language-r">summarize(iris_df, mean(Petal.Length))
</code></pre>
<pre><code>## # A tibble: 1 × 1
##   `mean(Petal.Length)`
##                  &lt;dbl&gt;
## 1                 3.76
</code></pre>
<p>Summary calculation on many columns</p>
<pre><code class="language-r">summarize_all(iris_df[,1:4], mean)
</code></pre>
<pre><code>## # A tibble: 1 × 4
##   Sepal.Length Sepal.Width Petal.Length Petal.Width
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
## 1         5.84        3.06         3.76        1.20
</code></pre>
<p>Summarize by grouping column</p>
<pre><code class="language-r">summarize(group_by(iris_df, Species), mean(Petal.Length))
</code></pre>
<pre><code>## # A tibble: 3 × 2
##   Species    `mean(Petal.Length)`
##   &lt;chr&gt;                     &lt;dbl&gt;
## 1 setosa                     1.46
## 2 versicolor                 4.26
## 3 virginica                  5.55
</code></pre>
<p>Aggregate summaries</p>
<pre><code class="language-r">summarize_all(group_by(iris_df, Species), mean) 
</code></pre>
<pre><code>## # A tibble: 3 × 5
##   Species    Sepal.Length Sepal.Width Petal.Length Petal.Width
##   &lt;chr&gt;             &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
## 1 setosa             5.01        3.43         1.46       0.246
## 2 versicolor         5.94        2.77         4.26       1.33 
## 3 virginica          6.59        2.97         5.55       2.03
</code></pre>
<p>Note: <code>group_by</code> does the looping for the user similar to <code>aggregate</code> or <code>tapply</code>.</p>
<h3 id="merging-tibbles">Merging tibbles</h3>
<p>The <code>dplyr</code> package provides several join functions for merging <code>tibbles</code> by a common key column
similar to the <code>merge</code> function in base R. These <code>*_join</code> functions include:</p>
<ul>
<li><code>inner_join()</code>: returns join only for rows matching among both <code>tibbles</code></li>
<li><code>full_join()</code>: returns join for all (matching and non-matching) rows of two <code>tibbles</code></li>
<li><code>left_join()</code>: returns join for all rows in first <code>tibble</code></li>
<li><code>right_join()</code>: returns join for all rows in second <code>tibble</code></li>
<li><code>anti_join()</code>: returns for first <code>tibble</code> only those rows that have no match in the second one</li>
</ul>
<p>Sample <code>tibbles</code> to illustrate <code>*.join</code> functions.</p>
<pre><code class="language-r">df1 &lt;- bind_cols(tibble(ids1=paste0(&quot;g&quot;, 1:10)), as_tibble(matrix(1:40, 10, 4, dimnames=list(1:10, paste0(&quot;CA&quot;, 1:4)))))
df1
</code></pre>
<pre><code>## # A tibble: 10 × 5
##    ids1    CA1   CA2   CA3   CA4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31
##  2 g2        2    12    22    32
##  3 g3        3    13    23    33
##  4 g4        4    14    24    34
##  5 g5        5    15    25    35
##  6 g6        6    16    26    36
##  7 g7        7    17    27    37
##  8 g8        8    18    28    38
##  9 g9        9    19    29    39
## 10 g10      10    20    30    40
</code></pre>
<pre><code class="language-r">df2 &lt;- bind_cols(tibble(ids2=paste0(&quot;g&quot;, c(2,5,11,12))), as_tibble(matrix(1:16, 4, 4, dimnames=list(1:4, paste0(&quot;CB&quot;, 1:4)))))
df2
</code></pre>
<pre><code>## # A tibble: 4 × 5
##   ids2    CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g2        1     5     9    13
## 2 g5        2     6    10    14
## 3 g11       3     7    11    15
## 4 g12       4     8    12    16
</code></pre>
<h4 id="inner-join">Inner join</h4>
<pre><code class="language-r">inner_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 2 × 9
##   ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g2        2    12    22    32     1     5     9    13
## 2 g5        5    15    25    35     2     6    10    14
</code></pre>
<h4 id="left-join">Left join</h4>
<pre><code class="language-r">left_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 10 × 9
##    ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31    NA    NA    NA    NA
##  2 g2        2    12    22    32     1     5     9    13
##  3 g3        3    13    23    33    NA    NA    NA    NA
##  4 g4        4    14    24    34    NA    NA    NA    NA
##  5 g5        5    15    25    35     2     6    10    14
##  6 g6        6    16    26    36    NA    NA    NA    NA
##  7 g7        7    17    27    37    NA    NA    NA    NA
##  8 g8        8    18    28    38    NA    NA    NA    NA
##  9 g9        9    19    29    39    NA    NA    NA    NA
## 10 g10      10    20    30    40    NA    NA    NA    NA
</code></pre>
<h4 id="right-join">Right join</h4>
<pre><code class="language-r">right_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 4 × 9
##   ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g2        2    12    22    32     1     5     9    13
## 2 g5        5    15    25    35     2     6    10    14
## 3 g11      NA    NA    NA    NA     3     7    11    15
## 4 g12      NA    NA    NA    NA     4     8    12    16
</code></pre>
<h4 id="full-join">Full join</h4>
<pre><code class="language-r">full_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 12 × 9
##    ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31    NA    NA    NA    NA
##  2 g2        2    12    22    32     1     5     9    13
##  3 g3        3    13    23    33    NA    NA    NA    NA
##  4 g4        4    14    24    34    NA    NA    NA    NA
##  5 g5        5    15    25    35     2     6    10    14
##  6 g6        6    16    26    36    NA    NA    NA    NA
##  7 g7        7    17    27    37    NA    NA    NA    NA
##  8 g8        8    18    28    38    NA    NA    NA    NA
##  9 g9        9    19    29    39    NA    NA    NA    NA
## 10 g10      10    20    30    40    NA    NA    NA    NA
## 11 g11      NA    NA    NA    NA     3     7    11    15
## 12 g12      NA    NA    NA    NA     4     8    12    16
</code></pre>
<h4 id="anti-join">Anti join</h4>
<pre><code class="language-r">anti_join(df1, df2, by=c(&quot;ids1&quot;=&quot;ids2&quot;))
</code></pre>
<pre><code>## # A tibble: 8 × 5
##   ids1    CA1   CA2   CA3   CA4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g1        1    11    21    31
## 2 g3        3    13    23    33
## 3 g4        4    14    24    34
## 4 g6        6    16    26    36
## 5 g7        7    17    27    37
## 6 g8        8    18    28    38
## 7 g9        9    19    29    39
## 8 g10      10    20    30    40
</code></pre>
<p>For additional join options users want to cosult the <code>*_join</code> help pages.</p>
<h3 id="chaining">Chaining</h3>
<p>To simplify chaining of serveral operations, <code>dplyr</code> uses the <code>%&gt;%</code>
operator from <code>magrittr</code>, where <code>x %&gt;% f(y)</code> turns into <code>f(x, y)</code>. This way one can pipe
together multiple operations by writing them from left-to-right or
top-to-bottom. This makes for easy to type and readable code. Since R-4.1.0, a native
pipe <code>|&gt;</code> operator is available that works largely the same as <code>%&gt;%</code>.</p>
<h4 id="example-1">Example 1</h4>
<p>Series of data manipulations and export</p>
<pre><code class="language-r">read_tsv(&quot;iris.txt&quot;) %&gt;% # Import with read_tbv from readr package
    as_tibble() %&gt;% # Declare to use tibble
    select(Sepal.Length:Species) %&gt;% # Select columns
    filter(Species==&quot;setosa&quot;) %&gt;% # Filter rows by some value
    arrange(Sepal.Length) %&gt;% # Sort by some column
    mutate(Subtract=Petal.Length - Petal.Width) # Calculate and append
</code></pre>
<pre><code>## # A tibble: 50 × 6
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Subtract
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;
##  1          4.3         3            1.1         0.1 setosa       1  
##  2          4.4         2.9          1.4         0.2 setosa       1.2
##  3          4.4         3            1.3         0.2 setosa       1.1
##  4          4.4         3.2          1.3         0.2 setosa       1.1
##  5          4.5         2.3          1.3         0.3 setosa       1  
##  6          4.6         3.1          1.5         0.2 setosa       1.3
##  7          4.6         3.4          1.4         0.3 setosa       1.1
##  8          4.6         3.6          1           0.2 setosa       0.8
##  9          4.6         3.2          1.4         0.2 setosa       1.2
## 10          4.7         3.2          1.3         0.2 setosa       1.1
## # … with 40 more rows
</code></pre>
<pre><code class="language-r">    # write_tsv(&quot;iris.txt&quot;) # Export to file, omitted here to show result 
</code></pre>
<h4 id="example-2">Example 2</h4>
<p>Series of summary calculations for grouped data (<code>group_by</code>)</p>
<pre><code class="language-r">iris_df %&gt;% # Declare tibble to use 
    group_by(Species) %&gt;% # Group by species
    summarize(Mean_Sepal.Length=mean(Sepal.Length), 
              Max_Sepal.Length=max(Sepal.Length),
              Min_Sepal.Length=min(Sepal.Length),
              SD_Sepal.Length=sd(Sepal.Length),
              Total=n()) 
</code></pre>
<pre><code>## # A tibble: 3 × 6
##   Species    Mean_Sepal.Length Max_Sepal.Length Min_Sepal.Length SD_Sepal.Length Total
##   &lt;chr&gt;                  &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt; &lt;int&gt;
## 1 setosa                  5.01              5.8              4.3           0.352    50
## 2 versicolor              5.94              7                4.9           0.516    50
## 3 virginica               6.59              7.9              4.9           0.636    50
</code></pre>
<h4 id="example-3">Example 3</h4>
<p>Combining <code>dplyr</code> chaining with <code>ggplot</code></p>
<pre><code class="language-r">iris_df %&gt;% 
    group_by(Species) %&gt;% 
    summarize_all(mean) %&gt;% 
    reshape2::melt(id.vars=c(&quot;Species&quot;), variable.name = &quot;Samples&quot;, value.name=&quot;Values&quot;) %&gt;%
    ggplot(aes(Samples, Values, fill = Species)) + 
            geom_bar(position=&quot;dodge&quot;, stat=&quot;identity&quot;)
</code></pre>
<img src="../../../en/tutorials/dplyr/dplyr_files/figure-html/plyr_chaining3-1.png" width="672" />
<h2 id="sqlite-databases">SQLite Databases</h2>
<p><code>SQLite</code> is a lightweight relational database solution. The <code>RSQLite</code> package provides an easy to use interface to create, manage and query <code>SQLite</code> databases directly from R. Basic instructions
for using <code>SQLite</code> from the command-line are available <a href="https://www.sqlite.org/cli.html">here</a>. A short introduction to <code>RSQLite</code> is available <a href="https://github.com/rstats-db/RSQLite/blob/master/vignettes/RSQLite.Rmd">here</a>.</p>
<h2 id="loading-data-into-sqlite-databases">Loading data into SQLite databases</h2>
<p>The following loads two <code>data.frames</code> derived from the <code>iris</code> data set (here <code>mydf1</code> and <code>mydf2</code>)
into an SQLite database (here <code>test.db</code>).</p>
<pre><code class="language-r">library(RSQLite)
unlink(&quot;test.db&quot;) # Delete any existing test.db
mydb &lt;- dbConnect(SQLite(), &quot;test.db&quot;) # Creates database file test.db
mydf1 &lt;- data.frame(ids=paste0(&quot;id&quot;, seq_along(iris[,1])), iris)
mydf2 &lt;- mydf1[sample(seq_along(mydf1[,1]), 10),]
dbWriteTable(mydb, &quot;mydf1&quot;, mydf1)
dbWriteTable(mydb, &quot;mydf2&quot;, mydf2)
</code></pre>
<h3 id="list-names-of-tables-in-database">List names of tables in database</h3>
<pre><code class="language-r">dbListTables(mydb)
</code></pre>
<pre><code>## [1] &quot;mydf1&quot; &quot;mydf2&quot;
</code></pre>
<h3 id="import-table-into-dataframe">Import table into <code>data.frame</code></h3>
<pre><code class="language-r">dbGetQuery(mydb, 'SELECT * FROM mydf2')
</code></pre>
<pre><code>##      ids Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
## 1   id95          5.6         2.7          4.2         1.3 versicolor
## 2   id79          6.0         2.9          4.5         1.5 versicolor
## 3  id124          6.3         2.7          4.9         1.8  virginica
## 4   id33          5.2         4.1          1.5         0.1     setosa
## 5   id48          4.6         3.2          1.4         0.2     setosa
## 6  id122          5.6         2.8          4.9         2.0  virginica
## 7  id104          6.3         2.9          5.6         1.8  virginica
## 8   id56          5.7         2.8          4.5         1.3 versicolor
## 9  id105          6.5         3.0          5.8         2.2  virginica
## 10  id43          4.4         3.2          1.3         0.2     setosa
</code></pre>
<h3 id="query-database">Query database</h3>
<pre><code class="language-r">dbGetQuery(mydb, 'SELECT * FROM mydf1 WHERE &quot;Sepal.Length&quot; &lt; 4.6')
</code></pre>
<pre><code>##    ids Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1  id9          4.4         2.9          1.4         0.2  setosa
## 2 id14          4.3         3.0          1.1         0.1  setosa
## 3 id39          4.4         3.0          1.3         0.2  setosa
## 4 id42          4.5         2.3          1.3         0.3  setosa
## 5 id43          4.4         3.2          1.3         0.2  setosa
</code></pre>
<h3 id="join-tables">Join tables</h3>
<p>The two tables can be joined on the shared <code>ids</code> column as follows.</p>
<pre><code class="language-r">dbGetQuery(mydb, 'SELECT * FROM mydf1, mydf2 WHERE mydf1.ids = mydf2.ids')
</code></pre>
<pre><code>##      ids Sepal.Length Sepal.Width Petal.Length Petal.Width    Species   ids Sepal.Length
## 1   id33          5.2         4.1          1.5         0.1     setosa  id33          5.2
## 2   id43          4.4         3.2          1.3         0.2     setosa  id43          4.4
## 3   id48          4.6         3.2          1.4         0.2     setosa  id48          4.6
## 4   id56          5.7         2.8          4.5         1.3 versicolor  id56          5.7
## 5   id79          6.0         2.9          4.5         1.5 versicolor  id79          6.0
## 6   id95          5.6         2.7          4.2         1.3 versicolor  id95          5.6
## 7  id104          6.3         2.9          5.6         1.8  virginica id104          6.3
## 8  id105          6.5         3.0          5.8         2.2  virginica id105          6.5
## 9  id122          5.6         2.8          4.9         2.0  virginica id122          5.6
## 10 id124          6.3         2.7          4.9         1.8  virginica id124          6.3
##    Sepal.Width Petal.Length Petal.Width    Species
## 1          4.1          1.5         0.1     setosa
## 2          3.2          1.3         0.2     setosa
## 3          3.2          1.4         0.2     setosa
## 4          2.8          4.5         1.3 versicolor
## 5          2.9          4.5         1.5 versicolor
## 6          2.7          4.2         1.3 versicolor
## 7          2.9          5.6         1.8  virginica
## 8          3.0          5.8         2.2  virginica
## 9          2.8          4.9         2.0  virginica
## 10         2.7          4.9         1.8  virginica
</code></pre>
<h2 id="session-info">Session Info</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.2.0 (2022-04-22)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 11 (bullseye)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] RSQLite_2.2.14    data.table_1.14.2 forcats_0.5.1     stringr_1.4.0     dplyr_1.0.9      
##  [6] purrr_0.3.4       readr_2.1.2       tidyr_1.2.0       tibble_3.1.7      tidyverse_1.3.1  
## [11] ggplot2_3.3.6     limma_3.52.0      BiocStyle_2.24.0 
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.8.3        lubridate_1.8.0     assertthat_0.2.1    digest_0.6.29      
##  [5] utf8_1.2.2          plyr_1.8.7          R6_2.5.1            cellranger_1.1.0   
##  [9] backports_1.4.1     reprex_2.0.1        evaluate_0.15       highr_0.9          
## [13] httr_1.4.3          pillar_1.7.0        rlang_1.0.2         readxl_1.4.0       
## [17] rstudioapi_0.13     blob_1.2.3          jquerylib_0.1.4     rmarkdown_2.14     
## [21] labeling_0.4.2      bit_4.0.4           munsell_0.5.0       broom_0.8.0        
## [25] compiler_4.2.0      modelr_0.1.8        xfun_0.30           pkgconfig_2.0.3    
## [29] htmltools_0.5.2     tidyselect_1.1.2    codetools_0.2-18    fansi_1.0.3        
## [33] crayon_1.5.1        tzdb_0.3.0          dbplyr_2.1.1        withr_2.5.0        
## [37] grid_4.2.0          jsonlite_1.8.0      gtable_0.3.0        lifecycle_1.0.1    
## [41] DBI_1.1.2           magrittr_2.0.3      scales_1.2.0        cachem_1.0.6       
## [45] cli_3.3.0           stringi_1.7.6       vroom_1.5.7         farver_2.1.0       
## [49] reshape2_1.4.4      fs_1.5.2            xml2_1.3.3          bslib_0.3.1        
## [53] ellipsis_0.3.2      generics_0.1.2      vctrs_0.4.1         tools_4.2.0        
## [57] bit64_4.0.5         glue_1.6.2          hms_1.1.1           parallel_4.2.0     
## [61] fastmap_1.1.0       yaml_2.3.5          colorspace_2.0-3    BiocManager_1.30.17
## [65] rvest_1.0.2         memoise_2.0.1       knitr_1.39          haven_2.5.0        
## [69] sass_0.4.1
</code></pre>
<h2 id="references">References</h2>
<div id="refs" class="references hanging-indent">
<div id="ref-noauthor_undated-kc">
<p>“Learn the tidyverse.” n.d. <a href="https://www.tidyverse.org/learn/">https://www.tidyverse.org/learn/</a>. <a href="https://www.tidyverse.org/learn/">https://www.tidyverse.org/learn/</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2023-05-30: <a  href="https://github.com/tgirke/GEN242//commit/a9e9175fc8c4c51dd0022adcdf6d5d328fe07d7a">some edits (a9e9175fc)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2024 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
