#############
#### STAR ###
#############

### Read mapping with `STAR`

```{r intialize_sal_for_testing}
library(systemPipeR)
# sal_test <- SPRproject(logs.dir= ".SPRproject_test") # use this line when .SPRproject_test doesn't exist yet
sal_test <- SPRproject(overwrite = TRUE, logs.dir= ".SPRproject_test")
```

```{r load_SPR, message=FALSE, eval=FALSE, spr=TRUE}
appendStep(sal_test) <- LineWise(code = {
                library(systemPipeR)
                }, step_name = "load_SPR")
```

## Read preprocessing

### Preprocessing with `preprocessReads` function

```{r preprocessing, message=FALSE, eval=FALSE, spr=TRUE}
appendStep(sal_test) <- SYSargsList(
    step_name = "preprocessing",
    targets = "targetsPE.txt", dir = TRUE,
    wf_file = "preprocessReads/preprocessReads-pe.cwl",
    input_file = "preprocessReads/preprocessReads-pe.yml",
    dir_path = system.file("extdata/cwl", package = "systemPipeR"),
    inputvars = c(
        FileName1 = "_FASTQ_PATH1_",
        FileName2 = "_FASTQ_PATH2_",
        SampleName = "_SampleName_"
    ),
    dependency = c("load_SPR"))
```

## Alignments with `STAR`

### `STAR` Indexing

```{r hisat2_index, eval=FALSE, spr=TRUE}
appendStep(sal_test) <- SYSargsList(
    step_name = "star_index", 
    dir = FALSE, 
    targets=NULL, 
    wf_file = "star/star-index.cwl", 
    input_file="star/star-index.yml",
    dir_path="param/cwl", 
    dependency = "load_SPR"
)
```

### `STAR` mapping

```{r hisat2_mapping, eval=FALSE, spr=TRUE}
appendStep(sal_test) <- SYSargsList(
    step_name = "star_mapping",
    dir = TRUE, 
    targets ="preprocessing", 
    wf_file = "star-mapping-pe.cwl",
    input_file = "star-mapping-pe.yml",
    dir_path = "param/star_test",
    inputvars = c(preprocessReads_1 = "_FASTQ_PATH1_", preprocessReads_2 = "_FASTQ_PATH2_", 
                  SampleName = "_SampleName_"),
    rm_targets_col = c("FileName1", "FileName2"), 
    dependency = c("preprocessing", "star_index")
)
```

```{r some_useful_checks, eval=FALSE}
## Return command-line calls for STAR
cmdlist(sal_test, step="star_mapping", targets=1)

## BAM outpaths required for read counting below
outpaths <- getColumn(sal_test, step = "star_mapping", "outfiles", column = "Aligned_toTranscriptome_out_bam")
file.exists(outpaths) # Will not return TRUE until STAR completed sucessfully
```

```{r run_step_wise_for_testing, eval=FALSE}
## To run sal_test stepwise, make sure you have constructed your 
## sal_test object step-by-step starting from an empty sal_test
## as shown above under chunk: intialize_sal_for_testing 
sal_test <- runWF(sal_test, steps=c(1)) # increment step number one by one just for checking
sal_test
outpaths <- getColumn(sal_test, step = "star_mapping", "outfiles", column = "Aligned_toTranscriptome_out_bam")
outpaths
file.exists(outpaths) # Will not return TRUE until STAR completed sucessfully
```

```{r alternative_for_wf_setup, eval=FALSE}
## The following can be used for setting up things initial testing
starPE <- loadWorkflow(targets = "targetsPE.txt", wf_file = "star-mapping-pe.cwl", 
                       input_file = "star-mapping-pe.yml", dir_path = "./param/star_test")
starPE <- renderWF(starPE, inputvars = c(FileName1 = "_FASTQ_PATH1_", FileName2 = "_FASTQ_PATH2_", 
                                         SampleName = "_SampleName_"))
cmdlist(starPE)
runCommandline(starPE, make_bam = FALSE)
```
