<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>Building R Packages | GEN242</title><meta property="og:title" content="Building R Packages" />
<meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ Slides ] [ .Rmd ] [ .R ]
 Overview Motivation for building R packages  Organization  Consolidate functions with related utilties in single place Interdepencies among less complex functions make coding more efficient Minimizes duplications   Documentation  Help page infrastructure improves documentation of functions Big picture of utilties provided by package vignettes (manuals)   Sharability  Package can be easier shared with colleagues and public Increases code accessibilty for other users    Package development environments This page introduces two approaches for building R packages:" />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/rpackages/rpackages/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2022-03-22T12:54:22-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="Building R Packages">
<meta itemprop="description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ Slides ] [ .Rmd ] [ .R ]
 Overview Motivation for building R packages  Organization  Consolidate functions with related utilties in single place Interdepencies among less complex functions make coding more efficient Minimizes duplications   Documentation  Help page infrastructure improves documentation of functions Big picture of utilties provided by package vignettes (manuals)   Sharability  Package can be easier shared with colleagues and public Increases code accessibilty for other users    Package development environments This page introduces two approaches for building R packages:">
<meta itemprop="dateModified" content="2022-03-22T12:54:22-07:00" />
<meta itemprop="wordCount" content="1972">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Building R Packages"/>
<meta name="twitter:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ Slides ] [ .Rmd ] [ .R ]
 Overview Motivation for building R packages  Organization  Consolidate functions with related utilties in single place Interdepencies among less complex functions make coding more efficient Minimizes duplications   Documentation  Help page infrastructure improves documentation of functions Big picture of utilties provided by package vignettes (manuals)   Sharability  Package can be easier shared with colleagues and public Increases code accessibilty for other users    Package development environments This page introduces two approaches for building R packages:"/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S22)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T19)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2022/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.7c6678c4955a8a1b810f7dbec715ab66.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.7c6678c4955a8a1b810f7dbec715ab66.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../tutorials/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Tutorials</a>
  </li>
  <ul>
    <li class="collapse show" id="tutorials">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsgithubgithub" href="../../../tutorials/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialslinuxlinux" href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrbasicsrbasics" href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrprogrammingrprogramming" href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrparallelrparallel" href="../../../tutorials/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrsequencesrsequences" href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempiper" href="../../../tutorials/systempiper/systempiper/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseq" href="../../../tutorials/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspchipseqspchipseq" href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsscrnaseqscrnaseq" href="../../../tutorials/scrnaseq/scrnaseq/">scRNA-Seq Embedding Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrmarkdownrmarkdown" href="../../../tutorials/rmarkdown/rmarkdown/">R Markdown Tutorial</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrfearfea" href="../../../tutorials/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrclusteringrclustering" href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrgraphicsrgraphics" href="../../../tutorials/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappsshinyapps" href="../../../tutorials/shinyapps/shinyapps/">Shiny Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-tutorialsrpackagesrpackages" href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsdplyrdplyr" href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialscmdtocwlcmdtocwl" href="../../../tutorials/cmdtocwl/cmdtocwl/">Automate Creation of CWL Instructions</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempiperrnaseqsystempipernaseq" href="../../../tutorials/systempiper/rnaseq/systempipernaseq/">RNA-Seq Workflow Template</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rpackages/rpackages.Rmd" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/tutorials/rpackages/rpackages.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/tutorials/rpackages/rpackages.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=Building%20R%20Packages" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#overview">Overview</a>
      <ul>
        <li><a href="#motivation-for-building-r-packages">Motivation for building R packages</a></li>
        <li><a href="#package-development-environments">Package development environments</a></li>
      </ul>
    </li>
    <li><a href="#1-r-base-approach">1. R Base Approach</a>
      <ul>
        <li><a href="#11-create-package-skeleton">1.1 Create package skeleton</a></li>
        <li><a href="#12-build-package">1.2 Build package</a></li>
        <li><a href="#13-check-package">1.3 Check package</a></li>
        <li><a href="#14-install-package">1.4 Install package</a></li>
        <li><a href="#15-maintain-package">1.5 Maintain package</a></li>
        <li><a href="#16-submit-package-to-public-repository">1.6 Submit package to public repository</a></li>
      </ul>
    </li>
    <li><a href="#2-r-devtools-approach">2. R <code>devtools</code> Approach</a></li>
    <li><a href="#workflow-for-building-r-packages">Workflow for building R packages</a>
      <ul>
        <li><a href="#21-create-package-skeleton">2.1 Create package skeleton</a></li>
        <li><a href="#22-add-r-functions">2.2 Add R functions</a></li>
        <li><a href="#23-auto-generate-roxygen-comment-lines">2.3 Auto-generate roxygen comment lines</a></li>
        <li><a href="#24-autogenerate-help-files">2.4 Autogenerate help files</a></li>
        <li><a href="#25-add-a-vignette">2.5 Add a vignette</a></li>
        <li><a href="#26-check-install-and-build-package">2.6 Check, install and build package</a></li>
        <li><a href="#27-using-the-new-package">2.7 Using the new package</a></li>
        <li><a href="#28-share-package-on-github">2.8 Share package on GitHub</a></li>
      </ul>
    </li>
    <li><a href="#session-info">Session Info</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../tutorials/">Tutorials</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>Building R Packages</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 10 minute read</p>
           
	<!--
- Compile from command-line
Rscript -e "rmarkdown::render('rpackages.Rmd', c('html_document'), clean=F); knitr::knit('rpackages.Rmd', tangle=TRUE)"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div style="text-align: right">
<p>Source code downloads:    
[ <a href="https://girke.bioinformatics.ucr.edu/GEN242/slides/slides_19/">Slides</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rpackages/rpackages.Rmd">.Rmd</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rpackages/rpackages.R">.R</a> ]</p>
</div>
<h2 id="overview">Overview</h2>
<h3 id="motivation-for-building-r-packages">Motivation for building R packages</h3>
<ol>
<li>Organization
<ul>
<li>Consolidate functions with related utilties in single place</li>
<li>Interdepencies among less complex functions make coding more efficient</li>
<li>Minimizes duplications</li>
</ul>
</li>
<li>Documentation
<ul>
<li>Help page infrastructure improves documentation of functions</li>
<li>Big picture of utilties provided by package vignettes (manuals)</li>
</ul>
</li>
<li>Sharability
<ul>
<li>Package can be easier shared with colleagues and public</li>
<li>Increases code accessibilty for other users</li>
</ul>
</li>
</ol>
<h3 id="package-development-environments">Package development environments</h3>
<p>This page introduces two approaches for building R packages:</p>
<ol>
<li><code>R Base</code> and related functionalities</li>
<li><code>devtools</code> and related packages (<em>e.g.</em> <code>usethis</code>, <code>roxygen2</code> and <code>sinew</code>)</li>
</ol>
<p>The sample code provided below creates for each of the two methods a simple test package
that can be installed and loaded on a user’s system. The instructions for the
second appoach are more detailed since it is likely to provide the most
practical solution for newer users of R.</p>
<h2 id="1-r-base-approach">1. R Base Approach</h2>
<p>R packages can be built with the <code>package.skeleton</code> function. The most
comprehensive documentation on package development is provided by the <a href="https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Package-Dependencies">Writing
R
Extensions</a>
page on CRAN. The basic workflow example below will create a directory named
<code>mypackage</code> containing the skeleton of the package for all functions, methods
and classes defined in the R script(s) passed on to the <code>code_files</code> argument.
The basic structure of the package directory is described
<a href="http://cran.fhcrc.org/doc/manuals/R-exts.html#Package-structure">here</a>.
The package directory will also contain a file named <code>Read-and-delete-me</code> with
instructions for completing the package:</p>
<h3 id="11-create-package-skeleton">1.1 Create package skeleton</h3>
<pre><code class="language-r">## Download R script (here pkg_build_fct.R) containing two sample functions
download.file(&quot;https://raw.githubusercontent.com/tgirke/GEN242/main/content/en/tutorials/rpackages/helper_functions/pkg_build_fct.R&quot;, &quot;pkg_build_fct.R&quot;)
## Build package skeleton based on functions in pkg_build_fct.R
package.skeleton(name=&quot;mypackage&quot;, code_files=c(&quot;pkg_build_fct.R&quot;))
</code></pre>
<p>The given example will create a directory named <code>mypackage</code> containing the
skeleton of the package for all functions, methods and classes defined in the R
script(s) passed on to the <code>code_files</code> argument. The basic structure of the
package directory is described <a href="http://cran.fhcrc.org/doc/manuals/R-exts.html#Package-structure">here</a>.
The package directory will also contain a file named ‘Read-and-delete-me’ with the following
instructions for completing the package:</p>
<ul>
<li>Edit the help file skeletons in man, possibly combining help files for multiple functions.</li>
<li>Edit the exports in NAMESPACE, and add necessary imports.</li>
<li>Put any C/C++/Fortran code in src.</li>
<li>If you have compiled code, add a <code>useDynLib()</code> directive to <code>NAMESPACE</code>.</li>
<li>Run R CMD build to build the package tarball.</li>
<li>Run R CMD check to check the package tarball.</li>
<li>Read Writing R Extensions for more information.</li>
</ul>
<h3 id="12-build-package">1.2 Build package</h3>
<p>Once a package skeleton is available one can build the package from the
command-line (Linux/OS X), or from within R by executing the command-line calls
with R’s <code>system(&quot;...&quot;)</code> command. For instance, the command-line call <code>R CMD build ...</code>
can be run from within R with <code>system(&quot;R CMD build ...&quot;)</code>. The following examples
refer to the R console.</p>
<pre><code class="language-r">system(&quot;R CMD build mypackage&quot;)
</code></pre>
<h3 id="13-check-package">1.3 Check package</h3>
<p>This will create a tarball of the package with its version number encoded in
the file name (here <code>mypackage_1.0.tar.gz</code>). Subsequently, the package tarball
needs to be checked for errors with <code>R CMD check</code>.</p>
<pre><code class="language-r">system(&quot;R CMD check mypackage_1.0.tar.gz&quot;)
</code></pre>
<p>All issues in a package’s source code and documentation should be addressed
until <code>R CMD check</code> returns no error or warning messages anymore.</p>
<h3 id="14-install-package">1.4 Install package</h3>
<p>Install package from source on Linux or OS X systems.</p>
<pre><code class="language-r">install.packages(&quot;mypackage_1.0.tar.gz&quot;, repos=NULL) # On OS X include the argument type=&quot;source&quot;
</code></pre>
<p>Windows requires a zip archive for installing R packages, which can be most
conveniently created from the command-line (Linux/OS X) by installing the
package in a local directory (here <code>tempdir</code>) and then creating a zip archive
from the installed package directory:</p>
<pre><code class="language-r">$ mkdir tempdir
$ R CMD INSTALL -l tempdir mypackage_1.0.tar.gz
$ cd tempdir
$ zip -r mypackage mypackage
## The resulting mypackage.zip archive can be installed from R under Windows like this:
install.packages(&quot;mypackage.zip&quot;, repos=NULL)
</code></pre>
<p>This procedure only works for packages which do not rely on compiled code (C/C++).
Instructions to fully build an R package under Windows can be found <a href="http://www.murdoch-sutherland.com/Rtools/">here</a> and
<a href="https://www.r-bloggers.com/2010/06/how-to-create-an-r-package-in-windows/">here</a>.</p>
<h3 id="15-maintain-package">1.5 Maintain package</h3>
<p>Several useful helper utilities exist for maintaing and extending packages. Typical package development routines
include:</p>
<ul>
<li>Adding new functions, methods and classes to the script files in the ./R directory in your package</li>
<li>Adding their names to the NAMESPACE file of the package</li>
<li>Additional <code>.Rd</code> help templates can be generated with the <code>prompt()</code> function family like this:</li>
</ul>
<pre><code class="language-r">source(&quot;myscript.R&quot;) # imports functions, methods and classes from myscript.R
prompt(myfct) # writes help file myfct.Rd
promptClass(&quot;myclass&quot;) # writes file myclass-class.Rd
promptMethods(&quot;mymeth&quot;) # writes help file mymeth.Rd
</code></pre>
<p>The resulting <code>.Rd</code> help files can be edited in a text editor, and properly rendered and viewed from within R with help of
the following functions.</p>
<pre><code class="language-r">library(tools)
Rd2txt(&quot;./mypackage/man/myfct.Rd&quot;) # renders *.Rd files as they look in final help pages
checkRd(&quot;./mypackage/man/myfct.Rd&quot;) # checks *.Rd help file for problems
</code></pre>
<h3 id="16-submit-package-to-public-repository">1.6 Submit package to public repository</h3>
<p>The best way of sharing an R package with the community is to submit it to one
of the main R package repositories, such as CRAN or Bioconductor. The details
about the submission process are given on the corresponding repository
submission pages:</p>
<ul>
<li><a href="http://www.bioconductor.org/developers/package-submission/">Submitting to Bioconductor</a></li>
<li><a href="https://cran.r-project.org/">Submitting to CRAN</a></li>
</ul>
<h2 id="2-r-devtools-approach">2. R <code>devtools</code> Approach</h2>
<p>Several package develpment routines of the traditional method outlined above
are manual, such as updating the NAMESPACE file and documenting functions in
separate help (<code>*.Rd</code>) files. This process can be simplified and partially
automated by taking advantage of a more recent R package development
environment composed of several helper packages including <code>devtools</code>,
<code>usethis</code>, <code>roxygen2</code> and <code>sinew</code> (Wickham and Bryan, n.d.). Many books and web sites document this process
in more detail. Here is a small selection of useful online documentation about
R package development:</p>
<ul>
<li>Book: <a href="https://r-pkgs.org/index.html">R Packages</a> by Hadley Wickham and Jenny Bryan</li>
<li><a href="https://tinyheero.github.io/jekyll/update/2015/07/26/making-your-first-R-package.html">My First R Package</a> by Fong Chun Chan</li>
<li><a href="https://www.amitkohli.com/2020/01/07/2020-01-07-how-to-create-an-r-package-my-way/">How to Creat an R Package, Easy Mode</a> by Amit Kohli</li>
<li><a href="https://rawgit.com/rstudio/cheatsheets/master/package-development.pdf">Package Development Cheat Sheet</a></li>
<li>Automating <code>roxygen2</code> documentation with <code>sinew</code> by Jonathan Sidi: <a href="https://yonicd.github.io/2017-09-18-sinew/">Blog</a> and <a href="https://cran.r-project.org/web/packages/sinew/index.html">CRAN</a></li>
</ul>
<h2 id="workflow-for-building-r-packages">Workflow for building R packages</h2>
<p>The following outlines the basic workflow for building, testing and extending R packages with
the package development environment functionalities outlined above.</p>
<h3 id="21-create-package-skeleton">2.1 Create package skeleton</h3>
<pre><code class="language-r">library(&quot;devtools&quot;); library(&quot;roxygen2&quot;); library(&quot;usethis&quot;); library(sinew) # If not availble install these packages with 'install.packages(...)'
create(&quot;myfirstpkg&quot;) # Creates package skeleton. The chosen name (here myfirstpkg) will be the name of the package.
setwd(&quot;myfirstpkg&quot;) # Set working directory of R session to package directory 'myfirstpkg'
use_mit_license() # Add license information to description file (here MIT). To look up alternatives, do ?use_mit_license
</code></pre>
<h3 id="22-add-r-functions">2.2 Add R functions</h3>
<p>Next, R functions can be added to <code>*.R</code> file(s) under the R directory of the new
package. Several functions can be organized in one <code>*.R</code> file, each in its own
file or any combination. For demonstration purposes, the following will
download an R file (<code>pkg_build_fct.R</code> from <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/content/en/tutorials/rpackages/helper_functions/pkg_build_fct.R">here</a>) defining two functions (named:<code>myMAcomp</code>
and <code>talkToMe</code>) and save it to the R directory of the package.</p>
<pre><code class="language-r">download.file(&quot;https://raw.githubusercontent.com/tgirke/GEN242/main/content/en/tutorials/rpackages/helper_functions/pkg_build_fct.R&quot;, &quot;R/pkg_build_fct.R&quot;)
</code></pre>
<h3 id="23-auto-generate-roxygen-comment-lines">2.3 Auto-generate roxygen comment lines</h3>
<p>The <code>makeOxygen</code> function from the <code>sinew</code> package creates <code>roxygen2</code> comment
skeletons based on the information from each function (below for <code>myMAcomp</code> example).
The roxygen comment lines need to be added above the code of each function.
This can be done by copy and paste from the R console or by writing the output
to a temporary file (below via <code>writeLines</code>). Alternatively, the <code>makeOxyFile</code>
function can be used to create a roxygenized copy of an R source file, where the
roxygen comment lines have been added above all functions automatically. Next,
the default text in the comment lines needs to be replaced by meaningful text
describing the utility and usage of each function. This editing process of documentation
can be completed and/or revised any time.</p>
<pre><code class="language-r">load_all() # Loads package in a simulated way without installing it. 
writeLines(makeOxygen(myMAcomp), &quot;myroxylines&quot;) # This creates a 'myroxylines' file in current directory. Delete this file after adding its content to the corresponding functions.
</code></pre>
<h3 id="24-autogenerate-help-files">2.4 Autogenerate help files</h3>
<p>The <code>document</code> function autogenerates for each function one <code>*.Rd</code> file in the
<code>man</code> directory of the package. The content in the <code>*.Rd</code> help files is based
on the information in the roxygen comment lines generated in the
previous step. In addition, all relevant export/import instructions are added
to the <code>NAMESPACE</code> file. Importantly, when using roxygen-based documentation in a package
then the <code>NAMESPACE</code> and <code>*.Rd</code> files should not be manually edited since this information
will be lost during the automation routines provided by <code>roxygen2</code>.</p>
<pre><code class="language-r">document() # Auto-generates/updates *.Rd files under man directory (here: myMAcomp.Rd and talkToMe.Rd)
tools::Rd2txt(&quot;man/myMAcomp.Rd&quot;) # Renders Rd file from source
tools::checkRd(&quot;man/myMAcomp.Rd&quot;) # Checks Rd file for problems
</code></pre>
<h3 id="25-add-a-vignette">2.5 Add a vignette</h3>
<p>A vignette template can be auto-generated with the <code>use_vignette</code> function from
the <code>usethis</code> package. The <code>*.Rmd</code> source file of the vignette will be located
under a new <code>vignette</code> directory. Additional vignettes can be manually added to
this directory as needed.</p>
<pre><code class="language-r">use_vignette(&quot;introduction&quot;, title=&quot;Introduction to this package&quot;)
</code></pre>
<h3 id="26-check-install-and-build-package">2.6 Check, install and build package</h3>
<p>Now the package can be checked for problems. All warnings and errors should be
addressed prior to submission to a public repository. After this it can be
installed on a user’s system with the <code>install</code> command. In addition, the
<code>build</code> function allows to assemble the package in a <code>*.tar.gz</code> file. The
latter is often important for sharing packages and/or submitting them to public
repositories.</p>
<pre><code class="language-r">setwd(&quot;..&quot;) # Redirect R session to parent directory
check(&quot;myfirstpkg&quot;) # Check package for problems, when in pkg dir one can just use check()
# remove.packages(&quot;myfirstpkg&quot;) # Optional. Removes test package if already installed
install(&quot;myfirstpkg&quot;, build_vignettes=TRUE) # Installs package  
build(&quot;myfirstpkg&quot;) # Creates *.tar.gz file for package required to for submission to CRAN/Bioc
</code></pre>
<h3 id="27-using-the-new-package">2.7 Using the new package</h3>
<p>After installing and loading the package its functions, help files and
vignettes can be accessed as follows.</p>
<pre><code class="language-r">library(&quot;myfirstpkg&quot;)
library(help=&quot;myfirstpkg&quot;)
?myMAcomp
vignette(&quot;introduction&quot;, &quot;myfirstpkg&quot;)
</code></pre>
<p>Another very useful development function is <code>test</code> for evaluating the test code of a package.</p>
<h3 id="28-share-package-on-github">2.8 Share package on GitHub</h3>
<p>To host and share the new package <code>myfirstpkg</code> on GitHub, one can use the following steps:</p>
<ol>
<li>Create an empty target GitHub repos online (<em>e.g.</em> named <code>mypkg_repos</code>) as outlined <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/github/github/#exercise">here</a>.</li>
<li>Clone the new GitHub repos to local system with <code>git clone https://github.com/&lt;github_username&gt;/&lt;repo name&gt;</code> (here from command-line)</li>
<li>Copy the root directory of the package into the downloaded repos with <code>cp -r myfirstpkg mypkg_repos</code></li>
<li>Next <code>cd</code> into <code>mypkg_repos</code>, and then add all files and directories of the package to the staging area with <code>git add -A :/</code>.</li>
<li>Commit and push the changes to GitHub with: <code>git commit -am &quot;first commit&quot;; git push</code>.</li>
<li>After this the package should be life on the corresponding GitHub web page.</li>
<li>Assuming the package is public, it can be installed directly from GitHub by anyone as shown below (from within R). Installs of
private packages require a personal access token (PAT) that needs to be assigned to the <code>auth_token</code> argument. PATs can be created
<a href="https://github.com/settings/tokens">here</a>.</li>
</ol>
<pre><code class="language-r">devtools::install_github(&quot;&lt;github_user_name&gt;/&lt;mypkg_repos&gt;&quot;, subdir=&quot;myfirstpkg&quot;) # If the package is in the root directory of the repos, then the 'subdir' argument can be dropped.
</code></pre>
<h2 id="session-info">Session Info</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.1.0 (2021-05-18)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] ggplot2_3.3.3    limma_3.48.0     BiocStyle_2.20.0
## 
## loaded via a namespace (and not attached):
##  [1] bslib_0.2.5.1       compiler_4.1.0      pillar_1.6.1        BiocManager_1.30.15
##  [5] jquerylib_0.1.4     tools_4.1.0         digest_0.6.27       jsonlite_1.7.2     
##  [9] evaluate_0.14       lifecycle_1.0.0     tibble_3.1.2        gtable_0.3.0       
## [13] pkgconfig_2.0.3     rlang_0.4.11        DBI_1.1.1           yaml_2.2.1         
## [17] blogdown_1.3        xfun_0.23           withr_2.4.2         stringr_1.4.0      
## [21] dplyr_1.0.6         knitr_1.33          generics_0.1.0      sass_0.4.0         
## [25] vctrs_0.3.8         tidyselect_1.1.1    grid_4.1.0          glue_1.4.2         
## [29] R6_2.5.0            fansi_0.4.2         rmarkdown_2.8       bookdown_0.22      
## [33] purrr_0.3.4         magrittr_2.0.1      codetools_0.2-18    scales_1.1.1       
## [37] htmltools_0.5.1.1   ellipsis_0.3.2      assertthat_0.2.1    colorspace_2.0-1   
## [41] utf8_1.2.1          stringi_1.6.2       munsell_0.5.0       crayon_1.4.1
</code></pre>
<h2 id="references">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Wickham_undated-ei" class="csl-entry">
<p>Wickham, Hadley, and Jennifer Bryan. n.d. “R Packages.” <a href="https://r-pkgs.org/index.html">https://r-pkgs.org/index.html</a>. <a href="https://r-pkgs.org/index.html">https://r-pkgs.org/index.html</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2022-03-22: <a  href="https://github.com/tgirke/GEN242//commit/e20be329eebff908398ed70829f0b7d773432a6f">some edits (e20be329)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2022 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
