<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>Programming in R | GEN242</title><meta property="og:title" content="Programming in R" />
<meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Overview One of the main attractions of using the R (http://cran.at.r-project.org) environment is the ease with which users can write their own programs and custom functions. The R programming syntax is extremely easy to learn, even for users with no previous programming experience." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/rprogramming/rprogramming/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2022-04-14T12:27:01-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="Programming in R">
<meta itemprop="description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Overview One of the main attractions of using the R (http://cran.at.r-project.org) environment is the ease with which users can write their own programs and custom functions. The R programming syntax is extremely easy to learn, even for users with no previous programming experience.">
<meta itemprop="dateModified" content="2022-04-14T12:27:01-07:00" />
<meta itemprop="wordCount" content="6143">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Programming in R"/>
<meta name="twitter:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Overview One of the main attractions of using the R (http://cran.at.r-project.org) environment is the ease with which users can write their own programs and custom functions. The R programming syntax is extremely easy to learn, even for users with no previous programming experience."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Slides (S22)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T19)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2022/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.f6370abdfad64803865e837a1e71634f.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.f6370abdfad64803865e837a1e71634f.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../tutorials/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Tutorials</a>
  </li>
  <ul>
    <li class="collapse show" id="tutorials">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsgithubgithub" href="../../../tutorials/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialslinuxlinux" href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrbasicsrbasics" href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-tutorialsrprogrammingrprogramming" href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrparallelrparallel" href="../../../tutorials/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrsequencesrsequences" href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempiper" href="../../../tutorials/systempiper/systempiper/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseq" href="../../../tutorials/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseqnew" href="../../../tutorials/sprnaseq/sprnaseqnew/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspchipseqspchipseq" href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsscrnaseqscrnaseq" href="../../../tutorials/scrnaseq/scrnaseq/">scRNA-Seq Embedding Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrmarkdownrmarkdown" href="../../../tutorials/rmarkdown/rmarkdown/">R Markdown Tutorial</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrfearfea" href="../../../tutorials/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrclusteringrclustering" href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrgraphicsrgraphics" href="../../../tutorials/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappsshinyapps" href="../../../tutorials/shinyapps/shinyapps/">Shiny Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrpackagesrpackages" href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsdplyrdplyr" href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialscmdtocwlcmdtocwl" href="../../../tutorials/cmdtocwl/cmdtocwl/">Automate Creation of CWL Instructions</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempiperrnaseqsystempipernaseq" href="../../../tutorials/systempiper/rnaseq/systempipernaseq/">RNA-Seq Workflow Template</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rprogramming/rprogramming.Rmd" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/tutorials/rprogramming/rprogramming.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/tutorials/rprogramming/rprogramming.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=Programming%20in%20R" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#overview">Overview</a></li>
    <li><a href="#why-programming-in-r">Why Programming in R?</a></li>
    <li><a href="#r-basics">R Basics</a></li>
    <li><a href="#code-editors-for-r">Code Editors for R</a></li>
    <li><a href="#finding-help">Finding Help</a></li>
    <li><a href="#control-structures">Control Structures</a>
      <ul>
        <li><a href="#important-operators">Important Operators</a>
          <ul>
            <li><a href="#comparison-operators">Comparison operators</a></li>
            <li><a href="#logical-operators">Logical operators</a></li>
          </ul>
        </li>
        <li><a href="#conditional-executions-if-statements">Conditional Executions: <code>if</code> Statements</a></li>
        <li><a href="#conditional-executions-ifelse-statements">Conditional Executions: <code>ifelse</code> Statements</a></li>
      </ul>
    </li>
    <li><a href="#loops">Loops</a>
      <ul>
        <li><a href="#for-loop"><code>for</code> loop</a>
          <ul>
            <li><a href="#conditional-stop-of-loops">Conditional Stop of Loops</a></li>
          </ul>
        </li>
        <li><a href="#while-loop"><code>while</code> loop</a></li>
        <li><a href="#the-apply-function-family">The <code>apply</code> Function Family</a>
          <ul>
            <li><a href="#apply"><code>apply</code></a></li>
            <li><a href="#tapply"><code>tapply</code></a></li>
            <li><a href="#sapply-lapply-and-vapply"><code>sapply</code>, <code>lapply</code> and <code>vapply</code></a></li>
          </ul>
        </li>
        <li><a href="#improving-speed-performance-of-loops">Improving Speed Performance of Loops</a>
          <ul>
            <li><a href="#1-for-loop-with-append-versus-inject-approach">1. <code>for</code> loop with append versus inject approach</a></li>
            <li><a href="#2-apply-loop-versus-rowmeans">2. <code>apply</code> loop versus <code>rowMeans</code></a></li>
            <li><a href="#3-apply-loop-versus-vectorized-approach">3. <code>apply</code> loop versus vectorized approach</a></li>
            <li><a href="#4-example-of-fast-querying-routine-applied-to-a-large-matrix">4. Example of fast querying routine applied to a large matrix</a>
              <ul>
                <li><a href="#a-create-a-sample-matrix">(a) Create a sample matrix</a></li>
                <li><a href="#b-organize-results-in-list">(b) Organize results in <code>list</code></a></li>
                <li><a href="#c-combinatorial-filter">(c) Combinatorial filter</a></li>
                <li><a href="#d-extract-query-results">(d) Extract query results</a></li>
              </ul>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#functions">Functions</a>
      <ul>
        <li><a href="#function-overview">Function Overview</a></li>
        <li><a href="#function-syntax-rules">Function Syntax Rules</a></li>
        <li><a href="#examples">Examples</a></li>
      </ul>
    </li>
    <li><a href="#useful-utilities">Useful Utilities</a>
      <ul>
        <li><a href="#debugging-utilities">Debugging Utilities</a></li>
        <li><a href="#regular-expressions">Regular Expressions</a>
          <ul>
            <li><a href="#string-matching-with-grep">String matching with <code>grep</code></a></li>
            <li><a href="#string-substitution-with-gsub">String substitution with <code>gsub</code></a></li>
          </ul>
        </li>
        <li><a href="#interpreting-a-character-string-as-expression">Interpreting a Character String as Expression</a></li>
        <li><a href="#replacement-split-and-paste-functions-for-strings">Replacement, Split and Paste Functions for Strings</a></li>
        <li><a href="#time-date-and-sleep">Time, Date and Sleep</a>
          <ul>
            <li><a href="#example">Example</a>
              <ul>
                <li><a href="#import-of-specific-file-lines-with-regular-expression">Import of Specific File Lines with Regular Expression</a></li>
              </ul>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#calling-external-software">Calling External Software</a></li>
    <li><a href="#running-r-scripts">Running R Scripts</a>
      <ul>
        <li><a href="#possibilities-for-executing-r-scripts">Possibilities for Executing R Scripts</a>
          <ul>
            <li><a href="#r-console">R console</a></li>
            <li><a href="#command-line">Command-line</a></li>
            <li><a href="#passing-arguments-from-command-line-to-r">Passing arguments from command-line to R</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#building-r-packages">Building R Packages</a></li>
    <li><a href="#programming-exercises">Programming Exercises</a>
      <ul>
        <li><a href="#exercise-1">Exercise 1</a>
          <ul>
            <li><a href="#for-loop-1"><code>for</code> loop</a></li>
            <li><a href="#while-loop-1"><code>while</code> loop</a></li>
            <li><a href="#apply-loop"><code>apply</code> loop</a></li>
            <li><a href="#avoiding-loops">Avoiding loops</a></li>
          </ul>
        </li>
        <li><a href="#exercise-2">Exercise 2</a>
          <ul>
            <li><a href="#custom-functions">Custom functions</a></li>
          </ul>
        </li>
        <li><a href="#exercise-3">Exercise 3</a>
          <ul>
            <li><a href="#nested-loops-to-generate-similarity-matrices">Nested loops to generate similarity matrices</a></li>
          </ul>
        </li>
        <li><a href="#exercise-4">Exercise 4</a>
          <ul>
            <li><a href="#build-your-own-r-package">Build your own R package</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#homework-5">Homework 5</a></li>
    <li><a href="#additional-exercises">Additional Exercises</a>
      <ul>
        <li><a href="#pattern-matching-and-positional-parsing-of-equences">Pattern matching and positional parsing of equences</a></li>
        <li><a href="#identify-over-represented-strings-in-sequence-sets">Identify over-represented strings in sequence sets</a></li>
      </ul>
    </li>
    <li><a href="#object-oriented-programming-oop">Object-Oriented Programming (OOP)</a>
      <ul>
        <li><a href="#define-s4-classes">Define S4 Classes</a>
          <ul>
            <li><a href="#1-define-s4-classes-with-setclass-and-new">1. Define S4 Classes with <code>setClass()</code> and <code>new()</code></a></li>
            <li><a href="#2-create-new-class-instance">2. Create new class instance</a></li>
            <li><a href="#3-initialization-method">3. Initialization method</a></li>
            <li><a href="#4-usage-and-helper-functions">4. Usage and helper functions</a></li>
            <li><a href="#5-inheritance">5. Inheritance</a></li>
            <li><a href="#6-coerce-objects-to-another-class">6. Coerce objects to another class</a></li>
            <li><a href="#7-virtual-classes">7. Virtual classes</a></li>
            <li><a href="#8-introspection-of-classes">8. Introspection of classes</a></li>
          </ul>
        </li>
        <li><a href="#assign-generics-and-methods">Assign Generics and Methods</a>
          <ul>
            <li><a href="#1-accessor-functions">1. Accessor functions</a></li>
            <li><a href="#2-replacement-methods">2. Replacement methods</a></li>
            <li><a href="#3-behavior-of-bracket-operator">3. Behavior of bracket operator</a></li>
            <li><a href="#4-print-behavior">4. Print behavior</a></li>
            <li><a href="#5-define-custom-methods">5. Define custom methods</a></li>
            <li><a href="#6-plotting-method">6. Plotting method</a></li>
            <li><a href="#7-utilities-to-inspect-methods">7. Utilities to inspect methods</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#session-info">Session Info</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../tutorials/">Tutorials</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>Programming in R</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 29 minute read</p>
           
	<!--
- Compile from command-line
Rscript -e "rmarkdown::render('rprogramming.Rmd', c('html_document'), clean=F); knitr::knit('rprogramming.Rmd', tangle=TRUE)"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div style="text-align: right">
<p>Source code downloads:    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rprogramming/rprogramming.Rmd">.Rmd</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rprogramming/rprogramming.R">.R</a> ]</p>
</div>
<h2 id="overview">Overview</h2>
<p>One of the main attractions of using the R
(<a href="http://cran.at.r-project.org">http://cran.at.r-project.org</a>) environment is
the ease with which users can write their own programs and custom functions.
The R programming syntax is extremely easy to learn, even for users with no
previous programming experience. Once the basic R programming control
structures are understood, users can use the R language as a powerful
environment to perform complex custom analyses of almost any type of data (Gentleman 2008).</p>
<h2 id="why-programming-in-r">Why Programming in R?</h2>
<ul>
<li>Powerful statistical environment and programming language</li>
<li>Facilitates reproducible research</li>
<li>Efficient data structures make programming very easy</li>
<li>Ease of implementing custom functions</li>
<li>Powerful graphics</li>
<li>Access to fast growing number of analysis packages</li>
<li>One of the most widely used languages in bioinformatics</li>
<li>Is standard for data mining and biostatistical analysis</li>
<li>Technical advantages: free, open-source, available for all OSs</li>
</ul>
<h2 id="r-basics">R Basics</h2>
<p>The previous <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/rbasics/rbasics/">Rbasics</a> tutorial provides a general introduction to the usage of the R environment and its basic command syntax.
More details can be found in the R &amp; BioConductor manual <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual">here</a>.</p>
<h2 id="code-editors-for-r">Code Editors for R</h2>
<p>Several excellent code editors are available that provide functionalities like R syntax highlighting, auto code indenting and utilities to send code/functions to the R console.</p>
<ul>
<li><a href="https://www.rstudio.com/products/rstudio/features/">RStudio</a>: GUI-based IDE for R</li>
<li><a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/linux/linux/#nvim-r-tmux-essentials">Vim-R-Tmux</a>: R working environment based on vim and tmux</li>
<li><a href="http://www.xemacs.org/Download/index.html">Emacs</a> (<a href="http://ess.r-project.org/">ESS add-on package</a>)</li>
<li><a href="https://wiki.gnome.org/Apps/Gedit">gedit</a> and <a href="https://wiki.gnome.org/Apps/Gedit">Rgedit</a></li>
<li><a href="https://rkward.kde.org/">RKWard</a></li>
<li><a href="http://www.walware.de/goto/statet">Eclipse</a></li>
<li><a href="https://sourceforge.net/projects/tinn-r/files/Tinn-R%20portable/">Tinn-R</a></li>
<li><a href="https://sourceforge.net/projects/npptor/">Notepad++ (NppToR)</a></li>
</ul>
<center>
<p>Programming in R using RStudio</p>
</center>
<center>
<img title="R_Interfaces" src="../images/rstudio.png"/>
</center>
<center>
<p>Programming in R using Vim or Emacs</p>
</center>
<center>
<img title="vim-r" src="../images/vimR.png"/>
</center>
<h2 id="finding-help">Finding Help</h2>
<p>Reference list on R programming (selection)</p>
<ul>
<li><a href="http://adv-r.had.co.nz/">Advanced R</a>, by Hadley Wickham</li>
<li><a href="http://master.bioconductor.org/help/publications/books/r-programming-for-bioinformatics/">R Programming for Bioinformatics</a>, by Robert Gentleman</li>
<li><a href="http://www.stats.ox.ac.uk/pub/MASS3/Sprog/">S Programming</a>, by W. N. Venables and B. D. Ripley</li>
<li><a href="http://www.amazon.com/Programming-Data-Language-Lecture-Economics/dp/0387985034">Programming with Data</a>, by John M. Chambers</li>
<li><a href="http://www1.maths.lth.se/help/R/">R Help</a> &amp; <a href="http://www1.maths.lth.se/help/R/RCC/">R Coding Conventions</a>, Henrik Bengtsson, Lund University</li>
<li><a href="http://zoonek2.free.fr/UNIX/48_R/02.html">Programming in R</a> (Vincent Zoonekynd)</li>
<li><a href="http://www2.warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r">Peter’s R Programming Pages</a>, University of Warwick</li>
<li><a href="http://pj.freefaculty.org/R/statsRus.html">Rtips</a>, Paul Johnsson, University of Kansas</li>
<li><a href="http://heather.cs.ucdavis.edu/~matloff/r.html">R for Programmers</a>, Norm Matloff, UC Davis</li>
<li><a href="http://www.statistik.uni-dortmund.de/useR-2008/tutorials/useR2008introhighperfR.pdf">High-Performance R</a>, Dirk Eddelbuettel tutorial presented at useR-2008</li>
<li><a href="http://www.stat.harvard.edu/ccr2005/index.html">C/C++ level programming for R</a>, Gopi Goswami</li>
</ul>
<h2 id="control-structures">Control Structures</h2>
<h3 id="important-operators">Important Operators</h3>
<h4 id="comparison-operators">Comparison operators</h4>
<ul>
<li><code>==</code> (equal)</li>
<li><code>!=</code> (not equal)</li>
<li><code>&gt;</code> (greater than)</li>
<li><code>&gt;=</code> (greater than or equal)</li>
<li><code>&lt;</code> (less than)</li>
<li><code>&lt;=</code> (less than or equal)</li>
</ul>
<h4 id="logical-operators">Logical operators</h4>
<ul>
<li><code>&amp;</code> (and)</li>
<li><code>&amp;&amp;</code> (and)</li>
<li><code>|</code> (or)</li>
<li><code>||</code> (or)</li>
<li><code>!</code> (not)</li>
</ul>
<p>Note: <code>&amp;</code> and <code>&amp;&amp;</code> indicate logical AND, while <code>|</code> and <code>||</code> indicate logical OR. The shorter form performs element-wise comparisons of same-length vectors.
The longer form evaluates left to right examining only the first element of each vector (can be of different lengths). Evaluation proceeds only until the result
is determined. The longer form is preferred for programming control-flow, <em>e.g.</em> via <code>if</code> clauses.</p>
<h3 id="conditional-executions-if-statements">Conditional Executions: <code>if</code> Statements</h3>
<p>An <code>if</code> statement operates on length-one logical vectors.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">if (TRUE) { 
    statements_1 
} else { 
    statements_2 
}
</code></pre>
<p>In the <code>else</code> component, avoid inserting newlines between <code>} else</code>. For details on how to best and consistently format R code,
this <a href="http://adv-r.had.co.nz/Style.html">style guide</a> is a good start. In addition, the <a href="https://yihui.org/formatr/"><code>formatR</code></a> package can be helpful.</p>
<p><strong>Example</strong></p>
<pre><code class="language-r">if (1==0) { 
    print(1) 
} else { 
    print(2) 
}
</code></pre>
<pre><code>## [1] 2
</code></pre>
<p><strong>Example 2</strong></p>
<pre><code class="language-r">if (1==0) { 
    print(1) 
} else if (1==2) { 
    print(2) 
} else {
    print(3)
}
</code></pre>
<pre><code>## [1] 3
</code></pre>
<h3 id="conditional-executions-ifelse-statements">Conditional Executions: <code>ifelse</code> Statements</h3>
<p>The <code>ifelse</code> statement operates on vectors.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">ifelse(test, true_value, false_value)
</code></pre>
<p><strong>Example</strong></p>
<pre><code class="language-r">x &lt;- 1:10 
ifelse(x&lt;5, sqrt(x), 0)
</code></pre>
<pre><code>##  [1] 1.000000 1.414214 1.732051 2.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000
</code></pre>
<h2 id="loops">Loops</h2>
<h3 id="for-loop"><code>for</code> loop</h3>
<p><code>for</code> loops iterate over elements of a looping vector.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">for(variable in sequence) { 
    statements 
}
</code></pre>
<p><strong>Example</strong></p>
<pre><code class="language-r">mydf &lt;- iris
myve &lt;- NULL
for(i in seq(along=mydf[,1])) {
    myve &lt;- c(myve, mean(as.numeric(mydf[i,1:3])))
}
myve[1:8]
</code></pre>
<pre><code>## [1] 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000
</code></pre>
<p><strong>Note:</strong> Inject into objecs is much faster than append approach with <code>c</code>, <code>cbind</code>, etc.</p>
<p><strong>Example</strong></p>
<pre><code class="language-r">myve &lt;- numeric(length(mydf[,1]))
for(i in seq(along=myve)) {
    myve[i] &lt;- mean(as.numeric(mydf[i,1:3]))
}
myve[1:8]
</code></pre>
<pre><code>## [1] 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000
</code></pre>
<h4 id="conditional-stop-of-loops">Conditional Stop of Loops</h4>
<p>The <code>stop</code> function can be used to break out of a loop (or a function) when a condition becomes <code>TRUE</code>. In addition, an error message will be printed.</p>
<p><strong>Example</strong></p>
<pre><code class="language-r">x &lt;- 1:10
z &lt;- NULL
for(i in seq(along=x)) { 
    if (x[i] &lt; 5) { 
        z &lt;- c(z, x[i]-1)  
        print(z)
    } else { 
        stop(&quot;values need to be &lt; 5&quot;) 
    }
}
</code></pre>
<h3 id="while-loop"><code>while</code> loop</h3>
<p>Iterates as long as a condition is true.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">while(condition) {
    statements
}
</code></pre>
<p><strong>Example</strong></p>
<pre><code class="language-r">z &lt;- 0
while(z&lt;5) { 
    z &lt;- z + 2
    print(z)  
}
</code></pre>
<pre><code>## [1] 2
## [1] 4
## [1] 6
</code></pre>
<h3 id="the-apply-function-family">The <code>apply</code> Function Family</h3>
<h4 id="apply"><code>apply</code></h4>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">apply(X, MARGIN, FUN, ARGs)
</code></pre>
<p><strong>Arguments</strong></p>
<ul>
<li><code>X</code>: <code>array</code>, <code>matrix</code> or <code>data.frame</code></li>
<li><code>MARGIN</code>: <code>1</code> for rows, <code>2</code> for columns</li>
<li><code>FUN</code>: one or more functions</li>
<li><code>ARGs</code>: possible arguments for functions</li>
</ul>
<p><strong>Example</strong></p>
<pre><code class="language-r">apply(iris[1:8,1:3], 1, mean)
</code></pre>
<pre><code>##        1        2        3        4        5        6        7        8 
## 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000
</code></pre>
<h4 id="tapply"><code>tapply</code></h4>
<p>Applies a function to vector components that are defined by a factor.</p>
<p><strong>Syntax</strong></p>
<pre><code class="language-r">tapply(vector, factor, FUN)
</code></pre>
<p><strong>Example</strong></p>
<pre><code class="language-r">iris[1:2,]
</code></pre>
<pre><code>##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
</code></pre>
<pre><code class="language-r">tapply(iris$Sepal.Length, iris$Species, mean)
</code></pre>
<pre><code>##     setosa versicolor  virginica 
##      5.006      5.936      6.588
</code></pre>
<h4 id="sapply-lapply-and-vapply"><code>sapply</code>, <code>lapply</code> and <code>vapply</code></h4>
<p>The iterator functions <code>sapply</code>, <code>lapply</code> and <code>vapply</code> apply a function to
vectors or lists. The <code>lapply</code> function always returns a list, while <code>sapply</code>
returns <code>vector</code> or <code>matrix</code> objects when possible. If not then a list is
returned. The <code>vapply</code> function returns a vector or array of type matching the
<code>FUN.VALUE</code>. Compared to <code>sapply</code>, <code>vapply</code> is a safer choice with respect to
controlling specific output types to avoid exception handling problems.</p>
<p><strong>Examples</strong></p>
<pre><code class="language-r">l &lt;- list(a = 1:10, beta = exp(-3:3), logic = c(TRUE,FALSE,FALSE,TRUE))
lapply(l, mean)
</code></pre>
<pre><code>## $a
## [1] 5.5
## 
## $beta
## [1] 4.535125
## 
## $logic
## [1] 0.5
</code></pre>
<pre><code class="language-r">sapply(l, mean)
</code></pre>
<pre><code>##        a     beta    logic 
## 5.500000 4.535125 0.500000
</code></pre>
<pre><code class="language-r">vapply(l, mean, FUN.VALUE=numeric(1))
</code></pre>
<pre><code>##        a     beta    logic 
## 5.500000 4.535125 0.500000
</code></pre>
<p>Often used in combination with a function definition:</p>
<pre><code class="language-r">lapply(names(l), function(x) mean(l[[x]]))
sapply(names(l), function(x) mean(l[[x]]))
vapply(names(l), function(x) mean(l[[x]]), FUN.VALUE=numeric(1))
</code></pre>
<h3 id="improving-speed-performance-of-loops">Improving Speed Performance of Loops</h3>
<p>Looping over very large data sets can become slow in R. However, this
limitation can be overcome by eliminating certain operations in loops or
avoiding loops over the data intensive dimension in an object altogether. The
latter can be achieved by performing mainly vector-to-vecor or
matrix-to-matrix computations. These vectorized operations run in R often over
100 times faster than the corresponding <code>for()</code> or <code>apply()</code> loops. In
addition, one can make use of the existing speed-optimized C-level functions in
R, such as <code>rowSums</code>, <code>rowMeans</code>, <code>table</code>, and <code>tabulate</code>. Moreover, one can
design custom functions that avoid expensive R loops by using vector- or
matrix-based approaches. Alternatively, one can write programs that will
perform all time consuming computations on the C-level.</p>
<p>The following code samples illustrate the time-performance differences among
the different approaches of running iterative operations in R.</p>
<h4 id="1-for-loop-with-append-versus-inject-approach">1. <code>for</code> loop with append versus inject approach</h4>
<p>The following runs a <code>for</code> loop where the result is appended in each iteration
with the <code>c()</code> function. The corresponding <code>cbind</code> and <code>rbind</code> for two dimensional
data objects would have a similar performance impact as <code>c()</code>.</p>
<pre><code class="language-r">myMA &lt;- matrix(rnorm(1000000), 100000, 10, dimnames=list(1:100000, paste(&quot;C&quot;, 1:10, sep=&quot;&quot;)))
results &lt;- NULL
system.time(for(i in seq(along=myMA[,1])) results &lt;- c(results, mean(myMA[i,])))
   user  system elapsed
 39.156   6.369  45.559
</code></pre>
<p>Now the for loop is run with an inject approach for storing the results in each iteration.</p>
<pre><code class="language-r">results &lt;- numeric(length(myMA[,1]))
system.time(for(i in seq(along=myMA[,1])) results[i] &lt;- mean(myMA[i,]))
   user  system elapsed
  1.550   0.005   1.556 
</code></pre>
<p>As one can see from the output of <code>system.time</code>, the inject approach is 20-50 times faster.</p>
<h4 id="2-apply-loop-versus-rowmeans">2. <code>apply</code> loop versus <code>rowMeans</code></h4>
<p>The following performs a row-wise mean calculation on a large matrix first with an <code>apply</code>
loop and then with the <code>rowMeans</code> function.</p>
<pre><code class="language-r">system.time(myMAmean &lt;- apply(myMA, 1, mean))
  user  system elapsed
 1.452   0.005   1.456

system.time(myMAmean &lt;- rowMeans(myMA))
   user  system elapsed
  0.005   0.001   0.006
</code></pre>
<p>Based on the results from <code>system.time</code>, the <code>rowMeans</code> approach is over 200 times faster
than the <code>apply</code> loop.</p>
<h4 id="3-apply-loop-versus-vectorized-approach">3. <code>apply</code> loop versus vectorized approach</h4>
<p>In this example row-wise standard deviations are computed with an <code>apply</code> loop and then
in a vectorized manner.</p>
<pre><code class="language-r">system.time(myMAsd &lt;- apply(myMA, 1, sd))
   user  system elapsed
  3.707   0.014   3.721
myMAsd[1:4]
        1         2         3         4
0.8505795 1.3419460 1.3768646 1.3005428

system.time(myMAsd &lt;- sqrt((rowSums((myMA-rowMeans(myMA))^2)) / (length(myMA[1,])-1)))
   user  system elapsed
  0.020   0.009   0.028

myMAsd[1:4]
        1         2         3         4
0.8505795 1.3419460 1.3768646 1.3005428
</code></pre>
<p>The vector-based approach in the last step is over 200 times faster than the apply loop.</p>
<h4 id="4-example-of-fast-querying-routine-applied-to-a-large-matrix">4. Example of fast querying routine applied to a large matrix</h4>
<h5 id="a-create-a-sample-matrix">(a) Create a sample matrix</h5>
<p>The following <code>lfcPvalMA</code> function creates a test <code>matrix</code> containing randomly generated log2 fold changes (LFCs)
and p-values (here: pval or FDRs) for variable numbers of samples or test results. In biology this dataset mimics the
results of an analysis of differentially expressed genes (DEGs) from several contrasts arranged in a
single <code>matrix</code> (or <code>data.frame</code>).</p>
<pre><code class="language-r">lfcPvalMA &lt;- function(Nrow=200, Ncol=4, stats_labels=c(&quot;lfc&quot;, &quot;pval&quot;)) {
    set.seed(1410)
    assign(stats_labels[1], runif(n = Nrow * Ncol, min = -4, max = 4))
    assign(stats_labels[2], runif(n = Nrow * Ncol, min = 0, max = 1))
    lfc_ma &lt;- matrix(lfc, Nrow, Ncol, dimnames=list(paste(&quot;g&quot;, 1:Nrow, sep=&quot;&quot;), paste(&quot;t&quot;, 1:Ncol, &quot;_&quot;, stats_labels[1], sep=&quot;&quot;))) 
    pval_ma &lt;- matrix(pval, Nrow, Ncol, dimnames=list(paste(&quot;g&quot;, 1:Nrow, sep=&quot;&quot;), paste(&quot;t&quot;, 1:Ncol, &quot;_&quot;, stats_labels[2], sep=&quot;&quot;))) 
    statsMA &lt;- cbind(lfc_ma, pval_ma)
    return(statsMA[, order(colnames(statsMA))])
}
degMA &lt;- lfcPvalMA(Nrow=200, Ncol=4, stats_labels=c(&quot;lfc&quot;, &quot;pval&quot;))
dim(degMA) 
</code></pre>
<pre><code>## [1] 200   8
</code></pre>
<pre><code class="language-r">degMA[1:4,] # Prints first 4 rows of DEG matrix generated as a test data set 
</code></pre>
<pre><code>##        t1_lfc    t1_pval    t2_lfc   t2_pval     t3_lfc   t3_pval     t4_lfc   t4_pval
## g1 -1.8476368 0.39486484  1.879310 0.7785999  0.1769551 0.9904342  0.1747932 0.9536679
## g2  0.2542926 0.04188993 -1.629778 0.6379570 -1.9280792 0.6106041 -2.3599518 0.1950022
## g3  3.4703657 0.73881357  2.047794 0.3129176 -3.8891714 0.1508787  3.7811606 0.2560303
## g4 -2.8548158 0.99201512 -2.710385 0.1772805  1.0920515 0.2826038  3.9313225 0.5519854
</code></pre>
<h5 id="b-organize-results-in-list">(b) Organize results in <code>list</code></h5>
<p>To filter the results efficiently, it is usually best to store the two different
stats (here <code>lfc</code> and <code>pval</code>) in separate matrices (here two) where each has the
same dimensions and row/column ordering. Note, in this case a <code>list</code>
is used to store the two <code>matrices</code>.</p>
<pre><code class="language-r">degList &lt;- list(lfc=degMA[ , grepl(&quot;lfc&quot;, colnames(degMA))], pval=degMA[ , grepl(&quot;pval&quot;, colnames(degMA))])
names(degList)
</code></pre>
<pre><code>## [1] &quot;lfc&quot;  &quot;pval&quot;
</code></pre>
<pre><code class="language-r">sapply(degList, dim)
</code></pre>
<pre><code>##      lfc pval
## [1,] 200  200
## [2,]   4    4
</code></pre>
<h5 id="c-combinatorial-filter">(c) Combinatorial filter</h5>
<p>With the above generated data structure of two complementary matrices it is
easy to apply combinatorial filtering routines that are both flexible and
time-efficient (fast). The following example queries for fold changes of at
least 2 (here <code>lfc &gt;= 1 | lfc &lt;= -1</code>) plus p-values of 0.5 or lower. Note, all
intermediate and final results are stored in logical matrices. In addition to
boolean comparisons, one can apply basic mathematical operations, such as
calculating the sum of each cell across many matrices. This returns a numeric matix of
integers representing the counts of <code>TRUE</code> values in each position of the
considered logical matrices. Subsequently, one can perform summary and
filtering routines on these count-based matrices which is convenient when
working with large numbers of matrices. All these matrix-to-matrix comparisons
are very fast to compute and require zero looping instructions by the user.</p>
<pre><code class="language-r">queryResult &lt;- (degList$lfc &lt;= 1 | degList$lfc &lt;= -1) &amp; degList$pval &lt;= 0.5 
colnames(queryResult) &lt;- gsub(&quot;_.*&quot;, &quot;&quot;, colnames(queryResult)) # Adjust column names 
queryResult[1:4,]
</code></pre>
<pre><code>##       t1    t2    t3    t4
## g1  TRUE FALSE FALSE FALSE
## g2  TRUE FALSE FALSE  TRUE
## g3 FALSE FALSE  TRUE FALSE
## g4 FALSE  TRUE FALSE FALSE
</code></pre>
<h5 id="d-extract-query-results">(d) Extract query results</h5>
<ol>
<li>Retrieve row labels (genes) that match the query from the previous step in each column, and
store them in a <code>list</code>.</li>
</ol>
<!-- end list -->
<pre><code class="language-r">matchingIDlist &lt;- sapply(colnames(queryResult), function(x) names(queryResult[queryResult[ , x] , x]), simplify=FALSE)
matchingIDlist
</code></pre>
<pre><code>## $t1
##  [1] &quot;g1&quot;   &quot;g2&quot;   &quot;g5&quot;   &quot;g6&quot;   &quot;g11&quot;  &quot;g16&quot;  &quot;g18&quot;  &quot;g19&quot;  &quot;g21&quot;  &quot;g23&quot;  &quot;g24&quot;  &quot;g31&quot;  &quot;g36&quot; 
## [14] &quot;g37&quot;  &quot;g41&quot;  &quot;g46&quot;  &quot;g60&quot;  &quot;g61&quot;  &quot;g63&quot;  &quot;g70&quot;  &quot;g71&quot;  &quot;g72&quot;  &quot;g75&quot;  &quot;g81&quot;  &quot;g83&quot;  &quot;g84&quot; 
## [27] &quot;g88&quot;  &quot;g91&quot;  &quot;g97&quot;  &quot;g98&quot;  &quot;g100&quot; &quot;g102&quot; &quot;g103&quot; &quot;g104&quot; &quot;g110&quot; &quot;g111&quot; &quot;g112&quot; &quot;g113&quot; &quot;g114&quot;
## [40] &quot;g120&quot; &quot;g121&quot; &quot;g123&quot; &quot;g124&quot; &quot;g126&quot; &quot;g130&quot; &quot;g134&quot; &quot;g135&quot; &quot;g139&quot; &quot;g140&quot; &quot;g145&quot; &quot;g147&quot; &quot;g153&quot;
## [53] &quot;g157&quot; &quot;g158&quot; &quot;g159&quot; &quot;g160&quot; &quot;g162&quot; &quot;g170&quot; &quot;g171&quot; &quot;g172&quot; &quot;g173&quot; &quot;g175&quot; &quot;g178&quot; &quot;g183&quot; &quot;g184&quot;
## [66] &quot;g187&quot; &quot;g190&quot; &quot;g192&quot; &quot;g196&quot; &quot;g199&quot;
## 
## $t2
##  [1] &quot;g4&quot;   &quot;g5&quot;   &quot;g7&quot;   &quot;g9&quot;   &quot;g10&quot;  &quot;g12&quot;  &quot;g16&quot;  &quot;g23&quot;  &quot;g34&quot;  &quot;g35&quot;  &quot;g39&quot;  &quot;g41&quot;  &quot;g44&quot; 
## [14] &quot;g46&quot;  &quot;g47&quot;  &quot;g48&quot;  &quot;g49&quot;  &quot;g50&quot;  &quot;g51&quot;  &quot;g52&quot;  &quot;g56&quot;  &quot;g66&quot;  &quot;g75&quot;  &quot;g80&quot;  &quot;g81&quot;  &quot;g85&quot; 
## [27] &quot;g88&quot;  &quot;g89&quot;  &quot;g90&quot;  &quot;g94&quot;  &quot;g99&quot;  &quot;g102&quot; &quot;g112&quot; &quot;g115&quot; &quot;g116&quot; &quot;g118&quot; &quot;g119&quot; &quot;g120&quot; &quot;g129&quot;
## [40] &quot;g144&quot; &quot;g145&quot; &quot;g148&quot; &quot;g152&quot; &quot;g155&quot; &quot;g156&quot; &quot;g160&quot; &quot;g164&quot; &quot;g165&quot; &quot;g167&quot; &quot;g168&quot; &quot;g170&quot; &quot;g172&quot;
## [53] &quot;g178&quot; &quot;g186&quot; &quot;g187&quot; &quot;g194&quot; &quot;g197&quot;
## 
## $t3
##  [1] &quot;g3&quot;   &quot;g6&quot;   &quot;g7&quot;   &quot;g9&quot;   &quot;g12&quot;  &quot;g15&quot;  &quot;g23&quot;  &quot;g25&quot;  &quot;g26&quot;  &quot;g27&quot;  &quot;g38&quot;  &quot;g43&quot;  &quot;g52&quot; 
## [14] &quot;g53&quot;  &quot;g54&quot;  &quot;g58&quot;  &quot;g66&quot;  &quot;g69&quot;  &quot;g72&quot;  &quot;g76&quot;  &quot;g77&quot;  &quot;g80&quot;  &quot;g84&quot;  &quot;g85&quot;  &quot;g86&quot;  &quot;g88&quot; 
## [27] &quot;g89&quot;  &quot;g90&quot;  &quot;g91&quot;  &quot;g99&quot;  &quot;g100&quot; &quot;g107&quot; &quot;g110&quot; &quot;g122&quot; &quot;g124&quot; &quot;g125&quot; &quot;g129&quot; &quot;g134&quot; &quot;g139&quot;
## [40] &quot;g141&quot; &quot;g143&quot; &quot;g144&quot; &quot;g146&quot; &quot;g148&quot; &quot;g154&quot; &quot;g163&quot; &quot;g165&quot; &quot;g171&quot; &quot;g173&quot; &quot;g178&quot; &quot;g180&quot; &quot;g182&quot;
## [53] &quot;g188&quot; &quot;g190&quot; &quot;g193&quot; &quot;g195&quot;
## 
## $t4
##  [1] &quot;g2&quot;   &quot;g5&quot;   &quot;g7&quot;   &quot;g8&quot;   &quot;g9&quot;   &quot;g12&quot;  &quot;g13&quot;  &quot;g15&quot;  &quot;g20&quot;  &quot;g21&quot;  &quot;g26&quot;  &quot;g28&quot;  &quot;g30&quot; 
## [14] &quot;g31&quot;  &quot;g36&quot;  &quot;g37&quot;  &quot;g38&quot;  &quot;g47&quot;  &quot;g63&quot;  &quot;g64&quot;  &quot;g65&quot;  &quot;g67&quot;  &quot;g68&quot;  &quot;g69&quot;  &quot;g76&quot;  &quot;g77&quot; 
## [27] &quot;g80&quot;  &quot;g85&quot;  &quot;g90&quot;  &quot;g98&quot;  &quot;g99&quot;  &quot;g101&quot; &quot;g105&quot; &quot;g106&quot; &quot;g120&quot; &quot;g123&quot; &quot;g125&quot; &quot;g126&quot; &quot;g129&quot;
## [40] &quot;g131&quot; &quot;g134&quot; &quot;g137&quot; &quot;g140&quot; &quot;g143&quot; &quot;g147&quot; &quot;g148&quot; &quot;g153&quot; &quot;g155&quot; &quot;g157&quot; &quot;g165&quot; &quot;g167&quot; &quot;g170&quot;
## [53] &quot;g171&quot; &quot;g172&quot; &quot;g174&quot; &quot;g175&quot; &quot;g176&quot; &quot;g178&quot; &quot;g181&quot; &quot;g182&quot; &quot;g188&quot; &quot;g192&quot; &quot;g199&quot;
</code></pre>
<ol start="2">
<li>Return all row labels (genes) that match the above query across a specified number of columns
(here 2). Note, the <code>rowSums</code> function is used for this, which performs the row-wise looping
internally and extremely fast.</li>
</ol>
<!-- end list -->
<pre><code class="language-r">matchingID &lt;- rowSums(queryResult) &gt; 2 
queryResult[matchingID, , drop=FALSE]
</code></pre>
<pre><code>##         t1    t2    t3    t4
## g5    TRUE  TRUE FALSE  TRUE
## g7   FALSE  TRUE  TRUE  TRUE
## g9   FALSE  TRUE  TRUE  TRUE
## g12  FALSE  TRUE  TRUE  TRUE
## g23   TRUE  TRUE  TRUE FALSE
## g80  FALSE  TRUE  TRUE  TRUE
## g85  FALSE  TRUE  TRUE  TRUE
## g88   TRUE  TRUE  TRUE FALSE
## g90  FALSE  TRUE  TRUE  TRUE
## g99  FALSE  TRUE  TRUE  TRUE
## g120  TRUE  TRUE FALSE  TRUE
## g129 FALSE  TRUE  TRUE  TRUE
## g134  TRUE FALSE  TRUE  TRUE
## g148 FALSE  TRUE  TRUE  TRUE
## g165 FALSE  TRUE  TRUE  TRUE
## g170  TRUE  TRUE FALSE  TRUE
## g171  TRUE FALSE  TRUE  TRUE
## g172  TRUE  TRUE FALSE  TRUE
## g178  TRUE  TRUE  TRUE  TRUE
</code></pre>
<pre><code class="language-r">names(matchingID[matchingID])
</code></pre>
<pre><code>##  [1] &quot;g5&quot;   &quot;g7&quot;   &quot;g9&quot;   &quot;g12&quot;  &quot;g23&quot;  &quot;g80&quot;  &quot;g85&quot;  &quot;g88&quot;  &quot;g90&quot;  &quot;g99&quot;  &quot;g120&quot; &quot;g129&quot; &quot;g134&quot;
## [14] &quot;g148&quot; &quot;g165&quot; &quot;g170&quot; &quot;g171&quot; &quot;g172&quot; &quot;g178&quot;
</code></pre>
<p>As demonstrated in the above query examples, by setting up the proper data structures (here two
<code>matrices</code> with same dimensions), and utilizing vectorized (matrix-to-matrix) operations
along with R’s built-in <code>row*</code> and <code>col*</code> stats function family (e.g. <code>rowSums</code>) one can
design with very little code flexible query routines that also run very time-efficient.</p>
<h2 id="functions">Functions</h2>
<h3 id="function-overview">Function Overview</h3>
<p>A very useful feature of the R environment is the possibility to expand existing functions and to easily write custom functions. In fact, most of the R software can be viewed as a series of R functions.</p>
<p><strong>Syntax</strong> to define function</p>
<pre><code class="language-r">myfct &lt;- function(arg1, arg2, ...) { 
    function_body 
}
</code></pre>
<p><strong>Syntax</strong> to call functions</p>
<pre><code class="language-r">myfct(arg1=..., arg2=...)
</code></pre>
<p>The value returned by a function is the value of the function body, which is usually an unassigned final expression, <em>e.g.</em>: <code>return()</code></p>
<h3 id="function-syntax-rules">Function Syntax Rules</h3>
<p><strong>General</strong></p>
<ul>
<li>Functions are defined by
<ol>
<li>The assignment with the keyword <code>function</code></li>
<li>The declaration of arguments/variables (<code>arg1, arg2, ...</code>)</li>
<li>The definition of operations (<code>function_body</code>) that perform computations on the provided arguments. A function name needs to be assigned to call the function.</li>
</ol>
</li>
</ul>
<p><strong>Naming</strong></p>
<ul>
<li>Function names can be almost anything. However, the usage of names of existing functions should be avoided.</li>
</ul>
<p><strong>Arguments</strong></p>
<ul>
<li>It is often useful to provide default values for arguments (<em>e.g.</em>: <code>arg1=1:10</code>). This way they don’t need to be provided in a function call. The argument list can also be left empty (<code>myfct &lt;- function() { fct_body }</code>) if a function is expected to return always the same value(s). The argument <code>...</code> can be used to allow one function to pass on argument settings to another.</li>
</ul>
<p><strong>Body</strong></p>
<ul>
<li>The actual expressions (commands/operations) are defined in the function body which should be enclosed by braces. The individual commands are separated by semicolons or new lines (preferred).</li>
</ul>
<p><strong>Usage</strong></p>
<ul>
<li>Functions are called by their name followed by parentheses containing possible argument names. Empty parenthesis after the function name will result in an error message when a function requires certain arguments to be provided by the user. The function name alone will print the definition of a function.</li>
</ul>
<p><strong>Scope</strong></p>
<ul>
<li>Variables created inside a function exist only for the life time of a function. Thus, they are not accessible outside of the function. To force variables in functions to exist globally, one can use the double assignment operator: <code>&lt;&lt;-</code></li>
</ul>
<h3 id="examples">Examples</h3>
<p><strong>Define sample function</strong></p>
<pre><code class="language-r">myfct &lt;- function(x1, x2=5) { 
    z1 &lt;- x1 / x1
    z2 &lt;- x2 * x2
        myvec &lt;- c(z1, z2) 
        return(myvec)
} 
</code></pre>
<p><strong>Function usage</strong></p>
<p>Apply function to values <code>2</code> and <code>5</code></p>
<pre><code class="language-r">myfct(x1=2, x2=5) 
</code></pre>
<pre><code>## [1]  1 25
</code></pre>
<p>Run without argument names</p>
<pre><code class="language-r">myfct(2, 5) 
</code></pre>
<pre><code>## [1]  1 25
</code></pre>
<p>Makes use of default value <code>5</code></p>
<pre><code class="language-r">myfct(x1=2) 
</code></pre>
<pre><code>## [1]  1 25
</code></pre>
<p>Print function definition (often unintended)</p>
<pre><code class="language-r">myfct 
</code></pre>
<pre><code>## function(x1, x2=5) { 
##  z1 &lt;- x1 / x1
##  z2 &lt;- x2 * x2
##         myvec &lt;- c(z1, z2) 
##         return(myvec)
## }
## &lt;bytecode: 0x56dd086c6988&gt;
</code></pre>
<h2 id="useful-utilities">Useful Utilities</h2>
<h3 id="debugging-utilities">Debugging Utilities</h3>
<p>Several debugging utilities are available for R. They include:</p>
<ul>
<li><code>traceback</code></li>
<li><code>browser</code></li>
<li><code>options(error=recover)</code></li>
<li><code>options(error=NULL)</code></li>
<li><code>debug</code></li>
</ul>
<p>The <a href="https://adv-r.hadley.nz/debugging.html">Debugging in R page</a> provides an overview of the available resources.</p>
<h3 id="regular-expressions">Regular Expressions</h3>
<p>R’s regular expression utilities work similar as in other languages. To learn how to use them in R, one can consult the main help page on this topic with <code>?regexp</code>.</p>
<h4 id="string-matching-with-grep">String matching with <code>grep</code></h4>
<p>The grep function can be used for finding patterns in strings, here letter <code>A</code> in vector <code>month.name</code>.</p>
<pre><code class="language-r">month.name[grep(&quot;A&quot;, month.name)] 
</code></pre>
<pre><code>## [1] &quot;April&quot;  &quot;August&quot;
</code></pre>
<h4 id="string-substitution-with-gsub">String substitution with <code>gsub</code></h4>
<p>Example for using regular expressions to substitute a pattern by another one using a back reference. Remember: single escapes <code>\</code> need to be double escaped <code>\\</code> in R.</p>
<pre><code class="language-r">gsub('(i.*a)', 'xxx_\\1', &quot;virginica&quot;, perl = TRUE) 
</code></pre>
<pre><code>## [1] &quot;vxxx_irginica&quot;
</code></pre>
<h3 id="interpreting-a-character-string-as-expression">Interpreting a Character String as Expression</h3>
<p>Some useful examples</p>
<p>Generates vector of object names in session</p>
<pre><code class="language-r">myfct &lt;- function(x) x^2
mylist &lt;- ls()
n &lt;- which(mylist %in% &quot;myfct&quot;)
mylist[n] 
</code></pre>
<pre><code>## [1] &quot;myfct&quot;
</code></pre>
<p>Executes entry in position <code>n</code> as expression</p>
<pre><code class="language-r">get(mylist[n])
</code></pre>
<pre><code>## function(x) x^2
</code></pre>
<pre><code class="language-r">get(mylist[n])(2)
</code></pre>
<pre><code>## [1] 4
</code></pre>
<p>Alternative approach</p>
<pre><code class="language-r">eval(parse(text=mylist[n])) 
</code></pre>
<pre><code>## function(x) x^2
</code></pre>
<h3 id="replacement-split-and-paste-functions-for-strings">Replacement, Split and Paste Functions for Strings</h3>
<p><strong>Selected examples</strong></p>
<p>Substitution with back reference which inserts in this example <code>_</code> character</p>
<pre><code class="language-r">x &lt;- gsub(&quot;(a)&quot;,&quot;\\1_&quot;, month.name[1], perl=T) 
x
</code></pre>
<pre><code>## [1] &quot;Ja_nua_ry&quot;
</code></pre>
<p>Split string on inserted character from above</p>
<pre><code class="language-r">strsplit(x,&quot;_&quot;)
</code></pre>
<pre><code>## [[1]]
## [1] &quot;Ja&quot;  &quot;nua&quot; &quot;ry&quot;
</code></pre>
<p>Reverse a character string by splitting first all characters into vector fields</p>
<pre><code class="language-r">paste(rev(unlist(strsplit(x, NULL))), collapse=&quot;&quot;) 
</code></pre>
<pre><code>## [1] &quot;yr_aun_aJ&quot;
</code></pre>
<h3 id="time-date-and-sleep">Time, Date and Sleep</h3>
<p><strong>Selected examples</strong></p>
<p>Return CPU (and other) times that an expression used (here ls)</p>
<pre><code class="language-r">system.time(ls()) 
</code></pre>
<pre><code>##    user  system elapsed 
##       0       0       0
</code></pre>
<p>Return the current system date and time</p>
<pre><code class="language-r">date() 
</code></pre>
<pre><code>## [1] &quot;Thu Apr 14 09:21:47 2022&quot;
</code></pre>
<p>Pause execution of R expressions for a given number of seconds (e.g. in loop)</p>
<pre><code class="language-r">Sys.sleep(1) 
</code></pre>
<h4 id="example">Example</h4>
<h5 id="import-of-specific-file-lines-with-regular-expression">Import of Specific File Lines with Regular Expression</h5>
<p>The following example demonstrates the retrieval of specific lines from an external file with a regular expression. First, an external file is created with the <code>cat</code> function, all lines of this file are imported into a vector with <code>readLines</code>, the specific elements (lines) are then retieved with the <code>grep</code> function, and the resulting lines are split into vector fields with <code>strsplit</code>.</p>
<pre><code class="language-r">cat(month.name, file=&quot;zzz.txt&quot;, sep=&quot;\n&quot;)
x &lt;- readLines(&quot;zzz.txt&quot;)
x[1:6] 
</code></pre>
<pre><code>## [1] &quot;January&quot;  &quot;February&quot; &quot;March&quot;    &quot;April&quot;    &quot;May&quot;      &quot;June&quot;
</code></pre>
<pre><code class="language-r">x &lt;- x[c(grep(&quot;^J&quot;, as.character(x), perl = TRUE))]
t(as.data.frame(strsplit(x, &quot;u&quot;)))
</code></pre>
<pre><code>##                 [,1]  [,2] 
## c..Jan....ary.. &quot;Jan&quot; &quot;ary&quot;
## c..J....ne..    &quot;J&quot;   &quot;ne&quot; 
## c..J....ly..    &quot;J&quot;   &quot;ly&quot;
</code></pre>
<h2 id="calling-external-software">Calling External Software</h2>
<p>External command-line software can be called with <code>system</code>. The following example calls <code>blastall</code> from R</p>
<pre><code class="language-r">system(&quot;blastall -p blastp -i seq.fasta -d uniprot -o seq.blastp&quot;)
</code></pre>
<h2 id="running-r-scripts">Running R Scripts</h2>
<h3 id="possibilities-for-executing-r-scripts">Possibilities for Executing R Scripts</h3>
<h4 id="r-console">R console</h4>
<pre><code class="language-r">source(&quot;my_script.R&quot;)
</code></pre>
<h4 id="command-line">Command-line</h4>
<pre><code class="language-sh">Rscript my_script.R # or just ./myscript.R after making it executable
R CMD BATCH my_script.R # Alternative way 1 
R --slave &lt; my_script.R # Alternative way 2
</code></pre>
<h4 id="passing-arguments-from-command-line-to-r">Passing arguments from command-line to R</h4>
<p>Create an R script named <code>test.R</code> with the following content:</p>
<pre><code class="language-sh">myarg &lt;- commandArgs()
print(iris[1:myarg[6], ])
</code></pre>
<p>Then run it from the command-line like this:</p>
<pre><code class="language-sh">Rscript test.R 10
</code></pre>
<p>In the given example the number <code>10</code> is passed on from the command-line as an argument to the R script which is used to return to <code>STDOUT</code> the first 10 rows of the <code>iris</code> sample data. If several arguments are provided, they will be interpreted as one string and need to be split in R with the strsplit function. A more detailed example can be found <a href="http://manuals.bioinformatics.ucr.edu/home/ht-seq#TOC-Quality-Reports-of-FASTQ-Files-">here</a>.</p>
<h2 id="building-r-packages">Building R Packages</h2>
<p>This section has been moved to a dedicated tutorial on R package development <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/rpackages/rpackages/">here</a>.</p>
<h2 id="programming-exercises">Programming Exercises</h2>
<h3 id="exercise-1">Exercise 1</h3>
<h4 id="for-loop-1"><code>for</code> loop</h4>
<p><strong>Task 1.1</strong>: Compute the mean of each row in <code>myMA</code> by applying the mean function in a <code>for</code> loop.</p>
<pre><code class="language-r">myMA &lt;- matrix(rnorm(500), 100, 5, dimnames=list(1:100, paste(&quot;C&quot;, 1:5, sep=&quot;&quot;)))
myve_for &lt;- NULL
for(i in seq(along=myMA[,1])) {
    myve_for &lt;- c(myve_for, mean(as.numeric(myMA[i, ])))
}
myResult &lt;- cbind(myMA, mean_for=myve_for)
myResult[1:4, ]
</code></pre>
<pre><code>##          C1         C2         C3         C4         C5    mean_for
## 1 -1.202619 -0.1053735 -1.1074995  0.4269033 -1.3932608 -0.67636998
## 2  1.174138 -0.8991985 -1.2658862  1.0151751 -0.4075398 -0.07666225
## 3  1.357860 -1.2370991  0.7402124 -1.0277818  0.4851062  0.06365953
## 4  1.173512 -1.2724114  0.2270477 -1.5421106  0.7484669 -0.13309912
</code></pre>
<h4 id="while-loop-1"><code>while</code> loop</h4>
<p><strong>Task 1.2</strong>: Compute the mean of each row in <code>myMA</code> by applying the mean function in a <code>while</code> loop.</p>
<pre><code class="language-r">z &lt;- 1
myve_while &lt;- NULL
while(z &lt;= length(myMA[,1])) {
    myve_while &lt;- c(myve_while, mean(as.numeric(myMA[z, ])))
    z &lt;- z + 1
}
myResult &lt;- cbind(myMA, mean_for=myve_for, mean_while=myve_while)
myResult[1:4, -c(1,2)]
</code></pre>
<pre><code>##           C3         C4         C5    mean_for  mean_while
## 1 -1.1074995  0.4269033 -1.3932608 -0.67636998 -0.67636998
## 2 -1.2658862  1.0151751 -0.4075398 -0.07666225 -0.07666225
## 3  0.7402124 -1.0277818  0.4851062  0.06365953  0.06365953
## 4  0.2270477 -1.5421106  0.7484669 -0.13309912 -0.13309912
</code></pre>
<p><strong>Task 1.3</strong>: Confirm that the results from both mean calculations are identical</p>
<pre><code class="language-r">all(myResult[,6] == myResult[,7])
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<h4 id="apply-loop"><code>apply</code> loop</h4>
<p><strong>Task 1.4</strong>: Compute the mean of each row in myMA by applying the mean function in an <code>apply</code> loop</p>
<pre><code class="language-r">myve_apply &lt;- apply(myMA, 1, mean)
myResult &lt;- cbind(myMA, mean_for=myve_for, mean_while=myve_while, mean_apply=myve_apply)
myResult[1:4, -c(1,2)]
</code></pre>
<pre><code>##           C3         C4         C5    mean_for  mean_while  mean_apply
## 1 -1.1074995  0.4269033 -1.3932608 -0.67636998 -0.67636998 -0.67636998
## 2 -1.2658862  1.0151751 -0.4075398 -0.07666225 -0.07666225 -0.07666225
## 3  0.7402124 -1.0277818  0.4851062  0.06365953  0.06365953  0.06365953
## 4  0.2270477 -1.5421106  0.7484669 -0.13309912 -0.13309912 -0.13309912
</code></pre>
<h4 id="avoiding-loops">Avoiding loops</h4>
<p><strong>Task 1.5</strong>: When operating on large data sets it is much faster to use the <code>rowMeans</code> function</p>
<pre><code class="language-r">mymean &lt;- rowMeans(myMA)
myResult &lt;- cbind(myMA, mean_for=myve_for, mean_while=myve_while, mean_apply=myve_apply, mean_int=mymean)
myResult[1:4, -c(1,2,3)]
</code></pre>
<pre><code>##           C4         C5    mean_for  mean_while  mean_apply    mean_int
## 1  0.4269033 -1.3932608 -0.67636998 -0.67636998 -0.67636998 -0.67636998
## 2  1.0151751 -0.4075398 -0.07666225 -0.07666225 -0.07666225 -0.07666225
## 3 -1.0277818  0.4851062  0.06365953  0.06365953  0.06365953  0.06365953
## 4 -1.5421106  0.7484669 -0.13309912 -0.13309912 -0.13309912 -0.13309912
</code></pre>
<p>To find out which other built-in functions for basic calculations exist, type <code>?rowMeans</code>.</p>
<h3 id="exercise-2">Exercise 2</h3>
<h4 id="custom-functions">Custom functions</h4>
<p><strong>Task 2.1</strong>: Use the following code as basis to implement a function that allows the user to compute the mean for any combination of columns in a matrix or data frame. The first argument of this function should specify the input data set, the second the mathematical function to be passed on (<em>e.g.</em> <code>mean</code>, <code>sd</code>, <code>max</code>) and the third one should allow the selection of the columns by providing a grouping vector.</p>
<pre><code class="language-r">myMA &lt;- matrix(rnorm(100000), 10000, 10, dimnames=list(1:10000, paste(&quot;C&quot;, 1:10, sep=&quot;&quot;)))
myMA[1:2,]
</code></pre>
<pre><code>##           C1         C2        C3            C4        C5        C6         C7         C8
## 1  0.1750576 -0.2826930 0.9290137  0.0830523803 1.4113331 -1.221849 -0.4663186  0.6170635
## 2 -0.8056727  0.1885557 0.4891758 -0.0003536156 0.8356054  1.406484  1.0363437 -0.1351562
##           C9       C10
## 1 0.58305534 0.1716817
## 2 0.09596605 0.4551591
</code></pre>
<pre><code class="language-r">myList &lt;- tapply(colnames(myMA), c(1,1,1,2,2,2,3,3,4,4), list) 
names(myList) &lt;- sapply(myList, paste, collapse=&quot;_&quot;)
myMAmean &lt;- sapply(myList, function(x) apply(myMA[,x], 1, mean))
myMAmean[1:4,] 
</code></pre>
<pre><code>##      C1_C2_C3      C4_C5_C6       C7_C8     C9_C10
## 1  0.27379276  0.0908453513  0.07537248  0.3773685
## 2 -0.04264706  0.7472451361  0.45059378  0.2755626
## 3  0.49411343 -0.4486259202 -0.08519687  1.0929435
## 4  0.06575562 -0.0007494971  0.07968924 -0.7374505
</code></pre>
<!---
Solution

-->
<h3 id="exercise-3">Exercise 3</h3>
<h4 id="nested-loops-to-generate-similarity-matrices">Nested loops to generate similarity matrices</h4>
<p><strong>Task 3.1</strong>: Create a sample list populated with character vectors of different lengths</p>
<pre><code class="language-r">setlist &lt;- lapply(11:30, function(x) sample(letters, x, replace=TRUE))
names(setlist) &lt;- paste(&quot;S&quot;, seq(along=setlist), sep=&quot;&quot;) 
setlist[1:6]
</code></pre>
<pre><code>## $S1
##  [1] &quot;b&quot; &quot;s&quot; &quot;o&quot; &quot;w&quot; &quot;w&quot; &quot;x&quot; &quot;s&quot; &quot;z&quot; &quot;d&quot; &quot;f&quot; &quot;b&quot;
## 
## $S2
##  [1] &quot;e&quot; &quot;w&quot; &quot;l&quot; &quot;y&quot; &quot;i&quot; &quot;l&quot; &quot;m&quot; &quot;q&quot; &quot;s&quot; &quot;r&quot; &quot;d&quot; &quot;s&quot;
## 
## $S3
##  [1] &quot;h&quot; &quot;i&quot; &quot;t&quot; &quot;d&quot; &quot;f&quot; &quot;c&quot; &quot;u&quot; &quot;r&quot; &quot;v&quot; &quot;a&quot; &quot;n&quot; &quot;t&quot; &quot;v&quot;
## 
## $S4
##  [1] &quot;o&quot; &quot;w&quot; &quot;c&quot; &quot;x&quot; &quot;c&quot; &quot;l&quot; &quot;a&quot; &quot;z&quot; &quot;c&quot; &quot;f&quot; &quot;n&quot; &quot;i&quot; &quot;j&quot; &quot;u&quot;
## 
## $S5
##  [1] &quot;a&quot; &quot;z&quot; &quot;b&quot; &quot;t&quot; &quot;p&quot; &quot;l&quot; &quot;u&quot; &quot;z&quot; &quot;q&quot; &quot;x&quot; &quot;y&quot; &quot;h&quot; &quot;l&quot; &quot;j&quot; &quot;e&quot;
## 
## $S6
##  [1] &quot;z&quot; &quot;v&quot; &quot;n&quot; &quot;v&quot; &quot;b&quot; &quot;m&quot; &quot;p&quot; &quot;p&quot; &quot;z&quot; &quot;g&quot; &quot;l&quot; &quot;j&quot; &quot;z&quot; &quot;z&quot; &quot;t&quot; &quot;e&quot;
</code></pre>
<p><strong>Task 3.2</strong>: Compute the length for all pairwise intersects of the vectors stored in <code>setlist</code>. The intersects can be determined with the <code>%in%</code> function like this: <code>sum(setlist[[1]] %in% setlist[[2]])</code></p>
<pre><code class="language-r">setlist &lt;- sapply(setlist, unique)
olMA &lt;- sapply(names(setlist), function(x) sapply(names(setlist), 
               function(y) sum(setlist[[x]] %in% setlist[[y]])))
olMA[1:12,] 
</code></pre>
<pre><code>##     S1 S2 S3 S4 S5 S6 S7 S8 S9 S10 S11 S12 S13 S14 S15 S16 S17 S18 S19 S20
## S1   8  3  2  5  3  2  4  4  4   5   8   5   5   5   6   5   4   6   4   4
## S2   3 10  3  3  4  3  5  3  3   5   4   6   6   8   7   5   5   8   7   8
## S3   2  3 11  6  4  3  5  4  4   7   4   8   4   5   8   6   6   6   7   8
## S4   5  3  6 12  6  4  8  4  4   8   8   7   7   7   8   9   6   8   9   9
## S5   3  4  4  6 13  7  3  6  6   8   7   8   5  10   8   9   6  10  10  10
## S6   2  3  3  4  7 11  4  5  5   5   5   7   4   8   7   6   8   8   9   8
## S7   4  5  5  8  3  4 11  5  6   5   6   6   7   6   9   6   7   7   8   9
## S8   4  3  4  4  6  5  5 12  7   4   7   6   5   7  10   7   8   8   9   7
## S9   4  3  4  4  6  5  6  7 11   6   8   6   3   8  10   6   6   8   8   8
## S10  5  5  7  8  8  5  5  4  6  14   7   9   5   9   9  10   7  11   9   8
## S11  8  4  4  8  7  5  6  7  8   7  13   7   6  10  10   9   8   9   8   8
## S12  5  6  8  7  8  7  6  6  6   9   7  15   5  10  10   9   9  11  10  10
</code></pre>
<p><strong>Task 3.3</strong> Plot the resulting intersect matrix as heat map.
The <code>image</code> or the <code>pheatmap</code> functions can be used for this.</p>
<pre><code class="language-r">library(pheatmap); library(&quot;RColorBrewer&quot;)
pheatmap(olMA, color=brewer.pal(9,&quot;Blues&quot;), cluster_rows=FALSE, cluster_cols=FALSE, display_numbers=TRUE, number_format=&quot;%.0f&quot;, fontsize_number=10)
</code></pre>
<img src="../../../en/tutorials/rprogramming/rprogramming_files/figure-html/nested_loops3-1.png" width="672" />
<pre><code class="language-r"># image(olMA) 
</code></pre>
<h3 id="exercise-4">Exercise 4</h3>
<h4 id="build-your-own-r-package">Build your own R package</h4>
<p><strong>Task 4.1</strong>: Save one or more of your functions to a file called <code>script.R</code> and build the package with the <code>package.skeleton</code> function.</p>
<pre><code class="language-r">package.skeleton(name=&quot;mypackage&quot;, code_files=c(&quot;script1.R&quot;))
</code></pre>
<p><strong>Task 4.2</strong>: Build tarball of the package</p>
<pre><code class="language-r">system(&quot;R CMD build mypackage&quot;)
</code></pre>
<p><strong>Task 4.3</strong>: Install and use package</p>
<pre><code class="language-r">install.packages(&quot;mypackage_1.0.tar.gz&quot;, repos=NULL, type=&quot;source&quot;)
library(mypackage)
?myMAcomp # Opens help for function defined by mypackage
</code></pre>
<h2 id="homework-5">Homework 5</h2>
<p>See homework section <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw05/hw05/">here</a>.</p>
<h2 id="additional-exercises">Additional Exercises</h2>
<h3 id="pattern-matching-and-positional-parsing-of-equences">Pattern matching and positional parsing of equences</h3>
<p>The following sample script <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/content/en/tutorials/rprogramming/scripts/patternSearch.R">patternSearch.R</a> defines
functions for importing sequences into R, retrieving reverse and complement of nucleotide sequences, pattern searching, positional parsing and exporting
search results in HTML format. Sourcing the script will return usage instructions of its functions.</p>
<pre><code class="language-r">source(&quot;https://raw.githubusercontent.com/tgirke/GEN242/main/content/en/tutorials/rprogramming/scripts/patternSearch.R&quot;)
</code></pre>
<h3 id="identify-over-represented-strings-in-sequence-sets">Identify over-represented strings in sequence sets</h3>
<p>Example functions for finding over-represented words in sets of DNA, RNA or protein sequences
are defined in this script: <a href="https://raw.githubusercontent.com/tgirke/GEN242/main/content/en/tutorials/rprogramming/scripts/wordFinder.R">wordFinder.R</a>.
Sourcing the script will return usage instructions of its functions.</p>
<pre><code class="language-r">source(&quot;https://raw.githubusercontent.com/tgirke/GEN242/main/content/en/tutorials/rprogramming/scripts/wordFinder.R&quot;)
</code></pre>
<h2 id="object-oriented-programming-oop">Object-Oriented Programming (OOP)</h2>
<p>R supports several systems for object-oriented programming (OOP). This includes
an older S3 system, and the more recently introduced R6 and S4 systems. The latter is the
most formal version that supports multiple inheritance, multiple dispatch and
introspection. Many of these features are not available in the older S3
system. In general, the OOP approach taken by R is to separate the class
specifications from the specifications of generic functions (function-centric
system). The following introduction is restricted to the S4 system since it is
nowadays the preferred OOP method for package development in Bioconductor. More
information about OOP in R can be found in the following introductions:</p>
<ul>
<li><a href="http://zoonek2.free.fr/UNIX/48_R/02.html#4">Vincent Zoonekynd’s introduction to S3 Classes</a></li>
<li><a href="https://cran.r-project.org/doc/contrib/Genolini-S4tutorialV0-5en.pdf">Christophe Genolini’s S4 Intro</a></li>
<li><a href="http://master.bioconductor.org/help/course-materials/2008/advanced_R/">Advanced Bioconductor Courses</a></li>
<li><a href="https://www.springer.com/gp/book/9780387759357">Programming with R by John Chambers</a></li>
<li><a href="http://www.bioconductor.org/help/publications/books/r-programming-for-bioinformatics/">R Programming for Bioinformatics by Robert Gentleman</a></li>
<li><a href="https://adv-r.hadley.nz/r6.html">Advanced R online book by Hadley Wichham</a></li>
</ul>
<h3 id="define-s4-classes">Define S4 Classes</h3>
<h4 id="1-define-s4-classes-with-setclass-and-new">1. Define S4 Classes with <code>setClass()</code> and <code>new()</code></h4>
<pre><code class="language-r">y &lt;- matrix(1:10, 2, 5) # Sample data set
setClass(Class=&quot;myclass&quot;,
    representation=representation(a=&quot;ANY&quot;),
    prototype=prototype(a=y[1:2,]), # Defines default value (optional)
    validity=function(object) { # Can be defined in a separate step using setValidity
        if(class(object@a)[1]!=&quot;matrix&quot;) {
            return(paste(&quot;expected matrix, but obtained&quot;, class(object@a)))
        } else {
            return(TRUE)
        }
    }
)
</code></pre>
<p>The setClass function defines classes. Its most important arguments are</p>
<ul>
<li><code>Class</code>: the name of the class</li>
<li><code>representation</code>: the slots that the new class should have and/or other classes that this class extends.</li>
<li><code>prototype</code>: an object providing default data for the slots.</li>
<li><code>contains</code>: the classes that this class extends.</li>
<li><code>validity</code>, <code>access</code>, <code>version</code>: control arguments included for compatibility with S-Plus.</li>
<li><code>where</code>: the environment to use to store or remove the definition as meta data.</li>
</ul>
<h4 id="2-create-new-class-instance">2. Create new class instance</h4>
<p>The function <code>new</code> creates an instance of a class (here <code>myclass</code>).</p>
<pre><code class="language-r">myobj &lt;- new(&quot;myclass&quot;, a=y)
myobj
</code></pre>
<pre><code>## An object of class &quot;myclass&quot;
## Slot &quot;a&quot;:
##      [,1] [,2] [,3] [,4] [,5]
## [1,]    1    3    5    7    9
## [2,]    2    4    6    8   10
</code></pre>
<p>If evaluated the following would return an error due to wrong input type (<code>data.frame</code> instead of <code>matrix</code>).</p>
<pre><code class="language-r">new(&quot;myclass&quot;, a=iris) # Returns error due to wrong input  
</code></pre>
<p>The arguments of <code>new</code> are:</p>
<ul>
<li><code>Class</code>: the name of the class</li>
<li><code>...</code>: data to include in the new object with arguments according to slots in class definition</li>
</ul>
<h4 id="3-initialization-method">3. Initialization method</h4>
<p>A more generic way of creating class instances is to define an initialization
method (more details below).</p>
<pre><code class="language-r">setMethod(&quot;initialize&quot;, &quot;myclass&quot;, function(.Object, a) {
    .Object@a &lt;- a/a
    .Object
})
new(&quot;myclass&quot;, a = y)
</code></pre>
<pre><code>## An object of class &quot;myclass&quot;
## Slot &quot;a&quot;:
##      [,1] [,2] [,3] [,4] [,5]
## [1,]    1    1    1    1    1
## [2,]    1    1    1    1    1
</code></pre>
<h4 id="4-usage-and-helper-functions">4. Usage and helper functions</h4>
<p>The ‘@’ operator extracts the contents of a slot. Its usage should be limited to internal
functions.</p>
<pre><code class="language-r">myobj@a 
</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,]    1    3    5    7    9
## [2,]    2    4    6    8   10
</code></pre>
<p>Create a new S4 object from an old one.</p>
<pre><code class="language-r">initialize(.Object=myobj, a=as.matrix(cars[1:2,])) 
</code></pre>
<pre><code>## An object of class &quot;myclass&quot;
## Slot &quot;a&quot;:
##   speed dist
## 1     1    1
## 2     1    1
</code></pre>
<p>If evaluated the <code>removeClass</code> function removes an object from the current session.
This does not apply to associated methods.</p>
<pre><code class="language-r">removeClass(&quot;myclass&quot;) 
</code></pre>
<h4 id="5-inheritance">5. Inheritance</h4>
<p>Inheritance allows to define new classes that inherit all properties (e.g. data slots, methods)
from their existing parent classes. The <code>contains</code> argument used below allows to extend
existing classes. This propagates all slots of parent classes.</p>
<pre><code class="language-r">setClass(&quot;myclass1&quot;, representation(a = &quot;character&quot;, b = &quot;character&quot;))
setClass(&quot;myclass2&quot;, representation(c = &quot;numeric&quot;, d = &quot;numeric&quot;))
setClass(&quot;myclass3&quot;, contains=c(&quot;myclass1&quot;, &quot;myclass2&quot;))
new(&quot;myclass3&quot;, a=letters[1:4], b=letters[1:4], c=1:4, d=4:1)
</code></pre>
<pre><code>## An object of class &quot;myclass3&quot;
## Slot &quot;a&quot;:
## [1] &quot;a&quot; &quot;b&quot; &quot;c&quot; &quot;d&quot;
## 
## Slot &quot;b&quot;:
## [1] &quot;a&quot; &quot;b&quot; &quot;c&quot; &quot;d&quot;
## 
## Slot &quot;c&quot;:
## [1] 1 2 3 4
## 
## Slot &quot;d&quot;:
## [1] 4 3 2 1
</code></pre>
<pre><code class="language-r">getClass(&quot;myclass1&quot;)
</code></pre>
<pre><code>## Class &quot;myclass1&quot; [in &quot;.GlobalEnv&quot;]
## 
## Slots:
##                           
## Name:          a         b
## Class: character character
## 
## Known Subclasses: &quot;myclass3&quot;
</code></pre>
<pre><code class="language-r">getClass(&quot;myclass2&quot;)
</code></pre>
<pre><code>## Class &quot;myclass2&quot; [in &quot;.GlobalEnv&quot;]
## 
## Slots:
##                       
## Name:        c       d
## Class: numeric numeric
## 
## Known Subclasses: &quot;myclass3&quot;
</code></pre>
<pre><code class="language-r">getClass(&quot;myclass3&quot;)
</code></pre>
<pre><code>## Class &quot;myclass3&quot; [in &quot;.GlobalEnv&quot;]
## 
## Slots:
##                                               
## Name:          a         b         c         d
## Class: character character   numeric   numeric
## 
## Extends: &quot;myclass1&quot;, &quot;myclass2&quot;
</code></pre>
<h4 id="6-coerce-objects-to-another-class">6. Coerce objects to another class</h4>
<p>The following defines a coerce method. After this the standard <code>as(..., &quot;...&quot;)</code>
syntax can be used to coerce the new class to another one.</p>
<pre><code class="language-r">setAs(from=&quot;myclass&quot;, to=&quot;character&quot;, def=function(from) as.character(as.matrix(from@a)))
as(myobj, &quot;character&quot;)
</code></pre>
<pre><code>##  [1] &quot;1&quot;  &quot;2&quot;  &quot;3&quot;  &quot;4&quot;  &quot;5&quot;  &quot;6&quot;  &quot;7&quot;  &quot;8&quot;  &quot;9&quot;  &quot;10&quot;
</code></pre>
<h4 id="7-virtual-classes">7. Virtual classes</h4>
<p>Virtual classes are constructs for which no instances will be or can be
created. They are used to link together classes which may have distinct
representations (e.g. cannot inherit from each other) but for which one wants
to provide similar functionality. Often it is desired to create a virtual class
and to then have several other classes extend it. Virtual classes can be
defined by leaving out the representation argument or including the class
<code>VIRTUAL</code> as illustrated here:</p>
<pre><code class="language-r">setClass(&quot;myVclass&quot;)
setClass(&quot;myVclass&quot;, representation(a = &quot;character&quot;, &quot;VIRTUAL&quot;))
</code></pre>
<h4 id="8-introspection-of-classes">8. Introspection of classes</h4>
<p>Useful functions to introspect classes include:</p>
<ul>
<li><code>getClass(&quot;myclass&quot;)</code></li>
<li><code>getSlots(&quot;myclass&quot;)</code></li>
<li><code>slotNames(&quot;myclass&quot;)</code></li>
<li><code>extends(&quot;myclass2&quot;)</code></li>
</ul>
<h3 id="assign-generics-and-methods">Assign Generics and Methods</h3>
<p>Generics and methods can be assigned with the methods <code>setGeneric()</code> and <code>setMethod()</code>.</p>
<h4 id="1-accessor-functions">1. Accessor functions</h4>
<p>This avoids the usage of the <code>@</code> operator.</p>
<pre><code class="language-r">setGeneric(name=&quot;acc&quot;, def=function(x) standardGeneric(&quot;acc&quot;))
</code></pre>
<pre><code>## [1] &quot;acc&quot;
</code></pre>
<pre><code class="language-r">setMethod(f=&quot;acc&quot;, signature=&quot;myclass&quot;, definition=function(x) {
    return(x@a)
})
acc(myobj)
</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5]
## [1,]    1    3    5    7    9
## [2,]    2    4    6    8   10
</code></pre>
<h4 id="2-replacement-methods">2. Replacement methods</h4>
<ol>
<li>Using custom accessor function with <code>acc &lt;-</code> syntax.</li>
</ol>
<!-- end list -->
<pre><code class="language-r">setGeneric(name=&quot;acc&lt;-&quot;, def=function(x, value) standardGeneric(&quot;acc&lt;-&quot;))
</code></pre>
<pre><code>## [1] &quot;acc&lt;-&quot;
</code></pre>
<pre><code class="language-r">setReplaceMethod(f=&quot;acc&quot;, signature=&quot;myclass&quot;, definition=function(x, value) {
                 x@a &lt;- value
                 return(x)
})
## After this the following replace operations with 'acc' work on new object class
acc(myobj)[1,1] &lt;- 999 # Replaces first value
colnames(acc(myobj)) &lt;- letters[1:5] # Assigns new column names
rownames(acc(myobj)) &lt;- letters[1:2] # Assigns new row names
myobj
</code></pre>
<pre><code>## An object of class &quot;myclass&quot;
## Slot &quot;a&quot;:
##     a b c d  e
## a 999 3 5 7  9
## b   2 4 6 8 10
</code></pre>
<ol start="2">
<li>Replacement method using <code>[</code> operator, here <code>[...] &lt;-</code> syntax.</li>
</ol>
<!-- end list -->
<pre><code class="language-r">setReplaceMethod(f=&quot;[&quot;, signature=&quot;myclass&quot;, definition=function(x, i, j, value) {
                 x@a[i,j] &lt;- value
                 return(x)
})
myobj[1,2] &lt;- 999
myobj
</code></pre>
<pre><code>## An object of class &quot;myclass&quot;
## Slot &quot;a&quot;:
##     a   b c d  e
## a 999 999 5 7  9
## b   2   4 6 8 10
</code></pre>
<h4 id="3-behavior-of-bracket-operator">3. Behavior of bracket operator</h4>
<p>The behavior of the bracket <code>[</code> subsetting operator can be defined as follows.</p>
<pre><code class="language-r">setMethod(f=&quot;[&quot;, signature=&quot;myclass&quot;,
          definition=function(x, i, j, ..., drop) {
          x@a &lt;- x@a[i,j]
          return(x)
})
myobj[1:2, 1:3] # Standard subsetting works now on new class
</code></pre>
<pre><code>## An object of class &quot;myclass&quot;
## Slot &quot;a&quot;:
##     a   b c
## a 999 999 5
## b   2   4 6
</code></pre>
<h4 id="4-print-behavior">4. Print behavior</h4>
<p>A convient summary printing behavior for a new class should always be defined.</p>
<pre><code class="language-r">setMethod(f=&quot;show&quot;, signature=&quot;myclass&quot;, definition=function(object) {
          cat(&quot;An instance of &quot;, &quot;\&quot;&quot;, class(object), &quot;\&quot;&quot;, &quot; with &quot;, length(acc(object)[,1]), &quot; elements&quot;, &quot;\n&quot;, sep=&quot;&quot;)
          if(length(acc(object)[,1])&gt;=5) {
                print(as.data.frame(rbind(acc(object)[1:2,], ...=rep(&quot;...&quot;, length(acc(object)[1,])), acc(object)[(length(acc(object)[,1])-1):length(acc(object)[,1]),])))
          } else {
                print(acc(object))
          }
})
myobj # Prints object with custom method
</code></pre>
<pre><code>## An instance of &quot;myclass&quot; with 2 elements
##     a   b c d  e
## a 999 999 5 7  9
## b   2   4 6 8 10
</code></pre>
<h4 id="5-define-custom-methods">5. Define custom methods</h4>
<p>The following gives an example for defining a data specific method, here randomizing row
order of matrix stored in new S4 class.</p>
<pre><code class="language-r">setGeneric(name=&quot;randomize&quot;, def=function(x) standardGeneric(&quot;randomize&quot;))
</code></pre>
<pre><code>## [1] &quot;randomize&quot;
</code></pre>
<pre><code class="language-r">setMethod(f=&quot;randomize&quot;, signature=&quot;myclass&quot;, definition=function(x) {
          acc(x)[sample(1:length(acc(x)[,1]), length(acc(x)[,1])), ]
})
randomize(myobj)
</code></pre>
<pre><code>##     a   b c d  e
## b   2   4 6 8 10
## a 999 999 5 7  9
</code></pre>
<h4 id="6-plotting-method">6. Plotting method</h4>
<p>Define a graphical plotting method for new class and allow users to access it with
R’s generic <code>plot</code> function.</p>
<pre><code class="language-r">setMethod(f=&quot;plot&quot;, signature=&quot;myclass&quot;, definition=function(x, ...) {
          barplot(as.matrix(acc(x)), ...)
})
plot(myobj)
</code></pre>
<img src="../../../en/tutorials/rprogramming/rprogramming_files/figure-html/s4_plot_methods-1.png" width="672" />
<h4 id="7-utilities-to-inspect-methods">7. Utilities to inspect methods</h4>
<p>Important inspection methods for classes include:</p>
<ul>
<li><code>showMethods(class=&quot;myclass&quot;)</code></li>
<li><code>findMethods(&quot;randomize&quot;)</code></li>
<li><code>getMethod(&quot;randomize&quot;, signature=&quot;myclass&quot;)</code></li>
<li><code>existsMethod(&quot;randomize&quot;, signature=&quot;myclass&quot;)</code></li>
</ul>
<h2 id="session-info">Session Info</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.1.3 (2022-03-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] RColorBrewer_1.1-2 pheatmap_1.0.12    BiocStyle_2.22.0  
## 
## loaded via a namespace (and not attached):
##  [1] knitr_1.37          magrittr_2.0.2      munsell_0.5.0       colorspace_2.0-2   
##  [5] R6_2.5.1            rlang_1.0.2         fastmap_1.1.0       highr_0.9          
##  [9] stringr_1.4.0       tools_4.1.3         grid_4.1.3          gtable_0.3.0       
## [13] xfun_0.30           cli_3.1.0           jquerylib_0.1.4     htmltools_0.5.2    
## [17] yaml_2.3.5          digest_0.6.29       lifecycle_1.0.1     bookdown_0.24      
## [21] BiocManager_1.30.16 codetools_0.2-18    sass_0.4.0          evaluate_0.15      
## [25] rmarkdown_2.13      blogdown_1.8.2      stringi_1.7.6       compiler_4.1.3     
## [29] bslib_0.3.1         scales_1.1.1        jsonlite_1.8.0
</code></pre>
<h2 id="references">References</h2>
<div id="refs" class="references">
<div id="ref-Gentleman2008-xo">
<p>Gentleman, Robert. 2008. <em>R Programming for Bioinformatics (Chapman &amp; Hall/CRC Computer Science &amp; Data Analysis)</em>. 1 edition. Chapman; Hall/CRC. <a href="http://www.amazon.com/Programming-Bioinformatics-Chapman-Computer-Analysis/dp/1420063677">http://www.amazon.com/Programming-Bioinformatics-Chapman-Computer-Analysis/dp/1420063677</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2022-04-14: <a  href="https://github.com/tgirke/GEN242//commit/f3a05e82719c971074afcf6d8d0699f1f2e4e976">some edits (f3a05e82)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2022 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
