
---
title: Student Paper Presentations
linkTitle: "Paper Presentations"
description: >
type: docs
weight: 501
---


## Overview

Each student has been assigned one journal publication to present in class. The
expected structure of the paper presentations is outlined in this [Slideshow Template](https://docs.google.com/presentation/d/14Jdss8C5YvwkBsTxL-Z4HLE5W8RkfxprOpK-_-dzSCY/edit?usp=sharing). 
A detailed presentation schedule is available on the internal [Course Schedule](https://elearn.ucr.edu/courses/134360). 
The grading of both the paper and project presentations (2-3) includes
anonymous feedback from all students as well as the instructor, where
understanding of the material, clarity of the oral presentations and critical
thinking are the main grading criteria. The grading form is [here](). 
The following lists the assigned papers organized by course project topics.


## Publications organized by course projects

All references in [Paperpile](https://paperpile.com/shared/yWql4M)

### RNA-Seq Aligners 

1. Kim D, Pertea G, Trapnell C, Pimentel H, Kelley R, Salzberg SL (2013) TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions. Genome Biol 14: R36. [PubMed](https://pubmed.ncbi.nlm.nih.gov/23618408/)
2. Kim D, Langmead B, Salzberg SL (2015) HISAT: a fast spliced aligner with low memory requirements. Nat Methods 12: 357–360. [PubMed](https://pubmed.ncbi.nlm.nih.gov/25751142/)

### DEG Methods  

3. Guo Y, Li C-I, Ye F, Shyr Y (2013) Evaluation of read count based RNAseq analysis methods. BMC Genomics 14 Suppl 8: S2. [PubMed](https://pubmed.ncbi.nlm.nih.gov/24564449/)
4. Love MI, Huber W, Anders S (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol 15: 550. [PubMed](https://www.ncbi.nlm.nih.gov/pubmed/25516281)

### Cluster Analysis

5. Abu-Jamous B, Kelly S (2018) Clust: automatic extraction of optimal co-expressed gene clusters from gene expression data. Genome Biol 19: 172. [PubMed](https://pubmed.ncbi.nlm.nih.gov/30359297/)
6. Rodriguez MZ, Comin CH, Casanova D, Bruno OM, Amancio DR, Costa L da F, Rodrigues FA (2019) Clustering algorithms: A comparative approach. PLoS One 14: e0210236. [PubMed](https://pubmed.ncbi.nlm.nih.gov/30645617/)

### Differentially Expressed Transcript Variants 

7. Anders S, Reyes A, Huber W (2012) Detecting differential usage of exons from RNA-seq data. Genome Res 22: 2008–2017. [PubMed](https://pubmed.ncbi.nlm.nih.gov/22722343/)
8. Pimentel H, Bray NL, Puente S, Melsted P, Pachter L (2017) Differential analysis of RNA-seq incorporating quantification uncertainty. Nat Methods 14: 687–690. [PubMed](https://pubmed.ncbi.nlm.nih.gov/28581496/)
9. Bray NL, Pimentel H, Melsted P, Pachter L (2016) Near-optimal probabilistic RNA-seq quantification. Nat Biotechnol 34: 525–527. [PubMed](https://pubmed.ncbi.nlm.nih.gov/27043002/)
10. Trapnell C, Roberts A, Goff L, Pertea G, Kim D, Kelley DR, Pimentel H, Salzberg SL, Rinn JL, Pachter L (2012) Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks. Nat Protoc 7: 562–578. [PubMed](https://pubmed.ncbi.nlm.nih.gov/22383036/)

### Clustering and Embedding of High-dimensional scRNA-Seq Data

11. Duò A, Robinson MD, Soneson C (2018) A systematic performance evaluation of clustering methods for single-cell RNA-seq data. F1000Res 7: 1141. [PubMed](https://pubmed.ncbi.nlm.nih.gov/30271584/)
12. Kiselev VY, Kirschner K, Schaub MT, Andrews T, Yiu A, Chandra T, Natarajan KN, Reik W, Barahona M, Green AR, et al (2017) SC3: consensus clustering of single-cell RNA-seq data. Nat Methods 14: 483–486. [PubMed](https://pubmed.ncbi.nlm.nih.gov/28346451/)
13. Sun S, Zhu J, Ma Y, Zhou X (2019) Accuracy, robustness and scalability of dimensionality reduction methods for single-cell RNA-seq analysis. Genome Biol 20: 269. [PubMed](https://pubmed.ncbi.nlm.nih.gov/31823809/) 
14. Huang H, Wang Y, Rudin C, Browne EP (2022) Towards a comprehensive evaluation of dimension reduction methods for transcriptomic data visualization. Commun Biol 5: 719. [PubMed](https://pubmed.ncbi.nlm.nih.gov/35853932/)
15. Grabski IN, Street K, Irizarry RA (2023) Significance analysis for clustering with single-cell RNA-sequencing data. Nat Methods 20: 1196–1202. [PubMed](https://www.nature.com/articles/s41592-023-01933-9)

### VAR-Seq Analysis

16. Cooke DP, Wedge DC, Lunter G (2021) A unified haplotype-based method for accurate and comprehensive variant calling. Nat Biotechnol 39: 885–892. [PubMed](https://pubmed.ncbi.nlm.nih.gov/33782612/)
17. Poplin R, Chang P-C, Alexander D, Schwartz S, Colthurst T, Ku A, Newburger D, Dijamco J, Nguyen N, Afshar PT, et al (2018) A universal SNP and small-indel variant caller using deep neural networks. Nat Biotechnol 36: 983–987. [PubMed](https://pubmed.ncbi.nlm.nih.gov/30247488/)
18. DePristo MA, Banks E, Poplin R, Garimella KV, Maguire JR, Hartl C, Philippakis AA, del Angel G, Rivas MA, Hanna M, et al (2011) A framework for variation discovery and genotyping using next-generation DNA sequencing data. Nat Genet 43: 491–498. [PubMed](https://pubmed.ncbi.nlm.nih.gov/21478889/)
19. Li H (2011) A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data. Bioinformatics 27: 2987–2993. [PubMed](https://pubmed.ncbi.nlm.nih.gov/21903627/) 
