---
title: Student Paper Presentations
linkTitle: "Paper Presentations"
description: >
type: docs
weight: 501
---


## Overview

Each student has been assigned one journal publication to present in class. The
expected structure of the paper presentations is outlined in this [Slideshow Template](https://docs.google.com/presentation/d/14Jdss8C5YvwkBsTxL-Z4HLE5W8RkfxprOpK-_-dzSCY/edit?usp=sharing). 
A detailed presentation schedule is available on the internal [Course Schedule](https://elearn.ucr.edu/courses/134360). 
The grading of both the paper and project presentations (2-3) includes
anonymous feedback from all students as well as the instructor, where
understanding of the material, clarity of the oral presentations and critical
thinking are the main grading criteria. The grading form is [here](). 
The following lists the assigned papers organized by course project topics.


## Publications organized by course projects

All references in [Paperpile](https://paperpile.com/shared/FOShC0)

### RNA-Seq Aligners 

1. Kim D, Langmead B, Salzberg SL (2015) HISAT: a fast spliced aligner with low memory requirements. Nat Methods 12: 357–360. [PubMed](https://pubmed.ncbi.nlm.nih.gov/25751142/)
2. Liao Y, Smyth GK, Shi W (2013) The Subread aligner: fast, accurate and scalable read mapping by seed-and-vote. Nucleic Acids Res 41: e108. [PubMed](https://pubmed.ncbi.nlm.nih.gov/23558742/)

### DEG Methods  

3. Love MI, Huber W, Anders S (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol 15: 550. [PubMed](https://pubmed.ncbi.nlm.nih.gov/25516281/)
4. Liu R, Holik AZ, Su S, Jansz N, Chen K, Leong HS, Blewitt ME, Asselin-Labat M-L, Smyth GK, Ritchie ME (2015) Why weight? Modelling sample and observational level variability improves power in RNA-seq analyses. Nucleic Acids Res. [PubMed](https://pubmed.ncbi.nlm.nih.gov/25925576/)

### ChIP-Seq Motif Analysis 

5. Machanick P, Bailey TL (2011) MEME-ChIP: motif analysis of large DNA datasets. Bioinformatics 27: 1696–1697. [PubMed](https://pubmed.ncbi.nlm.nih.gov/21486936/)

### VAR-Seq 

6. Ng PC, Henikoff S (2003) SIFT: Predicting amino acid changes that affect protein function. Nucleic Acids Res 31: 3812–3814. [PubMed](https://pubmed.ncbi.nlm.nih.gov/12824425/)
7. Cotto KC, Feng Y-Y, Ramu A, Richters M, Freshour SL, Skidmore ZL, Xia H, McMichael JF, Kunisaki J, Campbell KM, et al (2023) Integrated analysis of genomic and transcriptomic data for the discovery of splice-associated variants in cancer. Nat Commun 14: 1589. [PubMed](https://pubmed.ncbi.nlm.nih.gov/36949070/)

### Methyl-Seq

8. Piao Y, Xu W, Park KH, Ryu KH, Xiang R (2021) Comprehensive Evaluation of Differential Methylation Analysis Methods for Bisulfite Sequencing Data. Int J Environ Res Public Health. [PubMed](https://pubmed.ncbi.nlm.nih.gov/34360271/)
9. Akalin A, Kormaksson M, Li S, Garrett-Bakelman FE, Figueroa ME, Melnick A, Mason CE (2012) methylKit: a comprehensive R package for the analysis of genome-wide DNA methylation profiles. Genome Biol 13: R87. [PubMed](https://pubmed.ncbi.nlm.nih.gov/23034086/)

### Genome Graphics

10. Debès C, Papadakis A, Grönke S, Karalay Ö, Tain LS, Mizi A, Nakamura S, Hahn O, Weigelt C, Josipovic N, et al (2023) Ageing-associated changes in transcriptional elongation influence longevity. Nature. [PubMed](https://pubmed.ncbi.nlm.nih.gov/37046086/)

### Drug-Target Analysis

12. Kumar A, Zhang KYJ (2018) Advances in the Development of Shape Similarity Methods and Their Application in Drug Discovery. Front Chem 6: 315. [PubMed](https://pubmed.ncbi.nlm.nih.gov/30090808/) 
13. Rodriguez MZ, Comin CH, Casanova D, Bruno OM, Amancio DR, Costa L da F, Rodrigues FA (2019) Clustering algorithms: A comparative approach. PLoS One 14: e0210236. [PubMed](https://pubmed.ncbi.nlm.nih.gov/30645617/)
14. Vamathevan J, Clark D, Czodrowski P, Dunham I, Ferran E, Lee G, Li B, Madabhushi A, Shah P, Spitzer M, et al (2019) Applications of machine learning in drug discovery and development. Nat Rev Drug Discov 18: 463–477. [PubMed](https://pubmed.ncbi.nlm.nih.gov/30976107/)

