<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.95.0" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>Cluster Analysis in R | GEN242</title><meta property="og:title" content="Cluster Analysis in R" />
<meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Introduction  What is Clustering?  Clustering is the classification of data objects into similarity groups (clusters) according to a defined distance measure. It is used in many fields, such as machine learning, data mining, pattern recognition, image analysis, genomics, systems biology, etc." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/rclustering/rclustering/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2023-05-10T18:08:32-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="Cluster Analysis in R">
<meta itemprop="description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Introduction  What is Clustering?  Clustering is the classification of data objects into similarity groups (clusters) according to a defined distance measure. It is used in many fields, such as machine learning, data mining, pattern recognition, image analysis, genomics, systems biology, etc.">
<meta itemprop="dateModified" content="2023-05-10T18:08:32-07:00" />
<meta itemprop="wordCount" content="2869">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Cluster Analysis in R"/>
<meta name="twitter:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Introduction  What is Clustering?  Clustering is the classification of data objects into similarity groups (clusters) according to a defined distance measure. It is used in many fields, such as machine learning, data mining, pattern recognition, image analysis, genomics, systems biology, etc."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Git and GitHub (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux Cluster Basics (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Overview (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Enrichment Analysis (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics in R (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (S22)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_23/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflow Review (S23)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_24/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Final Discussion (S24)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_25/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Additional Slide Shows (S25)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spvarseq/spvarseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 VAR-Seq Workflow (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T20)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8: R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_dex/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEXseq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/09_drug_target_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Drug-target Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/10_lncrna_orf_discovery/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 lncRNA Discovery
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2023/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.15eef89a4418b5ef13f243860f0d169d.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.15eef89a4418b5ef13f243860f0d169d.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../tutorials/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Tutorials</a>
  </li>
  <ul>
    <li class="collapse show" id="tutorials">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsgithubgithub" href="../../../tutorials/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialslinuxlinux" href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrbasicsrbasics" href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrprogrammingrprogramming" href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrparallelrparallel" href="../../../tutorials/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrsequencesrsequences" href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempiper" href="../../../tutorials/systempiper/systempiper/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseq" href="../../../tutorials/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspvarseqspvarseq" href="../../../tutorials/spvarseq/spvarseq/">VAR-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspchipseqspchipseq" href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsscrnaseqscrnaseq" href="../../../tutorials/scrnaseq/scrnaseq/">scRNA-Seq Embedding Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrmarkdownrmarkdown" href="../../../tutorials/rmarkdown/rmarkdown/">R Markdown Tutorial</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrfearfea" href="../../../tutorials/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-tutorialsrclusteringrclustering" href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrgraphicsrgraphics" href="../../../tutorials/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappslezhangshinyapps" href="../../../tutorials/shinyappslezhang/shinyapps/">Shiny Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappsshinyapps" href="../../../tutorials/shinyapps/shinyapps/">Shiny Web Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrpackagesrpackages" href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsdplyrdplyr" href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialscmdtocwlcmdtocwl" href="../../../tutorials/cmdtocwl/cmdtocwl/">Automate Creation of CWL Instructions</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempiperrnaseqsystempipernaseq" href="../../../tutorials/systempiper/rnaseq/systempipernaseq/">RNA-Seq Workflow Template</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rclustering/rclustering.Rmd" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/tutorials/rclustering/rclustering.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/tutorials/rclustering/rclustering.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=Cluster%20Analysis%20in%20R" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#introduction">Introduction</a></li>
    <li><a href="#data-preprocessing">Data Preprocessing</a>
      <ul>
        <li><a href="#data-transformations">Data Transformations</a></li>
        <li><a href="#distance-methods">Distance Methods</a></li>
        <li><a href="#cluster-linkage">Cluster Linkage</a></li>
      </ul>
    </li>
    <li><a href="#clustering-algorithms">Clustering Algorithms</a>
      <ul>
        <li><a href="#hierarchical-clustering">Hierarchical Clustering</a>
          <ul>
            <li><a href="#overview-of-algorithm">Overview of algorithm</a>
              <ul>
                <li><a href="#hierarchical-clustering-agglomerative-approach">Hierarchical clustering: agglomerative Approach</a></li>
                <li><a href="#hierarchical-clustering-with-heatmap">Hierarchical Clustering with Heatmap</a></li>
              </ul>
            </li>
            <li><a href="#hierarchical-clustering-approaches">Hierarchical Clustering Approaches</a></li>
            <li><a href="#tree-cutting-to-obtain-discrete-clusters">Tree Cutting to Obtain Discrete Clusters</a></li>
            <li><a href="#examples">Examples</a>
              <ul>
                <li><a href="#using-hclust-and-heatmap2">Using <code>hclust</code> and <code>heatmap.2</code></a></li>
                <li><a href="#stepwise-approach-with-tree-cutting">Stepwise Approach with Tree Cutting</a></li>
              </ul>
            </li>
          </ul>
        </li>
        <li><a href="#k-means-clustering">K-Means Clustering</a>
          <ul>
            <li><a href="#overview-of-algorithm-1">Overview of algorithm</a></li>
            <li><a href="#examples-1">Examples</a></li>
          </ul>
        </li>
        <li><a href="#fuzzy-c-means-clustering">Fuzzy C-Means Clustering</a>
          <ul>
            <li><a href="#examples-2">Examples</a>
              <ul>
                <li><a href="#fuzzy-clustering-with-fanny">Fuzzy Clustering with <code>fanny</code></a></li>
              </ul>
            </li>
          </ul>
        </li>
        <li><a href="#principal-component-analysis-pca">Principal Component Analysis (PCA)</a>
          <ul>
            <li><a href="#basic-steps">Basic Steps</a></li>
            <li><a href="#example">Example</a></li>
          </ul>
        </li>
        <li><a href="#multidimensional-scaling-mds">Multidimensional Scaling (MDS)</a>
          <ul>
            <li><a href="#example-1">Example</a></li>
          </ul>
        </li>
        <li><a href="#biclustering">Biclustering</a></li>
        <li><a href="#similarity-measures-for-clusters">Similarity Measures for Clusters</a>
          <ul>
            <li><a href="#example-2">Example:</a>
              <ul>
                <li><a href="#jaccard-index-for-cluster-sets">Jaccard index for cluster sets</a></li>
                <li><a href="#clustering-cluster-sets-with-jaccard-index">Clustering cluster sets with Jaccard index</a></li>
              </ul>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#clustering-exercises">Clustering Exercises</a>
      <ul>
        <li><a href="#data-preprocessing-1">Data Preprocessing</a>
          <ul>
            <li><a href="#scaling">Scaling</a></li>
          </ul>
        </li>
        <li><a href="#distance-matrices">Distance Matrices</a>
          <ul>
            <li><a href="#euclidean-distance-matrix">Euclidean distance matrix</a></li>
            <li><a href="#correlation-based-distance-matrix">Correlation-based distance matrix</a></li>
          </ul>
        </li>
        <li><a href="#hierarchical-clustering-with-hclust">Hierarchical Clustering with <code>hclust</code></a>
          <ul>
            <li><a href="#tree-plotting-i">Tree plotting I</a></li>
            <li><a href="#tree-plotting-ii">Tree plotting II</a></li>
          </ul>
        </li>
        <li><a href="#tree-cutting">Tree Cutting</a></li>
        <li><a href="#heatmaps">Heatmaps</a>
          <ul>
            <li><a href="#with-heatmap2">With <code>heatmap.2</code></a></li>
            <li><a href="#with-pheatmap">With <code>pheatmap</code></a></li>
            <li><a href="#customizing-heatmaps">Customizing heatmaps</a></li>
            <li><a href="#complex-heatmaps">Complex heatmaps</a></li>
          </ul>
        </li>
        <li><a href="#k-means-clustering-with-pam">K-Means Clustering with PAM</a></li>
        <li><a href="#k-means-fuzzy-clustering">K-Means Fuzzy Clustering</a></li>
        <li><a href="#multidimensional-scaling-mds-1">Multidimensional Scaling (MDS)</a></li>
        <li><a href="#principal-component-analysis-pca-1">Principal Component Analysis (PCA)</a></li>
        <li><a href="#additional-exercises">Additional Exercises</a></li>
      </ul>
    </li>
    <li><a href="#version-information">Version Information</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../tutorials/">Tutorials</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>Cluster Analysis in R</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 14 minute read</p>
           
	<!--
- Compile from command-line
Rscript -e "rmarkdown::render('rclustering.Rmd', c('html_document'), clean=F); knitr::knit('rclustering.Rmd', tangle=TRUE)"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div style="text-align: right">
<p>Source code downloads:    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rclustering/rclustering.Rmd">.Rmd</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rclustering/rclustering.R">.R</a> ]</p>
</div>
<h2 id="introduction">Introduction</h2>
<ul>
<li>What is Clustering?
<ul>
<li>Clustering is the classification of data objects into similarity groups (clusters) according to a defined distance measure.</li>
<li>It is used in many fields, such as machine learning, data mining, pattern recognition, image analysis, genomics, systems biology, etc.</li>
<li>Machine learning typically regards data clustering as a form of unsupervised learning.</li>
</ul>
</li>
<li>Why Clustering and Data Mining in R?}
<ul>
<li>Efficient data structures and functions for clustering</li>
<li>Reproducible and programmable</li>
<li>Comprehensive set of clustering and machine learning libraries</li>
<li>Integration with many other data analysis tools</li>
</ul>
</li>
<li>Useful Links
<ul>
<li><a href="https://cran.r-project.org/web/views/Cluster.html">Cluster Task Views</a></li>
<li><a href="https://cran.r-project.org/web/views/MachineLearning.html">Machine Learning Task Views</a></li>
<li><a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual#TOC-Clustering-and-Data-Mining-in-R">UCR Manual</a></li>
</ul>
</li>
</ul>
<h2 id="data-preprocessing">Data Preprocessing</h2>
<h3 id="data-transformations">Data Transformations</h3>
<p>Choice depends on data set!</p>
<ul>
<li>Center and standardize
<ol>
<li>Center: subtract from each value the mean of the corresponding vector</li>
<li>Standardize: devide by standard deviation</li>
</ol>
<!-- end list -->
<ul>
<li>Result: <em>Mean = 0</em> and <em>STDEV = 1</em></li>
</ul>
</li>
<li>Center and scale with the <code>scale()</code> function
<ol>
<li>Center: subtract from each value the mean of the corresponding vector</li>
<li>Scale: divide centered vector by their <em>root mean square</em> (<em>rms</em>):
[ x_{rms} = \sqrt[]{\frac{1}{n-1}\sum_{i=1}^{n}{x_{i}{^2}}} ]</li>
</ol>
<!-- end list -->
<ul>
<li>Result: <em>Mean = 0</em> and <em>STDEV = 1</em></li>
</ul>
</li>
<li>Log transformation</li>
<li>Rank transformation: replace measured values by ranks</li>
<li>No transformation</li>
</ul>
<h3 id="distance-methods">Distance Methods</h3>
<p>List of most common ones!</p>
<ul>
<li>Euclidean distance for two profiles <em>X</em> and <em>Y</em>:
[ d(X,Y) = \sqrt[]{ \sum_{i=1}^{n}{(x_{i}-y_{i})^2} }]
<ul>
<li><strong>Disadvantages</strong>: not scale invariant, not for negative correlations</li>
</ul>
</li>
<li>Maximum, Manhattan, Canberra, binary, Minowski, …</li>
<li>Correlation-based distance: <em>1-r</em>
<ul>
<li>Pearson correlation coefficient (PCC):
<code>$$r = \frac{n\sum_{i=1}^{n}{x_{i}y_{i}} - \sum_{i=1}^{n}{x_{i}} \sum_{i=1}^{n}{y_{i}}}{ \sqrt[]{(\sum_{i=1}^{n}{x_{i}^2} - (\sum_{i=1}^{n}{x_{i})^2}) (\sum_{i=1}^{n}{y_{i}^2} - (\sum_{i=1}^{n}{y_{i})^2})} }$$</code>
<ul>
<li><strong>Disadvantage</strong>: outlier sensitive</li>
</ul>
</li>
<li>Spearman correlation coefficient (SCC)
<ul>
<li>Same calculation as PCC but with ranked values!</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>There are many more distance measures</p>
<ul>
<li>If the distances among items are quantifiable, then clustering is possible.</li>
<li>Choose the most accurate and meaningful distance measure for a given field of application.</li>
<li>If uncertain then choose several distance measures and compare the results.</li>
</ul>
<h3 id="cluster-linkage">Cluster Linkage</h3>
<center>
<img title="cluster linkage" src="../images/linkage.png" >
</center>
<h2 id="clustering-algorithms">Clustering Algorithms</h2>
<h3 id="hierarchical-clustering">Hierarchical Clustering</h3>
<h4 id="overview-of-algorithm">Overview of algorithm</h4>
<ol>
<li>Identify clusters (items) with closest distance</li>
<li>Join them to new clusters</li>
<li>Compute distance between clusters (items)</li>
<li>Return to step 1</li>
</ol>
<h5 id="hierarchical-clustering-agglomerative-approach">Hierarchical clustering: agglomerative Approach</h5>
<center>
<img title="hierarchical clustering" src="../images/hierarchical.png" >
</center>
<h5 id="hierarchical-clustering-with-heatmap">Hierarchical Clustering with Heatmap</h5>
<center>
<img title="heatmap" src="../images/heatmap.png" >
</center>
<ul>
<li>A heatmap is a color coded table. To visually identify patterns, the rows and columns of a heatmap are often sorted by hierarchical clustering trees.</li>
<li>In case of gene expression data, the row tree usually represents the genes, the column tree the treatments and the colors in the heat table represent the intensities or ratios of the underlying gene expression data set.</li>
</ul>
<h4 id="hierarchical-clustering-approaches">Hierarchical Clustering Approaches</h4>
<ol>
<li>Agglomerative approach (bottom-up)
<ul>
<li>R functions: <code>hclust()</code> and <code>agnes()</code></li>
</ul>
</li>
<li>Divisive approach (top-down)
<ul>
<li>R function: <code>diana()</code></li>
</ul>
</li>
</ol>
<h4 id="tree-cutting-to-obtain-discrete-clusters">Tree Cutting to Obtain Discrete Clusters</h4>
<ol>
<li>Node height in tree</li>
<li>Number of clusters</li>
<li>Search tree nodes by distance cutoff</li>
</ol>
<h4 id="examples">Examples</h4>
<h5 id="using-hclust-and-heatmap2">Using <code>hclust</code> and <code>heatmap.2</code></h5>
<p>Note, with large data sets consider using <a href="https://cran.r-project.org/web/packages/flashClust/index.html">flashClust</a> which is a fast implementation of hierarchical clustering.</p>
<pre><code class="language-r">library(gplots) 
y &lt;- matrix(rnorm(500), 100, 5, dimnames=list(paste(&quot;g&quot;, 1:100, sep=&quot;&quot;), paste(&quot;t&quot;, 1:5, sep=&quot;&quot;))) 
heatmap.2(y) # Shortcut to final result
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/hclust_heatmap_example-1.png" width="672" />
<h5 id="stepwise-approach-with-tree-cutting">Stepwise Approach with Tree Cutting</h5>
<pre><code class="language-r">## Row- and column-wise clustering 
hr &lt;- hclust(as.dist(1-cor(t(y), method=&quot;pearson&quot;)), method=&quot;complete&quot;)
hc &lt;- hclust(as.dist(1-cor(y, method=&quot;spearman&quot;)), method=&quot;complete&quot;) 
## Tree cutting
mycl &lt;- cutree(hr, h=max(hr$height)/1.5); mycolhc &lt;- rainbow(length(unique(mycl)), start=0.1, end=0.9); mycolhc &lt;- mycolhc[as.vector(mycl)] 
## Plot heatmap 
mycol &lt;- colorpanel(40, &quot;darkblue&quot;, &quot;yellow&quot;, &quot;white&quot;) # or try redgreen(75)
heatmap.2(y, Rowv=as.dendrogram(hr), Colv=as.dendrogram(hc), col=mycol, scale=&quot;row&quot;, density.info=&quot;none&quot;, trace=&quot;none&quot;, RowSideColors=mycolhc) 
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/hclust_heatmap_example_setpwise-1.png" width="672" />
<h3 id="k-means-clustering">K-Means Clustering</h3>
<h4 id="overview-of-algorithm-1">Overview of algorithm</h4>
<ol>
<li>Choose the number of k clusters</li>
<li>Randomly assign items to the k clusters</li>
<li>Calculate new centroid for each of the k clusters</li>
<li>Calculate the distance of all items to the k centroids</li>
<li>Assign items to closest centroid</li>
<li>Repeat until clusters assignments are stable</li>
</ol>
<center>
<img title="kmeans" src="../images/kmeans.png" >
</center>
<h4 id="examples-1">Examples</h4>
<pre><code class="language-r">km &lt;- kmeans(t(scale(t(y))), 3)
km$cluster 
</code></pre>
<pre><code>##   g1   g2   g3   g4   g5   g6   g7   g8   g9  g10  g11  g12  g13  g14  g15  g16  g17  g18  g19  g20 
##    1    3    1    1    1    1    3    2    1    3    1    3    1    2    3    2    2    3    1    1 
##  g21  g22  g23  g24  g25  g26  g27  g28  g29  g30  g31  g32  g33  g34  g35  g36  g37  g38  g39  g40 
##    1    2    1    2    2    3    2    1    1    3    3    3    2    3    1    1    3    3    1    1 
##  g41  g42  g43  g44  g45  g46  g47  g48  g49  g50  g51  g52  g53  g54  g55  g56  g57  g58  g59  g60 
##    2    2    1    2    2    2    1    3    3    2    2    2    3    3    2    2    1    2    3    2 
##  g61  g62  g63  g64  g65  g66  g67  g68  g69  g70  g71  g72  g73  g74  g75  g76  g77  g78  g79  g80 
##    2    3    3    3    3    3    2    1    1    2    1    3    1    1    1    2    3    3    1    2 
##  g81  g82  g83  g84  g85  g86  g87  g88  g89  g90  g91  g92  g93  g94  g95  g96  g97  g98  g99 g100 
##    2    1    3    3    2    2    1    2    1    1    2    3    3    2    2    2    3    1    1    1
</code></pre>
<h3 id="fuzzy-c-means-clustering">Fuzzy C-Means Clustering</h3>
<ul>
<li>In contrast to strict (hard) clustering approaches, fuzzy (soft) clustering methods allow multiple cluster memberships of the clustered items (Hathaway, Bezdek, and Pal 1996).</li>
<li>This is commonly achieved by assigning to each item a weight of belonging to each cluster.</li>
<li>Thus, items at the edge of a cluster, may be in a cluster to a lesser degree than items at the center of a cluster.</li>
<li>Typically, each item has as many coefficients (weights) as there are clusters that sum up for each item to one.</li>
</ul>
<h4 id="examples-2">Examples</h4>
<h5 id="fuzzy-clustering-with-fanny">Fuzzy Clustering with <code>fanny</code></h5>
<pre><code class="language-r">library(cluster) # Loads the cluster library.
fannyy &lt;- fanny(y, k=4, metric = &quot;euclidean&quot;, memb.exp = 1.2)
round(fannyy$membership, 2)[1:4,]
</code></pre>
<pre><code>##    [,1] [,2] [,3] [,4]
## g1 0.97 0.02 0.02 0.00
## g2 0.07 0.91 0.01 0.00
## g3 0.28 0.08 0.44 0.20
## g4 0.84 0.03 0.12 0.01
</code></pre>
<pre><code class="language-r">fannyy$clustering # Hard clustering result
</code></pre>
<pre><code>##   g1   g2   g3   g4   g5   g6   g7   g8   g9  g10  g11  g12  g13  g14  g15  g16  g17  g18  g19  g20 
##    1    2    3    1    3    3    1    2    1    2    3    4    1    2    4    2    2    4    1    4 
##  g21  g22  g23  g24  g25  g26  g27  g28  g29  g30  g31  g32  g33  g34  g35  g36  g37  g38  g39  g40 
##    1    4    3    2    2    2    2    1    3    4    1    1    4    4    1    3    1    1    4    1 
##  g41  g42  g43  g44  g45  g46  g47  g48  g49  g50  g51  g52  g53  g54  g55  g56  g57  g58  g59  g60 
##    2    4    1    2    2    3    3    1    2    2    3    4    4    1    4    3    3    4    2    2 
##  g61  g62  g63  g64  g65  g66  g67  g68  g69  g70  g71  g72  g73  g74  g75  g76  g77  g78  g79  g80 
##    3    2    3    2    4    2    2    3    1    2    3    1    3    1    1    3    4    1    3    3 
##  g81  g82  g83  g84  g85  g86  g87  g88  g89  g90  g91  g92  g93  g94  g95  g96  g97  g98  g99 g100 
##    3    1    2    2    2    2    3    2    1    3    4    2    2    2    2    2    2    1    3    1
</code></pre>
<pre><code class="language-r">(fannyyMA &lt;- fannyy$membership &gt; 0.20)[1:4,] # Soft clustering result
</code></pre>
<pre><code>##     [,1]  [,2]  [,3]  [,4]
## g1  TRUE FALSE FALSE FALSE
## g2 FALSE  TRUE FALSE FALSE
## g3  TRUE FALSE  TRUE FALSE
## g4  TRUE FALSE FALSE FALSE
</code></pre>
<pre><code class="language-r">apply(fannyyMA, 1, which)[1:4] # First 4 clusters
</code></pre>
<pre><code>## $g1
## [1] 1
## 
## $g2
## [1] 2
## 
## $g3
## [1] 1 3
## 
## $g4
## [1] 1
</code></pre>
<h3 id="principal-component-analysis-pca">Principal Component Analysis (PCA)</h3>
<p>Principal components analysis (PCA) is a data reduction technique that allows to simplify multidimensional data sets to 2 or 3 dimensions for plotting purposes and visual variance analysis.</p>
<h4 id="basic-steps">Basic Steps</h4>
<ul>
<li>Center (and standardize) data</li>
<li>First principal component axis
<ul>
<li>Across centroid of data cloud</li>
<li>Distance of each point to that line is minimized, so that it crosses the maximum variation of the data cloud</li>
</ul>
</li>
<li>Second principal component axis
<ul>
<li>Orthogonal to first principal component</li>
<li>Along maximum variation in the data</li>
</ul>
</li>
<li>First PCA axis becomes x-axis and second PCA axis y-axis</li>
<li>Continue process until the necessary number of principal components is obtained</li>
</ul>
<center>
<img title="pca" src="../images/pca.png" >
</center>
<h4 id="example">Example</h4>
<pre><code class="language-r">pca &lt;- prcomp(y, scale=T)
summary(pca) # Prints variance summary for all principal components
</code></pre>
<pre><code>## Importance of components:
##                          PC1    PC2    PC3    PC4    PC5
## Standard deviation     1.179 1.1177 0.9965 0.8492 0.8041
## Proportion of Variance 0.278 0.2498 0.1986 0.1442 0.1293
## Cumulative Proportion  0.278 0.5279 0.7265 0.8707 1.0000
</code></pre>
<pre><code class="language-r">plot(pca$x, pch=20, col=&quot;blue&quot;, type=&quot;n&quot;) # To plot dots, drop type=&quot;n&quot;
text(pca$x, rownames(pca$x), cex=0.8)
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/pca_example-1.png" width="672" />
1st and 2nd principal components explain x% of variance in data.
<h3 id="multidimensional-scaling-mds">Multidimensional Scaling (MDS)</h3>
<ul>
<li>Alternative dimensionality reduction approach</li>
<li>Represents distances in 2D or 3D space</li>
<li>Starts from distance matrix (PCA uses data points)</li>
</ul>
<h4 id="example-1">Example</h4>
<p>The following example performs MDS analysis with <code>cmdscale</code> on the geographic distances among European cities.</p>
<pre><code class="language-r">loc &lt;- cmdscale(eurodist) 
plot(loc[,1], -loc[,2], type=&quot;n&quot;, xlab=&quot;&quot;, ylab=&quot;&quot;, main=&quot;cmdscale(eurodist)&quot;)
text(loc[,1], -loc[,2], rownames(loc), cex=0.8) 
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/mds_example-1.png" width="672" />
<h3 id="biclustering">Biclustering</h3>
<p>Finds in matrix subgroups of rows and columns which are as similar as possible to each other and as different as possible to the remaining data points.</p>
<center>
<img title="pca" src="../images/biclust.png"  >
</center>
<center>
<p>Unclustered ————————–&gt; Clustered</p>
</center>
<h3 id="similarity-measures-for-clusters">Similarity Measures for Clusters</h3>
<ul>
<li>Compare the numbers of identical and unique item pairs appearing in cluster sets</li>
<li>Achieved by counting the number of item pairs found in both clustering sets <em>(a)</em> as well as the pairs appearing only in the first <em>(b)</em> or the second <em>(c)</em> set.</li>
<li>With this a similarity coefficient, such as the Jaccard index, can be computed. The latter is defined as the size of the intersect divided by the size of the union of two sample sets: <em>a/(a+b+c)</em>.</li>
<li>In case of partitioning results, the Jaccard Index measures how frequently pairs of items are joined together in two clustering data sets and how often pairs are observed only in one set.</li>
<li>Related coefficient are the Rand Index and the Adjusted Rand Index. These indices also consider the number of pairs <em>(d)</em> that are not joined together in any of the clusters in both sets.</li>
</ul>
<h4 id="example-2">Example:</h4>
<h5 id="jaccard-index-for-cluster-sets">Jaccard index for cluster sets</h5>
<p>The following imports the <code>cindex()</code> function and computes the Jaccard Index for two sample cluster results.</p>
<pre><code class="language-r">source(&quot;http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/clusterIndex.R&quot;) 
library(cluster); y &lt;- matrix(rnorm(5000), 1000, 5, dimnames=list(paste(&quot;g&quot;, 1:1000, sep=&quot;&quot;), paste(&quot;t&quot;, 1:5, sep=&quot;&quot;))); clarax &lt;- clara(y, 49); clV1 &lt;- clarax$clustering; clarax &lt;- clara(y, 50); clV2 &lt;- clarax$clustering 
ci &lt;- cindex(clV1=clV1, clV2=clV2, self=FALSE, minSZ=1, method=&quot;jaccard&quot;)
ci[2:3] # Returns Jaccard index and variables used to compute it 
</code></pre>
<pre><code>## $variables
##    a    b    c 
## 6905 7441 7145 
## 
## $Jaccard_Index
## [1] 0.3212973
</code></pre>
<h5 id="clustering-cluster-sets-with-jaccard-index">Clustering cluster sets with Jaccard index</h5>
<p>The following example shows how one can cluster entire cluster result sets. First, 10 sample cluster results are created with Clara using k-values from 3 to 12. The results are stored as named clustering vectors in a list object. Then a nested sapply loop is used to generate a similarity matrix of Jaccard Indices for the clustering results. After converting the result into a distance matrix, hierarchical clustering is performed with <code>hclust</code>.}</p>
<pre><code class="language-r">clVlist &lt;- lapply(3:12, function(x) clara(y[1:30, ], k=x)$clustering); names(clVlist) &lt;- paste(&quot;k&quot;, &quot;=&quot;, 3:12)
d &lt;- sapply(names(clVlist), function(x) sapply(names(clVlist), function(y) cindex(clV1=clVlist[[y]], clV2=clVlist[[x]], method=&quot;jaccard&quot;)[[3]]))
hv &lt;- hclust(as.dist(1-d))
plot(as.dendrogram(hv), edgePar=list(col=3, lwd=4), horiz=T, main=&quot;Similarities of 10 Clara Clustering Results for k: 3-12&quot;) 
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/jaccard_index_clustering-1.png" width="672" />
<ul>
<li>Remember: there are many additional clustering algorithms.</li>
<li>Additional details can be found in the Clustering Section of the <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual#TOC-Clustering-and-Data-Mining-in-R">R/Bioconductor Manual</a>.</li>
</ul>
<h2 id="clustering-exercises">Clustering Exercises</h2>
<h3 id="data-preprocessing-1">Data Preprocessing</h3>
<h4 id="scaling">Scaling</h4>
<pre><code class="language-r">## Sample data set
set.seed(1410)
y &lt;- matrix(rnorm(50), 10, 5, dimnames=list(paste(&quot;g&quot;, 1:10, sep=&quot;&quot;), 
            paste(&quot;t&quot;, 1:5, sep=&quot;&quot;)))
dim(y)
</code></pre>
<pre><code>## [1] 10  5
</code></pre>
<pre><code class="language-r">## Scaling
yscaled &lt;- t(scale(t(y))) # Centers and scales y row-wise
apply(yscaled, 1, sd)
</code></pre>
<pre><code>##  g1  g2  g3  g4  g5  g6  g7  g8  g9 g10 
##   1   1   1   1   1   1   1   1   1   1
</code></pre>
<h3 id="distance-matrices">Distance Matrices</h3>
<h4 id="euclidean-distance-matrix">Euclidean distance matrix</h4>
<pre><code class="language-r">dist(y[1:4,], method = &quot;euclidean&quot;)
</code></pre>
<pre><code>##          g1       g2       g3
## g2 4.793697                  
## g3 4.932658 6.354978         
## g4 4.033789 4.788508 1.671968
</code></pre>
<h4 id="correlation-based-distance-matrix">Correlation-based distance matrix</h4>
<p>Correlation matrix</p>
<pre><code class="language-r">c &lt;- cor(t(y), method=&quot;pearson&quot;) 
as.matrix(c)[1:4,1:4]
</code></pre>
<pre><code>##             g1         g2          g3         g4
## g1  1.00000000 -0.2965885 -0.00206139 -0.4042011
## g2 -0.29658847  1.0000000 -0.91661118 -0.4512912
## g3 -0.00206139 -0.9166112  1.00000000  0.7435892
## g4 -0.40420112 -0.4512912  0.74358925  1.0000000
</code></pre>
<p>Correlation-based distance matrix</p>
<pre><code class="language-r">d &lt;- as.dist(1-c)
as.matrix(d)[1:4,1:4]
</code></pre>
<pre><code>##          g1       g2        g3        g4
## g1 0.000000 1.296588 1.0020614 1.4042011
## g2 1.296588 0.000000 1.9166112 1.4512912
## g3 1.002061 1.916611 0.0000000 0.2564108
## g4 1.404201 1.451291 0.2564108 0.0000000
</code></pre>
<h3 id="hierarchical-clustering-with-hclust">Hierarchical Clustering with <code>hclust</code></h3>
<p>Hierarchical clustering with complete linkage and basic tree plotting</p>
<pre><code class="language-r">hr &lt;- hclust(d, method = &quot;complete&quot;, members=NULL)
names(hr)
</code></pre>
<pre><code>## [1] &quot;merge&quot;       &quot;height&quot;      &quot;order&quot;       &quot;labels&quot;      &quot;method&quot;      &quot;call&quot;       
## [7] &quot;dist.method&quot;
</code></pre>
<pre><code class="language-r">par(mfrow = c(1, 2)); plot(hr, hang = 0.1); plot(hr, hang = -1) 
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/hclust1-1.png" width="672" />
<h4 id="tree-plotting-i">Tree plotting I</h4>
<pre><code class="language-r">plot(as.dendrogram(hr), edgePar=list(col=3, lwd=4), horiz=T) 
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/hclust_plot_tree1-1.png" width="672" />
<h4 id="tree-plotting-ii">Tree plotting II</h4>
<p>The <code>ape</code> library provides more advanced features for tree plotting</p>
<pre><code class="language-r">library(ape) 
plot.phylo(as.phylo(hr), type=&quot;p&quot;, edge.col=4, edge.width=2, 
           show.node.label=TRUE, no.margin=TRUE)
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/hclust_plot_tree2-1.png" width="672" />
<h3 id="tree-cutting">Tree Cutting</h3>
<p>Accessing information in <em>hclust</em> objects</p>
<pre><code class="language-r">hr
</code></pre>
<pre><code>## 
## Call:
## hclust(d = d, method = &quot;complete&quot;, members = NULL)
## 
## Cluster method   : complete 
## Number of objects: 10
</code></pre>
<pre><code class="language-r">## Print row labels in the order they appear in the tree
hr$labels[hr$order] 
</code></pre>
<pre><code>##  [1] &quot;g10&quot; &quot;g3&quot;  &quot;g4&quot;  &quot;g2&quot;  &quot;g9&quot;  &quot;g6&quot;  &quot;g7&quot;  &quot;g1&quot;  &quot;g5&quot;  &quot;g8&quot;
</code></pre>
<p>Tree cutting with <code>cutree</code></p>
<pre><code class="language-r">mycl &lt;- cutree(hr, h=max(hr$height)/2)
mycl[hr$labels[hr$order]] 
</code></pre>
<pre><code>## g10  g3  g4  g2  g9  g6  g7  g1  g5  g8 
##   3   3   3   2   2   5   5   1   4   4
</code></pre>
<h3 id="heatmaps">Heatmaps</h3>
<h4 id="with-heatmap2">With <code>heatmap.2</code></h4>
<p>All in one step: clustering and heatmap plotting</p>
<pre><code class="language-r">library(gplots)
heatmap.2(y, col=redgreen(75))
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/heatmap2a-1.png" width="672" />
<h4 id="with-pheatmap">With <code>pheatmap</code></h4>
<p>All in one step: clustering and heatmap plotting</p>
<pre><code class="language-r">library(pheatmap); library(&quot;RColorBrewer&quot;)
pheatmap(y, color=brewer.pal(9,&quot;Blues&quot;))
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/pheatmap-1.png" width="672" />
<h4 id="customizing-heatmaps">Customizing heatmaps</h4>
<p>Customizes row and column clustering and shows tree cutting result in row color bar. Additional color schemes can be found <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual#clustering_hc_submap">here</a>.</p>
<pre><code class="language-r">hc &lt;- hclust(as.dist(1-cor(y, method=&quot;spearman&quot;)), method=&quot;complete&quot;)
mycol &lt;- colorpanel(40, &quot;darkblue&quot;, &quot;yellow&quot;, &quot;white&quot;)
heatmap.2(y, Rowv=as.dendrogram(hr), Colv=as.dendrogram(hc), col=mycol,
          scale=&quot;row&quot;, density.info=&quot;none&quot;, trace=&quot;none&quot;, 
          RowSideColors=as.character(mycl))
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/heatmap2_custom-1.png" width="672" />
<h4 id="complex-heatmaps">Complex heatmaps</h4>
<p>For plotting complex heatmaps, the <code>ComplexHeatmap</code> provides useful functionalities. The following creates a sample heatmap. For additional examples, visit the manual <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/">here</a>.</p>
<pre><code class="language-r">library(ComplexHeatmap)
set.seed(123)
mat = matrix(rnorm(100), 10)
rownames(mat) = paste0(&quot;R&quot;, 1:10)
colnames(mat) = paste0(&quot;C&quot;, 1:10)
column_ha = HeatmapAnnotation(foo1 = runif(10), bar1 = anno_barplot(runif(10)))
row_ha = rowAnnotation(foo2 = runif(10), bar2 = anno_barplot(runif(10)))
Heatmap(mat, name = &quot;mat&quot;, top_annotation = column_ha, right_annotation = row_ha)
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/complex_heatmap-1.png" width="672" />
<h3 id="k-means-clustering-with-pam">K-Means Clustering with PAM</h3>
<p>Runs K-means clustering with PAM (partitioning around medoids) algorithm and shows result in color bar of hierarchical clustering result from before.</p>
<pre><code class="language-r">library(cluster)
pamy &lt;- pam(d, 4)
(kmcol &lt;- pamy$clustering)
</code></pre>
<pre><code>##  g1  g2  g3  g4  g5  g6  g7  g8  g9 g10 
##   1   2   3   3   4   4   4   4   2   3
</code></pre>
<pre><code class="language-r">heatmap.2(y, Rowv=as.dendrogram(hr), Colv=as.dendrogram(hc), col=mycol,
          scale=&quot;row&quot;, density.info=&quot;none&quot;, trace=&quot;none&quot;, 
          RowSideColors=as.character(kmcol))
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/kmeans2-1.png" width="672" />
<h3 id="k-means-fuzzy-clustering">K-Means Fuzzy Clustering</h3>
<p>Performs k-means fuzzy clustering</p>
<pre><code class="language-r">library(cluster)
fannyy &lt;- fanny(d, k=4, memb.exp = 1.5)
round(fannyy$membership, 2)[1:4,]
</code></pre>
<pre><code>##    [,1] [,2] [,3] [,4]
## g1 1.00 0.00 0.00 0.00
## g2 0.00 0.99 0.00 0.00
## g3 0.02 0.01 0.95 0.03
## g4 0.00 0.00 0.99 0.01
</code></pre>
<pre><code class="language-r">fannyy$clustering 
</code></pre>
<pre><code>##  g1  g2  g3  g4  g5  g6  g7  g8  g9 g10 
##   1   2   3   3   4   4   4   4   2   3
</code></pre>
<pre><code class="language-r">## Returns multiple cluster memberships for coefficient above a certain 
## value (here &gt;0.1)
fannyyMA &lt;- round(fannyy$membership, 2) &gt; 0.10 
apply(fannyyMA, 1, function(x) paste(which(x), collapse=&quot;_&quot;))
</code></pre>
<pre><code>##    g1    g2    g3    g4    g5    g6    g7    g8    g9   g10 
##   &quot;1&quot;   &quot;2&quot;   &quot;3&quot;   &quot;3&quot;   &quot;4&quot;   &quot;4&quot;   &quot;4&quot; &quot;2_4&quot;   &quot;2&quot;   &quot;3&quot;
</code></pre>
<h3 id="multidimensional-scaling-mds-1">Multidimensional Scaling (MDS)</h3>
<p>Performs MDS analysis on the geographic distances between European cities</p>
<pre><code class="language-r">loc &lt;- cmdscale(eurodist) 
## Plots the MDS results in 2D plot. The minus is required in this example to 
## flip the plotting orientation.
plot(loc[,1], -loc[,2], type=&quot;n&quot;, xlab=&quot;&quot;, ylab=&quot;&quot;, main=&quot;cmdscale(eurodist)&quot;)
text(loc[,1], -loc[,2], rownames(loc), cex=0.8) 
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/cmdscale2-1.png" width="672" />
<h3 id="principal-component-analysis-pca-1">Principal Component Analysis (PCA)</h3>
<p>Performs PCA analysis after scaling the data. It returns a list with class <code>prcomp</code> that contains five components: (1) the standard deviations (sdev) of the principal components, (2) the matrix of eigenvectors (rotation), (3) the principal component data (x), (4) the centering (center) and (5) scaling (scale) used.</p>
<pre><code class="language-r">library(scatterplot3d)
pca &lt;- prcomp(y, scale=TRUE)
names(pca)
</code></pre>
<pre><code>## [1] &quot;sdev&quot;     &quot;rotation&quot; &quot;center&quot;   &quot;scale&quot;    &quot;x&quot;
</code></pre>
<pre><code class="language-r">summary(pca) # Prints variance summary for all principal components.
</code></pre>
<pre><code>## Importance of components:
##                           PC1    PC2    PC3     PC4    PC5
## Standard deviation     1.3611 1.1777 1.0420 0.69264 0.4416
## Proportion of Variance 0.3705 0.2774 0.2172 0.09595 0.0390
## Cumulative Proportion  0.3705 0.6479 0.8650 0.96100 1.0000
</code></pre>
<pre><code class="language-r">scatterplot3d(pca$x[,1:3], pch=20, color=&quot;blue&quot;) 
</code></pre>
<img src="../../../en/tutorials/rclustering/rclustering_files/figure-html/pca2-1.png" width="672" />
<h3 id="additional-exercises">Additional Exercises</h3>
<p>See <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual#TOC-Clustering-Exercises">here</a></p>
<h2 id="version-information">Version Information</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.1.3 (2022-03-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] scatterplot3d_0.3-41 RColorBrewer_1.1-2   pheatmap_1.0.12      cluster_2.1.2       
## [5] gplots_3.1.1         ape_5.5              ggplot2_3.3.5        BiocStyle_2.22.0    
## 
## loaded via a namespace (and not attached):
##  [1] gtools_3.9.2        tidyselect_1.1.1    xfun_0.30           bslib_0.3.1        
##  [5] purrr_0.3.4         lattice_0.20-45     colorspace_2.0-2    vctrs_0.3.8        
##  [9] generics_0.1.1      htmltools_0.5.2     yaml_2.3.5          utf8_1.2.2         
## [13] rlang_1.0.2         jquerylib_0.1.4     pillar_1.6.4        glue_1.6.2         
## [17] withr_2.4.3         DBI_1.1.1           lifecycle_1.0.1     stringr_1.4.0      
## [21] munsell_0.5.0       blogdown_1.8.2      gtable_0.3.0        caTools_1.18.2     
## [25] codetools_0.2-18    evaluate_0.15       knitr_1.37          fastmap_1.1.0      
## [29] parallel_4.1.3      fansi_0.5.0         highr_0.9           Rcpp_1.0.8.2       
## [33] KernSmooth_2.23-20  scales_1.1.1        BiocManager_1.30.16 jsonlite_1.8.0     
## [37] digest_0.6.29       stringi_1.7.6       bookdown_0.24       dplyr_1.0.7        
## [41] grid_4.1.3          cli_3.1.0           tools_4.1.3         bitops_1.0-7       
## [45] magrittr_2.0.2      sass_0.4.0          tibble_3.1.6        crayon_1.4.2       
## [49] pkgconfig_2.0.3     ellipsis_0.3.2      assertthat_0.2.1    rmarkdown_2.13     
## [53] R6_2.5.1            nlme_3.1-155        compiler_4.1.3
</code></pre>
<h2 id="references">References</h2>
<div id="refs" class="references hanging-indent">
<div id="ref-Hathaway1996-hu">
<p>Hathaway, R J, J C Bezdek, and N R Pal. 1996. “Sequential Competitive Learning and the Fuzzy c-Means Clustering Algorithms.” <em>Neural Netw.</em> 9 (5): 787–96. <a href="http://www.hubmed.org/display.cgi?uids=12662563">http://www.hubmed.org/display.cgi?uids=12662563</a>.</p>
</div>
</div>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2023-05-10: <a  href="https://github.com/tgirke/GEN242//commit/982465c80b87f2fca9bd80e26edc9b63d94ac37a">some edits (982465c80)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2024 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
