<!doctype html>
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="generator" content="Hugo 0.94.2" /><META NAME="ROBOTS" CONTENT="INDEX, FOLLOW">


<link rel="shortcut icon" href="../../../favicons/favicon.ico" >
<link rel="apple-touch-icon" href="../../../favicons/apple-touch-icon-180x180.png" sizes="180x180">
<link rel="icon" type="image/png" href="../../../favicons/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="../../../favicons/favicon-32x32.png" sizes="32x32">
<link rel="icon" type="image/png" href="../../../favicons/android-36x36.png" sizes="36x36">
<link rel="icon" type="image/png" href="../../../favicons/android-48x48.png" sizes="48x48">
<link rel="icon" type="image/png" href="../../../favicons/android-72x72.png" sizes="72x72">
<link rel="icon" type="image/png" href="../../../favicons/android-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="../../../favicons/android-144x144.png" sizes="144x144">
<link rel="icon" type="image/png" href="../../../favicons/android-192x192.png" sizes="192x192">

<title>Introduction to R | GEN242</title><meta property="og:title" content="Introduction to R" />
<meta property="og:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Overview What is R? R is a powerful statistical environment and programming language for the analysis and visualization of data. The associated Bioconductor and CRAN package repositories provide many additional R packages for statistical data analysis for a wide array of research areas." />
<meta property="og:type" content="article" />
<meta property="og:url" content="/tutorials/rbasics/rbasics/" /><meta property="article:section" content="tutorials" />

<meta property="article:modified_time" content="2023-05-03T17:47:20-07:00" /><meta property="og:site_name" content="GEN242" />

<meta itemprop="name" content="Introduction to R">
<meta itemprop="description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Overview What is R? R is a powerful statistical environment and programming language for the analysis and visualization of data. The associated Bioconductor and CRAN package repositories provide many additional R packages for statistical data analysis for a wide array of research areas.">
<meta itemprop="dateModified" content="2023-05-03T17:47:20-07:00" />
<meta itemprop="wordCount" content="4492">
<meta itemprop="keywords" content="" /><meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="Introduction to R"/>
<meta name="twitter:description" content="document.addEventListener(&#34;DOMContentLoaded&#34;, function() { document.querySelector(&#34;h1&#34;).className = &#34;title&#34;; });  document.addEventListener(&#34;DOMContentLoaded&#34;, function() { var links = document.links; for (var i = 0, linksLength = links.length; i Source code downloads: [ .Rmd ] [ .R ]
 Overview What is R? R is a powerful statistical environment and programming language for the analysis and visualization of data. The associated Bioconductor and CRAN package repositories provide many additional R packages for statistical data analysis for a wide array of research areas."/>


<script type="application/javascript">
var doNotTrack = false;
if (!doNotTrack) {
	window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
	ga('create', 'UA-00000000-0', 'auto');
	
	ga('send', 'pageview');
}
</script>
<script async src='https://www.google-analytics.com/analytics.js'></script>






<link rel="preload" href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" as="style">
<link href="../../../scss/main.min.299fcc8a804d9dd8538b845649d3793be4c1fbef1adf9ec85636bbda3364c9cf.css" rel="stylesheet" integrity="">


<script
  src="https://code.jquery.com/jquery-3.5.1.min.js"
  integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0="
  crossorigin="anonymous"></script>

<script
  src="https://unpkg.com/lunr@2.3.8/lunr.min.js"
  integrity="sha384-vRQ9bDyE0Wnu+lMfm57BlYLO0/XauFuKpVsZPs7KEDwYKktWi5+Kz3MP8++DFlRY"
  crossorigin="anonymous"></script>



<link rel="stylesheet" href="../../../css/prism.css"/>


<script src="../../../js/driver0.9.8.min.js"></script>
<link rel="stylesheet" href="../../../css/driver0.9.8.min.css"/>


<link rel="stylesheet" href="../../../css/site.css"/>
<script src="../../../js/site.js"></script>



  </head>
  <body class="td-page line-numbers">
    <header>
     
<nav class="js-navbar-scroll navbar navbar-expand navbar-dark flex-column flex-md-row td-navbar">
  <a class="navbar-brand" href="../../../">
		<span class="font-weight-bold">GEN242</span>
	</a>
	<row class="header-switch">
	  	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="sidebarSwitch" checked=true>
        <label class="custom-control-label" for="sidebarSwitch">sidebar</label>
      </div>
    	<div class="custom-control custom-switch">
        <input type="checkbox" class="custom-control-input" id="tocSwitch" checked=true>
        <label class="custom-control-label" for="tocSwitch">toc</label>
      </div>
	</row>
	<div class="td-navbar-nav-scroll ml-md-auto" id="main_navbar">
		<ul class="navbar-nav mt-2 mt-lg-0">
			
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../about/" id="navbarDropdown" role="button">
          	About
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/introduction/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/syllabus/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Syllabus
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/schedule/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Schedule
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../about/internal/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Internal
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/internal_resources/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Resources
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../about/internal/install/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Installation
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../slides/" id="navbarDropdown" role="button">
          	Slides
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_01"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Course Introduction (S1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_02"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Genome Basics (S2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_03"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Databases and Software (S3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_04"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Git and GitHub (S4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_05"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux Cluster Basics (S5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_06"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 DNA Sequencing (S6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_07"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (S7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_08"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Sequence Alignments (S8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_09"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (S9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_10"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Multiple Alignments (S10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_11"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Short Read Alignments (S11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_12"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R on HPC (S12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_13"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (S13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_14"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Gene Expression Analysis (S14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_15"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Workflows (S15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_16/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Overview (S16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_17/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Enrichment Analysis (S17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_18/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis (S18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_19/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (S19)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_20/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics in R (S20)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_21/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (S21)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_22/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (S22)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_23/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflow Review (S23)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_24/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Final Discussion (S24)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../slides/slides_25/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Additional Slide Shows (S25)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../tutorials/" id="navbarDropdown" role="button">
          	Tutorials
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/github/github/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Introduction (T1)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/linux/linux/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Linux and HPCC Cluster (T2)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rbasics/rbasics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Introduction to R (T3)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rprogramming/rprogramming/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Programming in R (T4)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rparallel/rparallel/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Parallel R (T5)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rsequences/rsequences/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 NGS Analysis Basics (T6)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/systempiper/systempiper/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Workflows (T7)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/sprnaseq/sprnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 RNA-Seq Workflow (T8)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/spchipseq/spchipseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 ChIP-Seq Workflow (T9)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/scrnaseq/scrnaseq/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 scRNA-Seq Embedding Methods (T10)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rmarkdown/rmarkdown/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 R Markdown (T11)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rfea/rfea/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Functional Enrichment Analysis (T12)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rclustering/rclustering/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Cluster Analysis and Data Mining (T13)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Project Data (T14)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rgraphics/rgraphics/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Graphics and Visualization (T15)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/shinyapps/shinyapps/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Shiny Apps (T16)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/rpackages/rpackages/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Building R Packages (T17)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/dplyr/dplyr/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 dplyr, tidyr and SQLite (T18)
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../tutorials/cmdtocwl/cmdtocwl/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Automate CWL Creation (T19)
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../assignments/" id="navbarDropdown" role="button">
          	Assignments
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/homework/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Homework
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw01/hw01/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW1: Intro &amp; GitHub
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw02/hw02/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW2: Linux &amp; HPC
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw03/hw03/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW3: R Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw04/hw04/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW4: Alignments
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw05/hw05"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW5: R Programming
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw06/hw06"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW6: NGS Analysis Basics
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw07/hw07"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW7: RNA-Seq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/homework/hw08/hw08"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 HW8: R Markdown
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/projects/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_overview/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Overview of Projects
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/01_rnaseq_aligners/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 RNA-Seq Aligners
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_deg/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEG Methods
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/02_rnaseq_dex/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 DEXseq Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/03_cluster_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Cluster Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/04_scrnaseq_embedding/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 scRNA-Seq Embedding
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/05_chipseq_peakcaller/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 ChIP-Seq Peak Callers
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/06_functional_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Functional Enrichment
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/07_motif_enrichment/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Motif Enrichment*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/08_genome_graphics/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Genome Graphics*
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/09_drug_target_analysis/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Drug-target Analysis
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/10_lncrna_orf_discovery/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 lncRNA Discovery
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/projects/project_data/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Project Data
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Presentations
              </a>
          	
          	  
              
                
              
              
              <a class="dropdown-item" href="../../../assignments/presentations/paper_presentations/"
              style="padding-left: 2rem; font-weight: calc(3/4*500)">
                 Paper Presentations
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
          <a class="nav-link dropdown-toggle" href="../../../external_resources/" id="navbarDropdown" role="button">
          	Links
          </a>
          <div class="dropdown-menu" aria-labelledby="navbarDropdownMenuLink">
          	
          	  
              
              
              <a class="dropdown-item" href="https://piazza.com/ucr/spring2023/gen242/home"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Piazza
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://classroom.github.com/classrooms/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Classroom
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://github.com/tgirke/GEN242"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 GitHub Repo
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://bioconductor.org/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Bioconductor
              </a>
          	
          	  
              
              
              <a class="dropdown-item" href="https://www.dcassol.com/docsy_Rmarkdown/rmarkdown/install/"
              style="padding-left: 1rem; font-weight: calc(3/3*500)">
                 Hugo, Docsy &amp; R
              </a>
          	
          </div>
				
			</li>
			
			<li class="nav-item mr-4 mb-2 mb-lg-0" style="position: relative">
				
				
				
				
				
				
				
					<a class="nav-link" href="../../../blog/" ><span>News</span></a>
				
			</li>
			
			
			
		</ul>
	</div>
	<div class="navbar-nav d-none d-lg-block">



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.76afbca159bd2c1a9dbb8637d05ec4a2.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>

</div>
</nav>

    </header>
    <div class="container-fluid td-outer">
      <div class="td-main">
        <div class="row flex-xl-nowrap">
          <div class="col-12 col-md-3 col-xl-2 td-sidebar d-print-none">
            




<div id="td-sidebar-menu" class="td-sidebar__inner">
  
  <form class="td-sidebar__search d-flex align-items-center">
    



<input
  type="search"
  class="form-control td-search-input"
  placeholder="&#xf002; Search this site…"
  aria-label="Search this site…"
  autocomplete="off"
  
  data-offline-search-index-json-src="../../../offline-search-index.76afbca159bd2c1a9dbb8637d05ec4a2.json"
  data-offline-search-base-href="../../../"
  data-offline-search-max-results="10"
>


    <button class="btn btn-link td-sidebar__toggle d-md-none p-0 ml-3 fas fa-bars" type="button" data-toggle="collapse" data-target="#td-section-nav" aria-controls="td-docs-nav" aria-expanded="false" aria-label="Toggle section navigation">
    </button>
  </form>
  
  <nav class="collapse td-sidebar-nav" id="td-section-nav">
    
    
    






<ul class="td-sidebar-nav__section pr-md-3">
  <li class="td-sidebar-nav__section-title">
    <a  href="../../../tutorials/" class="align-left pl-0 pr-2 active td-sidebar-link td-sidebar-link__section">Tutorials</a>
  </li>
  <ul>
    <li class="collapse show" id="tutorials">
      
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsgithubgithub" href="../../../tutorials/github/github/">GitHub</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialslinuxlinux" href="../../../tutorials/linux/linux/">Linux &amp; HPCC Cluster</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page  active" id="m-tutorialsrbasicsrbasics" href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrprogrammingrprogramming" href="../../../tutorials/rprogramming/rprogramming/">Programming in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrparallelrparallel" href="../../../tutorials/rparallel/rparallel/">Parallel Evaluations in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrsequencesrsequences" href="../../../tutorials/rsequences/rsequences/">NGS Analysis Basics</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssystempipersystempiper" href="../../../tutorials/systempiper/systempiper/">systemPipeR: Workflow Design and Reporting Environment</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialssprnaseqsprnaseq" href="../../../tutorials/sprnaseq/sprnaseq/">RNA-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsspchipseqspchipseq" href="../../../tutorials/spchipseq/spchipseq/">ChIP-Seq Workflow Template</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsscrnaseqscrnaseq" href="../../../tutorials/scrnaseq/scrnaseq/">scRNA-Seq Embedding Methods</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrmarkdownrmarkdown" href="../../../tutorials/rmarkdown/rmarkdown/">R Markdown Tutorial</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrfearfea" href="../../../tutorials/rfea/rfea/">Functional Enrichment Analysis</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrclusteringrclustering" href="../../../tutorials/rclustering/rclustering/">Cluster Analysis in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrgraphicsrgraphics" href="../../../tutorials/rgraphics/rgraphics/">Graphics and Data Visualization in R</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappslezhangshinyapps" href="../../../tutorials/shinyappslezhang/shinyapps/">Shiny Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsshinyappsshinyapps" href="../../../tutorials/shinyapps/shinyapps/">Shiny Web Apps</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsrpackagesrpackages" href="../../../tutorials/rpackages/rpackages/">Building R Packages</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialsdplyrdplyr" href="../../../tutorials/dplyr/dplyr/">Environments dplyr, tidyr and some SQLite</a>
        
      
      
      
        
          
          
          <a class="td-sidebar-link td-sidebar-link__page " id="m-tutorialscmdtocwlcmdtocwl" href="../../../tutorials/cmdtocwl/cmdtocwl/">Automate Creation of CWL Instructions</a>
        
      
      
    </li>
  </ul>
</ul>

  </nav>
</div>




          </div>
          
          

          
            
          

          
          
              <div class="d-none d-xl-block col-xl-2 td-toc d-print-none">
                
  
  
  
  
  
  <div class="td-page-meta ml-2 pb-1 pt-2 mb-0">
  
    
    
      
    
    
    
    
    
    
    
    


    
    
      <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rbasics/Rbasics.Rmd" target="_blank" class="source-link"><i class="fa fa-code fa-fw"></i> View Rmd source</a>
    

    <a href="https://github.com/tgirke/GEN242//edit/main/content/en/tutorials/rbasics/Rbasics.md" target="_blank"><i class="fa fa-edit fa-fw"></i> Edit this page</a>
    <a href="https://github.com/tgirke/GEN242//new/main/content/en/tutorials/rbasics/Rbasics.md?filename=change-me.md&amp;value=---%0Atitle%3A&#43;%22Long&#43;Page&#43;Title%22%0AlinkTitle%3A&#43;%22Short&#43;Nav&#43;Title%22%0Aweight%3A&#43;100%0Adescription%3A&#43;%3E-%0A&#43;&#43;&#43;&#43;&#43;Page&#43;description&#43;for&#43;heading&#43;and&#43;indexes.%0A---%0A%0A%23%23&#43;Heading%0A%0AEdit&#43;this&#43;template&#43;to&#43;create&#43;your&#43;new&#43;page.%0A%0A%2A&#43;Give&#43;it&#43;a&#43;good&#43;name%2C&#43;ending&#43;in&#43;%60.md%60&#43;-&#43;e.g.&#43;%60getting-started.md%60%0A%2A&#43;Edit&#43;the&#43;%22front&#43;matter%22&#43;section&#43;at&#43;the&#43;top&#43;of&#43;the&#43;page&#43;%28weight&#43;controls&#43;how&#43;its&#43;ordered&#43;amongst&#43;other&#43;pages&#43;in&#43;the&#43;same&#43;directory%3B&#43;lowest&#43;number&#43;first%29.%0A%2A&#43;Add&#43;a&#43;good&#43;commit&#43;message&#43;at&#43;the&#43;bottom&#43;of&#43;the&#43;page&#43;%28%3C80&#43;characters%3B&#43;use&#43;the&#43;extended&#43;description&#43;field&#43;for&#43;more&#43;detail%29.%0A%2A&#43;Create&#43;a&#43;new&#43;branch&#43;so&#43;you&#43;can&#43;preview&#43;your&#43;new&#43;file&#43;and&#43;request&#43;a&#43;review&#43;via&#43;Pull&#43;Request.%0A" target="_blank"><i class="fa fa-edit fa-fw"></i> Create child page</a>
    <a href="https://github.com/tgirke/GEN242//issues/new?title=Introduction%20to%20R" target="_blank"><i class="fab fa-github fa-fw"></i> Create documentation issue</a>
    
      
      <a href="https://github.com/tgirke/GEN242/issues/new" target="_blank"><i class="fas fa-tasks fa-fw"></i> Create project issue</a>
    

  
  
  </div>





<nav id="TableOfContents">
  <ul>
    <li><a href="#overview">Overview</a></li>
    <li><a href="#what-is-r">What is R?</a></li>
    <li><a href="#why-using-r">Why Using R?</a></li>
    <li><a href="#books-and-documentation">Books and Documentation</a></li>
    <li><a href="#r-working-environments">R Working Environments</a>
      <ul>
        <li><a href="#example-rstudio">Example: RStudio</a></li>
        <li><a href="#example-nvim-r-tmux">Example: Nvim-R-Tmux</a></li>
      </ul>
    </li>
    <li><a href="#r-package-repositories">R Package Repositories</a></li>
    <li><a href="#working-routine-for-tutorials">Working routine for tutorials</a></li>
    <li><a href="#installation-of-r-rstudio-and-r-packages">Installation of R, RStudio and R Packages</a></li>
    <li><a href="#getting-around">Getting Around</a>
      <ul>
        <li><a href="#startup-and-closing-behavior">Startup and Closing Behavior</a></li>
      </ul>
    </li>
    <li><a href="#navigating-directories">Navigating directories</a></li>
    <li><a href="#basic-syntax">Basic Syntax</a></li>
    <li><a href="#data-types">Data Types</a>
      <ul>
        <li><a href="#numeric-data">Numeric data</a></li>
        <li><a href="#character-data">Character data</a></li>
        <li><a href="#complex-data">Complex data</a></li>
        <li><a href="#logical-data">Logical data</a></li>
      </ul>
    </li>
    <li><a href="#data-objects">Data Objects</a>
      <ul>
        <li><a href="#object-types">Object types</a>
          <ul>
            <li><a href="#vectors-1d">Vectors (1D)</a></li>
            <li><a href="#factors-1d">Factors (1D)</a></li>
            <li><a href="#matrices-2d">Matrices (2D)</a></li>
            <li><a href="#data-frames-2d">Data Frames (2D)</a></li>
            <li><a href="#tibbles">Tibbles</a></li>
            <li><a href="#arrays">Arrays</a></li>
            <li><a href="#lists">Lists</a></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><a href="#functions">Functions</a></li>
    <li><a href="#subsetting-of-data-objects">Subsetting of data objects</a></li>
    <li><a href="#important-utilities">Important Utilities</a>
      <ul>
        <li><a href="#combining-objects">Combining Objects</a></li>
        <li><a href="#accessing-dimensions-of-objects">Accessing Dimensions of Objects</a></li>
        <li><a href="#accessing-name-slots-of-objects">Accessing Name Slots of Objects</a></li>
        <li><a href="#sorting-objects">Sorting Objects</a></li>
      </ul>
    </li>
    <li><a href="#operators-and-calculations">Operators and Calculations</a>
      <ul>
        <li><a href="#comparison-operators">Comparison Operators</a></li>
        <li><a href="#basic-calculations">Basic Calculations</a></li>
      </ul>
    </li>
    <li><a href="#reading-and-writing-external-data">Reading and Writing External Data</a>
      <ul>
        <li><a href="#import-of-tabular-data">Import of tabular data</a></li>
        <li><a href="#export-of-tabular-data">Export of tabular data</a></li>
        <li><a href="#line-wise-import">Line-wise import</a></li>
        <li><a href="#line-wise-export">Line-wise export</a></li>
        <li><a href="#export-r-object">Export R object</a></li>
        <li><a href="#import-r-object">Import R object</a></li>
        <li><a href="#copy-and-paste-into-r">Copy and paste into R</a></li>
        <li><a href="#copy-and-paste-from-r">Copy and paste from R</a></li>
        <li><a href="#homework-3a">Homework 3A</a></li>
      </ul>
    </li>
    <li><a href="#useful-r-functions">Useful R Functions</a>
      <ul>
        <li><a href="#unique-entries">Unique entries</a></li>
        <li><a href="#count-occurrences">Count occurrences</a></li>
        <li><a href="#aggregate-data">Aggregate data</a></li>
        <li><a href="#intersect-data">Intersect data</a></li>
        <li><a href="#merge-data-frames">Merge data frames</a></li>
      </ul>
    </li>
    <li><a href="#graphics-in-r">Graphics in R</a>
      <ul>
        <li><a href="#advantages">Advantages</a></li>
        <li><a href="#documentation-for-r-graphics">Documentation for R Graphics</a></li>
        <li><a href="#graphics-environments">Graphics Environments</a></li>
        <li><a href="#base-graphics-overview">Base Graphics: Overview</a>
          <ul>
            <li><a href="#preferred-object-types">Preferred Object Types</a></li>
          </ul>
        </li>
        <li><a href="#scatter-plots">Scatter Plots</a></li>
        <li><a href="#basic-scatter-plot">Basic Scatter Plot</a>
          <ul>
            <li><a href="#all-pairs">All pairs</a></li>
            <li><a href="#with-labels">With labels</a></li>
          </ul>
        </li>
        <li><a href="#more-examples">More examples</a>
          <ul>
            <li><a href="#add-regression-line">Add regression line</a></li>
            <li><a href="#log-scale">Log scale</a></li>
            <li><a href="#add-a-mathematical-expression">Add a mathematical expression</a></li>
          </ul>
        </li>
        <li><a href="#homework-3b">Homework 3B</a></li>
        <li><a href="#line-plots">Line Plots</a>
          <ul>
            <li><a href="#single-data-set">Single data set</a></li>
            <li><a href="#many-data-sets">Many Data Sets</a></li>
          </ul>
        </li>
        <li><a href="#bar-plots">Bar Plots</a>
          <ul>
            <li><a href="#basics">Basics</a></li>
            <li><a href="#error-bars">Error Bars</a></li>
          </ul>
        </li>
        <li><a href="#histograms">Histograms</a></li>
        <li><a href="#density-plots">Density Plots</a></li>
        <li><a href="#pie-charts">Pie Charts</a></li>
        <li><a href="#color-selection-utilities">Color Selection Utilities</a></li>
        <li><a href="#saving-graphics-to-file">Saving Graphics to File</a></li>
        <li><a href="#homework-3c">Homework 3C</a></li>
      </ul>
    </li>
    <li><a href="#analysis-routine">Analysis Routine</a>
      <ul>
        <li><a href="#overview-1">Overview</a></li>
        <li><a href="#analysis-routine-data-import">Analysis Routine: Data Import</a></li>
        <li><a href="#merging-data-frames">Merging Data Frames</a></li>
        <li><a href="#filtering-data">Filtering Data</a></li>
        <li><a href="#string-substitutions">String Substitutions</a></li>
        <li><a href="#calculations-on-data-frames">Calculations on Data Frames</a></li>
        <li><a href="#plotting-example">Plotting Example</a></li>
        <li><a href="#export-results-and-run-entire-exercise-as-script">Export Results and Run Entire Exercise as Script</a></li>
      </ul>
    </li>
    <li><a href="#miscellaneous-topics">Miscellaneous Topics</a>
      <ul>
        <li><a href="#upgrading-to-new-rbioc-versions">Upgrading to New R/Bioc Versions</a></li>
      </ul>
    </li>
    <li><a href="#session-info">Session Info</a></li>
    <li><a href="#references">References</a></li>
  </ul>
</nav>



              </div>
              <main class="col-12 col-md-9 col-xl-8 pl-md-5" role="main">
                
  

                <nav aria-label="breadcrumb" class="d-none d-md-block d-print-none">
	<ol class="breadcrumb spb-1">
		







<li class="breadcrumb-item" >
	<a href="../../../tutorials/">Tutorials</a>
</li>




<li class="breadcrumb-item active" aria-current="page">
	<a href="../../../tutorials/rbasics/rbasics/">Introduction to R</a>
</li>

	</ol>
</nav	>

                
<div class="td-content">
	<h1>Introduction to R</h1>
    
	
        <p class="reading-time"><i class="fa fa-clock" aria-hidden="true"></i> 22 minute read</p>
           
	<!---
- Compile from command-line
Rscript -e "rmarkdown::render('Rbasics.Rmd', c('html_document'), clean=FALSE); knitr::knit('Rbasics.Rmd', tangle=TRUE)"; Rscript ../md2jekyll.R Rbasics.knit.md 8; Rscript -e "rmarkdown::render('Rbasics.Rmd', c('pdf_document'))"
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>
<div style="text-align: right">
<p>Source code downloads:    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rbasics/Rbasics.Rmd">.Rmd</a> ]    
[ <a href="https://raw.githubusercontent.com/tgirke/GEN242//main/content/en/tutorials/rbasics/Rbasics.R">.R</a> ]</p>
</div>
<h2 id="overview">Overview</h2>
<h2 id="what-is-r">What is R?</h2>
<p><a href="http://cran.at.r-project.org">R</a> is a powerful statistical environment and
programming language for the analysis and visualization of data. The
associated <a href="http://bioconductor.org/">Bioconductor</a> and CRAN package
repositories provide many additional R packages for statistical data analysis
for a wide array of research areas. The R software is free and runs on all
common operating systems.</p>
<h2 id="why-using-r">Why Using R?</h2>
<ul>
<li>Complete statistical environment and programming language</li>
<li>Efficient functions and data structures for data analysis</li>
<li>Powerful graphics</li>
<li>Access to fast growing number of analysis packages</li>
<li>Most widely used language in bioinformatics</li>
<li>Is standard for data mining and biostatistical analysis</li>
<li>Technical advantages: free, open-source, available for all OSs</li>
</ul>
<h2 id="books-and-documentation">Books and Documentation</h2>
<ul>
<li>simpleR - Using R for Introductory Statistics (John Verzani, 2004) - <a href="http://cran.r-project.org/doc/contrib/Verzani-SimpleR.pdf">URL</a></li>
<li>Bioinformatics and Computational Biology Solutions Using R and Bioconductor (Gentleman et al., 2005) - <a href="http://www.bioconductor.org/help/publications/books/bioinformatics-and-computational-biology-solutions/">URL</a></li>
<li>More on this see “Finding Help” section in UCR Manual - <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual#TOC-Finding-Help">URL</a></li>
</ul>
<h2 id="r-working-environments">R Working Environments</h2>
<center>
<img title="R_Interfaces" src="../images/rinterface.png"/>
</center>
<center>
<p>R Projects and Interfaces</p>
</center>
<p>Some R working environments with support for syntax highlighting and utilities to send code
to the R console:</p>
<ul>
<li><a href="https://www.rstudio.com/products/rstudio/features">RStudio</a>: excellent choice for beginners (<a href="http://www.rstudio.com/wp-content/uploads/2016/01/rstudio-IDE-cheatsheet.pdf">Cheat Sheet</a>)</li>
<li>Basic R code editors provided by Rguis</li>
<li><a href="https://wiki.gnome.org/Apps/Gedit">gedit</a>, <a href="http://rgedit.sourceforge.net/">Rgedit</a>, <a href="https://rkward.kde.org/">RKWard</a>, <a href="http://www.walware.de/goto/statet">Eclipse</a>, <a href="http://www.sciviews.org/Tinn-R/">Tinn-R</a>, <a href="https://notepad-plus-plus.org/">Notepad++</a>, <a href="http://sourceforge.net/projects/npptor/">NppToR</a></li>
<li><a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r/vim-r">Vim-R-Tmux</a>: R working environment based on vim and tmux</li>
<li><a href="http://www.xemacs.org/Download/index.html">Emacs</a> (<a href="http://ess.r-project.org/">ESS add-on package</a>)</li>
</ul>
<h3 id="example-rstudio">Example: RStudio</h3>
<p>New integrated development environment (IDE) for <a href="http://www.rstudio.com/ide/download/">R</a>. Highly functional for both beginners and
advanced.</p>
<center>
<img title="RStudio" src="../images/rstudio.png"/>
</center>
<center>
<p>RStudio IDE</p>
</center>
<p>Some userful shortcuts: <code>Ctrl+Enter</code> (send code), <code>Ctrl+Shift+C</code> (comment/uncomment), <code>Ctrl+1/2</code> (switch window focus)</p>
<h3 id="example-nvim-r-tmux">Example: Nvim-R-Tmux</h3>
<p>Terminal-based Working Environment for R: <a href="https://girke.bioinformatics.ucr.edu/GEN242/custom/slides/R_for_HPC/NvimR.html#1">Nvim-R-Tmux</a>.</p>
<center>
<img title="Nvim-R-Tmux" src="../images/Nvim-R.gif" >
</center>
<center>
<p>Nvim-R-Tmux IDE for R</p>
</center>
<p>Install and usage instructions for Nvim-R are provided in this <a href="https://girke.bioinformatics.ucr.edu/GEN242/custom/slides/R_for_HPC/NvimR.html#11">slide show</a> and
<a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/linux/linux/#quick-configuration-in-user-accounts-of-ucrs-hpcc">this tutorial</a>. The most detailed instructions
can be found on the <a href="https://github.com/tgirke/Nvim-R_Tmux">Nvim-R_Tmux</a> GitHub repos.</p>
<h2 id="r-package-repositories">R Package Repositories</h2>
<ul>
<li>CRAN (&gt;14,000 packages) general data analysis - <a href="http://cran.at.r-project.org/">URL</a></li>
<li>Bioconductor (&gt;2,000 packages) bioscience data analysis - <a href="http://www.bioconductor.org/">URL</a></li>
<li>Omegahat (&gt;90 packages) programming interfaces - <a href="https://github.com/omegahat?tab=repositories">URL</a></li>
<li>RStudio packages - <a href="https://www.rstudio.com/products/rpackages/">URL</a></li>
</ul>
<h2 id="working-routine-for-tutorials">Working routine for tutorials</h2>
<p>This section provides a short overview of the standard working routine users should use to load R-based tutorials of this website into an R IDE (Nvim-R or RStudio).
For Nvim-R on HPCC users can visit the Quick Demo slide <a href="https://girke.bioinformatics.ucr.edu/GEN242/custom/slides/R_for_HPC/NvimR.html#11">here</a>.</p>
<ol>
<li>
<p>Download <code>*.Rmd</code> or <code>*.R</code> file. These so called source files are always linked on the top right corner of each tutorial. The ones for this tutorial are <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/rbasics/rbasics/">here</a>.
The file download can be accomplished via <code>wget</code> from the command-line or with the save function in a user’s web browser.</p>
</li>
<li>
<p>Load <code>*.Rmd</code> or <code>*.R</code> file in Neovim (Nvim-R) or RStudio.</p>
</li>
<li>
<p>Send code from code editor to R console by pushing <code>space bar</code> in Neovim (Nvim-R) or <code>Ctrl + Enter</code> in RStudio. In <code>*.Rmd</code> files the code lines are in so called <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/rmarkdown/rmarkdown/#r-code-chunks">code chunks</a> and only those
ones can be sent to the console. To obtain in Neovim a connected R session one has to initiate by pressing the <code>\rf</code> key combination. For details see <a href="https://girke.bioinformatics.ucr.edu/GEN242/custom/slides/R_for_HPC/NvimR.html#11">here</a>.</p>
</li>
</ol>
<h2 id="installation-of-r-rstudio-and-r-packages">Installation of R, RStudio and R Packages</h2>
<ol>
<li>
<p>Install R for your operating system from <a href="http://cran.at.r-project.org/">CRAN</a>.</p>
</li>
<li>
<p>Install RStudio from <a href="http://www.rstudio.com/ide/download">RStudio</a>.</p>
</li>
<li>
<p>Install CRAN Packages from R console like this:</p>
<pre><code class="language-r">install.packages(c(&quot;pkg1&quot;, &quot;pkg2&quot;)) 
install.packages(&quot;pkg.zip&quot;, repos=NULL)
</code></pre>
</li>
<li>
<p>Install <a href="https://bioconductor.org/packages/release/BiocViews.html#___Software">Bioconductor</a> packages as follows:</p>
<pre><code class="language-r">if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
    install.packages(&quot;BiocManager&quot;) # Installs BiocManager if not available yet
BiocManager::version() # Reports Bioconductor version
BiocManager::install(c(&quot;pkg1&quot;, &quot;pkg2&quot;)) # Installs packages specified under &quot;pkg1&quot; 
</code></pre>
</li>
<li>
<p>For more details consult the <a href="http://www.bioconductor.org/install/">Bioc Install page</a>
and <a href="http://www.bioconductor.org/packages/release/bioc/html/BiocInstaller.html">BiocInstaller</a> package.</p>
</li>
<li>
<p>Instructions for upgrading R and packages to newer versions are given at the end of this tutorial <a href="#upgrading-to-new-rbioc-versions">here</a>.</p>
</li>
</ol>
<h2 id="getting-around">Getting Around</h2>
<h3 id="startup-and-closing-behavior">Startup and Closing Behavior</h3>
<ul>
<li>
<p><strong>Starting R</strong>:
The R GUI versions, including RStudio, under Windows and Mac OS X can be
opened by double-clicking their icons. Alternatively, one can start it by
typing <code>R</code> in a terminal (default under Linux).</p>
</li>
<li>
<p><strong>Startup/Closing Behavior</strong>:
The R environment is controlled by hidden files in the startup directory:
<code>.RData</code>, <code>.Rhistory</code> and <code>.Rprofile</code> (optional).</p>
</li>
<li>
<p><strong>Closing R</strong>:</p>
</li>
</ul>
<!-- end list -->
<pre><code class="language-r">q()  
</code></pre>
<pre><code>Save workspace image? [y/n/c]:
</code></pre>
<ul>
<li><strong>Note</strong>:
When responding with <code>y</code>, then the entire R workspace will be written to
the <code>.RData</code> file which can become very large. Often it is better to select <code>n</code> here,
because a much better working pratice is to save an analysis protocol to an <code>R</code> or <code>Rmd</code> source file.
This way one can quickly regenerate all data sets and objects needed in a future session.</li>
</ul>
<h2 id="navigating-directories">Navigating directories</h2>
<p>List objects in current R session</p>
<pre><code class="language-r">ls()
</code></pre>
<p>Return content of current working directory</p>
<pre><code class="language-r">dir()
</code></pre>
<p>Return path of current working directory</p>
<pre><code class="language-r">getwd()
</code></pre>
<p>Change current working directory</p>
<pre><code class="language-r">setwd(&quot;/home/user&quot;)
</code></pre>
<h2 id="basic-syntax">Basic Syntax</h2>
<p>Create an object with the assignment operator <code>&lt;-</code> or <code>=</code></p>
<pre><code class="language-r">object &lt;- ...
</code></pre>
<p>General R command syntax</p>
<pre><code class="language-r">object &lt;- function_name(arguments) 
object &lt;- object[arguments] 
</code></pre>
<p>Instead of the assignment operator one can use the <code>assign</code> function</p>
<pre><code class="language-r">assign(&quot;x&quot;, function(arguments))
</code></pre>
<p>To simplify chaining of serveral operations, <code>dplyr</code> (<code>magrittr</code>) provides the <code>%&gt;%</code> (pipe) operator,
where <code>x %&gt;% f(y)</code> turns into <code>f(x, y)</code>. This way one can pipe together multiple
operations by writing them from left-to-right or top-to-bottom. This makes for
easy to type and readable code. Details on this are provided in the dplyr tutorial <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/dplyr/dplyr/">here</a>.</p>
<pre><code class="language-r">... %&gt;% ...
</code></pre>
<p>Finding help</p>
<pre><code class="language-r">?function_name
</code></pre>
<p>Load a library/package</p>
<pre><code class="language-r">library(&quot;my_library&quot;) 
</code></pre>
<p>List functions defined by a library</p>
<pre><code class="language-r">library(help=&quot;my_library&quot;)
</code></pre>
<p>Load library manual (PDF or HTML file)</p>
<pre><code class="language-r">vignette(&quot;my_library&quot;) 
</code></pre>
<p>Execute an R script from within R</p>
<pre><code class="language-r">source(&quot;my_script.R&quot;)
</code></pre>
<p>Execute an R script from command-line (the first of the three options is preferred)</p>
<pre><code class="language-sh">$ Rscript my_script.R
$ R CMD BATCH my_script.R 
$ R --slave &lt; my_script.R 
</code></pre>
<h2 id="data-types">Data Types</h2>
<h3 id="numeric-data">Numeric data</h3>
<p>Example: <code>1, 2, 3, ...</code></p>
<pre><code class="language-r">x &lt;- c(1, 2, 3)
x
</code></pre>
<pre><code>## [1] 1 2 3
</code></pre>
<pre><code class="language-r">is.numeric(x)
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<pre><code class="language-r">as.character(x)
</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;2&quot; &quot;3&quot;
</code></pre>
<h3 id="character-data">Character data</h3>
<p>Example: <code>&quot;a&quot;, &quot;b&quot;, &quot;c&quot;, ...</code></p>
<pre><code class="language-r">x &lt;- c(&quot;1&quot;, &quot;2&quot;, &quot;3&quot;)
x
</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;2&quot; &quot;3&quot;
</code></pre>
<pre><code class="language-r">is.character(x)
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<pre><code class="language-r">as.numeric(x)
</code></pre>
<pre><code>## [1] 1 2 3
</code></pre>
<h3 id="complex-data">Complex data</h3>
<p>Example: mix of both</p>
<pre><code class="language-r">c(1, &quot;b&quot;, 3)
</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;b&quot; &quot;3&quot;
</code></pre>
<h3 id="logical-data">Logical data</h3>
<p>Example: <code>TRUE</code> of <code>FALSE</code></p>
<pre><code class="language-r">x &lt;- 1:10 &lt; 5
x  
</code></pre>
<pre><code>##  [1]  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE
</code></pre>
<pre><code class="language-r">!x
</code></pre>
<pre><code>##  [1] FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
</code></pre>
<pre><code class="language-r">which(x) # Returns index for the 'TRUE' values in logical vector
</code></pre>
<pre><code>## [1] 1 2 3 4
</code></pre>
<h2 id="data-objects">Data Objects</h2>
<h3 id="object-types">Object types</h3>
<ul>
<li>List of common object types
<ul>
<li><code>vectors</code>: ordered collection of numeric, character, complex and logical values.</li>
<li><code>factors</code>: special type vectors with grouping information of its components</li>
<li><code>data.frames</code> including modern variants <code>DataFrame</code>, <code>tibbles</code>, etc.: two dimensional structures with different data types</li>
<li><code>matrices</code>: two dimensional structures with data of same type</li>
<li><code>arrays</code>: multidimensional arrays of vectors</li>
<li><code>lists</code>: general form of vectors with different types of elements</li>
<li><code>functions</code>: piece of code</li>
<li>Many more …</li>
</ul>
</li>
<li>Simple rules for naming objects and their components
<ul>
<li>Object, row and column names should not start with a number</li>
<li>Avoid spaces in object, row and column names</li>
<li>Avoid special characters like ‘#’</li>
</ul>
</li>
</ul>
<h4 id="vectors-1d">Vectors (1D)</h4>
<p>Definition: <code>numeric</code> or <code>character</code></p>
<pre><code class="language-r">myVec &lt;- 1:10; names(myVec) &lt;- letters[1:10]  
myVec &lt;- setNames(1:10, letters[1:10]) # Same as above in single step
myVec[1:5]
</code></pre>
<pre><code>## a b c d e 
## 1 2 3 4 5
</code></pre>
<pre><code class="language-r">myVec[c(2,4,6,8)]
</code></pre>
<pre><code>## b d f h 
## 2 4 6 8
</code></pre>
<pre><code class="language-r">myVec[c(&quot;b&quot;, &quot;d&quot;, &quot;f&quot;)]
</code></pre>
<pre><code>## b d f 
## 2 4 6
</code></pre>
<h4 id="factors-1d">Factors (1D)</h4>
<p>Definition: vectors with grouping information</p>
<pre><code class="language-r">factor(c(&quot;dog&quot;, &quot;cat&quot;, &quot;mouse&quot;, &quot;dog&quot;, &quot;dog&quot;, &quot;cat&quot;))
</code></pre>
<pre><code>## [1] dog   cat   mouse dog   dog   cat  
## Levels: cat dog mouse
</code></pre>
<h4 id="matrices-2d">Matrices (2D)</h4>
<p>Definition: two dimensional structures with data of same type</p>
<pre><code class="language-r">myMA &lt;- matrix(1:30, 3, 10, byrow = TRUE) 
class(myMA)
</code></pre>
<pre><code>## [1] &quot;matrix&quot; &quot;array&quot;
</code></pre>
<pre><code class="language-r">myMA[1:2,]
</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    2    3    4    5    6    7    8    9    10
## [2,]   11   12   13   14   15   16   17   18   19    20
</code></pre>
<pre><code class="language-r">myMA[1, , drop=FALSE]
</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    2    3    4    5    6    7    8    9    10
</code></pre>
<h4 id="data-frames-2d">Data Frames (2D)</h4>
<p>Definition: <code>data.frames</code> are two dimensional objects with data of variable types</p>
<pre><code class="language-r">myDF &lt;- data.frame(Col1=1:10, Col2=10:1) 
myDF[1:2, ]
</code></pre>
<pre><code>##   Col1 Col2
## 1    1   10
## 2    2    9
</code></pre>
<h4 id="tibbles">Tibbles</h4>
<p><code>Tibbles</code> are a more modern version of <code>data.frames</code>. Among many other advantages, one can see
here that <code>tibbles</code> have a nicer printing bahavior. Much more detailed information on this object
class is provided in the <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/dplyr/dplyr/"><code>dplyr/tidyverse</code></a> manual section.</p>
<pre><code class="language-r">library(tidyverse)
as_tibble(iris)
</code></pre>
<pre><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;fct&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre>
<h4 id="arrays">Arrays</h4>
<p>Definition: data structure with one, two or more dimensions</p>
<h4 id="lists">Lists</h4>
<p>Definition: containers for any object type</p>
<pre><code class="language-r">myL &lt;- list(name=&quot;Fred&quot;, wife=&quot;Mary&quot;, no.children=3, child.ages=c(4,7,9)) 
myL
</code></pre>
<pre><code>## $name
## [1] &quot;Fred&quot;
## 
## $wife
## [1] &quot;Mary&quot;
## 
## $no.children
## [1] 3
## 
## $child.ages
## [1] 4 7 9
</code></pre>
<pre><code class="language-r">myL[[4]][1:2] 
</code></pre>
<pre><code>## [1] 4 7
</code></pre>
<h2 id="functions">Functions</h2>
<p>Definition: piece of code</p>
<pre><code class="language-r">myfct &lt;- function(arg1, arg2, ...) { 
	function_body 
}
</code></pre>
<h2 id="subsetting-of-data-objects">Subsetting of data objects</h2>
<p><strong>(1.) Subsetting by positive or negative index/position numbers</strong></p>
<pre><code class="language-r">myVec &lt;- 1:26; names(myVec) &lt;- LETTERS 
myVec[1:4]
</code></pre>
<pre><code>## A B C D 
## 1 2 3 4
</code></pre>
<p><strong>(2.) Subsetting by same length logical vectors</strong></p>
<pre><code class="language-r">myLog &lt;- myVec &gt; 10
myVec[myLog] 
</code></pre>
<pre><code>##  K  L  M  N  O  P  Q  R  S  T  U  V  W  X  Y  Z 
## 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26
</code></pre>
<p><strong>(3.) Subsetting by field names</strong></p>
<pre><code class="language-r">myVec[c(&quot;B&quot;, &quot;K&quot;, &quot;M&quot;)]
</code></pre>
<pre><code>##  B  K  M 
##  2 11 13
</code></pre>
<p><strong>(4.) Subset with <code>$</code> sign</strong>: references a single column or list component by its name</p>
<pre><code class="language-r">iris$Species[1:8]
</code></pre>
<pre><code>## [1] setosa setosa setosa setosa setosa setosa setosa setosa
## Levels: setosa versicolor virginica
</code></pre>
<h2 id="important-utilities">Important Utilities</h2>
<h3 id="combining-objects">Combining Objects</h3>
<p>The <code>c</code> function combines vectors and lists</p>
<pre><code class="language-r">c(1, 2, 3)
</code></pre>
<pre><code>## [1] 1 2 3
</code></pre>
<pre><code class="language-r">x &lt;- 1:3; y &lt;- 101:103
c(x, y)
</code></pre>
<pre><code>## [1]   1   2   3 101 102 103
</code></pre>
<p>The <code>cbind</code> and <code>rbind</code> functions can be used to append columns and rows, respecively.</p>
<pre><code class="language-r">ma &lt;- cbind(x, y)
ma
</code></pre>
<pre><code>##      x   y
## [1,] 1 101
## [2,] 2 102
## [3,] 3 103
</code></pre>
<pre><code class="language-r">rbind(ma, ma)
</code></pre>
<pre><code>##      x   y
## [1,] 1 101
## [2,] 2 102
## [3,] 3 103
## [4,] 1 101
## [5,] 2 102
## [6,] 3 103
</code></pre>
<h3 id="accessing-dimensions-of-objects">Accessing Dimensions of Objects</h3>
<p>Length and dimension information of objects</p>
<pre><code class="language-r">length(iris$Species)
</code></pre>
<pre><code>## [1] 150
</code></pre>
<pre><code class="language-r">dim(iris)
</code></pre>
<pre><code>## [1] 150   5
</code></pre>
<h3 id="accessing-name-slots-of-objects">Accessing Name Slots of Objects</h3>
<p>Accessing row and column names of 2D objects</p>
<pre><code class="language-r">rownames(iris)[1:8]
</code></pre>
<pre><code>## [1] &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot; &quot;5&quot; &quot;6&quot; &quot;7&quot; &quot;8&quot;
</code></pre>
<pre><code class="language-r">colnames(iris)
</code></pre>
<pre><code>## [1] &quot;Sepal.Length&quot; &quot;Sepal.Width&quot;  &quot;Petal.Length&quot; &quot;Petal.Width&quot;  &quot;Species&quot;
</code></pre>
<p>Return name field of vectors and lists</p>
<pre><code class="language-r">names(myVec)
</code></pre>
<pre><code>##  [1] &quot;A&quot; &quot;B&quot; &quot;C&quot; &quot;D&quot; &quot;E&quot; &quot;F&quot; &quot;G&quot; &quot;H&quot; &quot;I&quot; &quot;J&quot; &quot;K&quot; &quot;L&quot; &quot;M&quot; &quot;N&quot; &quot;O&quot; &quot;P&quot; &quot;Q&quot; &quot;R&quot; &quot;S&quot; &quot;T&quot; &quot;U&quot; &quot;V&quot; &quot;W&quot; &quot;X&quot;
## [25] &quot;Y&quot; &quot;Z&quot;
</code></pre>
<pre><code class="language-r">names(myL)
</code></pre>
<pre><code>## [1] &quot;name&quot;        &quot;wife&quot;        &quot;no.children&quot; &quot;child.ages&quot;
</code></pre>
<h3 id="sorting-objects">Sorting Objects</h3>
<p>The function <code>sort</code> returns a vector in ascending or descending order</p>
<pre><code class="language-r">sort(10:1)
</code></pre>
<pre><code>##  [1]  1  2  3  4  5  6  7  8  9 10
</code></pre>
<p>The function <code>order</code> returns a sorting index for sorting an object</p>
<pre><code class="language-r">sortindex &lt;- order(iris[,1], decreasing = FALSE)
sortindex[1:12]
</code></pre>
<pre><code>##  [1] 14  9 39 43 42  4  7 23 48  3 30 12
</code></pre>
<pre><code class="language-r">iris[sortindex,][1:2,]
</code></pre>
<pre><code>##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 14          4.3         3.0          1.1         0.1  setosa
## 9           4.4         2.9          1.4         0.2  setosa
</code></pre>
<pre><code class="language-r">sortindex &lt;- order(-iris[,1]) # Same as decreasing=TRUE
</code></pre>
<p>Sorting multiple columns</p>
<pre><code class="language-r">iris[order(iris$Sepal.Length, iris$Sepal.Width),][1:2,]
</code></pre>
<pre><code>##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 14          4.3         3.0          1.1         0.1  setosa
## 9           4.4         2.9          1.4         0.2  setosa
</code></pre>
<h2 id="operators-and-calculations">Operators and Calculations</h2>
<h3 id="comparison-operators">Comparison Operators</h3>
<p>Comparison operators: <code>==</code>, <code>!=</code>, <code>&lt;</code>, <code>&gt;</code>, <code>&lt;=</code>, <code>&gt;=</code></p>
<pre><code class="language-r">1==1
</code></pre>
<pre><code>## [1] TRUE
</code></pre>
<p>Logical operators for boolean operations: AND: <code>&amp;</code>, OR: <code>|</code>, NOT: <code>!</code></p>
<pre><code class="language-r">x &lt;- 1:10; y &lt;- 10:1
x &gt; y &amp; x &gt; 5
</code></pre>
<pre><code>##  [1] FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE
</code></pre>
<h3 id="basic-calculations">Basic Calculations</h3>
<p>To look up math functions, see Function Index <a href="http://cran.at.r-project.org/doc/manuals/R-intro.html#Function-and-variable-index">here</a></p>
<pre><code class="language-r">x + y
</code></pre>
<pre><code>##  [1] 11 11 11 11 11 11 11 11 11 11
</code></pre>
<pre><code class="language-r">sum(x)
</code></pre>
<pre><code>## [1] 55
</code></pre>
<pre><code class="language-r">mean(x)
</code></pre>
<pre><code>## [1] 5.5
</code></pre>
<pre><code class="language-r">apply(iris[1:6,1:3], 1, mean) 
</code></pre>
<pre><code>##        1        2        3        4        5        6 
## 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667
</code></pre>
<h2 id="reading-and-writing-external-data">Reading and Writing External Data</h2>
<h3 id="import-of-tabular-data">Import of tabular data</h3>
<p>Import of a tab-delimited tabular file</p>
<pre><code class="language-r">myDF &lt;- read.delim(&quot;myData.xls&quot;, sep=&quot;\t&quot;)
</code></pre>
<p>Import of Google Sheets. The following example imports a sample Google Sheet from <a href="https://docs.google.com/spreadsheets/d/1U-32UcwZP1k3saKeaH1mbvEAOfZRdNHNkWK2GI1rpPM/edit#gid=472150521">here</a>.
Detailed instructions for interacting from R with Google Sheets with the required <code>googlesheets4</code> package are <a href="https://googlesheets4.tidyverse.org/">here</a>.</p>
<pre><code class="language-r">library(googlesheets4)
gs4_deauth() # Easiest method for reading public access sheets
mysheet &lt;- read_sheet(&quot;1U-32UcwZP1k3saKeaH1mbvEAOfZRdNHNkWK2GI1rpPM&quot;, skip=4)
myDF &lt;- as.data.frame(mysheet)
myDF
</code></pre>
<p>Import from Excel sheets works well with <code>readxl</code>. For details see the <code>readxl</code> package manual <a href="https://readxl.tidyverse.org/">here</a>. Note: working with tab- or comma-delimited files is more flexible and highly preferred for automated analysis workflows.</p>
<pre><code class="language-r">library(&quot;readxl&quot;)
mysheet &lt;- read_excel(targets_path, sheet=&quot;Sheet1&quot;)
</code></pre>
<p>Additional import functions are described in the <code>readr</code> package section <a href="https://girke.bioinformatics.ucr.edu/GEN242/tutorials/dplyr/dplyr/#reading-and-writing-tabular-files">here</a>.</p>
<h3 id="export-of-tabular-data">Export of tabular data</h3>
<pre><code class="language-r">write.table(myDF, file=&quot;myfile.xls&quot;, sep=&quot;\t&quot;, quote=FALSE, col.names=NA)
</code></pre>
<h3 id="line-wise-import">Line-wise import</h3>
<pre><code class="language-r">myDF &lt;- readLines(&quot;myData.txt&quot;)
</code></pre>
<h3 id="line-wise-export">Line-wise export</h3>
<pre><code class="language-r">writeLines(month.name, &quot;myData.txt&quot;)
</code></pre>
<h3 id="export-r-object">Export R object</h3>
<pre><code class="language-r">mylist &lt;- list(C1=iris[,1], C2=iris[,2]) # Example to export
saveRDS(mylist, &quot;mylist.rds&quot;)
</code></pre>
<h3 id="import-r-object">Import R object</h3>
<pre><code class="language-r">mylist &lt;- readRDS(&quot;mylist.rds&quot;)
</code></pre>
<h3 id="copy-and-paste-into-r">Copy and paste into R</h3>
<p>On Windows/Linux systems</p>
<pre><code class="language-r">read.delim(&quot;clipboard&quot;) 
</code></pre>
<p>On Mac OS X systems</p>
<pre><code class="language-r">read.delim(pipe(&quot;pbpaste&quot;)) 
</code></pre>
<h3 id="copy-and-paste-from-r">Copy and paste from R</h3>
<p>On Windows/Linux systems</p>
<pre><code class="language-r">write.table(iris, &quot;clipboard&quot;, sep=&quot;\t&quot;, col.names=NA, quote=FALSE) 
</code></pre>
<p>On Mac OS X systems</p>
<pre><code class="language-r">zz &lt;- pipe('pbcopy', 'w')
write.table(iris, zz, sep=&quot;\t&quot;, col.names=NA, quote=FALSE)
close(zz) 
</code></pre>
<h3 id="homework-3a">Homework 3A</h3>
<p>Homework 3A: <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw03/hw03/">Object Subsetting Routines and Import/Export</a></p>
<h2 id="useful-r-functions">Useful R Functions</h2>
<h3 id="unique-entries">Unique entries</h3>
<p>Make vector entries unique with <code>unique</code></p>
<pre><code class="language-r">length(iris$Sepal.Length)
</code></pre>
<pre><code>## [1] 150
</code></pre>
<pre><code class="language-r">length(unique(iris$Sepal.Length))
</code></pre>
<pre><code>## [1] 35
</code></pre>
<h3 id="count-occurrences">Count occurrences</h3>
<p>Count occurrences of entries with <code>table</code></p>
<pre><code class="language-r">table(iris$Species)
</code></pre>
<pre><code>## 
##     setosa versicolor  virginica 
##         50         50         50
</code></pre>
<h3 id="aggregate-data">Aggregate data</h3>
<p>Compute aggregate statistics with <code>aggregate</code></p>
<pre><code class="language-r">aggregate(iris[,1:4], by=list(iris$Species), FUN=mean, na.rm=TRUE)
</code></pre>
<pre><code>##      Group.1 Sepal.Length Sepal.Width Petal.Length Petal.Width
## 1     setosa        5.006       3.428        1.462       0.246
## 2 versicolor        5.936       2.770        4.260       1.326
## 3  virginica        6.588       2.974        5.552       2.026
</code></pre>
<h3 id="intersect-data">Intersect data</h3>
<p>Compute intersect between two vectors with <code>%in%</code></p>
<pre><code class="language-r">month.name %in% c(&quot;May&quot;, &quot;July&quot;)
</code></pre>
<pre><code>##  [1] FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
</code></pre>
<h3 id="merge-data-frames">Merge data frames</h3>
<p>Join two data frames by common field entries with <code>merge</code> (here row names <code>by.x=0</code>). To obtain only the common rows, change <code>all=TRUE</code> to <code>all=FALSE</code>. To merge on specific columns, refer to them by their position numbers or their column names.</p>
<pre><code class="language-r">frame1 &lt;- iris[sample(1:length(iris[,1]), 30), ]
frame1[1:2,]
</code></pre>
<pre><code>##     Sepal.Length Sepal.Width Petal.Length Petal.Width   Species
## 147          6.3         2.5          5.0         1.9 virginica
## 49           5.3         3.7          1.5         0.2    setosa
</code></pre>
<pre><code class="language-r">dim(frame1)
</code></pre>
<pre><code>## [1] 30  5
</code></pre>
<pre><code class="language-r">my_result &lt;- merge(frame1, iris, by.x = 0, by.y = 0, all = TRUE)
dim(my_result)
</code></pre>
<pre><code>## [1] 150  11
</code></pre>
<h2 id="graphics-in-r">Graphics in R</h2>
<h3 id="advantages">Advantages</h3>
<ul>
<li>Powerful environment for visualizing scientific data</li>
<li>Integrated graphics and statistics infrastructure</li>
<li>Publication quality graphics</li>
<li>Fully programmable</li>
<li>Highly reproducible</li>
<li>Full <a href="http://www.latex-project.org/">LaTeX</a> and Markdown support via <code>knitr</code> and <code>R markdown</code></li>
<li>Vast number of R packages with graphics utilities</li>
</ul>
<h3 id="documentation-for-r-graphics">Documentation for R Graphics</h3>
<p><strong>General</strong></p>
<ul>
<li>Graphics Task Page - <a href="http://cran.r-project.org/web/views/Graphics.html">URL</a></li>
<li>R Graph Gallery - <a href="http://addictedtor.free.fr/graphiques/allgraph.php">URL</a></li>
<li>R Graphical Manual - <a href="http://cged.genes.nig.ac.jp/RGM2/index.php">URL</a></li>
<li>Paul Murrell’s book R (Grid) Graphics - <a href="http://www.stat.auckland.ac.nz/~paul/RGraphics/rgraphics.html">URL</a></li>
</ul>
<p><strong>Interactive graphics</strong></p>
<ul>
<li>rggobi` (GGobi) - <a href="http://www.ggobi.org/">URL</a></li>
<li><code>iplots</code> - <a href="http://www.rosuda.org/iplots/">URL</a></li>
<li>Open GL (<code>rgl</code>) - <a href="http://rgl.neoscientists.org/gallery.shtml">URL</a></li>
</ul>
<h3 id="graphics-environments">Graphics Environments</h3>
<p><strong>Viewing and saving graphics in R</strong></p>
<ul>
<li>On-screen graphics</li>
<li>postscript, pdf, svg</li>
<li>jpeg, png, wmf, tiff, …</li>
</ul>
<p><strong>Four major graphic environments</strong></p>
<ol>
<li>Low-level infrastructure</li>
</ol>
<!-- end list -->
<ul>
<li>R Base Graphics (low- and high-level)</li>
<li><code>grid</code>: <a href="http://www.stat.auckland.ac.nz/~paul/grid/grid.html">Manual</a></li>
</ul>
<!-- end list -->
<ol start="2">
<li>High-level infrastructure
\begin{itemize}</li>
</ol>
<!-- end list -->
<ul>
<li><code>lattice</code>: <a href="http://lmdvr.r-forge.r-project.org">Manual</a>, <a href="http://www.his.sunderland.ac.uk/~cs0her/Statistics/UsingLatticeGraphicsInR.htm">Intro</a>, <a href="http://www.amazon.com/Lattice-Multivariate-Data-Visualization-Use/dp/0387759689">Book</a></li>
<li><code>ggplot2</code>: <a href="http://had.co.nz/ggplot2/">Manual</a>, <a href="http://www.ling.upenn.edu/~joseff/rstudy/summer2010_ggplot2_intro.html">Intro</a>, <a href="http://had.co.nz/ggplot2/book/">Book</a></li>
</ul>
<h3 id="base-graphics-overview">Base Graphics: Overview</h3>
<p><strong>Important high-level plotting functions</strong></p>
<ul>
<li><code>plot</code>: generic x-y plotting</li>
<li><code>barplot</code>: bar plots</li>
<li><code>boxplot</code>: box-and-whisker plot</li>
<li><code>hist</code>: histograms</li>
<li><code>pie</code>: pie charts</li>
<li><code>dotchart</code>: cleveland dot plots</li>
<li><code>image, heatmap, contour, persp</code>: functions to generate image-like plots</li>
<li><code>qqnorm, qqline, qqplot</code>: distribution comparison plots</li>
<li><code>pairs, coplot</code>: display of multivariant data</li>
</ul>
<p><strong>Help on graphics functions</strong></p>
<ul>
<li><code>?myfct</code></li>
<li><code>?plot</code></li>
<li><code>?par</code></li>
</ul>
<h4 id="preferred-object-types">Preferred Object Types</h4>
<ul>
<li>Matrices and data frames</li>
<li>Vectors</li>
<li>Named vectors</li>
</ul>
<h3 id="scatter-plots">Scatter Plots</h3>
<h3 id="basic-scatter-plot">Basic Scatter Plot</h3>
<p>Sample data set for subsequent plots</p>
<pre><code class="language-r">set.seed(1410)
y &lt;- matrix(runif(30), ncol=3, dimnames=list(letters[1:10], LETTERS[1:3]))
</code></pre>
<p>Plot data</p>
<pre><code class="language-r">plot(y[,1], y[,2]) 
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/basic_scatter_plot-1.png" width="672" />
<h4 id="all-pairs">All pairs</h4>
<pre><code class="language-r">pairs(y) 
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/pairs_scatter_plot-1.png" width="672" />
<h4 id="with-labels">With labels</h4>
<pre><code class="language-r">plot(y[,1], y[,2], pch=20, col=&quot;red&quot;, main=&quot;Symbols and Labels&quot;)
text(y[,1]+0.03, y[,2], rownames(y))
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/labels_scatter_plot-1.png" width="672" />
<h3 id="more-examples">More examples</h3>
<p><strong>Print instead of symbols the row names</strong></p>
<pre><code class="language-r">plot(y[,1], y[,2], type=&quot;n&quot;, main=&quot;Plot of Labels&quot;)
text(y[,1], y[,2], rownames(y)) 
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/row_scatter_plot-1.png" width="672" />
<p><strong>Usage of important plotting parameters</strong></p>
<pre><code class="language-r">grid(5, 5, lwd = 2) 
op &lt;- par(mar=c(8,8,8,8), bg=&quot;lightblue&quot;)
plot(y[,1], y[,2], type=&quot;p&quot;, col=&quot;red&quot;, cex.lab=1.2, cex.axis=1.2, 
     cex.main=1.2, cex.sub=1, lwd=4, pch=20, xlab=&quot;x label&quot;, 
     ylab=&quot;y label&quot;, main=&quot;My Main&quot;, sub=&quot;My Sub&quot;)
par(op)
</code></pre>
<p><strong>Important arguments</strong></p>
<ul>
<li><code>mar</code>: specifies the margin sizes around the plotting area in order: <code>c(bottom, left, top, right)</code></li>
<li><code>col</code>: color of symbols</li>
<li><code>pch</code>: type of symbols, samples: <code>example(points)</code></li>
<li><code>lwd</code>: size of symbols</li>
<li><code>cex.*</code>: control font sizes</li>
<li>For details see <code>?par</code></li>
</ul>
<h4 id="add-regression-line">Add regression line</h4>
<pre><code class="language-r">plot(y[,1], y[,2])
myline &lt;- lm(y[,2]~y[,1]); abline(myline, lwd=2) 
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_regression-1.png" width="672" />
<pre><code class="language-r">summary(myline) 
</code></pre>
<pre><code>## 
## Call:
## lm(formula = y[, 2] ~ y[, 1])
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -0.40357 -0.17912 -0.04299  0.22147  0.46623 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)  
## (Intercept)   0.5764     0.2110   2.732   0.0258 *
## y[, 1]       -0.3647     0.3959  -0.921   0.3839  
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 0.3095 on 8 degrees of freedom
## Multiple R-squared:  0.09589,	Adjusted R-squared:  -0.01712 
## F-statistic: 0.8485 on 1 and 8 DF,  p-value: 0.3839
</code></pre>
<h4 id="log-scale">Log scale</h4>
<p>Same plot as above, but on log scale</p>
<pre><code class="language-r">plot(y[,1], y[,2], log=&quot;xy&quot;) 
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_regression_log-1.png" width="672" />
<h4 id="add-a-mathematical-expression">Add a mathematical expression</h4>
<pre><code class="language-r">plot(y[,1], y[,2]); text(y[1,1], y[1,2], expression(sum(frac(1,sqrt(x^2*pi)))), cex=1.3) 
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_regression_math-1.png" width="672" />
<h3 id="homework-3b">Homework 3B</h3>
<p>Homework 3B: <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw03/hw03/">Scatter Plots</a></p>
<h3 id="line-plots">Line Plots</h3>
<h4 id="single-data-set">Single data set</h4>
<pre><code class="language-r">plot(y[,1], type=&quot;l&quot;, lwd=2, col=&quot;blue&quot;) 
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_line_single-1.png" width="672" />
<h4 id="many-data-sets">Many Data Sets</h4>
<p>Plots line graph for all columns in data frame <code>y</code>. The <code>split.screen</code> function is used in this example in a for loop to overlay several line graphs in the same plot.</p>
<pre><code class="language-r">split.screen(c(1,1)) 
</code></pre>
<pre><code>## [1] 1
</code></pre>
<pre><code class="language-r">plot(y[,1], ylim=c(0,1), xlab=&quot;Measurement&quot;, ylab=&quot;Intensity&quot;, type=&quot;l&quot;, lwd=2, col=1)
for(i in 2:length(y[1,])) { 
	screen(1, new=FALSE)
	plot(y[,i], ylim=c(0,1), type=&quot;l&quot;, lwd=2, col=i, xaxt=&quot;n&quot;, yaxt=&quot;n&quot;, ylab=&quot;&quot;, xlab=&quot;&quot;, main=&quot;&quot;, bty=&quot;n&quot;) 
}
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_line_many-1.png" width="672" />
<pre><code class="language-r">close.screen(all=TRUE) 
</code></pre>
<h3 id="bar-plots">Bar Plots</h3>
<h4 id="basics">Basics</h4>
<pre><code class="language-r">barplot(y[1:4,], ylim=c(0, max(y[1:4,])+0.3), beside=TRUE, legend=letters[1:4]) 
text(labels=round(as.vector(as.matrix(y[1:4,])),2), x=seq(1.5, 13, by=1) + sort(rep(c(0,1,2), 4)), y=as.vector(as.matrix(y[1:4,]))+0.04) 
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_bar_simple-1.png" width="672" />
<h4 id="error-bars">Error Bars</h4>
<pre><code class="language-r">bar &lt;- barplot(m &lt;- rowMeans(y) * 10, ylim=c(0, 10))
stdev &lt;- sd(t(y))
arrows(bar, m, bar, m + stdev, length=0.15, angle = 90)
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_bar_error-1.png" width="672" />
<h3 id="histograms">Histograms</h3>
<pre><code class="language-r">hist(y, freq=TRUE, breaks=10)
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_hist-1.png" width="672" />
<h3 id="density-plots">Density Plots</h3>
<pre><code class="language-r">plot(density(y), col=&quot;red&quot;)
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_dens-1.png" width="672" />
<h3 id="pie-charts">Pie Charts</h3>
<pre><code class="language-r">pie(y[,1], col=rainbow(length(y[,1]), start=0.1, end=0.8), clockwise=TRUE)
legend(&quot;topright&quot;, legend=row.names(y), cex=1.3, bty=&quot;n&quot;, pch=15, pt.cex=1.8, 
col=rainbow(length(y[,1]), start=0.1, end=0.8), ncol=1) 
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_pie-1.png" width="672" />
<h3 id="color-selection-utilities">Color Selection Utilities</h3>
<p>Default color palette and how to change it</p>
<pre><code class="language-r">palette()
</code></pre>
<pre><code>## [1] &quot;black&quot;   &quot;#DF536B&quot; &quot;#61D04F&quot; &quot;#2297E6&quot; &quot;#28E2E5&quot; &quot;#CD0BBC&quot; &quot;#F5C710&quot; &quot;gray62&quot;
</code></pre>
<pre><code class="language-r">palette(rainbow(5, start=0.1, end=0.2))
palette()
</code></pre>
<pre><code>## [1] &quot;#FF9900&quot; &quot;#FFBF00&quot; &quot;#FFE600&quot; &quot;#F2FF00&quot; &quot;#CCFF00&quot;
</code></pre>
<pre><code class="language-r">palette(&quot;default&quot;)
</code></pre>
<p>The <code>gray</code> function allows to select any type of gray shades by providing values from 0 to 1</p>
<pre><code class="language-r">gray(seq(0.1, 1, by= 0.2))
</code></pre>
<pre><code>## [1] &quot;#1A1A1A&quot; &quot;#4D4D4D&quot; &quot;#808080&quot; &quot;#B3B3B3&quot; &quot;#E6E6E6&quot;
</code></pre>
<p>Color gradients with <code>colorpanel</code> function from <code>gplots</code> library`</p>
<pre><code class="language-r">library(gplots)
colorpanel(5, &quot;darkblue&quot;, &quot;yellow&quot;, &quot;white&quot;)
</code></pre>
<pre><code>## [1] &quot;#00008B&quot; &quot;#808046&quot; &quot;#FFFF00&quot; &quot;#FFFF80&quot; &quot;#FFFFFF&quot;
</code></pre>
<p>Much more on colors in R see Earl Glynn’s color chart <a href="http://research.stowers-institute.org/efg/R/Color/Chart/">here</a></p>
<h3 id="saving-graphics-to-file">Saving Graphics to File</h3>
<p>After the <code>pdf()</code> command all graphs are redirected to file <code>test.pdf</code>. Works for all common formats similarly: jpeg, png, ps, tiff, …</p>
<pre><code class="language-r">pdf(&quot;test.pdf&quot;)
plot(1:10, 1:10)
dev.off() 
</code></pre>
<p>Generates Scalable Vector Graphics (SVG) files that can be edited in vector graphics programs, such as InkScape.</p>
<pre><code class="language-r">library(&quot;RSvgDevice&quot;)
devSVG(&quot;test.svg&quot;)
plot(1:10, 1:10)
dev.off() 
</code></pre>
<h3 id="homework-3c">Homework 3C</h3>
<p>Homework 3C: <a href="https://girke.bioinformatics.ucr.edu/GEN242/assignments/homework/hw03/hw03/">Bar Plots</a></p>
<h2 id="analysis-routine">Analysis Routine</h2>
<h3 id="overview-1">Overview</h3>
<p>The following exercise introduces a variety of useful data analysis utilities in R.</p>
<h3 id="analysis-routine-data-import">Analysis Routine: Data Import</h3>
<ul>
<li>
<p><strong>Step 1</strong>: To get started with this exercise, direct your R session to a dedicated workshop directory and download into this directory the following sample tables. Then import the files into Excel and save them as tab delimited text files.</p>
<ul>
<li><a href="https://cluster.hpcc.ucr.edu/~tgirke/Documents/R_BioCond/Samples/MolecularWeight_tair7.xls">MolecularWeight_tair7.xls</a></li>
<li><a href="https://cluster.hpcc.ucr.edu/~tgirke/Documents/R_BioCond/Samples/TargetP_analysis_tair7.xls">TargetP_analysis_tair7.xls</a></li>
</ul>
</li>
</ul>
<p><strong>Import the tables into R</strong></p>
<p>Import molecular weight table</p>
<pre><code class="language-r">my_mw &lt;- read.delim(file=&quot;MolecularWeight_tair7.xls&quot;, header=TRUE, sep=&quot;\t&quot;) 
my_mw[1:2,]
</code></pre>
<p>Import subcelluar targeting table</p>
<pre><code class="language-r">my_target &lt;- read.delim(file=&quot;TargetP_analysis_tair7.xls&quot;, header=TRUE, sep=&quot;\t&quot;) 
my_target[1:2,]
</code></pre>
<p>Online import of molecular weight table</p>
<pre><code class="language-r">my_mw &lt;- read.delim(file=&quot;http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/MolecularWeight_tair7.xls&quot;, header=TRUE, sep=&quot;\t&quot;) 
my_mw[1:2,]
</code></pre>
<pre><code>##   Sequence.id Molecular.Weight.Da. Residues
## 1 AT1G08520.1                83285      760
## 2 AT1G08530.1                27015      257
</code></pre>
<p>Online import of subcelluar targeting table</p>
<pre><code class="language-r">my_target &lt;- read.delim(file=&quot;http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/TargetP_analysis_tair7.xls&quot;, header=TRUE, sep=&quot;\t&quot;) 
my_target[1:2,]
</code></pre>
<pre><code>##      GeneName Loc   cTP   mTP    SP other
## 1 AT1G08520.1   C 0.822 0.137 0.029 0.039
## 2 AT1G08530.1   C 0.817 0.058 0.010 0.100
</code></pre>
<h3 id="merging-data-frames">Merging Data Frames</h3>
<ul>
<li><strong>Step 2</strong>: Assign uniform gene ID column titles</li>
</ul>
<!-- end list -->
<pre><code class="language-r">colnames(my_target)[1] &lt;- &quot;ID&quot;
colnames(my_mw)[1] &lt;- &quot;ID&quot; 
</code></pre>
<ul>
<li><strong>Step 3</strong>: Merge the two tables based on common ID field</li>
</ul>
<!-- end list -->
<pre><code class="language-r">my_mw_target &lt;- merge(my_mw, my_target, by.x=&quot;ID&quot;, by.y=&quot;ID&quot;, all.x=TRUE)
</code></pre>
<ul>
<li><strong>Step 4</strong>: Shorten one table before the merge and then remove the non-matching rows (NAs) in the merged file</li>
</ul>
<!-- end list -->
<pre><code class="language-r">my_mw_target2a &lt;- merge(my_mw, my_target[1:40,], by.x=&quot;ID&quot;, by.y=&quot;ID&quot;, all.x=TRUE)  # To remove non-matching rows, use the argument setting 'all=FALSE'.
my_mw_target2 &lt;- na.omit(my_mw_target2a) # Removes rows containing &quot;NAs&quot; (non-matching rows).
</code></pre>
<ul>
<li><strong>Homework 3D</strong>: How can the merge function in the previous step be executed so that only the common rows among the two data frames are returned? Prove that both methods - the two step version with <code>na.omit</code> and your method - return identical results.</li>
<li><strong>Homework 3E</strong>: Replace all <code>NAs</code> in the data frame <code>my_mw_target2a</code> with zeros.</li>
</ul>
<h3 id="filtering-data">Filtering Data</h3>
<ul>
<li><strong>Step 5</strong>: Retrieve all records with a value of greater than 100,000 in ‘MW’ column and ‘C’ value in ‘Loc’ column (targeted to chloroplast).</li>
</ul>
<!-- end list -->
<pre><code class="language-r">query &lt;- my_mw_target[my_mw_target[, 2] &gt; 100000 &amp; my_mw_target[, 4] == &quot;C&quot;, ] 
query[1:4, ]
</code></pre>
<pre><code>##              ID Molecular.Weight.Da. Residues Loc   cTP   mTP    SP other
## 219 AT1G02730.1               132588     1181   C 0.972 0.038 0.008 0.045
## 243 AT1G02890.1               136825     1252   C 0.748 0.529 0.011 0.013
## 281 AT1G03160.1               100732      912   C 0.871 0.235 0.011 0.007
## 547 AT1G05380.1               126360     1138   C 0.740 0.099 0.016 0.358
</code></pre>
<pre><code class="language-r">dim(query)
</code></pre>
<pre><code>## [1] 170   8
</code></pre>
<ul>
<li><strong>Homework 3F</strong>: How many protein entries in the <code>my</code>_mw<code>_target</code> data frame have a MW of greater then 4,000 and less then 5,000. Subset the data frame accordingly and sort it by MW to check that your result is correct.</li>
</ul>
<h3 id="string-substitutions">String Substitutions</h3>
<ul>
<li><strong>Step 6</strong>: Use a regular expression in a substitute function to generate a separate ID column that lacks the gene model extensions.</li>
</ul>
<!-- end list -->
<pre><code class="language-r">my_mw_target3 &lt;- data.frame(loci=gsub(&quot;\\..*&quot;, &quot;&quot;, as.character(my_mw_target[,1]), perl = TRUE), my_mw_target)
my_mw_target3[1:3,1:8]
</code></pre>
<pre><code>##        loci          ID Molecular.Weight.Da. Residues Loc  cTP   mTP    SP
## 1 AT1G01010 AT1G01010.1                49426      429   _ 0.10 0.090 0.075
## 2 AT1G01020 AT1G01020.1                28092      245   * 0.01 0.636 0.158
## 3 AT1G01020 AT1G01020.2                21711      191   * 0.01 0.636 0.158
</code></pre>
<ul>
<li><strong>Homework 3G</strong>: Retrieve those rows in <code>my_mw_target3</code> where the second column contains the following identifiers: <code>c(&quot;AT5G52930.1&quot;, &quot;AT4G18950.1&quot;, &quot;AT1G15385.1&quot;, &quot;AT4G36500.1&quot;, &quot;AT1G67530.1&quot;)</code>. Use the <code>%in%</code> function for this query. As an alternative approach, assign the second column to the row index of the data frame and then perform the same query again using the row index. Explain the difference of the two methods.</li>
</ul>
<h3 id="calculations-on-data-frames">Calculations on Data Frames</h3>
<ul>
<li><strong>Step 7</strong>: Count the number of duplicates in the loci column with the <code>table</code> function and append the result to the data frame with the <code>cbind</code> function.</li>
</ul>
<!-- end list -->
<pre><code class="language-r">mycounts &lt;- table(my_mw_target3[,1])[my_mw_target3[,1]]
my_mw_target4 &lt;- cbind(my_mw_target3, Freq=mycounts[as.character(my_mw_target3[,1])]) 
</code></pre>
<ul>
<li><strong>Step 8</strong>: Perform a vectorized devision of columns 3 and 4 (average AA weight per protein)</li>
</ul>
<!-- end list -->
<pre><code class="language-r">data.frame(my_mw_target4, avg_AA_WT=(my_mw_target4[,3] / my_mw_target4[,4]))[1:2,] 
</code></pre>
<pre><code>##        loci          ID Molecular.Weight.Da. Residues Loc  cTP   mTP    SP other Freq.Var1
## 1 AT1G01010 AT1G01010.1                49426      429   _ 0.10 0.090 0.075 0.925 AT1G01010
## 2 AT1G01020 AT1G01020.1                28092      245   * 0.01 0.636 0.158 0.448 AT1G01020
##   Freq.Freq avg_AA_WT
## 1         1  115.2121
## 2         2  114.6612
</code></pre>
<ul>
<li><strong>Step 9</strong>: Calculate for each row the mean and standard deviation across several columns</li>
</ul>
<!-- end list -->
<pre><code class="language-r">mymean &lt;- apply(my_mw_target4[,6:9], 1, mean)
mystdev &lt;- apply(my_mw_target4[,6:9], 1, sd, na.rm=TRUE)
data.frame(my_mw_target4, mean=mymean, stdev=mystdev)[1:2,5:12] 
</code></pre>
<pre><code>##   Loc  cTP   mTP    SP other Freq.Var1 Freq.Freq   mean
## 1   _ 0.10 0.090 0.075 0.925 AT1G01010         1 0.2975
## 2   * 0.01 0.636 0.158 0.448 AT1G01020         2 0.3130
</code></pre>
<h3 id="plotting-example">Plotting Example</h3>
<ul>
<li><strong>Step 10</strong>: Generate scatter plot columns: ‘MW’ and ‘Residues’</li>
</ul>
<!-- end list -->
<pre><code class="language-r">plot(my_mw_target4[1:500,3:4], col=&quot;red&quot;)
</code></pre>
<img src="../../../en/tutorials/rbasics/Rbasics_files/figure-html/plot_example-1.png" width="672" />
<h3 id="export-results-and-run-entire-exercise-as-script">Export Results and Run Entire Exercise as Script</h3>
<ul>
<li><strong>Step 11</strong>: Write the data frame <code>my_mw_target4</code> into a tab-delimited text file and inspect it in Excel.</li>
</ul>
<!-- end list -->
<pre><code class="language-r">write.table(my_mw_target4, file=&quot;my_file.xls&quot;, quote=FALSE, sep=&quot;\t&quot;, col.names = NA) 
</code></pre>
<ul>
<li><strong>Homework 3H</strong>: Write all commands from this exercise into an R script named <code>exerciseRbasics.R</code>, or download it from <a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/exerciseRbasics.R">here</a>. Then execute the script with the <code>source</code> function like this: <code>source(&quot;exerciseRbasics.R&quot;)</code>. This will run all commands of this exercise and generate the corresponding output files in the current working directory. For homework 3H it is not necessary to submit the result files generated by the <code>exerciseRbasics.R</code> script. Stating how the script was executed (<em>e.g.</em> <code>source</code> or <code>Rscript</code> command) will be sufficient.</li>
</ul>
<!-- end list -->
<pre><code class="language-r">source(&quot;exerciseRbasics.R&quot;)
</code></pre>
<p>Or run it from the command-line (not from R!) with <code>Rscript</code> like this:</p>
<pre><code>Rscript exerciseRbasics.R
</code></pre>
<h2 id="miscellaneous-topics">Miscellaneous Topics</h2>
<h3 id="upgrading-to-new-rbioc-versions">Upgrading to New R/Bioc Versions</h3>
<p>When upgrading to a new R version, it is important to understand that a reinstall of
all R packages is necessary because CRAN/Bioc packages are developed and tested for specific
R versions. This means when upgrading R, then the corresponding packages need to
be upgraded to the versions that match the new R version. The following steps will work
in many situations.</p>
<ol>
<li>
<p>Export a list of all packages installed in a current version of R to a file
(below named <code>my_R_pkgs.txt</code>) by running the following
commands from within R (or use <code>Rscript -e</code> from command-line)</p>
<pre><code class="language-r">my_R_pkgs &lt;- rownames(installed.packages())
writeLines(my_R_pkgs, &quot;my_R_pkgs.txt&quot;)
</code></pre>
</li>
<li>
<p>Install new version of R, and then from within the new R version all
packages one had installed before. The first install command below installs
first a series of packages that are useful to have in general no matter
what. Custom packages are then installed in the next lines. Note, this
can only install packages from CRAN and Bioconductor. Packages from custom
sources, including private GitHub accounts, need to be installed separately.
Usually, one can identify them by the report generated at the end of the below install
routine telling which packages are not available on CRAN or
Bioconductor.</p>
<pre><code class="language-r">install.packages(c(&quot;devtools&quot;, &quot;tidyverse&quot;, &quot;BiocManager&quot;))
BiocManager::install(version = &quot;3.17&quot;) # look up current Bioc version here: https://bit.ly/3NADnll
my_R_pkgs &lt;- readLines(&quot;my_R_pkgs.txt&quot;)
BiocManager::install(my_R_pkgs)
</code></pre>
</li>
</ol>
<h2 id="session-info">Session Info</h2>
<pre><code class="language-r">sessionInfo()
</code></pre>
<pre><code>## R version 4.1.3 (2022-03-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Debian GNU/Linux 10 (buster)
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.8.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.8.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] gplots_3.1.1     forcats_0.5.1    stringr_1.4.0    dplyr_1.0.7      purrr_0.3.4     
##  [6] readr_2.1.1      tidyr_1.1.4      tibble_3.1.6     tidyverse_1.3.1  ggplot2_3.3.5   
## [11] limma_3.50.0     BiocStyle_2.22.0
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.8.3        lubridate_1.8.0     gtools_3.9.2        assertthat_0.2.1   
##  [5] digest_0.6.29       utf8_1.2.2          R6_2.5.1            cellranger_1.1.0   
##  [9] backports_1.4.0     reprex_2.0.1        evaluate_0.15       highr_0.9          
## [13] httr_1.4.2          blogdown_1.8.2      pillar_1.6.4        rlang_1.0.2        
## [17] readxl_1.3.1        rstudioapi_0.13     jquerylib_0.1.4     rmarkdown_2.13     
## [21] munsell_0.5.0       broom_0.7.10        compiler_4.1.3      modelr_0.1.8       
## [25] xfun_0.30           pkgconfig_2.0.3     htmltools_0.5.2     tidyselect_1.1.1   
## [29] bookdown_0.25       codetools_0.2-18    fansi_0.5.0         crayon_1.4.2       
## [33] tzdb_0.2.0          dbplyr_2.1.1        withr_2.4.3         bitops_1.0-7       
## [37] grid_4.1.3          jsonlite_1.8.0      gtable_0.3.0        lifecycle_1.0.1    
## [41] DBI_1.1.1           magrittr_2.0.2      scales_1.1.1        KernSmooth_2.23-20 
## [45] cli_3.1.0           stringi_1.7.6       fs_1.5.2            xml2_1.3.3         
## [49] bslib_0.3.1         ellipsis_0.3.2      generics_0.1.1      vctrs_0.3.8        
## [53] tools_4.1.3         glue_1.6.2          hms_1.1.1           fastmap_1.1.0      
## [57] yaml_2.3.5          colorspace_2.0-2    BiocManager_1.30.16 caTools_1.18.2     
## [61] rvest_1.0.2         knitr_1.37          haven_2.4.3         sass_0.4.0
</code></pre>
<h2 id="references">References</h2>

	
	
	<div class="text-muted mt-5 pt-3 border-top">Last modified 2023-05-03: <a  href="https://github.com/tgirke/GEN242//commit/3344605280f54f5a782a002a230189df1e7a1ab9">some edits (334460528)</a>
</div>
</div>


              </main>
          


        </div>
      </div>
      
<footer class="bg-dark py-5 row d-print-none">
  <div class="container-fluid mx-sm-5">
    <div class="row">
      <div class="col-6 col-sm-4 text-xs-center order-sm-2">
        
      </div>
      <div class="col-6 col-sm-4 text-right text-xs-center order-sm-3">
        
      </div>
      <div class="col-12 col-sm-4 text-center py-2 order-sm-2">
        <small class="text-white">&copy; 2023 The GEN242 Instructors All Rights Reserved</small>
        <small class="ml-1"><a href="https://www.ucr.edu/privacy.html" target="_blank">Privacy Policy</a></small>
	
      </div>
    </div>
  </div>
</footer>


    </div>
    
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>


<script>
  (function() {
    var i, text, code, codes = document.getElementsByTagName('code');
    for (i = 0; i < codes.length;) {
      code = codes[i];
      if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
        text = code.textContent;
        if (/^\$[^$]/.test(text) && /[^$]\$$/.test(text)) {
          text = text.replace(/^\$/, '\\(').replace(/\$$/, '\\)');
          code.textContent = text;
        }
        if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
            /^\$(.|\s)+\$$/.test(text) ||
            /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
          code.outerHTML = code.innerHTML;  
          continue;
        }
      }
      i++;
    }
  })();
</script>

<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
    tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true,
        processEnvironments: true
    },
    // Center justify equations in code and markdown cells. Elsewhere
    // we use CSS to left justify single line equations in code cells.
    displayAlign: 'center',
    "HTML-CSS": {
        styles: {'.MathJax_Display': {"margin": 0}},
        linebreaks: { automatic: true }
    }
});
</script>













<script src="../../../js/main.min.b98a3473667defb32cda6c536707346795949f52932465ac5010615f974ccd60.js" integrity="sha256-uYo0c2Z977Ms2mxTZwc0Z5WUn1KTJGWsUBBhX5dMzWA=" crossorigin="anonymous"></script>



<script src='../../../js/prism.js'></script>



  </body>
</html>
